<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Multidimensional biodiversity data: management and analysis: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Multidimensional biodiversity data: management and analysis
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Multidimensional biodiversity data: management and analysis
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul>
</li>
      </ul>
</div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
<input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset>
</form>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Multidimensional biodiversity data: management and analysis
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="abundance-data.html">2. Abundance Data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="traits-data.html">3. Trait data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="popgen-data.html">4. Working with population genetic data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="phylo-data.html">5. Summarizing phylogenetic data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="connecting-data-types.html">6. Connecting different data types</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="finding_data.html">7. Finding multi-dimensional biodiversity data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="care.html">8. The CARE principles and open data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Introduction</a></p>
<hr>
<p> Last updated on 2023-06-05 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the four types of biodiversity data we’ll be covering in
this course?</li>
<li>What are the four steps in handling each type?</li>
<li>What do the outcomes of analyzing multidimensional biodiversity data
(MDBD) look like visualized?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants should be able to…</p>
<ol style="list-style-type: decimal">
<li>Explain the four types of biodiversity data and the four steps in
handling each</li>
<li>Find information about the rest of the lesson</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<ul>
<li>intro of instructors</li>
<li>expectations of learners</li>
<li>exciting data narrative</li>
<li>system reqs</li>
</ul></section><section id="exciting-data-narrative"><h2 class="section-heading">Exciting data narrative<a class="anchor" aria-label="anchor" href="#exciting-data-narrative"></a>
</h2>
<hr class="half-width">
<ul>
<li>young island dominated by immigration, neutrality</li>
<li>older island</li>
</ul>
<div id="challenge-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>What do you think this plot says about ecological process?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Output
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-abundance-data"><p>Content from <a href="abundance-data.html">Abundance Data</a></p>
<hr>
<p> Last updated on 2023-06-05 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/abundance-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How is organismal abundance data typically stored?</li>
<li>How is the species abundance distribution used to compare abundance
data from differing systems?</li>
<li>How do we use summary statistics - including Hill numbers - to
quantitatively compare species abundance distributions?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants should be able to…</p>
<ol style="list-style-type: decimal">
<li>Import and examine organismal abundance data from .csv files</li>
<li>Clean taxonomic names</li>
<li>Manipulate abundance data into different formats</li>
<li>Generate species abundance distribution plots from abundance
data</li>
<li>Summarize species abundance data using Hill numbers</li>
<li>Interpret how different Hill numbers correspond to different
patterns in species diversity</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h1>
<p>For this episode, we’ll be working with a couple of specialized
packages for biological data.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">vegan</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: permute</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: lattice</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>This is vegan 2.6-4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">tidyr</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level1">
<h1 id="data-import-and-cleaning">Data import and cleaning<a class="anchor" aria-label="anchor" href="#data-import-and-cleaning"></a>
</h1>
<div class="section level2">
<h2 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<p>We’ll be working with ecological species abundance data stored in
<code>.csv</code> files. For help working with other storage formats, <a href="https://datacarpentry.org/R-ecology-lesson/05-r-and-databases.html" class="external-link">the
Carpentries’ Ecological Data lesson materials on databases</a> are a
great place to start!</p>
<p>Let’s load the data:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"http://bit.ly/3INNA9V"</span><span class="op">)</span></span></code></pre>
</div>
<p>And look at what we’ve got:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  island     site               GenSp
1  Kauai Kauai_01 Xyleborus perforans
2  Kauai Kauai_01 Xyleborus perforans
3  Kauai Kauai_01 Xyleborus perforans
4  Kauai Kauai_01 Xyleborus perforans
5  Kauai Kauai_01 Xyleborus perforans
6  Kauai Kauai_01 Xyleborus perforans</code></pre>
</div>
<p><code>abundances</code> is a data frame with columns for
<code>island</code>, <code>site</code>, and <code>GenSp</code>. These
data are in a common format in which each row represents a single
individual observed.</p>
</div>
<div class="section level2">
<h2 id="cleaning-taxonomic-names">Cleaning taxonomic names<a class="anchor" aria-label="anchor" href="#cleaning-taxonomic-names"></a>
</h2>
<p>The first thing we’ll want to do is check for human error wherever we
can, in this case in the form of typos in data entry.</p>
<p>The <code>taxize</code> R package can help identify and resolve
simple typos in taxonomic names.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_list</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="va">abundances</span><span class="op">$</span><span class="va">GenSp</span><span class="op">)</span></span>
<span></span>
<span><span class="va">name_resolve</span> <span class="op">&lt;-</span> <span class="fu">gnr_resolve</span><span class="op">(</span><span class="va">species_list</span>, best_match_only <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                            canonical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># returns only name, not authority</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">name_resolve</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 5
  user_supplied_name        submitted_name data_source_title score matched_name2
  &lt;chr&gt;                     &lt;chr&gt;          &lt;chr&gt;             &lt;dbl&gt; &lt;chr&gt;        
1 Xyleborus perforans       Xyleborus per… National Center … 0.988 Xyleborus pe…
2 Agrotis chersotoides      Agrotis chers… Catalogue of Lif… 0.75  Agrotis      
3 Oligotoma saundersii      Oligotoma sau… Wikispecies       0.988 Oligotoma sa…
4 Drosophila humeralis      Drosophila hu… Wikispecies       0.988 Drosophila h…
5 Entomobrya griseoolivata  Entomobrya gr… National Center … 0.988 Entomobrya g…
6 Pseudosinella kalalauens… Pseudosinella… Encyclopedia of … 0.988 Pseudosinell…</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mismatches</span> <span class="op">&lt;-</span> <span class="va">name_resolve</span><span class="op">[</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span> <span class="op">!=</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">mismatches</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"user_supplied_name"</span>, <span class="st">"matched_name2"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 5 × 2
  user_supplied_name     matched_name2       
  &lt;chr&gt;                  &lt;chr&gt;               
1 Agrotis chersotoides   Agrotis             
2 Trigonidium flelectens Trigonidium flectens
3 Schrankia simplllex    Schrankia simplex   
4 Holcobius insignigis   Holcobius insignis  
5 Carposina mauiii       Carposina mauii     </code></pre>
</div>
<p>Four of these are just typos. But <code>Agrotis chersotoides</code>
in our data is resolved only to Agrotis. What’s up there?</p>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>Agrotis taxonomic resolution</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span><span class="op">[</span><span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span> <span class="op">==</span> <span class="st">"Agrotis chersotoides"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Peridroma chersotoides"</span></span></code></pre>
</div>
<p>Now we need to add the newly-resolved names to our
<code>abundnaces</code> data:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">abundances</span>, <span class="va">name_resolve</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"GenSp"</span> <span class="op">=</span> <span class="st">"user_supplied_name"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">abundances</span><span class="op">$</span><span class="va">final_name</span> <span class="op">&lt;-</span> <span class="va">abundances</span><span class="op">$</span><span class="va">matched_name2</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="wrangling-abundance-data">Wrangling abundance data<a class="anchor" aria-label="anchor" href="#wrangling-abundance-data"></a>
</h2>
<p>The data we have have one record per individual, but we’d like to
have the <em>total</em> number of individuals of each species in our
data.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundance_tallies</span> <span class="op">&lt;-</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">abundances</span>, <span class="va">site</span>, <span class="va">island</span>, <span class="va">final_name</span><span class="op">)</span></span>
<span><span class="va">abundance_tallies</span> <span class="op">&lt;-</span> <span class="fu">summarize</span><span class="op">(</span><span class="va">abundance_tallies</span>, abundance <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'site', 'island'. You can override using
the `.groups` argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">abundance_tallies</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 4
# Groups:   site, island [1]
  site            island       final_name                         abundance
  &lt;chr&gt;           &lt;chr&gt;        &lt;chr&gt;                                  &lt;int&gt;
1 HawaiiIsland_01 HawaiiIsland Acanthia procellaris                       3
2 HawaiiIsland_01 HawaiiIsland Actia eucosmae                             4
3 HawaiiIsland_01 HawaiiIsland Anomalochrysa fulvescens rhododora         2
4 HawaiiIsland_01 HawaiiIsland Anoplolepis gracilipes                     5
5 HawaiiIsland_01 HawaiiIsland Anurophorus lohi                           2
6 HawaiiIsland_01 HawaiiIsland Aphis citricola                            3</code></pre>
</div>
<p>Write out the tallied data for later use:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write.csv</span><span class="op">(</span><span class="va">abundance_tallies</span>,  <span class="st">"abundance_tallies.csv"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="site-by-species-matrix">Site-by-species matrix<a class="anchor" aria-label="anchor" href="#site-by-species-matrix"></a>
</h2>
<p>Often we’ll want to work with a <em>site by species matrix</em>,
which has sites as rows and species as columns. We can get there using
the <code>pivot_wider</code> function from the package
<code>tidyr</code>:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="fu">pivot_wider</span><span class="op">(</span><span class="va">abundance_tallies</span>, id_cols <span class="op">=</span> <span class="va">site</span>,</span>
<span>                              names_from <span class="op">=</span> <span class="va">final_name</span>,</span>
<span>                              values_from <span class="op">=</span> <span class="va">abundance</span>, </span>
<span>                              values_fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 10
# Groups:   site [3]
  site            `Acanthia procellaris` `Actia eucosmae` Anomalochrysa fulves…¹
  &lt;chr&gt;                            &lt;int&gt;            &lt;int&gt;                  &lt;int&gt;
1 HawaiiIsland_01                      3                4                      2
2 Kauai_01                             0                0                      0
3 Maui_01                              0                0                      0
# ℹ abbreviated name: ¹​`Anomalochrysa fulvescens rhododora`
# ℹ 6 more variables: `Anoplolepis gracilipes` &lt;int&gt;, `Anurophorus lohi` &lt;int&gt;,
#   `Aphis citricola` &lt;int&gt;, `Carphuroides pectinatus` &lt;int&gt;,
#   `Carposina achroana` &lt;int&gt;, `Carulaspis minima` &lt;int&gt;</code></pre>
</div>
<p>We’ll want this data to have <em>row names</em> based on the sites,
so we’ll need some more steps:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">row.names</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span><span class="op">$</span><span class="va">site</span></span>
<span></span>
<span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                Acanthia procellaris Actia eucosmae
HawaiiIsland_01                    3              4
Kauai_01                           0              0
Maui_01                            0              0
                Anomalochrysa fulvescens rhododora Anoplolepis gracilipes
HawaiiIsland_01                                  2                      5
Kauai_01                                         0                      0
Maui_01                                          0                      0
                Anurophorus lohi Aphis citricola Carphuroides pectinatus
HawaiiIsland_01                2               3                       7
Kauai_01                       0               0                       0
Maui_01                        0               0                       0
                Carposina achroana Carulaspis minima Chlorichaeta albipennis
HawaiiIsland_01                  1                 1                       2
Kauai_01                         0                 0                       0
Maui_01                          0                 0                       0
                Cillaeopeplus dubius Cis insulicola Cis nudipennis
HawaiiIsland_01                    2              1              2
Kauai_01                           0              0              0
Maui_01                            0              0              0
                Cydia obliqua Dicranomyia hawaiiensis Drosophila lepidobregma
HawaiiIsland_01             3                       1                       3
Kauai_01                    0                       0                       0
Maui_01                     0                       0                       0
                Drosophila perkinsi Echthromorpha fuscator Elenchus melanias
HawaiiIsland_01                   4                      4                 1
Kauai_01                          0                      0                 0
Maui_01                           0                      0                 0
                Epiphanis tristis Eupelmus paraleucothrix Eupodes sigmoidensis
HawaiiIsland_01                 1                      20                    8
Kauai_01                        0                       0                    0
Maui_01                         0                       0                    0
                Fannia canicularis Hedylepta blackburni Heteropsylla cubana
HawaiiIsland_01                  1                    5                   2
Kauai_01                         0                    0                   0
Maui_01                         10                    0                   0
                Hypogastrura pahiku Ilburnia coprosmicola Laupala nigra
HawaiiIsland_01                   4                     1             2
Kauai_01                          0                     0             0
Maui_01                           0                     0             0
                Lispocephala fuscobrunnea Listroderes costirostris
HawaiiIsland_01                         2                       13
Kauai_01                                0                        0
Maui_01                                 0                        0
                Mecyclothorax purpuripennis Mestolobes droseropa
HawaiiIsland_01                           7                    2
Kauai_01                                  0                    0
Maui_01                                   0                    0
                Metaphycus anneckei Myllaena pacifica Neoscona theisi
HawaiiIsland_01                   1                 1               5
Kauai_01                          0                 0               0
Maui_01                           0                 0               0
                Nesogonia blackburni Nesomimesa hawaiiensis Notiphila insularis
HawaiiIsland_01                   19                      1                   1
Kauai_01                           0                      0                   0
Maui_01                            0                      0                   0
                Notoedres cati Nysius montivagus Omiodes anastrepta
HawaiiIsland_01              9                10                 10
Kauai_01                     0                 0                  0
Maui_01                      0                 0                  0
                Ophiusa disjungens Orthoea vincta Paratrechina bourbonica
HawaiiIsland_01                  2              3                       6
Kauai_01                         0              0                       0
Maui_01                          0              0                       0
                Pauahiana lineata Pediobius wilderi Pedronia cibotii
HawaiiIsland_01                 1                 6                4
Kauai_01                        0                 0                0
Maui_01                         0                 0                0
                Phytoseius hawaiiensis Pilorhagidia hirsuta Piophila casei
HawaiiIsland_01                      2                    1              1
Kauai_01                             0                    0              0
Maui_01                              0                    0              0
                Plagithmysus laticollis Projectothrips trespinus Psilopa olga
HawaiiIsland_01                       1                        1            3
Kauai_01                              0                        0            0
Maui_01                               0                        0            0
                Scaptomyza adunca Syrphophagus aphidivorus Tegonotus hibiscella
HawaiiIsland_01                20                        1                    1
Kauai_01                        0                        0                    0
Maui_01                         0                        0                    0
                Trigonidium kona Tympanococcus tympanistus Xyleborus vulcanus
HawaiiIsland_01               16                         1                  1
Kauai_01                       0                         0                 12
Maui_01                        0                         0                  0
                Aedes nocturnus Agrotis xiphias Aleurothrixus floccosus
HawaiiIsland_01               0               0                       0
Kauai_01                      1              23                       3
Maui_01                       0               0                       6
                Bembidion munroi brevicolle Brontolaemus nudicornis
HawaiiIsland_01                           0                       0
Kauai_01                                  8                       5
Maui_01                                   0                       0
                Campsicnemus penicillatus Cardiocondyla venustula
HawaiiIsland_01                         0                       0
Kauai_01                                7                      12
Maui_01                                 0                       0
                Cardiocondyla wroughtonii Carposina olivaceonitens
HawaiiIsland_01                         0                        0
Kauai_01                               19                       13
Maui_01                                 7                        0
                Carposina viridis Carulaspis giffardi Cheyletus eruditus
HawaiiIsland_01                 0                   0                  0
Kauai_01                        6                   4                 23
Maui_01                         0                   0                  0
                Chrysomphalus dictyospermi Chrysoperla comanche
HawaiiIsland_01                          0                    0
Kauai_01                                 3                    5
Maui_01                                  0                    0
                Cligenes marianensis Clunio vagans Coccygomimus punicipes
HawaiiIsland_01                    0             0                      0
Kauai_01                          15            14                      2
Maui_01                            0             0                      0
                Conioscinella formosa Contarinia sorghicola
HawaiiIsland_01                     0                     0
Kauai_01                            1                     4
Maui_01                             0                     0
                Cryptophlebia ombrodelta Dacnitus currax
HawaiiIsland_01                        0               0
Kauai_01                               5              14
Maui_01                                0               0
                Decadarchis penicillata Dialeurodes kirkaldyi
HawaiiIsland_01                       0                     0
Kauai_01                              6                     8
Maui_01                               0                     0
                Diplonevra peregrina Disenochus erythropus Drosophila humeralis
HawaiiIsland_01                    0                     0                    0
Kauai_01                          19                    15                   18
Maui_01                            0                     0                    0
                Drosophila ornata Drosophila papaalai Dyscritomyia fulgens
HawaiiIsland_01                 0                   0                    0
Kauai_01                        5                  12                    4
Maui_01                         0                   0                    0
                Entomobrya griseoolivata Epiphyas postvittana Evaza javanensis
HawaiiIsland_01                        0                    0                0
Kauai_01                              22                   13                7
Maui_01                                0                    0                0
                Friesea oleia Hawaiodillo sharpi Hemiberlesia rapax
HawaiiIsland_01             0                  0                  0
Kauai_01                   11                 11                  8
Maui_01                     0                  0                  0
                Heterocrossa plumbeonitida Holcobius frater Holcobius insignis
HawaiiIsland_01                          0                0                  0
Kauai_01                                33                4                 18
Maui_01                                  0                0                  0
                Hypocala deflorata Hyposmocoma centralis Hyposmocoma divisa
HawaiiIsland_01                  0                     0                  0
Kauai_01                         4                    11                  4
Maui_01                          0                     0                  0
                Hyposmocoma palmivora Hyposmocoma petroptilota
HawaiiIsland_01                     0                        0
Kauai_01                           12                        7
Maui_01                             0                        0
                Hyposmocoma sordidella Laupala kanaele Laupala wailua
HawaiiIsland_01                      0               0              0
Kauai_01                             7              10              9
Maui_01                              0               0              0
                Lycaena boetica Mesothriscus optimus Mesovelia mulsanti
HawaiiIsland_01               0                    0                  0
Kauai_01                     15                    3                 11
Maui_01                       0                    0                  0
                Nabis blackburni Nabis lolupe Necrobia ruficollis
HawaiiIsland_01                0            0                   0
Kauai_01                      18           10                   8
Maui_01                        0            0                   0
                Neelus minutus Nesiomiris seminotatus Nesodynerus sandwichensis
HawaiiIsland_01              0                      0                         0
Kauai_01                     7                     19                         6
Maui_01                      0                      0                         0
                Nipponorthezia guadalcanalia Oceanides humeralis
HawaiiIsland_01                            0                   0
Kauai_01                                   8                   3
Maui_01                                    0                   0
                Oliarus kauaiensis Oligota kauaiensis Oligotoma saundersii
HawaiiIsland_01                  0                  0                    0
Kauai_01                         8                 18                    8
Maui_01                          0                  0                    0
                Onychiurus folsomi Orthellia caesarion Oxytelus advena
HawaiiIsland_01                  0                   0               0
Kauai_01                        12                  14               6
Maui_01                          0                   0               0
                Paralabella curvicauda Pararrhaptica dermatopa
HawaiiIsland_01                      0                       0
Kauai_01                             3                      15
Maui_01                              0                       0
                Peridroma chersotoides Peridroma coniotis Phidippus audax
HawaiiIsland_01                      0                  0               0
Kauai_01                            15                  3              29
Maui_01                              0                  0               0
                Plagithmysus sugawai Polypedilum nubiferum Proterhinus setiger
HawaiiIsland_01                    0                     0                   0
Kauai_01                           8                    19                  15
Maui_01                            0                     0                   0
                Psammoechus insularis Pseudopterocheilus relictus
HawaiiIsland_01                     0                           0
Kauai_01                           12                           8
Maui_01                             0                           0
                Pseudosinella kalalauensis Quadristruma emmae Scaptomyza hardyi
HawaiiIsland_01                          0                  0                 0
Kauai_01                                 2                 10                 8
Maui_01                                  0                  0                 0
                Scholastes bimaculatus Schrankia altivolans Schrankia simplex
HawaiiIsland_01                      0                    0                 0
Kauai_01                             5                    7                 6
Maui_01                              0                    0                 0
                Scotorythra arboricolans Spolas delicata
HawaiiIsland_01                        0               0
Kauai_01                              10               2
Maui_01                                0               0
                Syntormon distortitarsis Tempyra biguttula
HawaiiIsland_01                        0                 0
Kauai_01                              12                17
Maui_01                                3                 0
                Tenuipalpus pacificus Thrips hawaiiensis Thyrocopa brevipalpis
HawaiiIsland_01                     0                  0                     0
Kauai_01                           15                  8                     5
Maui_01                             0                  0                     0
                Thyrocopa usitata Tinostoma smaragditis Trigonidium flectens
HawaiiIsland_01                 0                     0                    0
Kauai_01                        9                    15                   24
Maui_01                         0                     0                    0
                Trigonidium novena Trigonidium octonalis Udea calliastra
HawaiiIsland_01                  0                     0               0
Kauai_01                        10                     5               8
Maui_01                          0                     0               0
                Xyleborus perforans Abanchogastra debilis
HawaiiIsland_01                   0                     0
Kauai_01                          6                     0
Maui_01                           0                     4
                Adelencyrtus odonaspidis Aedes albopictus Agrotis tephrias
HawaiiIsland_01                        0                0                0
Kauai_01                               0                0                0
Maui_01                                2                8                4
                Amblyseius largoensis Anax strenuus Anisolabis eteronoma
HawaiiIsland_01                     0             0                    0
Kauai_01                            0             0                    0
Maui_01                            10             2                    1
                Aphidius obscuripes Araecerus fasciculatus Araecerus vieillardi
HawaiiIsland_01                   0                      0                    0
Kauai_01                          0                      0                    0
Maui_01                           5                      2                    1
                Atelothrus limbatus Athesapeuta cyperi Atheta coriaria
HawaiiIsland_01                   0                  0               0
Kauai_01                          0                  0               0
Maui_01                           3                  8               1
                Athetis thoracica Balclutha timberlakei
HawaiiIsland_01                 0                     0
Kauai_01                        0                     0
Maui_01                         2                    10
                Barypristus incendiarius Batrachedrodes supercincta
HawaiiIsland_01                        0                          0
Kauai_01                               0                          0
Maui_01                                5                          5
                Brontolaemus currax mauiensis Buenoa pallipes
HawaiiIsland_01                             0               0
Kauai_01                                    0               0
Maui_01                                     4              11
                Campsicnemus ephydrus Campsicnemus macula
HawaiiIsland_01                     0                   0
Kauai_01                            0                   0
Maui_01                             1                   5
                Campsicnemus philohydratus Canaceoides angulatus
HawaiiIsland_01                          0                     0
Kauai_01                                 0                     0
Maui_01                                  3                     5
                Carpophilus maculatus Carposina mauii Caryedon serratus
HawaiiIsland_01                     0               0                 0
Kauai_01                            0               0                 0
Maui_01                             4              12                 2
                Cephalops titanus Cheiracanthium diversum
HawaiiIsland_01                 0                       0
Kauai_01                        0                       0
Maui_01                         9                       6
                Chlorococcus straussiae Circulifer tenellus Cis immaturus
HawaiiIsland_01                       0                   0             0
Kauai_01                              0                   0             0
Maui_01                               3                   3             7
                Colpodiscus lucipetens Cremastus flavoorbitalis Cydia plicata
HawaiiIsland_01                      0                        0             0
Kauai_01                             0                        0             0
Maui_01                              3                        5             4
                Diestota occidentalis Drosophila disticha Drosophila kikkawai
HawaiiIsland_01                     0                   0                   0
Kauai_01                            0                   0                   0
Maui_01                             5                   2                  13
                Drosophila lemniscata Drosophila mediana Drosophila spectabilis
HawaiiIsland_01                     0                  0                      0
Kauai_01                            0                  0                      0
Maui_01                             2                  3                      2
                Drosophila truncipenna Drosophila virilis
HawaiiIsland_01                      0                  0
Kauai_01                             0                  0
Maui_01                              3                  4
                Dyscritomyia grimshawi Emperoptera zimmermani
HawaiiIsland_01                      0                      0
Kauai_01                             0                      0
Maui_01                              4                      4
                Entomobrya albocincta Ephydra millbrae Eupetinus laevigatus
HawaiiIsland_01                     0                0                    0
Kauai_01                            0                0                    0
Maui_01                             4               12                    2
                Eupetinus latimargo Eurynogaster spiniger
HawaiiIsland_01                   0                     0
Kauai_01                          0                     0
Maui_01                           4                     3
                Eurynogaster variabilis Forcipomyia hardyi Forcipomyia ingrami
HawaiiIsland_01                       0                  0                   0
Kauai_01                              0                  0                   0
Maui_01                               4                  5                   1
                Fornax longicornis Gymnochiromyia hawaiiensis
HawaiiIsland_01                  0                          0
Kauai_01                         0                          0
Maui_01                         11                          4
                Heterocrossa atronotata Hierodula patellifera
HawaiiIsland_01                       0                     0
Kauai_01                              0                     0
Maui_01                               2                     3
                Hofmannophila pseudospretella Hylaeus facilis
HawaiiIsland_01                             0               0
Kauai_01                                    0               0
Maui_01                                     5              15
                Hyposmocoma bitincta Hyposmocoma diffusa
HawaiiIsland_01                    0                   0
Kauai_01                           0                   0
Maui_01                            5                   3
                Hyposmocoma scepticella Leialoha mauiensis Leucania striata
HawaiiIsland_01                       0                  0                0
Kauai_01                              0                  0                0
Maui_01                               2                 13               16
                Liancalus metallicus Lispinodes mauiensis Lispocephala ascita
HawaiiIsland_01                    0                    0                   0
Kauai_01                           0                    0                   0
Maui_01                            6                    6                   3
                Lispocephala deceptiva Lispocephala pollinosa
HawaiiIsland_01                      0                      0
Kauai_01                             0                      0
Maui_01                              3                      2
                Lispocephala subseminigra Lispocephala xanthopleura
HawaiiIsland_01                         0                         0
Kauai_01                                0                         0
Maui_01                                 3                         6
                Macrosiphum solanifolii Mecaphesa hiatus Mecyclothorax rusticus
HawaiiIsland_01                       0                0                      0
Kauai_01                              0                0                      0
Maui_01                               2                2                      5
                Mecyclothorax unctus Metaphycus lounsburyi Myllaena vicina
HawaiiIsland_01                    0                     0               0
Kauai_01                           0                     0               0
Maui_01                            2                     7               1
                Nabis koelensis Neoclytarlus railliardiae
HawaiiIsland_01               0                         0
Kauai_01                      0                         0
Maui_01                       2                         1
                Neoscatella hawaiiensis Neseis ampliata Neseis haleakalae
HawaiiIsland_01                       0               0                 0
Kauai_01                              0               0                 0
Maui_01                               2               3                 2
                Nesiomiris cheliferus Nesomicromus brunnescens
HawaiiIsland_01                     0                        0
Kauai_01                            0                        0
Maui_01                             4                       12
                Nesosydne blackburni Nysius beardsleyi Nysius kinbergi
HawaiiIsland_01                    0                 0               0
Kauai_01                           0                 0               0
Maui_01                            6                 5               1
                Nysius longicollis Odynerus erythrostactes Odynerus nigripennis
HawaiiIsland_01                  0                       0                    0
Kauai_01                         0                       0                    0
Maui_01                          2                       2                    4
                Oeobia heterodoxa Oligota glabra Omiodes blackburni
HawaiiIsland_01                 0              0                  0
Kauai_01                        0              0                  0
Maui_01                         1              1                  2
                Orthezia insignis Pachodynerus nasidens Pagiopalus atomarius
HawaiiIsland_01                 0                     0                    0
Kauai_01                        0                     0                    0
Maui_01                         3                     2                    6
                Pararrhaptica perkinsana Pectinophora gossypiella
HawaiiIsland_01                        0                        0
Kauai_01                               0                        0
Maui_01                                9                        8
                Pentamerismus oregonensis Phaeogramma vittipennis
HawaiiIsland_01                         0                       0
Kauai_01                                0                       0
Maui_01                                 8                       3
                Phenacoccus madeirensis Philodoria dubauticola
HawaiiIsland_01                       0                      0
Kauai_01                              0                      0
Maui_01                               9                      9
                Phlyctaenia liopis Plagithmysus medeirosi
HawaiiIsland_01                  0                      0
Kauai_01                         0                      0
Maui_01                          1                      4
                Plagithmysus simillimus Plagithmysus smilacis
HawaiiIsland_01                       0                     0
Kauai_01                              0                     0
Maui_01                               3                     6
                Prognathogryllus kukui Proterhinus comes Proterhinus haleakalae
HawaiiIsland_01                      0                 0                      0
Kauai_01                             0                 0                      0
Maui_01                              5                 3                     16
                Proterhinus ruficollis Proterhinus tuberculiceps Psallus luteus
HawaiiIsland_01                      0                         0              0
Kauai_01                             0                         0              0
Maui_01                              5                         3              1
                Remaudiereana nigriceps Scaptomyza exigua Scaptomyza nasalis
HawaiiIsland_01                       0                 0                  0
Kauai_01                              0                 0                  0
Maui_01                               5                 3                  3
                Scaptomyza pallifrons Scatella femoralis Scotorythra brunnea
HawaiiIsland_01                     0                  0                   0
Kauai_01                            0                  0                   0
Maui_01                            10                  2                   3
                Sierola mauiensis Sigmatineurum nigrum Sinella kukae
HawaiiIsland_01                 0                    0             0
Kauai_01                        0                    0             0
Maui_01                         2                    1             2
                Stoeberhinus testaceus Tetragnatha acuta Tetragnatha hawaiensis
HawaiiIsland_01                      0                 0                      0
Kauai_01                             0                 0                      0
Maui_01                              4                 3                      3
                Tetragnatha macracantha Tetragnatha quasimodo
HawaiiIsland_01                       0                     0
Kauai_01                              0                     0
Maui_01                               4                     2
                Tetramorium tonganum Trigonidium pseudokua Trigonidium wahoi
HawaiiIsland_01                    0                     0                 0
Kauai_01                           0                     0                 0
Maui_01                            2                     2                 6
                Udea heterodoxa Vanessa tameamea Vespula pensylvanica
HawaiiIsland_01               0                0                    0
Kauai_01                      0                0                    0
Maui_01                       7                1                   10
                Xyleborinus saxeseni
HawaiiIsland_01                    0
Kauai_01                           0
Maui_01                            1</code></pre>
</div>
<!-- ```{r or base r} -->
<!-- x <- aggregate(abundances[, 'final_name', drop = FALSE], abundances[, c('final_name', 'site')],  -->
<!--                length) -->
<!-- head(x) #is identical -->
<!-- ``` -->
</div>
</div>
<div class="section level1">
<h1 id="visualization-via-the-species-abundance-distribution">Visualization via the species abundance distribution<a class="anchor" aria-label="anchor" href="#visualization-via-the-species-abundance-distribution"></a>
</h1>
<div class="section level2">
<h2 id="separating-the-islands">Separating the islands<a class="anchor" aria-label="anchor" href="#separating-the-islands"></a>
</h2>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kauai_abund</span> <span class="op">&lt;-</span> <span class="va">abundance_tallies</span><span class="op">[</span> <span class="va">abundance_tallies</span><span class="op">$</span><span class="va">island</span> <span class="op">==</span> <span class="st">"Kauai"</span>, <span class="op">]</span></span>
<span><span class="va">maui_abund</span> <span class="op">&lt;-</span> <span class="va">abundance_tallies</span><span class="op">[</span> <span class="va">abundance_tallies</span><span class="op">$</span><span class="va">island</span> <span class="op">==</span> <span class="st">"Maui"</span>, <span class="op">]</span></span>
<span><span class="va">hawaiiisland_abund</span> <span class="op">&lt;-</span> <span class="va">abundance_tallies</span><span class="op">[</span> <span class="va">abundance_tallies</span><span class="op">$</span><span class="va">island</span> <span class="op">==</span> <span class="st">"HawaiiIsland"</span>, <span class="op">]</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="plotting-sads">Plotting SADs<a class="anchor" aria-label="anchor" href="#plotting-sads"></a>
</h2>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">hawaiiisland_abund</span><span class="op">$</span><span class="va">abundance</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/abundance-data-rendered-hawaii%20sad%20plot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>Plot the rank-abundance distributions for the remaining islands.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">kauai_abund</span><span class="op">$</span><span class="va">abundance</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/abundance-data-rendered-kauai%20sad%20plot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">maui_abund</span><span class="op">$</span><span class="va">abundance</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/abundance-data-rendered-maui%20sad%20plot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h1>
<div class="section level2">
<h2 id="diversity-indices">Diversity indices<a class="anchor" aria-label="anchor" href="#diversity-indices"></a>
</h2>
<p>The <code>vegan</code> package is a widely-used toolset for
calculating diversity indices on ecological abundance data.</p>
<div class="section level3">
<h3 id="species-richness">Species richness<a class="anchor" aria-label="anchor" href="#species-richness"></a>
</h3>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">richness</span> <span class="op">&lt;-</span> <span class="fu">specnumber</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span></span>
<span></span>
<span><span class="va">richness</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>HawaiiIsland_01        Kauai_01         Maui_01 
             59              94             136 </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="shannons-index">Shannon’s index<a class="anchor" aria-label="anchor" href="#shannons-index"></a>
</h3>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shannon</span> <span class="op">&lt;-</span> <span class="fu">diversity</span><span class="op">(</span><span class="va">abundance_wide</span>, index <span class="op">=</span> <span class="st">"shannon"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">shannon</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>HawaiiIsland_01        Kauai_01         Maui_01 
       3.600169        4.361321        4.677393 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp_shannon</span> <span class="op">&lt;-</span> <span class="fu">exp</span><span class="op">(</span><span class="va">shannon</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">exp_shannon</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>HawaiiIsland_01        Kauai_01         Maui_01 
       36.60441        78.36057       107.48945 </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="simpsons-evenness">Simpson’s evenness<a class="anchor" aria-label="anchor" href="#simpsons-evenness"></a>
</h3>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simpson</span> <span class="op">&lt;-</span> <span class="fu">diversity</span><span class="op">(</span><span class="va">abundance_wide</span>, index <span class="op">=</span> <span class="st">"simpson"</span><span class="op">)</span></span>
<span><span class="va">simpson</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>HawaiiIsland_01        Kauai_01         Maui_01 
      0.9610713       0.9853324       0.9886889 </code></pre>
</div>
<div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion3"></a>
</h3>
<div class="callout-content">
<p>Compute the <em>inverse</em> Simpson index.</p>
<p>(Hint: use <code>?diversity</code> to get more information about the
<code>diversity</code> function).</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">invsimpson</span> <span class="op">&lt;-</span> <span class="fu">diversity</span><span class="op">(</span><span class="va">abundance_wide</span>, index <span class="op">=</span> <span class="st">"invsimpson"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">invsimpson</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>HawaiiIsland_01        Kauai_01         Maui_01 
       25.68800        68.17769        88.40846 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="combining-values-into-a-dataframe">Combining values into a dataframe<a class="anchor" aria-label="anchor" href="#combining-values-into-a-dataframe"></a>
</h3>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diversity_metrics</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">exp_shannon</span>, </span>
<span>                           <span class="va">invsimpson</span>,</span>
<span>                           <span class="va">richness</span><span class="op">)</span></span>
<span></span>
<span><span class="va">diversity_metrics</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            HawaiiIsland_01 Kauai_01   Maui_01
exp_shannon        36.60441 78.36057 107.48945
invsimpson         25.68800 68.17769  88.40846
richness           59.00000 94.00000 136.00000</code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="hill-numbers">Hill numbers<a class="anchor" aria-label="anchor" href="#hill-numbers"></a>
</h2>
<div id="discussion4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion4"></a>
</h3>
<div class="callout-content">
<p>Hill numbers are a unifying framework.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="calculating-hill-numbers">Calculating Hill numbers<a class="anchor" aria-label="anchor" href="#calculating-hill-numbers"></a>
</h3>
<p>The <code>hillR</code> package allows us to calculate Hill numbers in
much the same way we calculate diversity indices in
<code>vegan</code>.</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_0</span> <span class="op">&lt;-</span> <span class="fu">hill_taxa</span><span class="op">(</span><span class="va">abundance_wide</span>, q <span class="op">=</span> <span class="fl">0</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">hill_0</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>HawaiiIsland_01        Kauai_01         Maui_01 
             59              94             136 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_1</span> <span class="op">&lt;-</span> <span class="fu">hill_taxa</span><span class="op">(</span><span class="va">abundance_wide</span>, q <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hill_1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>HawaiiIsland_01        Kauai_01         Maui_01 
       36.60441        78.36057       107.48945 </code></pre>
</div>
<div id="discussion5" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion5"></a>
</h3>
<div class="callout-content">
<p>Calculate the hill numbers for q = 2.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_2</span> <span class="op">&lt;-</span> <span class="fu">hill_taxa</span><span class="op">(</span><span class="va">abundance_wide</span>, q <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hill_2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>HawaiiIsland_01        Kauai_01         Maui_01 
       25.68800        68.17769        88.40846 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="combining-and-plotting-hill-numbers">Combining and plotting Hill numbers<a class="anchor" aria-label="anchor" href="#combining-and-plotting-hill-numbers"></a>
</h3>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_numbers</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">hill_0</span>, <span class="va">hill_1</span>, <span class="va">hill_2</span><span class="op">)</span></span>
<span><span class="va">hill_numbers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       HawaiiIsland_01 Kauai_01   Maui_01
hill_0        59.00000 94.00000 136.00000
hill_1        36.60441 78.36057 107.48945
hill_2        25.68800 68.17769  88.40846</code></pre>
</div>
<div id="discussion6" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion6"></a>
</h3>
<div class="callout-content">
<p>Compare the values for the Hill numbers to the values we generated
earlier for richness, Shannon’s index, and the inverse Simpson.</p>
<p>The following may help:</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diversity_metrics</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            HawaiiIsland_01 Kauai_01   Maui_01
exp_shannon        36.60441 78.36057 107.48945
invsimpson         25.68800 68.17769  88.40846
richness           59.00000 94.00000 136.00000</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_numbers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       HawaiiIsland_01 Kauai_01   Maui_01
hill_0        59.00000 94.00000 136.00000
hill_1        36.60441 78.36057 107.48945
hill_2        25.68800 68.17769  88.40846</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="relating-hill-numbers-to-patterns-in-diversity">Relating Hill numbers to patterns in diversity<a class="anchor" aria-label="anchor" href="#relating-hill-numbers-to-patterns-in-diversity"></a>
</h2>
<p>Let’s revisit the SAD plots we generated before, and think about
these in terms of Hill numbers.</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">hawaiiisland_abund</span><span class="op">$</span><span class="va">abundance</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">kauai_abund</span><span class="op">$</span><span class="va">abundance</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">maui_abund</span><span class="op">$</span><span class="va">abundance</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/abundance-data-rendered-sad%20again%20plots-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_numbers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       HawaiiIsland_01 Kauai_01   Maui_01
hill_0        59.00000 94.00000 136.00000
hill_1        36.60441 78.36057 107.48945
hill_2        25.68800 68.17769  88.40846</code></pre>
</div>
</div>
</div>
<div class="section level1">
<h1 id="recap">Recap<a class="anchor" aria-label="anchor" href="#recap"></a>
</h1>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Organismal abundance data are a fundamental data type for population
and community ecology.</li>
<li>The <code>taxise</code> package can help with data cleaning, but
quality checks are often ultimately dataset-specific.</li>
<li>The species abundance distribution (SAD) summarizes site-specific
abundance information to facilitate cross-site or over-time
comparisons.</li>
<li>We can quantify the shape of the SAD using <em>summary
statistics</em>. Specifically, <em>Hill numbers</em> provide a unified
framework for describing the diversity of a system.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section><section id="aio-traits-data"><p>Content from <a href="traits-data.html">Trait data</a></p>
<hr>
<p> Last updated on 2023-06-05 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/traits-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you read in and clean trait data?</li>
<li>What visualization techniques help you perform quality control and
understand the distribution of your data?</li>
<li>What do Hill numbers convey in the context of trait data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants should be able to…</p>
<ol style="list-style-type: decimal">
<li>Import trait data in a CSV format into the R environment</li>
<li>Clean taxonomic names using the <code>taxize</code> package</li>
<li>Aggregate traits</li>
<li>Visualize trait distributions</li>
<li>Calculate Hill numbers</li>
<li>Interpret Hill numbers using ranked trait plots</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<p>We will work now with trait (i.e., phenotypic) data for communities
in the Hawaiian islands. Similarly to abundance, investigating trait
data can give insights into the ecological processes shaping different
communities. Specifically, it would be interesting to answer questions
like: 1) how much does a specific trait vary in a community?; 2) Are all
species converging into a similar trait value or dispersing into a wide
range of values for the trait?; 3) do those patterns of variation change
across different communities? Answering these questions can help us
understand how species are interacting with the environment and/or with
each other.</p>
<p>Here, we will work with a dataset containing values of <strong>body
mass</strong> for the species in the Hawaiian islands. Similar to the
abundance data, each row represents one specimen: in this dataset, each
row contains the body mass measured for that specimen. Our overall goal
here is to clean this data and attach it to the species data, so we can
investigate trait patterns per community.</p>
<p>We’re starting with a challenge!</p>
<div id="import-and-clean-trait-data" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="import-and-clean-trait-data" class="callout-inner">
<h3 class="callout-title">Import and clean trait data<a class="anchor" aria-label="anchor" href="#import-and-clean-trait-data"></a>
</h3>
<div class="callout-content">
<p>In the abundance-data episode, you learned how to read in a
<code>CSV</code> file with the <code>read.csv()</code> function. In
addition, you learned how to clean and standardize the taxonomic
information in your data set using the <code>gnr_resolve()</code>
function from the <code>taxize</code> package.</p>
<p>You’re going to use those skills to import and clean the trait data
set. The trait data set contains two columns: <code>GenSp</code>, which
contains the binomial species names, and <code>mass_g</code>, which
contains the individual’s mass in grams, the metric of body size chosen
for the study. The data also contain typos and a problem with the
taxonomy that you must correct and investigate. Make sure to glance at
the data before cleaning!</p>
<p>Your trait data is located at: <a href="http://bit.ly/41GsgvE" class="external-link uri">http://bit.ly/41GsgvE</a></p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Read in the traits data using the <code>read.csv</code> function and
supplying it with the path to the data.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">'http://bit.ly/41GsgvE'</span><span class="op">)</span></span></code></pre>
</div>
<p>Check the names of the traits using the <code>gnr_resolve()</code>
function. To streamline your efforts, supply only the unique species
names in your data to <code>gnr_resolve()</code>.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="co"># only need to check the unique names</span></span>
<span><span class="va">species_list</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="va">traits</span><span class="op">$</span><span class="va">GenSp</span><span class="op">)</span></span>
<span></span>
<span><span class="va">name_resolve</span> <span class="op">&lt;-</span> <span class="fu">gnr_resolve</span><span class="op">(</span><span class="va">species_list</span>, best_match_only <span class="op">=</span> <span class="cn">TRUE</span>, canonical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">name_resolve</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 5
  user_supplied_name        submitted_name data_source_title score matched_name2
  &lt;chr&gt;                     &lt;chr&gt;          &lt;chr&gt;             &lt;dbl&gt; &lt;chr&gt;        
1 Xyleborus perforans       Xyleborus per… National Center … 0.988 Xyleborus pe…
2 Agrotis chersotoides      Agrotis chers… Catalogue of Lif… 0.75  Agrotis      
3 Oligotoma saundersii      Oligotoma sau… Wikispecies       0.988 Oligotoma sa…
4 Drosophila humeralis      Drosophila hu… Wikispecies       0.988 Drosophila h…
5 Entomobrya griseoolivata  Entomobrya gr… National Center … 0.988 Entomobrya g…
6 Pseudosinella kalalauens… Pseudosinella… Encyclopedia of … 0.988 Pseudosinell…</code></pre>
</div>
<p>To quickly see which taxa are in conflict with <code>taxize</code>’s,
use bracket subsetting and boolean matching.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mismatches_traits</span> <span class="op">&lt;-</span> <span class="va">name_resolve</span><span class="op">[</span><span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span> <span class="op">!=</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"user_supplied_name"</span>, <span class="st">"matched_name2"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">mismatches_traits</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 5 × 2
  user_supplied_name     matched_name2       
  &lt;chr&gt;                  &lt;chr&gt;               
1 Agrotis chersotoides   Agrotis             
2 Trigonidium flelectens Trigonidium flectens
3 Schrankia simplllex    Schrankia simplex   
4 Holcobius insignigis   Holcobius insignis  
5 Carposina mauiii       Carposina mauii     </code></pre>
</div>
<p>Fixing the names comes in two steps. First, we join the
<code>traits</code> dataframe with the <code>name_resolve</code>
dataframe using the <code>left_join()</code> function. Note, we indicate
that the <code>GenSp</code> and <code>user_supplied_name</code> columns
have the same information by supplying a named vector to the
<code>by =</code> argument.</p>

<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">traits</span>, <span class="va">name_resolve</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"user_supplied_name"</span>, <span class="st">"matched_name2"</span><span class="op">)</span><span class="op">]</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"GenSp"</span> <span class="op">=</span> <span class="st">"user_supplied_name"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                 GenSp   mass_g       matched_name2
1  Xyleborus perforans 10.13036 Xyleborus perforans
2  Xyleborus perforans 10.05684 Xyleborus perforans
3  Xyleborus perforans 10.14838 Xyleborus perforans
4  Xyleborus perforans 10.20234 Xyleborus perforans
5  Xyleborus perforans 10.14719 Xyleborus perforans
6 Agrotis chersotoides 24.25539             Agrotis</code></pre>
</div>
<p>Then, to replace <em>Agrotis chersotoides</em> with the updated
<em>Peridroma chersotoides</em>, which you discovered in abundance-data
episode, use bracketed indexing, boolean matching, and assignment.</p>
<p>In addition, although not necessary, changing the column name
<code>matched_name2</code> to <code>final_name</code> to give it a more
sensible name for later use is good practice.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span><span class="op">$</span><span class="va">matched_name2</span><span class="op">[</span><span class="va">traits</span><span class="op">$</span><span class="va">matched_name2</span> <span class="op">==</span> <span class="st">"Agrotis"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Peridroma chersotoides"</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span><span class="op">[</span><span class="fu">colnames</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span> <span class="op">==</span> <span class="st">"matched_name2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"final_name"</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                 GenSp   mass_g             final_name
1  Xyleborus perforans 10.13036    Xyleborus perforans
2  Xyleborus perforans 10.05684    Xyleborus perforans
3  Xyleborus perforans 10.14838    Xyleborus perforans
4  Xyleborus perforans 10.20234    Xyleborus perforans
5  Xyleborus perforans 10.14719    Xyleborus perforans
6 Agrotis chersotoides 24.25539 Peridroma chersotoides</code></pre>
</div>

<p>&lt;&lt;&lt;&lt;&lt;&lt;&lt; HEAD # Aggregation</p>
<div class="section level1">
<h1 id="when-analyzing-trait-data-it-often-needs-to-be-summarized-at-a-higher-level-than-the-individual--for-instance-many-community-assembly">When analyzing trait data, it often needs to be summarized at a
higher level than the individual. For instance, many community
assembly<a class="anchor" aria-label="anchor" href="#when-analyzing-trait-data-it-often-needs-to-be-summarized-at-a-higher-level-than-the-individual--for-instance-many-community-assembly"></a>
</h1>
</div>
</div>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">SH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode sh" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregation</span></span></code></pre>
</div>
<p>When analyzing trait data, it often needs to be summarized at a
higher level than the individual. For instance, many community assembly
&gt;&gt;&gt;&gt;&gt;&gt;&gt; c86449c88ae44948adde76c7c0a39200bae38e26
analyses require species-level summaries, rather than individual
measurements. So, we often want to calculate summary statistics of
traits for each species. For numeric measurements, body size, statistics
like the mean, median, and standard deviation give information about the
center and spread of the distribution of traits for the species. For
this section, you will aggregate the trait data to the species level,
calculating the mean, median, and mode of body size for each
species.</p>
<p>While there are methods to aggregate data in base R, the <a href="https://dplyr.tidyverse.org/" class="external-link">dplyr</a> makes this task and other
data wrangling tasks much more intuitive. The function
<code>group_by()</code> groups your data frame by a variable. The first
argument to <code>group_by()</code> is your data frame, and all
following unnamed arguments are the variables you want to group the data
by. In our case, we want to group by species name, so we’ll supply
<code>final_name</code>.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># group the data frame by species</span></span>
<span><span class="va">traits_sumstats</span> <span class="op">&lt;-</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">traits</span>, <span class="va">final_name</span><span class="op">)</span></span></code></pre>
</div>
<p><code>group_by()</code> just adds an index that tells any following
<code>dplyr</code> functions to perform their calculations on the group
the data frame was indexed by. So, to perform the actual calculations,
we will use the <code>summarize()</code> function
(<code>summarise()</code> works too for non-Americans). The first
argument you supply is the data frame, following by the calculations you
want to perform on the grouped data and what you want to name the
resulting variables. The structure is
<code>new_var_name = function(original_var_name)</code> for each
calculation. Here, we’re using the <code>mean()</code>,
<code>median()</code>, and <code>sd()</code> functions. Then, we’ll take
a look at the new data set with the <code>head()</code> function.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># summarize the grouped data frame, so you're calculating the summary statistics for each species</span></span>
<span><span class="va">traits_sumstats</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">summarize</span><span class="op">(</span></span>
<span>        <span class="va">traits_sumstats</span>,</span>
<span>        mean_mass_g <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">mass_g</span><span class="op">)</span>,</span>
<span>        median_mass_g <span class="op">=</span> <span class="fu">median</span><span class="op">(</span><span class="va">mass_g</span><span class="op">)</span>,</span>
<span>        sd_mass_g <span class="op">=</span> <span class="fu">sd</span><span class="op">(</span><span class="va">mass_g</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_sumstats</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 4
  final_name               mean_mass_g median_mass_g sd_mass_g
  &lt;chr&gt;                          &lt;dbl&gt;         &lt;dbl&gt;     &lt;dbl&gt;
1 Abanchogastra debilis           6.65          6.64    0.0890
2 Acanthia procellaris           12.9          12.8     0.0726
3 Actia eucosmae                 10.8          10.9     0.0945
4 Adelencyrtus odonaspidis        5.99          6.00    0.0411
5 Aedes albopictus               10.9          10.9     0.0548
6 Aedes nocturnus                 6.84          6.83    0.0331</code></pre>
</div>
<p>Finally, you need to add the aggregated species-level information
back to the abundance data, so you have the summary statistics for each
species at each site.</p>
<p>First, read in the abundance tallies dataframe you made in the
abundance episode.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundance_tallies</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"http://bit.ly/3L2OFxk"</span><span class="op">)</span></span></code></pre>
</div>
<p>To join the data, you’ll use <code>left_join()</code>, which is
hopefully getting familiar to you now!</p>
<p><code>abundance_tallies</code> should be on the left side, since we
want the species-information for each species in each community.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_sumstats</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">abundance_tallies</span>, <span class="va">traits_sumstats</span>, by <span class="op">=</span> <span class="st">"final_name"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_sumstats</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             site       island                         final_name abundance
1 HawaiiIsland_01 HawaiiIsland               Acanthia procellaris         3
2 HawaiiIsland_01 HawaiiIsland                     Actia eucosmae         4
3 HawaiiIsland_01 HawaiiIsland Anomalochrysa fulvescens rhododora         2
4 HawaiiIsland_01 HawaiiIsland             Anoplolepis gracilipes         5
5 HawaiiIsland_01 HawaiiIsland                   Anurophorus lohi         2
6 HawaiiIsland_01 HawaiiIsland                    Aphis citricola         3
  mean_mass_g median_mass_g  sd_mass_g
1   12.875661     12.840191 0.07255218
2   10.836548     10.864682 0.09449608
3    2.860058      2.855061 0.04633525
4   12.366928     12.358811 0.01926359
5    6.251880      6.234964 0.07793529
6    3.822070      3.819887 0.12376448</code></pre>
</div>
<div class="section level1">
<h1 id="visualize-trait-distributions">Visualize trait distributions<a class="anchor" aria-label="anchor" href="#visualize-trait-distributions"></a>
</h1>
<p>Histograms and density plots can help give you a quick look at the
distribution of your data. For quality control purposes, they are useful
to see if there are any suspicious values due to human error. Let’s
overlay the histogram and density plots to get an idea of what
aggregated (histogram) and smoothed (density) representations of the
data look like.</p>
<p>The first argument to the <code>hist()</code> function is your data.
<code>breaks</code> sets the number of histogram bars, which can be
tuned heuristically to find a number of bars that best represents the
distribution of your data. I found that <code>breaks = 40</code> is a
reasonable value. <code>xlab</code>, <code>ylab</code>, and
<code>main</code> are all used to specify labels for the x-axis, y-axis,
and title, respectively. When scaling the plot, sometimes base R
plotting functions aren’t as responsive to the data as we like. To fix
this, you can use the <code>xlim</code> or <code>ylim</code> functions
to set the value range of the x-axis and y-axis, respectively. Here, we
set <code>ylim</code> to a vector <code>c(0, 0.11)</code>, which
specifies the y-axis to range from 0 to 0.11.</p>
<p>If you’re used to seeing histograms, the y-axis may look unfamiliar.
To get the two plots to use the same scale, you set
<code>freq = FALSE</code>, which converts the histogram counts to a
density value. The density scales the distribution from 0-1, so the bar
height (and line height on the density plot) is a fraction of the total
distribution.</p>
<p>Finally, to add the density line to the plot, you calculate the
density of the trait distribution with the <code>density()</code>
function and wrap it with the <code>lines()</code> function.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span></span>
<span>    <span class="va">traits_sumstats</span><span class="op">$</span><span class="va">mean_mass_g</span>,</span>
<span>    breaks <span class="op">=</span> <span class="fl">40</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Average mass (g)"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"All species"</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.11</span><span class="op">)</span>,</span>
<span>    freq <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="fu">density</span><span class="op">(</span><span class="va">traits_sumstats</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/traits-data-rendered-density-all-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In addition to quality control, knowing the distribution of trait
data lends itself towards questions of what processes are shaping the
communities under study. Are the traits overdispersed? Underdispersed?
Does the average vary across communities? To get at these questions, you
need to plot each site and compare among sites. To facilitate
comparison, you’re going to layer the distributions on top of each other
in a single plot. Histograms get cluttered quickly, so let’s use density
plots for this task.</p>
<p>To plot each site, you need a separate dataframe for each site. The
easiest way to do this is with the <code>split()</code> function, which
takes a dataframe followed by a variable to split the dataframe by. It
then returns a list of data frames for each element in the variable.</p>
<p>Running <code>head()</code> on the list returns a lot of output, so
you can use the <code>names()</code> function to make sure the names of
the dataframes in your list are the sites you intended to split by.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># split by site</span></span>
<span><span class="va">traits_sumstats_split</span> <span class="op">&lt;-</span> <span class="fu">split</span><span class="op">(</span><span class="va">traits_sumstats</span>, <span class="va">traits_sumstats</span><span class="op">$</span><span class="va">site</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "HawaiiIsland_01" "Kauai_01"        "Maui_01"        </code></pre>
</div>
<p>To plot the data, you first need to initialize the plot with the
<code>plot()</code> function of a single site’s density. Here is also
where you add the plot aesthetics (labels, colors, line characteristics,
etc.). You’re labeling the title and axis labels with <code>main</code>,
<code>xlab</code>, and <code>ylab</code>. The <code>lwd</code>
(“linewidth”) function specifies the size of the line, and I found a
value of two is reasonable. You then supply the color of the line
(<code>col</code>) with a color name or hex code. I chose a hex
representing a purple color from the <a href="https://sjmgarnier.github.io/viridisLite/index.html" class="external-link">viridis</a>
color palette. Finally, you need to specify the x-axis limits and y-axis
limits with <code>xlim</code> and <code>ylim</code>. While a negative
mass doesn’t make biological sense, it is an artifact of the smoothing
done by the <code>density</code> function.</p>
<p>To add density lines to the plot, you use the <code>lines()</code>
function with the data for the desired site as the first argument. You
also need to specify the line-specific aesthetics for each line, which
are the <code>lwd</code> and <code>col</code>. I chose the green and
yellow colors from the <em>viridis</em> palette, but you can pick
whatever your heart desires!</p>
<p>Finally, you need a legend for the plot. Within the
<code>legend()</code> function, you specify the position of the legend
(“topright”), the legend labels, the line width, and colors to assign to
the labels.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span></span>
<span>    <span class="fu">density</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">HawaiiIsland_01</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Average mass per study site"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Average mass (g)"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#440154FF"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">40</span><span class="op">)</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.11</span><span class="op">)</span></span>
<span>    </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span></span>
<span>    <span class="fu">density</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">Kauai_01</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#21908CFF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span></span>
<span>    <span class="fu">density</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">Maui_01</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#FDE725FF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">legend</span><span class="op">(</span></span>
<span>    <span class="st">"topright"</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Hawaii 01"</span>, <span class="st">"Kauai 01"</span>, <span class="st">"Maui 01"</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#FDE725FF"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/traits-data-rendered-density-site-plot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>It looks like Kauai insects have a higher average mass and more
dispersed distribution of masses than Hawaii and Maui! Next, we’ll take
a look at using Hill numbers to summarize the distribution of
traits.</p>
</div>
<div class="section level1">
<h1 id="hill-numbers">Hill numbers<a class="anchor" aria-label="anchor" href="#hill-numbers"></a>
</h1>
<p>Hill numbers are a useful and informative summary of trait diversity,
in addition to other biodiversity variables (<a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Gaggiotti
et al. 2018</a>). They contain signatures of the processes underlying
community assembly. To calculate Hill numbers for traits, you’re going
to use the <code>hill_func()</code> function from the <code>hillR</code>
package. It requires two inputs- the site by species abundance dataframe
and a traits dataframe that has species as rows and traits as
columns.</p>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>Your first challenge, if you don’t have the site by species abundance
dataframe, is to recreate it here. Remember, you’ll need to use the
<code>pivot_wider()</code> function and the
<code>abundance_tallies</code> data!</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">tidyr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="fu">pivot_wider</span><span class="op">(</span></span>
<span>    <span class="va">abundance_tallies</span>,</span>
<span>    id_cols <span class="op">=</span> <span class="va">site</span>,</span>
<span>    names_from <span class="op">=</span> <span class="va">final_name</span>,</span>
<span>    values_from <span class="op">=</span> <span class="va">abundance</span>,</span>
<span>    values_fill <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># tibbles don't like row names</span></span>
<span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">row.names</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span><span class="op">$</span><span class="va">site</span></span>
<span></span>
<span><span class="co"># remove the site column</span></span>
<span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<pre><code>Now, to create the traits dataframe you first need to remove the `site`,</code></pre>
<p><code>island</code>, <code>abundance</code>,
<code>median_mass_g</code>, and <code>sd_mass_g</code> columns.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_simple</span> <span class="op">&lt;-</span> <span class="va">traits_sumstats</span><span class="op">[</span>, <span class="op">-</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                          final_name mean_mass_g
1               Acanthia procellaris   12.875661
2                     Actia eucosmae   10.836548
3 Anomalochrysa fulvescens rhododora    2.860058
4             Anoplolepis gracilipes   12.366928
5                   Anurophorus lohi    6.251880
6                    Aphis citricola    3.822070</code></pre>
</div>
<p>Next, you need to filter for unique species in the dataframe.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_simple</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                          final_name mean_mass_g
1               Acanthia procellaris   12.875661
2                     Actia eucosmae   10.836548
3 Anomalochrysa fulvescens rhododora    2.860058
4             Anoplolepis gracilipes   12.366928
5                   Anurophorus lohi    6.251880
6                    Aphis citricola    3.822070</code></pre>
</div>
<p>Finally, you need to set the species names to be row names and remove
the <code>final_name</code> column. Note, you have to use the
<code>drop=F</code> argument because R has the funny behavior that it
likes to convert single-column dataframes into a vector.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">row.names</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">traits_simple</span><span class="op">$</span><span class="va">final_name</span></span>
<span></span>
<span><span class="va">traits_simple</span> <span class="op">&lt;-</span> <span class="va">traits_simple</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span>, drop<span class="op">=</span><span class="cn">F</span><span class="op">]</span></span></code></pre>
</div>
<p>Next, you’ll use the <code>hill_func()</code> function from the
<code>hillR</code> package to calculate Hill numbers 0-2 of body size
across sites.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span></span>
<span><span class="va">traits_hill_0</span> <span class="op">&lt;-</span> <span class="fu">hill_func</span><span class="op">(</span>comm <span class="op">=</span> <span class="va">abundance_wide</span>, traits <span class="op">=</span> <span class="va">traits_simple</span>, q <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="va">traits_hill_1</span> <span class="op">&lt;-</span> <span class="fu">hill_func</span><span class="op">(</span>comm <span class="op">=</span> <span class="va">abundance_wide</span>, traits <span class="op">=</span> <span class="va">traits_simple</span>, q <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">traits_hill_2</span> <span class="op">&lt;-</span> <span class="fu">hill_func</span><span class="op">(</span>comm <span class="op">=</span> <span class="va">abundance_wide</span>, traits <span class="op">=</span> <span class="va">traits_simple</span>, q <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<p>The output of <code>hill_func()</code> returns quite a few Hill
number options, which are defined in <a href="https://www.annualreviews.org/doi/abs/10.1146/annurev-ecolsys-120213-091540" class="external-link">Chao
et al. 2014</a>. For simplicity’s sake, we will look at
<code>D_q</code>, which from the documentation is the:</p>
<pre><code>&gt; functional Hill number, the effective number of equally abundant and</code></pre>
<blockquote>
<p>functionally equally distinct species</p>
</blockquote>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_hill_1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     HawaiiIsland_01     Kauai_01      Maui_01
Q           4.288057     7.313327     4.807929
FDis        3.061469     5.136366     3.602256
D_q        38.330994    78.854310   103.950856
MD_q      164.365495   576.687383   499.788382
FD_q     6300.292831 45474.285521 51953.430115</code></pre>
</div>
<p>To gain an intuition for what this means, let’s plot our data. First,
you will wrangle the output into a single dataframe to work with. Since
Hill q = 0 is species richness, let’s focus on Hill q = 1 and Hill q =
2.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_hill</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span><span class="va">traits_hill_1</span><span class="op">[</span><span class="fl">3</span>,<span class="op">]</span>, <span class="va">traits_hill_2</span><span class="op">[</span><span class="fl">3</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">traits_hill</span> <span class="op">&lt;-</span> <span class="fu">do.call</span><span class="op">(</span><span class="va">cbind</span>, <span class="va">traits_hill</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">traits_hill</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"q1"</span>, <span class="st">"q2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert to a dataframe</span></span>
<span><span class="va">traits_hill</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">traits_hill</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># I don't like rownames for plotting, so making the rownames a column</span></span>
<span><span class="va">traits_hill</span><span class="op">$</span><span class="va">site</span> <span class="op">&lt;-</span> <span class="fu">row.names</span><span class="op">(</span><span class="va">traits_hill</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">row.names</span><span class="op">(</span><span class="va">traits_hill</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre>
</div>
<p>Let’s look at how Hill q = 1 compare across sites.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="fu">as.factor</span><span class="op">(</span><span class="va">traits_hill</span><span class="op">$</span><span class="va">site</span><span class="op">)</span>, <span class="va">traits_hill</span><span class="op">$</span><span class="va">q1</span>, ylab <span class="op">=</span> <span class="st">"Hill q = 1"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/traits-data-rendered-hill-plot-raw-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Hill q = 1 is smallest on Hawaii Island, largest on Maui, and in the
middle on Kauai. But what does this mean? To gain an intuition for what
a higher or lower trait Hill number means, you will make rank plots of
the trait data, to make a plot analogous to the species abundance
distribution.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">HawaiiIsland_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Average mass per study site"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Rank"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Average mass (mg)"</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#440154FF"</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">40</span><span class="op">)</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">135</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">Kauai_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#31688EFF"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">Maui_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#35B779FF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">legend</span><span class="op">(</span></span>
<span>    <span class="st">"topright"</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Hawaii 01"</span>, <span class="st">"Kauai 01"</span>, <span class="st">"Maui 01"</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#31688EFF"</span>, <span class="st">"#35B779FF"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/traits-data-rendered-hill-plot-rank-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>You can see clearly that the diversity differs across islands, where
Maui has many more species than the other islands and a seemingly more
even distribution than the others. However, species richness strongly
influences the calculation of a Hill number. To factor out the influence
of the number of species from the Hill number and get a stronger
indicator of “evenness”, you can divide the Hill number by the number of
species, which you will do for Hill q = 1.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp_counts</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">c</span><span class="op">(</span></span>
<span>        <span class="fu">nrow</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">HawaiiIsland_01</span><span class="op">)</span>,</span>
<span>        <span class="fu">nrow</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">Kauai_01</span><span class="op">)</span>,</span>
<span>        <span class="fu">nrow</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">Maui_01</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">traits_hill</span><span class="op">$</span><span class="va">q1</span> <span class="op">/</span> <span class="va">sp_counts</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.6496779 0.8388756 0.7643445</code></pre>
</div>
<p>Interestingly, we see that Kauai actually has a more even
distribution of trait values than Maui when accounting for species
richness!</p>
<p>What are some other interpretations from the plots, summary
statistics, and what you know fr</p>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section><section id="aio-popgen-data"><p>Content from <a href="popgen-data.html">Working with population genetic data</a></p>
<hr>
<p> Last updated on 2023-06-05 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/popgen-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<section id="pop-gen-data"><h2 class="section-heading">Pop gen data<a class="anchor" aria-label="anchor" href="#pop-gen-data"></a>
</h2>
<hr class="half-width">
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is genetic diversity?</li>
<li>What does genetic diversity say about community assembly?</li>
<li>What are common summaries of genetic diversity within species and
communities?</li>
<li>What are common sequence file formats?</li>
<li>How do you read in and manipulate sequence data?</li>
<li>How do you visualize and summarize sequence data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal">
<li>Understand what genetic diversity is</li>
<li>Visualize genetic diversity distributions</li>
<li>Connect genetic diversity distributions to Hill numbers</li>
<li>Identify key features of different sequence data formats</li>
<li>Import population genetic data in multiple formats into the R
environment</li>
<li>Manipulate pop gen data by adding and removing sequences</li>
<li>Calculate genetic diversity <span class="math inline">\(\pi\)</span>
</li>
<li>Calculate Hill numbers</li>
<li>Interpret Hill numbers</li>
</ol>
</div>
</div>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<div class="section level2">
<h2 id="what-is-genetic-diversity">What is genetic diversity?<a class="anchor" aria-label="anchor" href="#what-is-genetic-diversity"></a>
</h2>
<p>GEOBON, in their description of Essential Biodiversity Variables, <a href="https://geobon.org/ebvs/what-are-ebvs/" class="external-link">define intraspecific
genetic diversity</a> as:</p>
<blockquote>
<p>The variation in DNA sequences among individuals of the same
species.</p>
</blockquote>
<p>This is a broad definition…</p>
</div>
<div class="section level2">
<h2 id="what-does-it-say-about-community-assembly">What does it say about community assembly?<a class="anchor" aria-label="anchor" href="#what-does-it-say-about-community-assembly"></a>
</h2>
<p>Genetic diversity is impacted by demographic events and natural
selection that shape the composition of the local community…</p>
</div>
</div>
<div class="section level1">
<h1 id="visualizing-genetic-diversity">Visualizing genetic diversity<a class="anchor" aria-label="anchor" href="#visualizing-genetic-diversity"></a>
</h1>
<div class="section level2">
<h2 id="within-species">Within-species<a class="anchor" aria-label="anchor" href="#within-species"></a>
</h2>
<div class="section level3">
<h3 id="sfs-visualization">SFS visualization<a class="anchor" aria-label="anchor" href="#sfs-visualization"></a>
</h3>
<p>The site frequency spectrum (SFS) is a summary of genetic variants
across a population, similar to how abundances are summarized across a
community. Specifically, it is the distribution of the allele
frequencies of a given set of loci (generally single nucleotide
polymorphisms, SNPs), in a population or sample. The best way to gain an
intuition for what the SFS is, is to visualize it.</p>
<figure><img src="fig/popgen-data-rendered-plot-SFS-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The first bar of this plot indicates derived alleles (i.e. not
ancestral) that appear once across the population. Approximately 35% of
the sampled SNPs in the genome alignment are these “singletons”. The
second bar indicates derived alleles that appear twice across the
population, where ~18% of the sampled SNPs are these “doubletons” and so
forth.</p>
<p>The shape of the SFS can tell us important information about the
demographic history of a population and their history of selection. The
above plot corresponds with a population that has experienced
demographic stability. A couple other examples of the impact of
demographic history on the SFS include:</p>
<p><img src="fig/popgen-data-rendered-plot-SFS-demography-1.png" width="50%" style="display: block; margin: auto;" class="figure"><img src="fig/popgen-data-rendered-plot-SFS-demography-2.png" width="50%" style="display: block; margin: auto;" class="figure"></p>
<p>There are a variety of summary statistics that describe the shape of
the SFS. The average number of pairwise nucleotide differences, a.k.a.
Tajima’s <span class="math inline">\(\pi\)</span>, where a high value
indicates a higher level of genetic diversity in the population,
describes the average of the distribution. However, as you’ve seen
above, the shape of the distribution contains important information
about the population’s history!</p>
</div>
<div class="section level3">
<h3 id="hill-numbers-of-the-sfs">Hill numbers of the SFS<a class="anchor" aria-label="anchor" href="#hill-numbers-of-the-sfs"></a>
</h3>
<p>In the context of summarizing species abundances, Hill numbers are
the effective number of species in a community, where each exponent
(q-value) increasingly weights rare species more heavily. The
interpretation is similar when summarizing within-population genetic
diversity. Hill numbers of within-population allele frequencies convey
the <em>effective number of alleles</em> in a population, where each
q-value increasingly weights rarer alleles more heavily.</p>
</div>
</div>
<div class="section level2">
<h2 id="across-species">Across species<a class="anchor" aria-label="anchor" href="#across-species"></a>
</h2>
<p>{{pi across species neutral viz}}</p>
<p>{{pi across species filtering viz}}</p>
<p>{{pi across species competition viz}}</p>
<div class="section level3">
<h3 id="hill-numbers-of-genetic-diversity-across-species">Hill numbers of genetic diversity across species<a class="anchor" aria-label="anchor" href="#hill-numbers-of-genetic-diversity-across-species"></a>
</h3>
<p>In the context of summarizing species abundances, Hill numbers are
the effective number of species in a community, where each exponent
(q-value) increasingly weights rare species more heavily. The
interpretation is similar when summarizing genetic diversity. Hill
numbers are the effective number species in a community, where each
q-value increasingly weights species with lower genetic diversity more
heavily. Rather than measuring an effective diversity of abundances,
this is an effective diversity of genetic diversities.</p>
<p>Something about correcting for species richness to just get
evenness…</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="work-with-pop-gen-data">Work with pop gen data<a class="anchor" aria-label="anchor" href="#work-with-pop-gen-data"></a>
</h1>
<div class="section level2">
<h2 id="sequences">Sequences<a class="anchor" aria-label="anchor" href="#sequences"></a>
</h2>
<div class="section level3">
<h3 id="fasta">FASTA<a class="anchor" aria-label="anchor" href="#fasta"></a>
</h3>
<p>A FASTA file contains one or more DNA sequences, where the DNA
sequence is preceded by a line with a carat (“&gt;”) followed by an
unique sequence ID:</p>
<p>&gt; seq0010<br>
GATCCCCAATTGGGG</p>
<p>This is perhaps the most common way to store sequence information as
it is simple and has historical inertia. For more information, see the
<a href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat/#:~:text=In%20FASTA%20format%20the%20line,should%20not%20contain%20any%20spaces." class="external-link">the
NCBI page</a>.</p>
<div class="section level4">
<h4 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h4>
<p>Prior to analysis, sequence data must be aligned if they haven’t been
already. A sequence alignment is necessary to identify regions of DNA
that are similar or that vary. There are <a href="https://en.wikipedia.org/wiki/List_of_sequence_alignment_software" class="external-link">many
pieces of software to perform alignments</a> and all have their
advantages and disadvantages, but we will choose a single program, <a href="https://www.ebi.ac.uk/Tools/msa/clustalo/" class="external-link">Clustal Omega</a>,
which is fast and robust for short sequence alignments. While Clustal
Omega can be run from the command line, we will use the R package <a href="https://bioconductor.org/packages/release/bioc/html/msa.html" class="external-link">msa</a>
to keep everything in the R environment.</p>
<p>First, a FASTA file containing the sequences needs to be read in
using the <code>readDNAStringSet()</code> function.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">msa</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: Biostrings</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: stats4</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: IRanges</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: XVector</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'Biostrings'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:base':

    strsplit</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_path</span> <span class="op">&lt;-</span> <span class="st">"/Users/connorfrench/Dropbox/Old_Mac/School_Stuff/CUNY/Courses/Spring-2019/Machine-Learning/project/fasta/hypsiboas_seqs_aligned.fas"</span></span>
<span></span>
<span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">seq_path</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in .Call2("new_input_filexp", filepath, PACKAGE = "XVector"): cannot open file '/Users/connorfrench/Dropbox/Old_Mac/School_Stuff/CUNY/Courses/Spring-2019/Machine-Learning/project/fasta/hypsiboas_seqs_aligned.fas'</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqs</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'seqs' not found</code></pre>
</div>
<p>To align the sequences using Clustal Omega, we use the
<code>msa</code> function and specify the <code>method</code> as
“ClustalOmega”. There are more arguments that you can use to fine-tune
the alignment, but these are not necessary for the vast majority of
alignments.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'seqs' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'alignment' not found</code></pre>
</div>
<p>If after you perform your alignment you find that an individual does
not belong or is negatively impacting the alignment, you have to
manipulate your unaligned sequence data and then re-align the sequences.
For instance, say you didn’t want to include the individual
<code>ind_name</code> in the alignment. First, remove the individual
from the <code>seqs</code> object.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove individual</span></span>
<span><span class="va">seqs_filt</span> <span class="op">&lt;-</span> <span class="va">seqs</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span> <span class="op">%in%</span> <span class="st">"Itar128"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'seqs' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqs_filt</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'seqs_filt' not found</code></pre>
</div>
<p>Then, realign the new set of sequences.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs_filt</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'seqs_filt' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'alignment' not found</code></pre>
</div>
<p>There are quite a few packages in R to work with sequence data, for
example <a href="https://cran.r-project.org/web/packages/seqinr/index.html" class="external-link">seqinr</a>,
<a href="https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.html" class="external-link">phangorn</a>,
and <a href="https://cran.r-project.org/web/packages/ape/index.html" class="external-link">ape</a>.
<code>ape</code> is the most versatile of these. But before we can use
ape, the alignment must be converted to the <code>ape::DNAbin</code>
format.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ape_align</span> <span class="op">&lt;-</span> <span class="fu">msaConvert</span><span class="op">(</span><span class="va">alignment</span>, type <span class="op">=</span> <span class="st">"ape::DNAbin"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'alignment' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ape_align</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'ape_align' not found</code></pre>
</div>
<p>Although not necessary for this workshop, it’s generally advisable to
write your DNA alignment to a new file to use later or with other
programs.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ape</span><span class="fu">::</span><span class="fu">write.FASTA</span><span class="op">(</span><span class="va">ape_align</span>, <span class="st">"path/to/alignment.fas"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'ape_align' not found</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="ape">ape<a class="anchor" aria-label="anchor" href="#ape"></a>
</h4>
</div>
<div class="section level4">
<h4 id="check-alignment">Check alignment<a class="anchor" aria-label="anchor" href="#check-alignment"></a>
</h4>
<p><code>ape</code> has a useful utility that allows you perform a
series of diagnostics on your alignment to make sure nothing is fishy.
The function can output multiple plots, but let’s just visualize the
alignment plot. If the colors seem jumbled in a region, this indicates
that an alignment error may have occurred and to inspect your sequences
for any problems.</p>
<p>In addition, the function outputs a series of helpful statistics to
your console. Ideally, your alignment should have few gaps and a number
of segregating sites that is reasonable for the number of sites in your
alignment and the evolutionary divergence of the sequences included in
the alignment, e.g., for an alignment with 970 sites of a single bird
species, around 40 segregating sites is reasonable, but 200 segregating
sites may indicate that something went wrong with your alignment or a
species was misidentified. In addition, for diploid species a vast
majority, if not all of the sites should contain one or two observed
bases.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'ape'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:Biostrings':

    complement</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">checkAlignment</span><span class="op">(</span><span class="va">ape_align</span>, what <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'ape_align' not found</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="calculate-genetic-diversity">Calculate genetic diversity<a class="anchor" aria-label="anchor" href="#calculate-genetic-diversity"></a>
</h4>
<p>Tajima’s pi (“raw” pairwise difference)… Why pairwise deletion rather
than full deletion…</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendist</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">fas</span>, model <span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'fas' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendiv</span> <span class="op">&lt;-</span> <span class="fu">median</span><span class="op">(</span><span class="va">fas_gendist</span><span class="op">[</span><span class="fu">upper.tri</span><span class="op">(</span><span class="va">fas_gendist</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'median': object 'fas_gendist' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendist_pop</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">fas_split</span>, <span class="va">dist.dna</span>, model <span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'X' in selecting a method for function 'lapply': object 'fas_split' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">calc_pi</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">gd</span> <span class="op">&lt;-</span> <span class="fu">median</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fu">upper.tri</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="kw">return</span><span class="op">(</span><span class="va">gd</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>    </span>
<span><span class="va">fas_gendiv_pop</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">fas_gendist_pop</span>, <span class="va">calc_pi</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'X' in selecting a method for function 'sapply': object 'fas_gendist_pop' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">fas_gendiv_pop</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'plot': object 'fas_gendiv_pop' not found</code></pre>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="snps">SNPs<a class="anchor" aria-label="anchor" href="#snps"></a>
</h2>
<div class="section level3">
<h3 id="vcf">VCF<a class="anchor" aria-label="anchor" href="#vcf"></a>
</h3>
<p>Here are a few things about a VCF file… tab delimited… fields…
size…</p>
<div class="section level4">
<h4 id="vcfr">vcfR<a class="anchor" aria-label="anchor" href="#vcfr"></a>
</h4>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">vcfR</span><span class="op">)</span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">read.vcfR</span><span class="op">(</span><span class="va">vcf_path</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="manipulate-data">Manipulate data<a class="anchor" aria-label="anchor" href="#manipulate-data"></a>
</h4>
<p>Filter individuals Filter based on quality, missingness</p>
</div>
<div class="section level4">
<h4 id="export-to-genind">Export to genind<a class="anchor" aria-label="anchor" href="#export-to-genind"></a>
</h4>
<p>A bit about genind and <code>adegenet</code>… If you have many SNPs,
genlight may be a more appropriate format…</p>
<p><code>vcfR2genind</code> also takes arguments for the function
<code>df2genind</code> from the <code>adegenet</code> package. There are
a couple things we can do to speed up the conversion, like specifying
the ploidy of the file and setting the <code>check.ploidy</code> flag to
<code>FALSE</code>.</p>
<p>To add populations…</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">adegenet</span><span class="op">)</span></span>
<span><span class="va">snps_genind</span> <span class="op">&lt;-</span> <span class="fu">vcfR2genind</span><span class="op">(</span></span>
<span>    <span class="va">snps</span>, </span>
<span>    n.cores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    ploidy <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    check.ploidy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    pop <span class="op">=</span> <span class="va">pops</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">snps_genind</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="calculate-genetic-diversity-1">Calculate genetic diversity<a class="anchor" aria-label="anchor" href="#calculate-genetic-diversity-1"></a>
</h4>
<p>If you didn’t specify populations when converting or reading in the
genind file, do so now…</p>
<p>Individual-based genetic distances… Maybe mention PCA?</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_gendist_ind</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="va">snps_genind</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'snps_genind' not found</code></pre>
</div>
<p>Nei’s genetic distance…</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_genpop</span> <span class="op">&lt;-</span> <span class="fu">genind2genpop</span><span class="op">(</span><span class="va">snps_genind</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in genind2genpop(snps_genind): could not find function "genind2genpop"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_gendist_pop</span> <span class="op">&lt;-</span> <span class="fu">dist.genpop</span><span class="op">(</span><span class="va">snps_genpop</span>, method <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in dist.genpop(snps_genpop, method = 1): could not find function "dist.genpop"</code></pre>
</div>
<p>FST…</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hierfstat</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'hierfstat'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:ape':

    pcoa, varcomp</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_hierfstat</span> <span class="op">&lt;-</span> <span class="fu">genind2hierfstat</span><span class="op">(</span><span class="va">snps_genind</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'snps_genind' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_fst</span> <span class="op">&lt;-</span> <span class="fu">pairwise.fst</span><span class="op">(</span><span class="va">snps_hierfstat</span>, diploid <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pairwise.fst(snps_hierfstat, diploid = TRUE): could not find function "pairwise.fst"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">snps_fst</span><span class="op">[</span><span class="fu">upper.tri</span><span class="op">(</span><span class="va">snps_fst</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'snps_fst' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="summarize-genetic-diversity">Summarize genetic diversity<a class="anchor" aria-label="anchor" href="#summarize-genetic-diversity"></a>
</h1>
<div class="section level2">
<h2 id="calculate-hill-numbers-within-speciespopulations">Calculate Hill numbers within species/populations<a class="anchor" aria-label="anchor" href="#calculate-hill-numbers-within-speciespopulations"></a>
</h2>
<p><em>* Going to need a SNP matrix</em></p>
<p>Hill numbers of intraspecific allele frequencies are the
<em>effective number of SNPs</em> in the population/species.</p>
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># columns are SNPs, rows are individuals</span></span>
<span><span class="co"># treat it like abundances </span></span>
<span><span class="va">intra_hill_1</span> <span class="op">&lt;-</span> <span class="fu">hillR</span><span class="fu">::</span><span class="fu">hill_taxa</span><span class="op">(</span><span class="va">snp_matrix</span>, q <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="interpret-refer-back-to-visualizations">Interpret (refer back to visualizations)<a class="anchor" aria-label="anchor" href="#interpret-refer-back-to-visualizations"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="calculate-hill-numbers-across-species">Calculate Hill numbers across species<a class="anchor" aria-label="anchor" href="#calculate-hill-numbers-across-species"></a>
</h2>
<p><em>Need species alignments from ape</em></p>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate pairwise distance matrices</span></span>
<span><span class="va">per_species_dist</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">species_alignments</span>, <span class="fu">ape</span><span class="fu">::</span><span class="va">dist.dna</span>, model <span class="op">=</span> <span class="st">"raw"</span>, as.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mat_avg</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="fu">upper.tri</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span>    <span class="kw">return</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">per_species_pi</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">per_species_dist</span>, <span class="va">mat_avg</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">per_species_pi</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="interpret-refer-back-to-visualizations-1">Interpret (refer back to visualizations)<a class="anchor" aria-label="anchor" href="#interpret-refer-back-to-visualizations-1"></a>
</h3>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</div>
</div></section><section id="aio-phylo-data"><p>Content from <a href="phylo-data.html">Summarizing phylogenetic data</a></p>
<hr>
<p> Last updated on 2023-06-05 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/phylo-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the common phylogeny data formats?</li>
<li>How do you import and manipulate phylogeny data in R?</li>
<li>How do you visualize phylogenies in R?</li>
<li>How to summarize and interpret phylogenies using Hill numbers?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants will be able to:</p>
<ol style="list-style-type: decimal">
<li>Identify key features of different phylogenetic data formats</li>
<li>Import, clean and manipulate phylogenetic data in the R
environment</li>
<li>Visualize phylogenies</li>
<li>Calculate and interpret phylogenetic Hill numbers</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>In this episode, we will explore how to extract information from
phylogenetic trees in order to complement our hypotheses and inferences
about processes shaping biodiversity patterns. Phylogenetic trees have
information about how the species in a taxonomic group are related to
each other and how much relative evolutionary change has accumulated
among them. Since local communities differ in their phylogenetic
composition, this information can give insights on why communities are
how they are.</p>
<p>The relative phylogenetic distance among species in a community as
well as the distribution of the amount of evolutionary history
(represented by the length of the branches in a phylogeny) are a result
of different factors such as the age since the initial formation of the
community and the rate of macroevolutionary processes such as speciation
and extinction. For instance, young communities that are dominated by
closely related species and show very short branch lengths may suggest a
short history with few colonization events and high rates of local
speciation; alternatively, if the same young communities harbor
distantly related species with longer branch lengths, it may suggest
that most of the local diversity was generated by speciation elsewhere,
along with many colonization by species from different points in the
phylogeny. Coupled with information on ecological traits and rates of
macroevolutionary processes, these patterns also allow to test for
hypotheses regarding, for instance, ecological filtering or niche
conservatism.</p>
<!-- Maybe bring in the paragraph above explicit connections to our young vs older Hawaian islands examples. -->
<p>Summarizing this phylogenetic information (i.e., phylogenetic
distance and distribution of branch lenghts) is therefore crucial for
inference. As we have seen in previous episodes, the use of <a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Hill
numbers</a> is an informative approach to summarize biodiversity. In
this episode, we will then go over 1) how to import and manipulate
phylogenetic data in R; 2) how to calculate and interpret Hill numbers
for different communities.</p>
</div>
<div class="section level3">
<h3 id="importing-phylogenetic-data">Importing phylogenetic data<a class="anchor" aria-label="anchor" href="#importing-phylogenetic-data"></a>
</h3>
<p>Several file formats exist to store phylogenetic information. The
most common formats are the <a href="http://evomics.org/resources/tree-formats/" class="external-link">Newick and Nexus
formats</a>. Both these formats are plain text files storing different
levels of information about the taxa relationship and evolutionary
history. <code>Newick</code> files are the standard for representing
trees in a computer-readable form, as they can be extremely simple and
therefore do not take up much memory. <code>Nexus</code> file, on the
other hand, are composed by different blocks regarding different types
of information and can be used to store DNA alignments, phylogenetic
trees, pre-defined groupings of taxa, or everything at once. Since they
are a step ahead in complexity, we will stick with <code>Newick</code>
files for now.</p>
<p><code>Newick</code> files store the information about the clades in a
tree by representing each clade within two parentheses, with tips
separated by <code>,</code>. We also add the symbol <code>;</code> at
the end to represent the end of the information for that phylogenetic
tree.</p>
<p>The basic structure of a tree in a <code>Newick</code> format is
therefore as follows:</p>
<p>((A,B),C);</p>
<p>The notation above indicates that: 1) we have three taxa in our tree,
named <code>A</code>, <code>B</code> and <code>C</code>; 2)
<code>A</code> and <code>B</code> form one clade (A,B); 3) the (A,B)
clade is sister to the <code>C</code> clade (we represent that by adding
another set of parentheses and a <code>,</code> separating (A,B) from
C).</p>
<p>In addition, <code>Newick</code> files can also store information on
the branch leading to it tip and node. We do that by adding
<code>:</code> after each tip/node.</p>
<p>((A:0.5,B:0.5):0.5,C:1);</p>
<p>The notation above indicates that: 1) the branches containing A and B
have each a length = 0.5; 2) the branch that leads to the node
connecting A and B also has length = 0.5; 3) the branch leading to C has
length = 1.</p>
<p>The notation above is what we import into R to start working with and
manipulating our phylogenetic tree. For that goal, we will use the <a href="https://academic.oup.com/bioinformatics/article/20/2/289/204981" class="external-link">ape</a>
package:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span></code></pre>
</div>
<p>To import our tree, we will be using the function
<code>read.tree()</code> from the <code>ape</code> package. In the case
of simple trees as the one above, we could directly create them within R
by giving that notation as a <code>character</code> value to this
function, using the <code>text</code> argument, as shown below:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text <span class="op">=</span> <span class="st">'((A:0.5,B:0.5):0.5,C:1);'</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, we can visually inspect our tree using the <code>plot()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">example_tree</span><span class="op">)</span></span></code></pre>
</div>
<p>Can you visualize the text notation in that image? We can see the
same information: A is closer related to B than C, and the branches
leading to A and B have half the length of the branch leading to C.</p>
<p>The <code>read.tree()</code> function creates an object of class
<code>phylo</code>. We can further investigate this object by calling it
in our console:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span></span></code></pre>
</div>
<p>The printed information shows us that we have a phylogenetic tree
with 3 tips and 2 internal nodes, where the tip labels are “A, B, C”. We
also are informed that this tree is rooted and has branch lengths.</p>
<p>One way to access the components of this object and better explore it
is to use <code>$</code> after the object name. Here, it will be
important for us to know a little bit more about where the information
about tip labels and branch lengths are stored in that
<code>phylo</code> objecy. Easy enough, we can access that by calling
<code>tip.labels</code> and <code>edge.length</code> after
<code>$</code>.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="va">example_tree</span><span class="op">$</span><span class="va">edge.length</span></span></code></pre>
</div>
<p>Now that we learned how to import and visualize trees in R, let’s
bring the phylogeny for the communities we are working with in this
workshop. Our data so far consists of <a href="abundance-data.html">abundances</a> and <a href="traits-data.html">traits</a> of several taxa of arthropods
collected across three islands in the Hawaian archipelago. Here, we will
import a global phylogeny of arthropods, located at: &lt;&gt;</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylogeny</span> <span class="op">&lt;-</span> <span class="fu">read_tree</span><span class="op">(</span><span class="st">''</span><span class="op">)</span></span></code></pre>
</div>
<p>This new object is way larger than the previous once, being a “real”
phylogeny and all. You can inspect it again by directly calling the
object <code>phylogeny</code>. To plot it, we will use the
<code>type</code> argument to modify how our tree will be displayed.
Here, we used the option <code>'fan'</code>, to display a circular
phylogeny (slightly better to show such a large phylogeny in the
screen).</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">phylogeny</span>, type <span class="op">=</span> <span class="st">'fan'</span><span class="op">)</span></span></code></pre>
</div>
<p>How do we combine all this information with the community datasets we
have so far for our three islands? First, we will have to perform some
cleaning.</p>
</div>
<div class="section level3">
<h3 id="cleaning-phylogenetic-data">Cleaning phylogenetic data<a class="anchor" aria-label="anchor" href="#cleaning-phylogenetic-data"></a>
</h3>
<p>Often, when we want to include the phylogenetic relatedness of
organisms in our biodiversity analyses, we will run into published
phylogenetic trees for our focal taxon (that is the case of the tree we
are using here). A couple issues arise in these cases. First, and maybe
most importantly, we need to make sure that the taxonomic names used
across the different datasets in our analyses (phylogenies included) are
the same. A second issue is that, depending on the resolution of the
published phylogeny, it may be the case that some taxa in our community
dataset may not be explicitly represented in the phylogeny. In these
cases, we will want to assign those “missing taxa” to one of the tips
present in the tree, based on their known genus and/or family
assignments.</p>
<p>In this section, we will deal with these issues in order to prepare
our data for community inferences. Our first issue will be dealt with in
a challenge!</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>In previous episodes, we have used the package <code>taxize</code> to
clean and standardize the taxonomic names in out dataset. Here, we will
apply the same approach to the tip names of our phylogeny. We need to
make sure the names we will accept from <code>taxize</code> will match
the names we have accepted in previous episodes. Finally, after we
decide on the names, we need to add them to the <code>phylogeny</code>
objects, where the tip names are stored.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>First, we check the names using the <code>gnr_resolve()</code>
function. In <code>phylogeny</code>, tip names are stored in
<code>phylogeny$tip.label</code>, so we provide that vector to our
function.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="va">name_resolve</span> <span class="op">&lt;-</span> <span class="fu">gnr_resolve</span><span class="op">(</span><span class="va">phylogeny</span><span class="op">$</span><span class="va">tip.label</span>, best_match_only <span class="op">=</span> <span class="cn">TRUE</span>, canonical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">name_resolve</span><span class="op">)</span></span></code></pre>
</div>
<p>We can see what taxa are in conflict with the names retrieved by
<code>taxize</code> by subsetting <code>name_resolve</code> based on
boolean matching.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mismatches_phylo</span> <span class="op">&lt;-</span> <span class="va">name_resolve</span><span class="op">[</span><span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span> <span class="op">!=</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"user_supplied_name"</span>, <span class="st">"matched_name2"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">mismatches_phylo</span></span></code></pre>
</div>
<p>From here, we can see that apparently there are no misspellings in
our tree (as expected from a published phylogeny). Most of the
mismatches seem to be related to changes in the taxonomic treatment of
species. This is a common situation, since taxonomic treatments are
often revised after new trees are published; it might be the case that
<code>nameA</code> in our dataset was treated as <code>nameB</code> at
the time the phylogeny was created.</p>
<p>In the previous episodes, we used the names provided by
<code>taxize</code> directly. Therefore, here we can just add those
names directly to the phylogeny.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylogeny</span><span class="op">$</span><span class="va">tip.label</span> <span class="op">==</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The next issue we need to address is the fact that we may have
species in our dataset that are not present in the phylogenetic tree we
are using. We therefore need to 1) find out which species are not in the
tree; 2) assign them to tips that are present in the tree.</p>
<p>To check which species are not in the tree, we can use any of the
objects we created so far that has the list of all the species in our
communities. These could be: 1) the abundances we calculated per site
and island in the <a href="abundance-data.html">abundances episode</a>,
stored in the object <code>abundance_tallies</code>; 2) the <em>site by
species matrix</em> we have created in the abundance episode (stored in
the object <code>abundance_wide</code>); 3) the list of species with
their respective trait values, created in the <a href="traits-data.html">traits episode</a> and stored in the
<code>traits_simple</code> object. Since we are going to directly use
the site-by-species matrix to calculate Hill numbers (next section),
let’s focus on this object and retrieve the species name for it.</p>
<p>We do that by accessing the <code>colnames()</code> of that object.
We will also remove the first element, since it is just the header for
the column with the <code>site</code> name.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<p>The taxonomic names in both the abundance data and in the
phylogenetic tree were corrected using the same database from
<code>taxize</code>; this therefore means that any names from our
abundance data that are not found in the tree simply represent taxa that
are not included in this phylogeny. We can use the boolean operator
<code>%in%</code> coupled with <code>!</code> to check
<code>species_names</code> that are not included in the
<code>phylogeny$tip.label</code>. In summary, the expression
<code>A %in% B</code> would return the positio of the elements in vector
A that are present in vector B. The <code>!</code> (NOT) operator
returns the opposite of that expression, in a way that
<code>!(A %in% B)</code> will return the position of elements in vector
A that are <em>NOT</em> present in vector B (precisely what we need
right now).</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">not_present</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="op">(</span><span class="va">species_names</span> <span class="op">%in%</span> <span class="va">phylogeny</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span></span></code></pre>
</div>
<p>The object <code>not_present</code> is now a vector of the positions
(or indexes) of the elements in <code>species_name</code> that are not
found in <code>phylogeny$tip.label</code>. Let’s then check their names
by using these indexes to subset <code>species_name</code>:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_name</span><span class="op">[</span><span class="va">not_present</span><span class="op">]</span></span></code></pre>
</div>
<p>Here, we can decide how we are going to assign these species to tips
in our tree. For instance, we see that we do not have <em>XXX</em> in
our tree, but we have <em>XXX</em>, which belongs to the same genus.
Similarly…</p>
<p>We can then give new names to these species, based on what tips in
our tree we want them to match to. To do so, we will first duplicate our
<code>species_name</code> vector, so we can keep the original names
(just in case we want to check the original names again in the
future).</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_name</span><span class="op">[</span><span class="va">not_present</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">new_names</span><span class="op">[</span><span class="va">not_present</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>Now that we have a vector of updated named, all we need to do is
modify those names in our site-by-species matrix. Similarly, we will
first copy our matrix into another object (in order to keep the original
abundance matrix intact), and then we will rename the columns of this
new object based on the modified names.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_wide</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">phylo_wide</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">new_names</span></span></code></pre>
</div>

<p>After solving these issues, we now have 1) a phylogenetic tree with
tip labels taxonomically cleaned; 2) a site-by-species matrix where all
species are represented in our tree. These two objects are the two bits
of information that we need to summarize the phylogenetic diversity and
the distribution of branch lengths across our focal communities.</p>
</div>
<div class="section level3">
<h3 id="summarizing-with-hill-numbers">Summarizing with hill numbers<a class="anchor" aria-label="anchor" href="#summarizing-with-hill-numbers"></a>
</h3>
<p>In this section, we will extract some summary statistics about the
pattern of phylogenetic diversity in our communities. As we discussed
above, the relative phylogenetic distance among species and the
distribution of this distance can give insights into processes of
community assembly. Here, we will make use of <a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Hill
numbers</a> to extract summaries of phylogenetic distances. To
understand what they mean, let’s first explore their properties in a
simple example.</p>
<p>Let’s assume we have a community with eight taxa: A through H. We
will create two possibilities of phylogenetic relationship among
them.</p>
<p>Tree 1 will represent a totally unbalanced pattern…</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree1</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(A:7,(B:6,(C:5,(D:4,(E:3,(F:2,(G:1,H:1):1):1):1):1):1):1);'</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span></span></code></pre>
</div>
<p>whereas tree 2 will represent a totally balanced pattern.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree2</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(((A:1,B:1):1,(C:1,D:1):1):1,((E:1,F:1):1,(G:1,H:1):1):1);'</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tree2</span><span class="op">)</span></span></code></pre>
</div>
<p>Notice that both trees are ultrametric: all tips align at the end,
representing extant taxa, which means branch lengths here represent the
amount of time that has passed between each node. The difference between
the first and the second tree resides in the fate of each new lineage at
a node. In the first tree, at each diversification event, one of the
lineages always persists till the present with no change while the other
undergoes another round of diversification. In the second tree, both
lineages from each node undergo a new split. The consequence is that in
tree 2, all extant species result from recent diversification (i.e.,
they have a short evolutionary history before coalescing into their
ancestor), whereas in tree 1 we have a mix of old and recent
lineages.</p>
<p>To visualize how these different patterns reflect in the diversity
metric, we will calculate the phylogenetic Hill number using the
function <code>hill_phylo</code> from the <code>hillR</code> package.
This function takes in a global phylogeny as well as a site-by-species
matrix, and returns phylo Hill numbers for each site based on which
species are present there. For our fictional example, we will create a
matrix with only one site where all eight species in the tree are
present. We do that by creating a vector of the value <code>1</code>
repeated 8 times, transforming that to a dataframe, and then naming the
columns with the names of the species we take from our fictional
trees.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span>,<span class="st">'D'</span>,<span class="st">'E'</span>,<span class="st">'F'</span>,<span class="st">'G'</span>,<span class="st">'H'</span><span class="op">)</span></span></code></pre>
</div>
<p>Using that matrix, we will calculate Hill numbers of order 0 to 2 for
both trees:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span><span class="va">tree1_h0</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">tree1_h1</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tree1_h2</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">tree2_h0</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">tree2_h1</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tree2_h2</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s format this as a dataframe so we can visualize these numbers:
the first column will be the order (our X axis), the second column will
be the Hill values (our Y axis) and the third column will be the tree it
belongs to (which we will use to draw different colored lines for
comparison).</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_h</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>order <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                      hill <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="va">tree1_h0</span>,<span class="va">tree1_h1</span>,<span class="va">tree2_h2</span>,<span class="va">tree2_h0</span>,<span class="va">tree2_h1</span>,<span class="va">tree2_h2</span><span class="op">)</span>,</span>
<span>                      tree <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">'tree1'</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="st">'tree2'</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we will call <code>plot</code> to generate a line plot for the
values of the first tree, and then use <code>lines</code> to add a line
containing the values for the second tree (we used boolean matching
subsetting to choose what to plot at each time)</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Plotting values for tree 1</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkred'</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting values for tree 2</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'red'</span><span class="op">)</span></span></code></pre>
</div>
<p>From this plot, we can notice a few patterns:</p>
<ol style="list-style-type: decimal">
<li>The hill values for tree 1 are usually higher than those for tree 2,
indicating that our community contains a deeper evolutionary history
with tree 1 (i.e., lineages containg longer branches);</li>
<li>The value of the hill numbers drop as the order goes up. This
happens because higher orders are putting less and less emphasis on
branch lengths that are not so common (like, for instance, the short
branch lenghts in tree 1). In order words, higher order emphasize what
is more common.</li>
<li>The rate of change of the value as we increase the order is also
different among trees. Tree 1 is a more uneven tree; it is composed
mostly of somewhat long branches, and few short branches. So, when we
change the order, we notice a significant change in the value (at least
more significant than the change seen in tree2), because there’s a part
of the tree we are starting to disconsider (i.e., those short branch
lengths that are not so common). We observe a smaller change in tree 2
because all branches have very similar lenghts, so changing the order
does not affect the overall diversity (in other words, all branch
lenghts are equally common, so emphasize more or less common lenghts
does not change much the values).</li>
</ol>
<p>What would happen if the taxa in our communities had different
relative abundances? In this case, not only the relative frequency of a
branch length would affect our summary statistic, but also how abundant
is the taxon that has each branch length. For instance, in tree 1, even
though long branches are over represented, maybe the species with long
branches are actually super rare in our community, and the short-branch
species are actually super abundant. As you can probably realize, this
difference suggests something about the evolutionary history of our
community, in this case that the most abundant species have a very
recent evolutionary history. Similarly, even though branch lengths are
very even in tree 2, if some of the species is more abundant than
others, it suggests that the evolutionary history of the tree is
unevenly represented in our community.</p>
<p>Conveniently, the <code>hill_phylo</code> function will take into
account both the relative frequency of branch lenghts <em>AND</em> the
abundance of extant taxon in the phylogeny we provide. For our simple
example, we can see how that changes our results by creating
site-by-species matrix where abundance goes up from species A to species
H (for simplicity, we’ll just use values from 1 to 8).</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span>,<span class="st">'D'</span>,<span class="st">'E'</span>,<span class="st">'F'</span>,<span class="st">'G'</span>,<span class="st">'H'</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, let’s calculate the same hill numbers with this new community.
We save the values to objects with <code>_rel</code> at the end, to
highlight in this case the taxa had different <em>rel</em>ative
abundances.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree1_h0_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">tree1_h1_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tree1_h2_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">tree2_h0_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">tree2_h1_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tree2_h2_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<p>Then we add the new values to our <code>phylo_h</code> dataframe. We
will signal they belong to <code>tree1_rel</code> and
<code>tree2_rel</code>, meaning they came from a community with
different <em>rel</em>ative abundance.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_h</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">phylo_h</span>,</span>
<span>                 <span class="fu">data.frame</span><span class="op">(</span>order <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                            hill <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="va">tree1_h0_rel</span>,<span class="va">tree1_h1_rel</span>,<span class="va">tree1_h2_rel</span>,</span>
<span>                                     <span class="va">tree2_h0_rel</span>,<span class="va">tree2_h1_rel</span>,<span class="va">tree2_h2_rel</span><span class="op">)</span>,</span>
<span>                            tree <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">'tree1_rel'</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="st">'tree2_rel'</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Plotting values for tree 1</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkred'</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting values for tree 2</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'red'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting values for tree 2</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1_rel'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1_rel'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkblue'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting values for tree 2</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2_rel'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2_rel'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'lightblue'</span><span class="op">)</span></span></code></pre>
</div>
<p>Reminders: 1) red is equal abundance and blue is different
abundances; 2) darker colors represent <code>tree1</code> (uneven branch
length) and lighter colors represent <code>tree2</code> (equal branch
lengths).</p>
<p>From this second plot, we notice that:</p>
<ol style="list-style-type: decimal">
<li><p>Hill number values for tree 1 (darker colors) are still higher
than those for tree 2 (lighter colors). This makes sense since the
relative abundance is the same for both trees, so the overall pattern
(i.e., longer branches in tree 1) remains;</p></li>
<li><p>the Hill number of order 0 is the same regardless of whether you
have different relative abundances among species, since order 0 is
simply the sum of all branch lengths (i.e., it does not account for how
the relative abundance of taxa);</p></li>
<li><p>when we include different relative abundance among taxa, we have
a higher rate of decrease in the Hill number value as we increase the
order. This is because higher orders are now being calculated based on
how common/rare each branch is in our community (based on our
site-by-species matrix). If we take tree 1, for instance, we see that
the Hill value for order 1 including different taxa abundance (darker
blue line) is lower than the value where abundance is equal (darker red
line). We can interpret this as our metric putting less weight on the
longer branches of tree 1, which are rare in our community, and putting
more weight on the shorter branches, which are more common but represent
a more recent evolutionary history. We see a similar (although less
accentuated) pattern with tree2. Calculating different order of hill
numbers while incorporating species abundance therefore allows to have a
single metric accounting for evolutionary history and species abundance
distribution that is comparable across different regions.</p></li>
</ol>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>Now that you got an intuition on the information phylogenetic Hill
numbers can give us, let’s move on to our actual data. For this
challenge, you should calculate the phylogenetic Hill numbers of orders
0, 1 and 2 for three islands we are working with throughout these
episodes. The key objects here will be the <code>phylogeny</code> tree
we imported earlier, as well as the <code>phylo_wide</code> object
containing our site-by-species matrix modified to match the tips in our
phylogeny. You should also plot the hill numbers for each island and
discuss what the calculated values allow you to infer regarding the
history of each community.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<!-- We gotta simulate the data to finalize this solution section. -->
</div>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section><section id="aio-connecting-data-types"><p>Content from <a href="connecting-data-types.html">Connecting different data types</a></p>
<hr>
<p> Last updated on 2023-06-05 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/connecting-data-types.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What questions and analyses can we explore by linking multiple data
types together?</li>
<li>How do we determine if our data can be linked?</li>
<li>How do we actually link different data types together?</li>
<li>How do we analyze, visualize, and interpret multiple data types
concurrently?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal">
<li>Generate ideas for scientific questions that can be answered using 2
or more of the data types we have covered in previous episodes</li>
<li>Identify considerations regarding the cross-compatibility of data
from different sources</li>
<li>Use spatial and taxonomic tools to link data of different types</li>
<li>Calculate and visualize joint diversity indices for multiple data
types simultaneously</li>
<li>Interpret these results in scientific context</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="motivation-why-combine-data-types">Motivation: Why combine data types?<a class="anchor" aria-label="anchor" href="#motivation-why-combine-data-types"></a>
</h1>

<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>What kinds of questions can we imagine answering using multipe data
types at once?</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="show-demo-visualization-of-hawaii-sim-data">Show demo visualization of Hawaii sim data<a class="anchor" aria-label="anchor" href="#show-demo-visualization-of-hawaii-sim-data"></a>
</h2>
</div>
</div>
<div class="section level1">
<h1 id="linking-different-data-types">Linking different data types<a class="anchor" aria-label="anchor" href="#linking-different-data-types"></a>
</h1>
<div class="section level2">
<h2 id="checking-the-compatibility-of-data-types">Checking the compatibility of data types<a class="anchor" aria-label="anchor" href="#checking-the-compatibility-of-data-types"></a>
</h2>
<div class="section level3">
<h3 id="taxonomic-overlap">Taxonomic overlap<a class="anchor" aria-label="anchor" href="#taxonomic-overlap"></a>
</h3>
</div>
<div class="section level3">
<h3 id="taxonomic-consistency">Taxonomic consistency<a class="anchor" aria-label="anchor" href="#taxonomic-consistency"></a>
</h3>
</div>
<div class="section level3">
<h3 id="spatial-extent">Spatial extent<a class="anchor" aria-label="anchor" href="#spatial-extent"></a>
</h3>
</div>
<div class="section level3">
<h3 id="others">Others??<a class="anchor" aria-label="anchor" href="#others"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="linking-data-types">Linking data types<a class="anchor" aria-label="anchor" href="#linking-data-types"></a>
</h2>
<div class="section level3">
<h3 id="resources-for-resolving-inconsistencies">Resources for resolving inconsistencies<a class="anchor" aria-label="anchor" href="#resources-for-resolving-inconsistencies"></a>
</h3>
</div>
<div class="section level3">
<h3 id="left_join">left_join<a class="anchor" aria-label="anchor" href="#left_join"></a>
</h3>
<p>(being slightly flip here, but this is the heart of it, right?)</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="analyzing-and-visualizing-joint-data-types">Analyzing and visualizing joint data types<a class="anchor" aria-label="anchor" href="#analyzing-and-visualizing-joint-data-types"></a>
</h1>
<div class="section level2">
<h2 id="hill-numbers-for-multidim-data">Hill numbers for multidim data<a class="anchor" aria-label="anchor" href="#hill-numbers-for-multidim-data"></a>
</h2>
<div class="section level3">
<h3 id="discussionlecture">Discussion/lecture<a class="anchor" aria-label="anchor" href="#discussionlecture"></a>
</h3>
</div>
<div class="section level3">
<h3 id="calculating-hills">Calculating Hills<a class="anchor" aria-label="anchor" href="#calculating-hills"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-multiple-data-types">Visualizing multiple data types<a class="anchor" aria-label="anchor" href="#visualizing-multiple-data-types"></a>
</h2>
<div class="section level3">
<h3 id="joing-vis-plots">Joing vis plots<a class="anchor" aria-label="anchor" href="#joing-vis-plots"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h2>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>Discuss Hawaii interp</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="challenge-using-other-data-sets">Challenge: using other data sets<a class="anchor" aria-label="anchor" href="#challenge-using-other-data-sets"></a>
</h1>

</div>
<div class="section level1">
<h1 id="recap">Recap<a class="anchor" aria-label="anchor" href="#recap"></a>
</h1>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Considering multiple data types at once can answer a unique and
rapidly evolving set of questions</li>
<li>Data types do need to be (minimally) compatible; there are
resources/approaches for reconciling not-fully-compatible data</li>
<li>Hill numbers provide a unified framework for the analysis and
visualization of different data types as part of one question</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section><section id="aio-finding_data"><p>Content from <a href="finding_data.html">Finding multi-dimensional biodiversity data</a></p>
<hr>
<p> Last updated on 2023-06-05 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/finding_data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Where do multi-dimensional biodiversity data live online?</li>
<li>How can we use R APIs to quickly and reproducibly retrieve
data?</li>
<li>What are some patterns in how these APIs work?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal">
<li>Effectively search for and find trait, sequence, and abundance data
online given a taxon or research question</li>
<li>Retrieve data from GEOME &amp; GBIF using their R APIs and apply API
principles to other databases</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h1>
<p>For this episode, we’ll be working with some R packages that wrap the
APIs of online repositories. Let’s set those up now.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'rotl'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:spocc':

    inspect</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'taxize'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:rotl':

    synonyms, tax_name, tax_rank</code></pre>
</div>
</div>
<div class="section level1">
<h1 id="sources-of-multidimensional-biodiversity-data-large-open-access-databases">Sources of multidimensional biodiversity data: large open-access
databases<a class="anchor" aria-label="anchor" href="#sources-of-multidimensional-biodiversity-data-large-open-access-databases"></a>
</h1>
<p>Large open-access biological databases have been growing at an
incredible pace in recent years. Often they accumulate data from many
projects and organizations. These platforms are then used to share and
standardize data for wide use.</p>
<p>Many such databases exist, but here are some listed examples:</p>
<ul>
<li><p>GBIF contains occurrence records of individuals assembled from
dozens of smaller databases.</p></li>
<li><p>NCBI (National Center for Biotechnology Information) database,
which includes GenBank is the largest repository of genetic data in the
world, but includes a lot of biomedical data.</p></li>
<li><p>OTOL (Open Tree of Life) is a database that combines numerous
published trees and taxonomies into one supertree</p></li>
<li><p>GEOME (Genomic Observatories MetaDatabase) contains genetic data
associated with specimens or samples from the field.</p></li>
<li><p>BOLD (Barcode of Life) is a reference library and database for
DNA barcoding</p></li>
<li><p>EOL (Encyclopedia of Life) is a species-level platform containing
data like taxonomic descriptions, distribution maps, and images</p></li>
</ul>
<div class="section level2">
<h2 id="whats-an-api">What’s an API?<a class="anchor" aria-label="anchor" href="#whats-an-api"></a>
</h2>
<p>An API is a set of tools that allows users to interact with a
database, such as reading, writing, updating, and deleting data. We
can’t go over all of these databases in detail, but let’s cover some
examples that illustrate some principles of how these APIs work for
different types of biodiversity data.</p>
<p>Hawaii is known for its diversity of <em>Tetragnatha</em>, a genus of
orb weaver spiders. With this target in mind let’s download some
occurrence data from GBIF, some sequence data from NCBI, and a phylogeny
from OTOL.</p>
</div>
<div class="section level2">
<h2 id="searching-by-taxon-getting-species-names">Searching by taxon: getting species names<a class="anchor" aria-label="anchor" href="#searching-by-taxon-getting-species-names"></a>
</h2>
<p>We want records of only the genus <em>Tetragnatha</em>, but some APIs
will require species names rather than the names of broader taxa like
genus or family. We can use the taxize package for this.</p>
<p>First we use get_uid() to get the identification number for the genus
in the taxonomy (drawn from NCBI).</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="va">uid</span> <span class="op">&lt;-</span> <span class="fu">get_uid</span><span class="op">(</span><span class="st">"Tetragnatha"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>No ENTREZ API key provided
 Get one via taxize::use_entrez()
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>══  1 queries  ═══════════════</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Retrieving data for taxon 'Tetragnatha'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>✔  Found:  Tetragnatha
══  Results  ═════════════════

• Total: 1 
• Found: 1 
• Not Found: 0</code></pre>
</div>
<p>Then we use the downstream() function to get all species in
<em>Tetragnatha</em>. This also works on larger taxa like families and
orders.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_uids</span> <span class="op">&lt;-</span> <span class="fu">downstream</span><span class="op">(</span><span class="va">uid</span>,downto<span class="op">=</span><span class="st">"species"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>No ENTREZ API key provided
 Get one via taxize::use_entrez()
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/
No ENTREZ API key provided
 Get one via taxize::use_entrez()
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/
No ENTREZ API key provided
 Get one via taxize::use_entrez()
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="va">species_uids</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">childtaxa_name</span></span>
<span><span class="va">species_names</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Tetragnatha paludicola"              
 [2] "Tetragnatha boydi"                   
 [3] "Tetragnatha cavaleriei"              
 [4] "Tetragnatha virescens"               
 [5] "Tetragnatha josephi"                 
 [6] "Tetragnatha gui"                     
 [7] "Tetragnatha bituberculata"           
 [8] "Tetragnatha shinanoensis"            
 [9] "Tetragnatha chauliodus"              
[10] "Tetragnatha cf. tincochacae DDC-2018"
[11] "Tetragnatha similis"                 
[12] "Tetragnatha riveti"                  
[13] "Tetragnatha puella"                  
[14] "Tetragnatha paschae"                 
[15] "Tetragnatha caudicula"               
[16] "Tetragnatha tanigawai"               
[17] "Tetragnatha ceylonica"               
[18] "Tetragnatha makiharai"               
[19] "Tetragnatha iriomotensis"            
[20] "Tetragnatha yesoensis"               
[21] "Tetragnatha cf. vermiformis JJA-2017"
[22] "Tetragnatha cf. moua sc_02763"       
[23] "Tetragnatha lauta"                   
[24] "Tetragnatha hasselti"                
[25] "Tetragnatha geniculata"              
[26] "Tetragnatha vermiformis"             
[27] "Tetragnatha shoshone"                
[28] "Tetragnatha caudata"                 
[29] "Tetragnatha dearmata"                
[30] "Tetragnatha squamata"                
[31] "Tetragnatha praedonia"               
[32] "Tetragnatha pinicola"                
[33] "Tetragnatha extensa"                 
[34] "Tetragnatha viridis"                 
[35] "Tetragnatha javana"                  
[36] "Tetragnatha obtusa"                  
[37] "Tetragnatha nigrita"                 
[38] "Tetragnatha rava"                    
[39] "Tetragnatha punua"                   
[40] "Tetragnatha nitens"                  
[41] "Tetragnatha moua"                    
[42] "Tetragnatha marquesiana"             
[43] "Tetragnatha macilenta"               
[44] "Tetragnatha montana"                 
[45] "Tetragnatha straminea"               
[46] "Tetragnatha guatemalensis"           
[47] "Tetragnatha elongata"                
[48] "Tetragnatha stelarobusta"            
[49] "Tetragnatha perkinsi"                
[50] "Tetragnatha pallescens"              
[51] "Tetragnatha limu"                    
[52] "Tetragnatha laboriosa"               
[53] "Tetragnatha filiciphilia"            
[54] "Tetragnatha eurychasma"              
[55] "Tetragnatha obscura"                 
[56] "Tetragnatha kukuiki"                 
[57] "Tetragnatha macracantha"             
[58] "Tetragnatha kikokiko"                
[59] "Tetragnatha anuenue"                 
[60] "Tetragnatha maxillosa"               
[61] "Tetragnatha acuta"                   
[62] "Tetragnatha pilosa"                  
[63] "Tetragnatha quasimodo"               
[64] "Tetragnatha tantalus"                
[65] "Tetragnatha polychromata"            
[66] "Tetragnatha perreirai"               
[67] "Tetragnatha restricta"               
[68] "Tetragnatha brevignatha"             
[69] "Tetragnatha kamakou"                 
[70] "Tetragnatha kukuhaa"                 
[71] "Tetragnatha versicolor"              
[72] "Tetragnatha kauaiensis"              
[73] "Tetragnatha waikamoi"                
[74] "Tetragnatha mandibulata"             
[75] "Tetragnatha hawaiensis"              </code></pre>
</div>
</div>
<div class="section level2">
<h2 id="searching-by-taxon-querying-gbif-otol-and-ncbi">Searching by taxon: querying GBIF, OTOL, and NCBI<a class="anchor" aria-label="anchor" href="#searching-by-taxon-querying-gbif-otol-and-ncbi"></a>
</h2>
<p>The spocc package is all about occurrence data and lets you download
occurrences from all major occurrence databases. For GBIF access using
the spocc package, we use the occ() function.</p>
<p>We’ll put the species names we retrieved as the ‘query’ and specify
gbif as our target database in the ‘from’ field.</p>
<p>Since the occurrences from spocc are organized as a list of
dataframes, one per species, we use the bind_rows() function from dplyr
to combine this into a single dataframe.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">spocc</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">occurrences</span> <span class="op">&lt;-</span> <span class="fu">occ</span><span class="op">(</span>query <span class="op">=</span> <span class="va">species_names</span>, from <span class="op">=</span> <span class="st">'gbif'</span><span class="op">)</span></span>
<span><span class="va">occurrences_df</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">occurrences</span><span class="op">$</span><span class="va">gbif</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">occurrences_df</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 192
  name           longitude latitude issues prov  key   scientificName datasetKey
  &lt;chr&gt;              &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;          &lt;chr&gt;     
1 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
2 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
3 Tetragnatha p…     -156.     20.8 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
4 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
5 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
6 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
# ℹ 184 more variables: publishingOrgKey &lt;chr&gt;, installationKey &lt;chr&gt;,
#   hostingOrganizationKey &lt;chr&gt;, publishingCountry &lt;chr&gt;, protocol &lt;chr&gt;,
#   lastCrawled &lt;chr&gt;, lastParsed &lt;chr&gt;, crawlId &lt;int&gt;, basisOfRecord &lt;chr&gt;,
#   individualCount &lt;int&gt;, occurrenceStatus &lt;chr&gt;, taxonKey &lt;int&gt;,
#   kingdomKey &lt;int&gt;, phylumKey &lt;int&gt;, classKey &lt;int&gt;, orderKey &lt;int&gt;,
#   familyKey &lt;int&gt;, genusKey &lt;int&gt;, speciesKey &lt;int&gt;, acceptedTaxonKey &lt;int&gt;,
#   acceptedScientificName &lt;chr&gt;, kingdom &lt;chr&gt;, phylum &lt;chr&gt;, order &lt;chr&gt;, …</code></pre>
</div>
<p>The rotl package provides an interface to the Open Tree of Life. To
get a tree from this database, we have to match our species names to
OTOL’s ids.</p>
<p>We first use trns_match_names() to match our species names to OTOL’s
taxonomy.</p>
<p>Then we use ott_id() to get OTOL ids for these matched names.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resolved_names</span> <span class="op">&lt;-</span> <span class="fu">tnrs_match_names</span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span></span>
<span><span class="va">otol_ids</span> <span class="op">&lt;-</span> <span class="fu">ott_id</span><span class="op">(</span><span class="va">resolved_names</span><span class="op">)</span></span></code></pre>
</div>
<p>Finally, we get the tree containing these IDs as tips.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu">tol_induced_subtree</span><span class="op">(</span>ott_ids <span class="op">=</span> <span class="fu">ott_id</span><span class="op">(</span><span class="va">resolved_names</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tr</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/finding_data-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Unfortunately we can see that the phylogenetic relationships between
these species aren’t resolved in OTOL, so we may have to look elsewhere
for a phylogeny.</p>
<p>Let’s continue with the rentrez package, which provides an interface
to NCBI.</p>
<p>NCBI’s API works with genera and larger ranks, but has a harder time
with lists of species.</p>
<p>For this reason, let’s just search for <em>Tetragnatha</em> using the
‘term’ argument, which takes a string in a specific format describing
the query.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">search_results</span> <span class="op">&lt;-</span> <span class="fu">entrez_search</span><span class="op">(</span>db<span class="op">=</span><span class="st">"nucleotide"</span>, term<span class="op">=</span><span class="st">"Tetragnatha &amp; Genus[RANK]"</span><span class="op">)</span></span></code></pre>
</div>
<p>UNLIKE spocc but SIMILARLY to rotol, this first search returns a set
of IDs.</p>
<p>We then have to use entrez_fetch() on the IDs in search_results to
obtain the actual record information.</p>
<p>We specify ‘db’ as ‘nucleotide’ and ‘rettype’ as ‘fasta’ to get fasta
sequences</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu">entrez_fetch</span><span class="op">(</span>db <span class="op">=</span> <span class="st">"nucleotide"</span>, id <span class="op">=</span> <span class="va">search_results</span><span class="op">$</span><span class="va">ids</span>, rettype <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="searching-by-location-querying-gbif-and-ncbi">Searching by location: querying GBIF and NCBI<a class="anchor" aria-label="anchor" href="#searching-by-location-querying-gbif-and-ncbi"></a>
</h2>
<p>In addition to or instead of querying by species, we can search for
records in a given location in the world. Within spocc we need to pass
location queries directly to GBIF using ‘gbifopts’. We’ll search in a
lat-long bounding box around Hawaii.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#df &lt;- occ(query = species_names, from = 'gbif', has_coords=TRUE, gbifopts=list("continent"#="north_america"))</span></span>
<span><span class="va">occurences_df</span> <span class="op">&lt;-</span> <span class="fu">occ</span><span class="op">(</span>query <span class="op">=</span> <span class="va">species_names</span>, from <span class="op">=</span> <span class="st">'gbif'</span>, has_coords<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                     gbifopts<span class="op">=</span><span class="fu">list</span><span class="op">(</span><span class="st">"decimalLatitude"</span><span class="op">=</span><span class="st">'18.910361,28.402123'</span>,</span>
<span>                                   <span class="st">"decimalLongitude"</span><span class="op">=</span><span class="st">'-178.334698,-154.806773'</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
</div>
<p>For NCBI, we can add the state of Hawaii to the query string. HI is
the abbreviation for Hawaii.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">search_results</span> <span class="op">&lt;-</span> <span class="fu">entrez_search</span><span class="op">(</span>db<span class="op">=</span><span class="st">"nucleotide"</span>, term<span class="op">=</span><span class="st">"Tetragnatha &amp; HI[State]"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu">entrez_fetch</span><span class="op">(</span>db <span class="op">=</span> <span class="st">"nucleotide"</span>, id <span class="op">=</span> <span class="va">search_results</span><span class="op">$</span><span class="va">ids</span>, rettype <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span></code></pre>
</div>
<p>From this we obtain 5 nucleotide sequences from Hawaii
<em>Tetragnatha</em>.</p>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>What are some similarities and differences in how data is obtained
using the R APIs of GBIF, NCBI, and OTOL?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>NCBI and OTOL both require an initial search to obtain record IDs,
then a second fetch to obtain the actual record data for each ID</p>
<p>NCBI uses a string-based query where the search is a
specially-formated string, while GBIF uses an argument-based query where
the search is based on the various arguments you give to the R API
function</p>
<p>NCBI is capable of taking a genus and returning the records of all
its species, but OTOL and GBIF require obtaining the species names first
using a package like taxize</p>
<p>NCBI can take a state in the query, GBIF can take a country but not a
state.</p>
<p>GBIF data is formatted as a list of occurrences for each species
contained in the object we get from the query.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="sources-of-mdbd-small-data-attached-to-papers">Sources of MDBD: “small data” attached to papers<a class="anchor" aria-label="anchor" href="#sources-of-mdbd-small-data-attached-to-papers"></a>
</h1>
<p>A lot of useful data is not held in big databases, and is instead
attached to papers. Searching on Google and Google Scholar can be an
effective way to find such data. Let’s search for “mammal trait
database” on Google Scholar.</p>
<p>The first paper looks good, so let’s look at the Supplementary Data
which will contain their data.</p>
<p>Let’s download and unzip the data.</p>
<p>Reading the metadata file, we see that a certain data file contains
the traits. Great! Let’s read it into R.</p>
</div>
<div class="section level1">
<h1 id="recap">Recap<a class="anchor" aria-label="anchor" href="#recap"></a>
</h1>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Many sources exist online where multidimensional biodiversity data
can be obtained</li>
<li>APIs allow for fast and reproducible querying and downloading</li>
<li>While there are patterns in how these APIs work, there will always
be differences between databases that make using each API a bit
different.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section><section id="aio-care"><p>Content from <a href="care.html">The CARE principles and open data</a></p>
<hr>
<p> Last updated on 2023-06-05 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/care.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is Indigenous data sovereignty?</li>
<li>What are the CARE principles?</li>
<li>What is Local Contexts Hub, and how can we use it to implement the
CARE principles in our work?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal">
<li>Articulate the mission of Local Contexts Hub in the context of the
CARE and FAIR principles</li>
<li>Apply Bio-Cultural and Traditional Knowledge Notices appropriately
to datasets</li>
<li>Create an account on Local Contexts Hub</li>
<li>Create a project on Local Contexts Hub</li>
<li>Query the Local Contexts Hub API to link a dataset with project IDs
on Local Contexts Hub</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<section id="indigenous-data-soverginty"><h2 class="section-heading">Indigenous data soverginty<a class="anchor" aria-label="anchor" href="#indigenous-data-soverginty"></a>
</h2>
<hr class="half-width"></section><section id="the-care-principles"><h2 class="section-heading">The CARE Principles<a class="anchor" aria-label="anchor" href="#the-care-principles"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="collective-benefit">Collective benefit<a class="anchor" aria-label="anchor" href="#collective-benefit"></a>
</h3>
</div>
<div class="section level3">
<h3 id="authority-to-control">Authority to control<a class="anchor" aria-label="anchor" href="#authority-to-control"></a>
</h3>
</div>
<div class="section level3">
<h3 id="responsibility">Responsibility<a class="anchor" aria-label="anchor" href="#responsibility"></a>
</h3>
</div>
<div class="section level3">
<h3 id="ethics">Ethics<a class="anchor" aria-label="anchor" href="#ethics"></a>
</h3>
</div>
</section><section id="local-contexts-hub"><h2 class="section-heading">Local Contexts Hub<a class="anchor" aria-label="anchor" href="#local-contexts-hub"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="labels">Labels<a class="anchor" aria-label="anchor" href="#labels"></a>
</h3>
<p><a href="https://localcontexts.org/labels/traditional-knowledge-labels/" class="external-link uri">https://localcontexts.org/labels/traditional-knowledge-labels/</a></p>
</div>
<div class="section level3">
<h3 id="notices">Notices<a class="anchor" aria-label="anchor" href="#notices"></a>
</h3>
<p><a href="https://localcontexts.org/notices/aboutnotices/" class="external-link uri">https://localcontexts.org/notices/aboutnotices/</a></p>
<p><a href="https://localcontexts.org/notice-in-use-maine-edna/" class="external-link uri">https://localcontexts.org/notice-in-use-maine-edna/</a></p>
</div>
</section><section id="the-local-contexts-hub-api"><h2 class="section-heading">The Local Contexts Hub API<a class="anchor" aria-label="anchor" href="#the-local-contexts-hub-api"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>“Be FAIR and CARE”: The CARE principles (<strong>C</strong>ollective
benefit, <strong>A</strong>uthority to control,
<strong>R</strong>esponsibility, and <strong>E</strong>thics) are part
of a framework emphasizing Indigenous rights and interests in the
context of data sharing ecosystems.</li>
<li>Local Contexts Hub provides Labels and Notices for learning more
about and engaging with Indigenous data in accordance with the CARE
principles.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/role-model/multidim-biodiv-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.3" class="external-link">sandpaper (0.12.3)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://role-model.github.io/multidim-biodiv-data/aio.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://role-model.github.io/multidim-biodiv-data/aio.html",
  "identifier": "https://role-model.github.io/multidim-biodiv-data/aio.html",
  "dateCreated": "2023-06-05",
  "dateModified": "2023-06-05",
  "datePublished": "2023-06-05"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

