<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Multidimensional biodiversity data: management and analysis: Finding multi-dimensional biodiversity data</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback." style="text-decoration: unset">
          Â 
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link" style="color: #383838">Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="background: #FFC700; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/finding_data.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Multidimensional biodiversity data: management and analysis
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Multidimensional biodiversity data: management and analysis
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Multidimensional biodiversity data: management and analysis
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 71%" class="percentage">
    71%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 71%" aria-valuenow="71" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/finding_data.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="care.html">2. Indigenous Data Sovereignty and the CARE Principles</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="abundance-data.html">3. Abundance Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="traits-data.html">4. Trait data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="phylo-data.html">5. Summarizing phylogenetic data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="popgen-data.html">6. Working with population genetic data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="connecting-data-types.html">7. Connecting different data types</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        8. Finding multi-dimensional biodiversity data
        </span>
      
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="care-data-repos.html">9. CARE Principles and Data Repositories</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="connecting-data-types.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="care-data-repos.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="connecting-data-types.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Connecting different</a>
        <a class="chapter-link float-end" href="care-data-repos.html" rel="next">Next: CARE Principles and... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Finding multi-dimensional biodiversity data</h1>
        <p> Last updated on 2023-06-18 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/finding_data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>Where do multi-dimensional biodiversity data live online?</li>
<li>How can we use R APIs to quickly and reproducibly retrieve
data?</li>
<li>What about data thatâs not in a database?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal"><li>Effectively search for and find occurrence, phylogenetic, sequence,
and trait data online given a taxon or research question</li>
<li>Retrieve data from NCBI &amp; GBIF using their R APIs and apply API
principles to other databases</li>
</ol></div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="sources-of-multidimensional-biodiversity-data-large-open-access-databases">Sources of multidimensional biodiversity data: large open-access
databases<a class="anchor" aria-label="anchor" href="#sources-of-multidimensional-biodiversity-data-large-open-access-databases"></a></h1>
<p>Large open-access biological databases have been growing at an
incredible pace in recent years. Often they accumulate data from many
projects and organizations. These platforms are then used to share and
standardize data for wide use.</p>
<p>Many such databases exist, but here are some listed examples:</p>
<ul><li><p><a href="https://www.gbif.org/" class="external-link">GBIF</a> contains occurrence
records of individuals assembled from dozens of smaller
databases.</p></li>
<li><p><a href="https://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a> (National Center
for Biotechnology Information) database, which includes <a href="https://www.ncbi.nlm.nih.gov/genbank/" class="external-link">GenBank</a> is the largest
repository of genetic data in the world, but includes a lot of
biomedical data.</p></li>
<li><p><a href="https://tree.opentreeoflife.org/opentree/argus/opentree14.7@ott93302" class="external-link">OTOL</a>
(Open Tree of Life) is a database that combines numerous published trees
and taxonomies into one supertree</p></li>
<li><p><a href="https://geome-db.org/" class="external-link">GEOME</a> (Genomic Observatories
MetaDatabase) contains genetic data associated with specimens or samples
from the field.</p></li>
<li><p><a href="https://boldsystems.org/" class="external-link">BOLD</a> (Barcode of Life) is
a reference library and database for DNA barcoding</p></li>
<li><p><a href="https://eol.org/" class="external-link">EOL</a> (Encyclopedia of Life) is a
species-level platform containing data like taxonomic descriptions,
distribution maps, and images</p></li>
</ul><div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>Do you have a favorite database? It doesnât have to be in the list!
Why do you like it, or how does it fit your research?</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="whats-an-api">Whatâs an API?<a class="anchor" aria-label="anchor" href="#whats-an-api"></a></h2>
<p>An API is a set of tools that allows users to interact with a
database, such as reading, writing, updating, and deleting data. We
canât go over all of these databases in detail, but letâs cover some
examples that illustrate some principles of how these APIs work for
different types of biodiversity data.</p>
<p>We can do this using the example of Hawaiian arthropods. Hawaii is
known for its diversity of <em>Tetragnatha</em>, a genus of orb weaver
spiders. With this target in mind letâs download some
<em>occurrence</em> data from GBIF, some <em>sequence</em> data from
NCBI, and a <em>phylogeny</em> for <em>Tetragnatha</em> spiders from
OTOL.</p>
<div class="section level3">
<h3 id="using-apis-setup">Using APIs: Setup<a class="anchor" aria-label="anchor" href="#using-apis-setup"></a></h3>
<p>For this episode, weâll be working with some R packages that wrap the
APIs of online repositories. Letâs set those up now.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">spocc</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rentrez</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rotl</span><span class="op">)</span> </span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="getting-consistent-species-names-using-taxize">Getting consistent species names using <code>taxize</code>
<a class="anchor" aria-label="anchor" href="#getting-consistent-species-names-using-taxize"></a></h3>
<p>We will want to source records of the genus <em>Tetragnatha</em>, but
some APIs will require species names rather than the names of broader
taxa like genus or family. We can use the <code>taxize</code> package
for this. <code>taxize</code> links to a genus identification number,
drawn from NCBI, for each genus. We use the <code>get_uid()</code>
function to get this number:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uid</span> <span class="op">&lt;-</span> <span class="fu">get_uid</span><span class="op">(</span><span class="st">"Tetragnatha"</span><span class="op">)</span></span></code></pre>
</div>

<p>Then, we use the <code>downstream()</code> function to get a list of
all species in <em>Tetragnatha</em>.</p>
<div id="callout1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout1"></a>
</h3>
<div class="callout-content">
<p>This also works on larger taxa like families and orders.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_uids</span> <span class="op">&lt;-</span> <span class="fu">downstream</span><span class="op">(</span><span class="va">uid</span>,downto<span class="op">=</span><span class="st">"species"</span><span class="op">)</span></span>
<span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="va">species_uids</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">childtaxa_name</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="occurrence-data-from-gbif">Occurrence data from GBIF<a class="anchor" aria-label="anchor" href="#occurrence-data-from-gbif"></a></h3>
<p>The <code>spocc</code> package is all about occurrence data and lets
you download occurrences from all major occurrence databases. For GBIF
access using the spocc package, we use the <code>occ()</code>
function.</p>
<p>Weâll put the species names we retrieved as the <code>query</code>
and specify <code>gbif</code> as our target database in the
<code>from</code> field.</p>
<p>Since the occurrences from spocc are organized as a list of
dataframes, one per species, we use the <code>bind_rows()</code>
function from <code>dplyr</code> to combine this into a single
dataframe.</p>
<p>We can also specify a geographic bounding box of where weâd like to
look for occurrences. Here weâre using a bounding box that covers the
Hawaiian Islands.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">occurrences</span> <span class="op">&lt;-</span> <span class="fu">occ</span><span class="op">(</span>query <span class="op">=</span> <span class="va">species_names</span>, from <span class="op">=</span> <span class="st">'gbif'</span>, has_coords<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                   gbifopts<span class="op">=</span><span class="fu">list</span><span class="op">(</span><span class="st">"decimalLatitude"</span><span class="op">=</span><span class="st">'18.910361,28.402123'</span>,</span>
<span>                                 <span class="st">"decimalLongitude"</span><span class="op">=</span><span class="st">'-178.334698,-154.806773'</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># extract the data from gbif</span></span>
<span><span class="va">occurrences_gbif</span> <span class="op">&lt;-</span> <span class="va">occurrences</span><span class="op">$</span><span class="va">gbif</span><span class="op">$</span><span class="va">data</span> </span>
<span></span>
<span><span class="co"># the results in `$data` are a list with one element per species, </span></span>
<span><span class="co"># so we combine all those elements</span></span>
<span><span class="va">occurrences_df</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">occurrences_gbif</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot coordinates of occurrences colored by species</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">occurrences_df</span><span class="op">$</span><span class="va">longitude</span>,<span class="va">occurrences_df</span><span class="op">$</span><span class="va">latitude</span>,</span>
<span>     cex<span class="op">=</span><span class="fl">0.25</span>,ylab<span class="op">=</span><span class="st">"Latitude"</span>,xlab<span class="op">=</span><span class="st">"Longitude"</span>,</span>
<span>     col<span class="op">=</span><span class="fu">as.factor</span><span class="op">(</span><span class="va">occurrences_df</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/finding_map.png" class="figure mx-auto d-block"></figure><div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>How was the abundance data returned to us from the occ() function
organized?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>The occ() function returned an object of class occDat. It contains
members for each database you could have queried, like âgbifâ or
âebirdâ. Each of these members contains members called âdataâ which
holds the actual data, and âmetaâ which contains metadata on what we
searched for and how it was returned. The âdataâ object is a list where
each member of the list holds a dataframe of records for a different
species. As described we can use âbind_rowsâ from dplyr to create a
composite dataframe from this list.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="but-wait-what-is-occurrence-data">But wait, what is occurrence data?<a class="anchor" aria-label="anchor" href="#but-wait-what-is-occurrence-data"></a></h3>
<p>Occurrence is related to abundance in that it also describes how
species are distributed in space but is different in some important
ways.</p>
<p>A single abundance data point describes the amount of individuals of
a species in a certain area.</p>
<p>But a single occurrence data point describes the presence of a SINGLE
individual in a location.</p>
<p>Occurrence data is vastly more common than abundance data online, but
often comes with much greater bias in the way it captures the
distribution of biodiversity. It is often âincidentalâ. A citizen
scientist happening to record a new butterfly they saw on their morning
hike (occurrence) or a researcher electing to record a rare beetle they
were looking for (occurrence) are much less systematic and more biased
than the field surveys specifically intended to monitor abundance.</p>
<p>Due to this bias it is often used just to indicate presence-absence
of a species rather than the magnitude of its population size.</p>
<p>It can be possible to obtain âabundance-likeâ information from
occurrences, but it requires VERY careful consideration of bias and
uncertainty.</p>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>Have you used occurrence data in your work? What do you think are
some pros and cons?</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="phylogenetic-trees-from-the-open-tree-of-life">Phylogenetic trees from the Open Tree of Life<a class="anchor" aria-label="anchor" href="#phylogenetic-trees-from-the-open-tree-of-life"></a></h3>
<p>First weâll get a clean list of taxonomic names from the GBIF
results. We could use our initial names we got from <code>taxize</code>,
but using our GBIF results conveniently narrows things down to Hawaii
species only.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_names_hawaii</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="va">occurrences_df</span><span class="op">$</span><span class="va">name</span><span class="op">)</span></span>
<span><span class="va">species_names_resolved</span> <span class="op">&lt;-</span> <span class="fu">gnr_resolve</span><span class="op">(</span><span class="va">species_names_hawaii</span>, best_match_only <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                          canonical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">species_names_hawaii</span> <span class="op">&lt;-</span> <span class="va">species_names_resolved</span><span class="op">$</span><span class="va">matched_name2</span></span></code></pre>
</div>
<p>Now we can use the <code>rotl</code> package. The <code>rotl</code>
package provides an interface to the Open Tree of Life. To get a tree
from this database, we have to match our species names to OTOLâs
ids.</p>
<p>We first use <code>trns_match_names()</code> to match our species
names to OTOLâs taxonomy.</p>
<p>Then we use <code>ott_id()</code> to get OTOL ids for these matched
names.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_names_otol</span> <span class="op">&lt;-</span> <span class="fu">tnrs_match_names</span><span class="op">(</span><span class="va">species_names_hawaii</span><span class="op">)</span></span>
<span><span class="va">otol_ids</span> <span class="op">&lt;-</span> <span class="fu">ott_id</span><span class="op">(</span><span class="va">species_names_otol</span><span class="op">)</span></span></code></pre>
</div>
<p>Finally, we get the tree containing these IDs as tips.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu">tol_induced_subtree</span><span class="op">(</span>ott_ids <span class="op">=</span> <span class="va">otol_ids</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tr</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/finding_phylo.png" class="figure mx-auto d-block"></figure><p>Unfortunately we can see that the phylogenetic relationships between
these species arenât resolved in OTOL, so we may have to look elsewhere
for a phylogeny. This is an example of how not all the data we are
looking for are in large repositories, especially higher resolution
phylogenies.</p>
</div>
<div class="section level3">
<h3 id="sequence-data-from-ncbi">Sequence data from NCBI<a class="anchor" aria-label="anchor" href="#sequence-data-from-ncbi"></a></h3>
<p>Letâs continue with the <code>rentrez</code> package, which provides
an interface to NCBI. The two usual steps for querying from NCBI is an
ID search using entrez_search() and a records fetch using those IDs in
entrez_fetch(). Sound a bit familiar?</p>
<p>Weâll again search for the species in our Hawaii occurrences.</p>
<p>The main arguments of the entrez_search() function are âdbâ, the
database to look in, and âtermâ a string representing the whole query.
While the occ() function had multiple fields for different ways we are
narrowing our search, the âtermâ field wraps this up in a single
string.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we're using `species_names` that we made previously to make a long</span></span>
<span><span class="co"># string of names that will be sent to NCBI via the ENTREZ API</span></span>
<span><span class="va">termstring</span> <span class="op">&lt;-</span> <span class="fu">paste</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">'%s[ORGN]'</span>, <span class="va">species_names_hawaii</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">' OR '</span><span class="op">)</span></span>
<span></span>
<span><span class="va">search_results</span> <span class="op">&lt;-</span> <span class="fu">entrez_search</span><span class="op">(</span>db<span class="op">=</span><span class="st">"nucleotide"</span>, term <span class="op">=</span> <span class="va">termstring</span><span class="op">)</span></span></code></pre>
</div>
<div id="callout2" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout2"></a>
</h3>
<div class="callout-content">
<p>It is also possible to query for ranks higher than species in
entrez_search()</p>
</div>
</div>
</div>
<p><em>Unlike</em> <code>spocc</code> but <em>similarly</em> to
<code>rotl</code>, this first search returns a set of IDs.</p>
<p>We then have to use <code>entrez_fetch()</code> on the IDs in
search_results to obtain the actual record information.</p>
<p>We specify âdbâ as ânucleotideâ and ârettypeâ as âfastaâ to get FASTA
sequences</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu">entrez_fetch</span><span class="op">(</span>db <span class="op">=</span> <span class="st">"nucleotide"</span>, id <span class="op">=</span> <span class="va">search_results</span><span class="op">$</span><span class="va">ids</span>, </span>
<span>                          rettype <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># just look at the first part of this *long* returned string</span></span>
<span><span class="co"># substr gets characters 1 through 1148 of the FASTA, and cat simply prints it with line breaks</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="fu">substr</span><span class="op">(</span><span class="va">sequences</span>,<span class="fl">1</span>,<span class="fl">1148</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion3"></a>
</h3>
<div class="callout-content">
<p>What are some similarities and differences in how data is queried
using the R APIs of GBIF, NCBI, and OTOL?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>NCBI and OTOL both require an initial search to obtain record IDs,
then a second fetch to obtain the actual record data for each ID</p>
<p>NCBI uses a string-based query where the search is a
specially-formated string, while GBIF uses an argument-based query where
the search is based on the various arguments you give to the R
function.</p>
<p>NCBI is capable of taking a genus and returning the records of all
its species, but OTOL and GBIF require obtaining the species names first
using a package like taxize</p>
<p>GBIF data is formatted as a list of occurrences for each species
contained in the object we get from the query.</p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="sources-of-mdbd-databases-without-apis">Sources of MDBD: Databases without APIs<a class="anchor" aria-label="anchor" href="#sources-of-mdbd-databases-without-apis"></a></h1>
<p>Many excellent databases exist without specialized APIs or R packages
to facilitate their use.</p>
<p>For example, the BioTIME database is a compilation of timeseries of
species ocurrence, abundance, and biomass data. It is <a href="https://biotime.st-andrews.ac.uk/" class="external-link">openly available</a> via the
University of St.Â Andrews. The North American Breeding Bird Survey, run
by the USGS, contains data on bird species abundances across the United
States and Canada going back more than 50 years. These data are hosted
on <a href="https://www.sciencebase.gov/catalog/item/52b1dfa8e4b0d9b325230cd9" class="external-link">ScienceBase</a>.</p>
<p>Some of these databases can be accessed via the <a href="https://retriever.readthedocs.io/en/latest/" class="external-link">Data Retriever and
accompanying R package the <code>rdataretriever</code>.</a></p>
<p>Some are findable via <a href="https://search.dataone.org/data" class="external-link">DataONE</a> and the <a href="https://edirepository.org/" class="external-link">Ecological Data Initiative</a>.</p>
<p>Others can be downloaded directly from the source.</p>
</div>
<div class="section level1">
<h1 id="sources-of-mdbd-small-data-attached-to-papers">Sources of MDBD: âsmall dataâ attached to papers<a class="anchor" aria-label="anchor" href="#sources-of-mdbd-small-data-attached-to-papers"></a></h1>
<p>Finally, A lot of useful data is not held in databases, and is
instead attached to papers. Searching manuscript databases or Google
Scholar can be an effective way to find such data. You can filter Web of
Science to find entries with âassociated dataâ.</p>
<p>However, this can still be a somewhat heterogeneous and
time-consuming strategy!</p>

</div>
<div class="section level1">
<h1 id="recap">Recap<a class="anchor" aria-label="anchor" href="#recap"></a></h1>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Many sources exist online where multidimensional biodiversity data
can be obtained</li>
<li>APIs allow for fast and reproducible querying and downloading</li>
<li>While there are patterns in how these APIs work, there will always
be differences between databases that make using each API a bit
different.</li>
<li>Standalone databases and even the supplemental data in manuscripts
can also be re-used.</li>
</ul></div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="connecting-data-types.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="care-data-repos.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="connecting-data-types.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Connecting different</a>
        <a class="chapter-link float-end" href="care-data-repos.html" rel="next">Next: CARE Principles and... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/finding_data.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/role-model/multidim-biodiv-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:biodataworkshop2023@gmail.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.4" class="external-link">sandpaper (0.12.4)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://role-model.github.io/multidim-biodiv-data/finding_data.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Finding multi-dimensional biodiversity data",
  "creativeWorkStatus": "active",
  "url": "https://role-model.github.io/multidim-biodiv-data/finding_data.html",
  "identifier": "https://role-model.github.io/multidim-biodiv-data/finding_data.html",
  "dateCreated": "2023-06-18",
  "dateModified": "2023-06-18",
  "datePublished": "2023-06-18"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

