<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Multidimensional biodiversity data: management and analysis: Summarizing phylogenetic data</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/phylo-data.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Multidimensional biodiversity data: management and analysis
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Multidimensional biodiversity data: management and analysis
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Multidimensional biodiversity data: management and analysis
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 75%" class="percentage">
    75%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 75%" aria-valuenow="75" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/phylo-data.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="finding_data.html">2. Finding multi-dimensional biodiversity data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="care.html">3. The CARE principles and open data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="abundance-data.html">4. Abundance Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="traits-data.html">5. Trait data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="popgen-data.html">6. Working with population genetic data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        7. Summarizing phylogenetic data
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction-to-phylogenetic-data">Introduction to phylogenetic data</a></li>
<li><a href="#working-with-phylogenetic-data-in-r">Working with phylogenetic data in R</a></li>
<li><a href="#cleaning-phylogenetic-data">Cleaning phylogenetic data</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="connecting-data-types.html">8. Connecting different data types</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="popgen-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="connecting-data-types.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="popgen-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Working with</a>
        <a class="chapter-link float-end" href="connecting-data-types.html" rel="next">Next: Connecting different... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Summarizing phylogenetic data</h1>
        <p> Last updated on 2023-06-06 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/phylo-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What information is contained in phylogenies?</li>
<li>How to summarize and interpret phylogenies using Hill numbers?</li>
<li>What are the common phylogeny data formats?</li>
<li>How do you import and manipulate phylogeny data in R?</li>
<li>How do you visualize phylogenies in R?</li>
<li>How do you calculate Hill numbers with phylogenetic data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants will be able to:</p>
<ol style="list-style-type: decimal"><li>Identify key features of different phylogenetic data formats</li>
<li>Import, clean and manipulate phylogenetic data in the R
environment</li>
<li>Visualize phylogenies</li>
<li>Calculate and interpret phylogenetic Hill numbers</li>
</ol></div>
</div>
</div>
</div>
</div>
<section id="introduction-to-phylogenetic-data"><h2 class="section-heading">Introduction to phylogenetic data<a class="anchor" aria-label="anchor" href="#introduction-to-phylogenetic-data"></a>
</h2>
<hr class="half-width"><p>In this episode, we will explore how to extract information from
phylogenetic trees in order to complement our hypotheses and inferences
about processes shaping biodiversity patterns. Phylogenetic trees have
information about how the species in a taxonomic group are related to
each other and how much relative evolutionary change has accumulated
among them. Since local communities differ in their phylogenetic
composition, this information can give insights on why communities are
how they are.</p>
<p>[HERE INCLUDE VISUALS OF PHYLOGENIES]</p>
<p>The relative phylogenetic distance among species in a community as
well as the distribution of the amount of evolutionary history
(represented by the length of the branches in a phylogeny) are a result
of different factors such as the age since the initial formation of the
community and the rate of macroevolutionary processes such as speciation
and extinction. For instance, young communities that are dominated by
closely related species and show very short branch lengths may suggest a
short history with few colonization events and high rates of local
speciation; alternatively, if the same young communities harbor
distantly related species with longer branch lengths, it may suggest
that most of the local diversity was generated by speciation elsewhere,
along with many colonization by species from different points in the
phylogeny. Coupled with information on ecological traits and rates of
macroevolutionary processes, these patterns also allow to test for
hypotheses regarding, for instance, ecological filtering or niche
conservatism.</p>
<p>Summarizing this phylogenetic information (i.e., phylogenetic
distance and distribution of branch lenghts) is therefore crucial for
inference. As we have seen in previous episodes, the use of <a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Hill
numbers</a> is an informative approach to summarize biodiversity.</p>
<p>[MORE HERE ON MEANING OF PHYLOGENETIC HILL, RELATING TO PHYLO
CHARACTERISTICS]</p>
</section><section id="working-with-phylogenetic-data-in-r"><h2 class="section-heading">Working with phylogenetic data in R<a class="anchor" aria-label="anchor" href="#working-with-phylogenetic-data-in-r"></a>
</h2>
<hr class="half-width"><div class="section level3">
<h3 id="importing-phylogenetic-data">Importing phylogenetic data<a class="anchor" aria-label="anchor" href="#importing-phylogenetic-data"></a></h3>
<p>Several file formats exist to store phylogenetic information. The
most common formats are the <a href="https://evomics.org/resources/tree-formats/" class="external-link">Newick and Nexus
formats</a>. Both these formats are plain text files storing different
levels of information about the taxa relationship and evolutionary
history. <code>Newick</code> files are the standard for representing
trees in a computer-readable form, as they can be extremely simple and
therefore do not take up much memory. <code>Nexus</code> file, on the
other hand, are composed by different blocks regarding different types
of information and can be used to store DNA alignments, phylogenetic
trees, pre-defined groupings of taxa, or everything at once. Since they
are a step ahead in complexity, we will stick with <code>Newick</code>
files for now.</p>
<p><code>Newick</code> files store the information about the clades in a
tree by representing each clade within two parentheses, with tips
separated by <code>,</code>. We also add the symbol <code>;</code> at
the end to represent the end of the information for that phylogenetic
tree.</p>
<p>The basic structure of a tree in a <code>Newick</code> format is
therefore as follows:</p>
<p>((A,B),C);</p>
<p>The notation above indicates that: 1) we have three taxa in our tree,
named <code>A</code>, <code>B</code> and <code>C</code>; 2)
<code>A</code> and <code>B</code> form one clade (A,B); 3) the (A,B)
clade is sister to the <code>C</code> clade (we represent that by adding
another set of parentheses and a <code>,</code> separating (A,B) from
C).</p>
<p>In addition, <code>Newick</code> files can also store information on
the branch leading to it tip and node. We do that by adding
<code>:</code> after each tip/node.</p>
<p>((A:0.5,B:0.5):0.5,C:1);</p>
<p>The notation above indicates that: 1) the branches containing A and B
have each a length = 0.5; 2) the branch that leads to the node
connecting A and B also has length = 0.5; 3) the branch leading to C has
length = 1.</p>
<p>The notation above is what we import into R to start working with and
manipulating our phylogenetic tree. For that goal, we will use the <a href="https://academic.oup.com/bioinformatics/article/20/2/289/204981" class="external-link">ape</a>
package:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span></code></pre>
</div>
<p>To import our tree, we will be using the function
<code>read.tree()</code> from the <code>ape</code> package. In the case
of simple trees as the one above, we could directly create them within R
by giving that notation as a <code>character</code> value to this
function, using the <code>text</code> argument, as shown below:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text <span class="op">=</span> <span class="st">'((A:0.5,B:0.5):0.5,C:1);'</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, we can visually inspect our tree using the <code>plot()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">example_tree</span><span class="op">)</span></span></code></pre>
</div>
<p>Can you visualize the text notation in that image? We can see the
same information: A is closer related to B than C, and the branches
leading to A and B have half the length of the branch leading to C.</p>
<p>The <code>read.tree()</code> function creates an object of class
<code>phylo</code>. We can further investigate this object by calling it
in our console:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span></span></code></pre>
</div>
<p>The printed information shows us that we have a phylogenetic tree
with 3 tips and 2 internal nodes, where the tip labels are “A, B, C”. We
also are informed that this tree is rooted and has branch lengths.</p>
<p>One way to access the components of this object and better explore it
is to use <code>$</code> after the object name. Here, it will be
important for us to know a little bit more about where the information
about tip labels and branch lengths are stored in that
<code>phylo</code> objecy. Easy enough, we can access that by calling
<code>tip.labels</code> and <code>edge.length</code> after
<code>$</code>.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="va">example_tree</span><span class="op">$</span><span class="va">edge.length</span></span></code></pre>
</div>
<p>Now that we learned how to import and visualize trees in R, let’s
bring the phylogeny for the communities we are working with in this
workshop. Our data so far consists of <a href="abundance-data.html">abundances</a> and <a href="traits-data.html">traits</a> of several taxa of arthropods
collected across three islands in the Hawaian archipelago. Let’s work
though importing phylogenetic information for these species.</p>
<p>First, we will get the OTT IDs for the species in our data:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rotl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/role-model/multidim-biodiv-data/rmd-review/episodes/data/abundances_raw.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="va">abundances</span><span class="op">$</span><span class="va">GenSp</span></span>
<span></span>
<span><span class="va">resolved_names</span> <span class="op">&lt;-</span> <span class="fu">tnrs_match_names</span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove names that aren't a perfect match</span></span>
<span></span>
<span><span class="va">good_names</span> <span class="op">&lt;-</span> <span class="va">resolved_names</span><span class="op">[</span> <span class="op">!</span><span class="va">resolved_names</span><span class="op">$</span><span class="va">approximate_match</span>,<span class="op">]</span></span>
<span><span class="va">good_names</span> <span class="op">&lt;-</span> <span class="va">good_names</span><span class="op">[</span> <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">good_names</span><span class="op">$</span><span class="va">approximate_match</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove a couple of ott ids that seem to confuse tol_induced_subtree</span></span>
<span></span>
<span><span class="va">good_names</span> <span class="op">&lt;-</span> <span class="va">good_names</span><span class="op">[</span> <span class="va">good_names</span><span class="op">$</span><span class="va">ott_id</span> <span class="op">!=</span> <span class="st">"3574707"</span>, <span class="op">]</span></span>
<span><span class="va">good_names</span> <span class="op">&lt;-</span> <span class="va">good_names</span><span class="op">[</span> <span class="va">good_names</span><span class="op">$</span><span class="va">ott_id</span> <span class="op">!=</span> <span class="st">"737065"</span>, <span class="op">]</span></span></code></pre>
</div>
<p>Next, we can query the Open Tree of Life for those IDs using the
<code>tol_induced_subtree</code> function.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arthro_tree</span> <span class="op">&lt;-</span> <span class="fu">tol_induced_subtree</span><span class="op">(</span>ott_ids <span class="op">=</span> <span class="va">good_names</span><span class="op">$</span><span class="va">ott_id</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arthro_tree</span></span>
<span><span class="fu">class</span><span class="op">(</span><span class="va">arthro_tree</span><span class="op">)</span></span></code></pre>
</div>
<p>This new object is way larger than the previous once, being a “real”
phylogeny and all. You can inspect it again by directly calling the
object <code>phylogeny</code>. To plot it, we will use the
<code>type</code> argument to modify how our tree will be displayed.
Here, we used the option <code>'fan'</code>, to display a circular
phylogeny (slightly better to show such a large phylogeny in the
screen).</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">arthro_tree</span>, type <span class="op">=</span> <span class="st">'fan'</span>, show.tip.label <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
<p>How do we combine all this information with the community datasets we
have so far for our three islands? First, we will have to perform some
cleaning.</p>
</div>
</section><section id="cleaning-phylogenetic-data"><h2 class="section-heading">Cleaning phylogenetic data<a class="anchor" aria-label="anchor" href="#cleaning-phylogenetic-data"></a>
</h2>
<hr class="half-width"><p>Often, when we want to include the phylogenetic relatedness of
organisms in our biodiversity analyses, we will run into published
phylogenetic trees for our focal taxon (that is the case of the tree we
are using here). A couple issues arise in these cases. First, and maybe
most importantly, we need to make sure that the taxonomic names used
across the different datasets in our analyses (phylogenies included) are
the same. A second issue is that, depending on the resolution of the
published phylogeny, it may be the case that some taxa in our community
dataset may not be explicitly represented in the phylogeny. In these
cases, we will want to assign those “missing taxa” to one of the tips
present in the tree, based on their known genus and/or family
assignments.</p>
<p><strong>RMD commenting May 26. With taxize down, and using
<code>rotl</code> for both name resolution and getting the phylogeny,
there aren’t going to be mismatched names in the phylogeny.</strong></p>
<p><strong>But, there is the potential for missing names…</strong></p>
<p><strong>For now, RMD is commenting out the phylo cleaning section. If
we end up continuing with <code>rotl</code> for both cleaning + phylo, I
think there will be an opportunity to do name/broken taxon fixing. Some
of this is covered in an R Open Tree of Life Carpentries lesson here: <a href="https://mctavishlab.github.io/R_OpenTree_tutorials/03-broken-taxa/index.html" class="external-link uri">https://mctavishlab.github.io/R_OpenTree_tutorials/03-broken-taxa/index.html</a>
(which is what RMD is working from).</strong></p>
<!-- In this section, we will deal with these issues in order to prepare our data for community inferences. Our first issue will be dealt with in a challenge! -->
<!-- ### Resolving taxonomic names -->
<!-- ::: challenge -->
<!-- In previous episodes, we have used the package `taxize` to clean and standardize the taxonomic names in out dataset. Here, we will apply the same approach to the tip names of our phylogeny. We need to make sure the names we will accept from `taxize` will match the names we have accepted in previous episodes. Finally, after we decide on the names, we need to add them to the `phylogeny` objects, where the tip names are stored. -->
<!-- ::: solution -->
<!-- First, we check the names using the `gnr_resolve()` function. In `phylogeny`, tip names are stored in `phylogeny$tip.label`, so we provide that vector to our function. -->
<!-- ```{r phylo-resolve-sol1} -->
<!-- library(taxize) -->
<!-- name_resolve <- gnr_resolve(phylogeny$tip.label, best_match_only = TRUE, canonical = TRUE) -->
<!-- head(name_resolve) -->
<!-- ``` -->
<!-- We can see what taxa are in conflict with the names retrieved by `taxize` by subsetting `name_resolve` based on boolean matching. -->
<!-- ```{r phylo-resolve-sol2} -->
<!-- mismatches_phylo <- name_resolve[name_resolve$user_supplied_name != name_resolve$matched_name2, c("user_supplied_name", "matched_name2")] -->
<!-- mismatches_phylo -->
<!-- ``` -->
<!-- From here, we can see that apparently there are no misspellings in our tree (as expected from a published phylogeny). Most of the mismatches seem to be related to changes in the taxonomic treatment of species. This is a common situation, since taxonomic treatments are often revised after new trees are published; it might be the case that `nameA` in our dataset was treated as `nameB` at the time the phylogeny was created. -->
<!-- In the previous episodes, we used the names provided by `taxize` directly. Therefore, here we can just add those names directly to the phylogeny. -->
<!-- ```{r phylo-resolve-sol3} -->
<!-- phylogeny$tip.label == name_resolve$matched_name2 -->
<!-- ``` -->
<!-- :::::::::::: -->
<!-- :::::::::::: -->
<!-- ### Resolving missing tip names -->
<!-- The next issue we need to address is the fact that we may have species in our dataset that are not present in the phylogenetic tree we are using. We therefore need to 1) find out which species are not in the tree; 2) assign them to tips that are present in the tree. -->
<!-- For this, let's refer back to our original species list and cross-check it against the list of names we were able to find good OTT IDs for. -->
<!-- ```{r} -->
<!-- all_names <- tolower(abundances$GenSp) -->
<!-- found_names <- good_names$search_string -->
<!-- unfound_names <- setdiff(all_names, found_names) -->
<!-- ``` -->
<!-- The taxonomic names in both the abundance data and in the phylogenetic tree were corrected using the same database from `taxize`; this therefore means that any names from our abundance data that are not found in the tree simply represent taxa that are not included in this phylogeny. We can use the boolean operator `%in%` coupled with `!` to check `species_names` that are not included in the `phylogeny$tip.label`. In summary, the expression `A %in% B` would return the positio of the elements in vector A that are present in vector B. The `!` (NOT) operator returns the opposite of that expression, in a way that `!(A %in% B)` will return the position of elements in vector A that are *NOT* present in vector B (precisely what we need right now). -->
<!-- ```{r} -->
<!-- not_present <- !(species_names %in% phylogeny$tip.label) -->
<!-- ``` -->
<!-- The object `not_present` is now a vector of the positions (or indexes) of the elements in `species_name` that are not found in `phylogeny$tip.label`. Let's then check their names by using these indexes to subset `species_name`: -->
<!-- ```{r} -->
<!-- species_name[not_present] -->
<!-- ``` -->
<!-- Here, we can decide how we are going to assign these species to tips in our tree. For instance, we see that we do not have _XXX_ in our tree, but we have _XXX_, which belongs to the same genus. Similarly... -->
<!-- We can then give new names to these species, based on what tips in our tree we want them to match to. To do so, we will first duplicate our `species_name` vector, so we can keep the original names (just in case we want to check the original names again in the future). -->
<!-- ```{r} -->
<!-- species_name[not_present] <- c() -->
<!-- new_names[not_present] <- c() -->
<!-- ``` -->
<!-- Now that we have a vector of updated named, all we need to do is modify those names in our site-by-species matrix. Similarly, we will first copy our matrix into another object (in order to keep the original abundance matrix intact), and then we will rename the columns of this new object based on the modified names. -->
<!-- ```{r} -->
<!-- phylo_wide <- abundance_wide -->
<!-- colnames(phylo_wide) <- new_names -->
<!-- ``` -->
<!-- :::::::::::::::::: instructor -->
<!--     Another issue that can arise with phylogenies is that when we assign species to a tip, we may end up with duplicates in the site-by-species matrix. We will not show this here, but it might be something important to bring up. -->
<!-- ::::::::::::::::::::::::::::::::::::::::::::::::::: -->
<!-- After solving these issues, we now have 1) a phylogenetic tree with tip labels taxonomically cleaned; 2) a site-by-species matrix where all species are represented in our tree. These two objects are the two bits of information that we need to summarize the phylogenetic diversity and the distribution of branch lengths across our focal communities. -->
<div class="section level3">
<h3 id="summarizing-with-hill-numbers">Summarizing with hill numbers<a class="anchor" aria-label="anchor" href="#summarizing-with-hill-numbers"></a></h3>
<p>In this section, we will extract some summary statistics about the
pattern of phylogenetic diversity in our communities. As we discussed
above, the relative phylogenetic distance among species and the
distribution of this distance can give insights into processes of
community assembly. Here, we will make use of <a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Hill
numbers</a> to extract summaries of phylogenetic distances. To
understand what they mean, let’s first explore their properties in a
simple example.</p>
<p>Let’s assume we have a community with eight taxa: A through H. We
will create two possibilities of phylogenetic relationship among
them.</p>
<p>Tree 1 will represent a totally unbalanced pattern…</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree1</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(A:7,(B:6,(C:5,(D:4,(E:3,(F:2,(G:1,H:1):1):1):1):1):1):1);'</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span></span></code></pre>
</div>
<p>whereas tree 2 will represent a totally balanced pattern.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree2</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(((A:1,B:1):1,(C:1,D:1):1):1,((E:1,F:1):1,(G:1,H:1):1):1);'</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tree2</span><span class="op">)</span></span></code></pre>
</div>
<p>Notice that both trees are ultrametric: all tips align at the end,
representing extant taxa, which means branch lengths here represent the
amount of time that has passed between each node. The difference between
the first and the second tree resides in the fate of each new lineage at
a node. In the first tree, at each diversification event, one of the
lineages always persists till the present with no change while the other
undergoes another round of diversification. In the second tree, both
lineages from each node undergo a new split. The consequence is that in
tree 2, all extant species result from recent diversification (i.e.,
they have a short evolutionary history before coalescing into their
ancestor), whereas in tree 1 we have a mix of old and recent
lineages.</p>
<p>To visualize how these different patterns reflect in the diversity
metric, we will calculate the phylogenetic Hill number using the
function <code>hill_phylo</code> from the <code>hillR</code> package.
This function takes in a global phylogeny as well as a site-by-species
matrix, and returns phylo Hill numbers for each site based on which
species are present there. For our fictional example, we will create a
matrix with only one site where all eight species in the tree are
present. We do that by creating a vector of the value <code>1</code>
repeated 8 times, transforming that to a dataframe, and then naming the
columns with the names of the species we take from our fictional
trees.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span>,<span class="st">'D'</span>,<span class="st">'E'</span>,<span class="st">'F'</span>,<span class="st">'G'</span>,<span class="st">'H'</span><span class="op">)</span></span></code></pre>
</div>
<p>Using that matrix, we will calculate Hill numbers of order 0 to 2 for
both trees:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span><span class="va">tree1_h0</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">tree1_h1</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tree1_h2</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">tree2_h0</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">tree2_h1</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tree2_h2</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s format this as a dataframe so we can visualize these numbers:
the first column will be the order (our X axis), the second column will
be the Hill values (our Y axis) and the third column will be the tree it
belongs to (which we will use to draw different colored lines for
comparison).</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_h</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>order <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                      hill <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="va">tree1_h0</span>,<span class="va">tree1_h1</span>,<span class="va">tree2_h2</span>,<span class="va">tree2_h0</span>,<span class="va">tree2_h1</span>,<span class="va">tree2_h2</span><span class="op">)</span>,</span>
<span>                      tree <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">'tree1'</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="st">'tree2'</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we will call <code>plot</code> to generate a line plot for the
values of the first tree, and then use <code>lines</code> to add a line
containing the values for the second tree (we used boolean matching
subsetting to choose what to plot at each time)</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Plotting values for tree 1</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkred'</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting values for tree 2</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'red'</span><span class="op">)</span></span></code></pre>
</div>
<p>From this plot, we can notice a few patterns:</p>
<ol style="list-style-type: decimal"><li>The hill values for tree 1 are usually higher than those for tree 2,
indicating that our community contains a deeper evolutionary history
with tree 1 (i.e., lineages containg longer branches);</li>
<li>The value of the hill numbers drop as the order goes up. This
happens because higher orders are putting less and less emphasis on
branch lengths that are not so common (like, for instance, the short
branch lenghts in tree 1). In order words, higher order emphasize what
is more common.</li>
<li>The rate of change of the value as we increase the order is also
different among trees. Tree 1 is a more uneven tree; it is composed
mostly of somewhat long branches, and few short branches. So, when we
change the order, we notice a significant change in the value (at least
more significant than the change seen in tree2), because there’s a part
of the tree we are starting to disconsider (i.e., those short branch
lengths that are not so common). We observe a smaller change in tree 2
because all branches have very similar lenghts, so changing the order
does not affect the overall diversity (in other words, all branch
lenghts are equally common, so emphasize more or less common lenghts
does not change much the values).</li>
</ol><p>What would happen if the taxa in our communities had different
relative abundances? In this case, not only the relative frequency of a
branch length would affect our summary statistic, but also how abundant
is the taxon that has each branch length. For instance, in tree 1, even
though long branches are over represented, maybe the species with long
branches are actually super rare in our community, and the short-branch
species are actually super abundant. As you can probably realize, this
difference suggests something about the evolutionary history of our
community, in this case that the most abundant species have a very
recent evolutionary history. Similarly, even though branch lengths are
very even in tree 2, if some of the species is more abundant than
others, it suggests that the evolutionary history of the tree is
unevenly represented in our community.</p>
<p>Conveniently, the <code>hill_phylo</code> function will take into
account both the relative frequency of branch lenghts <em>AND</em> the
abundance of extant taxon in the phylogeny we provide. For our simple
example, we can see how that changes our results by creating
site-by-species matrix where abundance goes up from species A to species
H (for simplicity, we’ll just use values from 1 to 8).</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span>,<span class="st">'D'</span>,<span class="st">'E'</span>,<span class="st">'F'</span>,<span class="st">'G'</span>,<span class="st">'H'</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, let’s calculate the same hill numbers with this new community.
We save the values to objects with <code>_rel</code> at the end, to
highlight in this case the taxa had different <em>rel</em>ative
abundances.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree1_h0_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">tree1_h1_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tree1_h2_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree1</span>,q<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">tree2_h0_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">tree2_h1_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tree2_h2_rel</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">mat</span>,<span class="va">tree2</span>,q<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<p>Then we add the new values to our <code>phylo_h</code> dataframe. We
will signal they belong to <code>tree1_rel</code> and
<code>tree2_rel</code>, meaning they came from a community with
different <em>rel</em>ative abundance.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_h</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">phylo_h</span>,</span>
<span>                 <span class="fu">data.frame</span><span class="op">(</span>order <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                            hill <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="va">tree1_h0_rel</span>,<span class="va">tree1_h1_rel</span>,<span class="va">tree1_h2_rel</span>,</span>
<span>                                     <span class="va">tree2_h0_rel</span>,<span class="va">tree2_h1_rel</span>,<span class="va">tree2_h2_rel</span><span class="op">)</span>,</span>
<span>                            tree <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">'tree1_rel'</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="st">'tree2_rel'</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Plotting values for tree 1</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkred'</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting values for tree 2</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'red'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting values for tree 2</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1_rel'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree1_rel'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkblue'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting values for tree 2</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">order</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2_rel'</span><span class="op">]</span>,</span>
<span>     <span class="va">phylo_h</span><span class="op">$</span><span class="va">hill</span><span class="op">[</span><span class="va">phylo_h</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'tree2_rel'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'lightblue'</span><span class="op">)</span></span></code></pre>
</div>
<p>Reminders: 1) red is equal abundance and blue is different
abundances; 2) darker colors represent <code>tree1</code> (uneven branch
length) and lighter colors represent <code>tree2</code> (equal branch
lengths).</p>
<p>From this second plot, we notice that:</p>
<ol style="list-style-type: decimal"><li><p>Hill number values for tree 1 (darker colors) are still higher
than those for tree 2 (lighter colors). This makes sense since the
relative abundance is the same for both trees, so the overall pattern
(i.e., longer branches in tree 1) remains;</p></li>
<li><p>the Hill number of order 0 is the same regardless of whether you
have different relative abundances among species, since order 0 is
simply the sum of all branch lengths (i.e., it does not account for how
the relative abundance of taxa);</p></li>
<li><p>when we include different relative abundance among taxa, we have
a higher rate of decrease in the Hill number value as we increase the
order. This is because higher orders are now being calculated based on
how common/rare each branch is in our community (based on our
site-by-species matrix). If we take tree 1, for instance, we see that
the Hill value for order 1 including different taxa abundance (darker
blue line) is lower than the value where abundance is equal (darker red
line). We can interpret this as our metric putting less weight on the
longer branches of tree 1, which are rare in our community, and putting
more weight on the shorter branches, which are more common but represent
a more recent evolutionary history. We see a similar (although less
accentuated) pattern with tree2. Calculating different order of hill
numbers while incorporating species abundance therefore allows to have a
single metric accounting for evolutionary history and species abundance
distribution that is comparable across different regions.</p></li>
</ol><div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>Now that you got an intuition on the information phylogenetic Hill
numbers can give us, let’s move on to our actual data. For this
challenge, you should calculate the phylogenetic Hill numbers of orders
0, 1 and 2 for three islands we are working with throughout these
episodes. The key objects here will be the <code>phylogeny</code> tree
we imported earlier, as well as the <code>phylo_wide</code> object
containing our site-by-species matrix modified to match the tips in our
phylogeny. You should also plot the hill numbers for each island and
discuss what the calculated values allow you to infer regarding the
history of each community.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<!-- We gotta simulate the data to finalize this solution section. -->
</div>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="popgen-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="connecting-data-types.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="popgen-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Working with</a>
        <a class="chapter-link float-end" href="connecting-data-types.html" rel="next">Next: Connecting different... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/phylo-data.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/role-model/multidim-biodiv-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.3" class="external-link">sandpaper (0.12.3)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://role-model.github.io/multidim-biodiv-data/phylo-data.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Summarizing phylogenetic data",
  "creativeWorkStatus": "active",
  "url": "https://role-model.github.io/multidim-biodiv-data/phylo-data.html",
  "identifier": "https://role-model.github.io/multidim-biodiv-data/phylo-data.html",
  "dateCreated": "2023-06-06",
  "dateModified": "2023-06-06",
  "datePublished": "2023-06-06"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

