<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Multidimensional biodiversity data: management and analysis: Summarizing phylogenetic data</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/phylo-data.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Multidimensional biodiversity data: management and analysis
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Multidimensional biodiversity data: management and analysis
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Multidimensional biodiversity data: management and analysis
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 71%" class="percentage">
    71%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 71%" aria-valuenow="71" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/phylo-data.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="finding_data.html">2. Finding multi-dimensional biodiversity data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="care.html">3. The CARE principles and open data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="abundance-data.html">4. Abundance Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="traits-data.html">5. Trait data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="popgen-data.html">6. Working with population genetic data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        7. Summarizing phylogenetic data
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction-to-phylogenetic-data">Introduction to phylogenetic data</a></li>
<li><a href="#working-with-phylogenetic-data-in-r">Working with phylogenetic data in R</a></li>
<li><a href="#cleaning-and-filtering-phylogenetic-data">Cleaning and filtering phylogenetic data</a></li>
<li><a href="#summarizing-with-hill-numbers">Summarizing with hill numbers</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="connecting-data-types.html">8. Connecting different data types</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="phylo-hill.html">9. phylo_hill</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="popgen-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="connecting-data-types.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="popgen-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Working with</a>
        <a class="chapter-link float-end" href="connecting-data-types.html" rel="next">Next: Connecting different... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Summarizing phylogenetic data</h1>
        <p> Last updated on 2023-06-14 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/phylo-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What information is contained in phylogenies?</li>
<li>How to summarize and interpret phylogenies using Hill numbers?</li>
<li>What are the common phylogeny data formats?</li>
<li>How do you import and manipulate phylogeny data in R?</li>
<li>How do you visualize phylogenies in R?</li>
<li>How do you calculate Hill numbers with phylogenetic data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants will be able to:</p>
<ol style="list-style-type: decimal"><li>Identify key features of the Newick format</li>
<li>Import and manipulate phylogenetic data in the R environment</li>
<li>Visualize phylogenies</li>
<li>Calculate and interpret phylogenetic Hill numbers</li>
</ol></div>
</div>
</div>
</div>
</div>
<section id="introduction-to-phylogenetic-data"><h2 class="section-heading">Introduction to phylogenetic data<a class="anchor" aria-label="anchor" href="#introduction-to-phylogenetic-data"></a>
</h2>
<hr class="half-width"><p>In this episode, we will explore how to extract information from
phylogenetic trees in order to complement our hypotheses and inferences
about processes shaping biodiversity patterns. Phylogenetic trees have
information about how the species in a taxonomic group are related to
each other and how much relative evolutionary change has accumulated
among them. Since local communities differ in their phylogenetic
composition, this information can give insights on why communities are
how they are.</p>
<p>[HERE INCLUDE VISUALS OF PHYLOGENIES]</p>
<p>The relative phylogenetic distance among species in a community as
well as the distribution of the amount of evolutionary history
(represented by the length of the branches in a phylogeny) are a result
of different factors such as the age since the initial formation of the
community and the rate of macroevolutionary processes such as speciation
and extinction. For instance, young communities that are dominated by
closely related species and show very short branch lengths may suggest a
short history with few colonization events and high rates of local
speciation; alternatively, if the same young communities harbor
distantly related species with longer branch lengths, it may suggest
that most of the local diversity was generated by speciation elsewhere
followed by colonization events involving distantly related species.
Coupled with information on ecological traits and rates of
macroevolutionary processes, these patterns also allow to test for
hypotheses regarding, for instance, ecological filtering or niche
conservatism.</p>
<p>Summarizing this phylogenetic information (i.e., phylogenetic
distance and distribution of branch lengths) is therefore important for
inference. As we have seen in previous episodes, the use of <a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Hill
numbers</a> is an informative approach to summarize biodiversity.</p>
<p>[MORE HERE ON MEANING OF PHYLOGENETIC HILL, RELATING TO PHYLO
CHARACTERISTICS]</p>
</section><section id="working-with-phylogenetic-data-in-r"><h2 class="section-heading">Working with phylogenetic data in R<a class="anchor" aria-label="anchor" href="#working-with-phylogenetic-data-in-r"></a>
</h2>
<hr class="half-width"><div class="section level3">
<h3 id="importing-phylogenetic-data">Importing phylogenetic data<a class="anchor" aria-label="anchor" href="#importing-phylogenetic-data"></a></h3>
<p>Several file formats exist to store phylogenetic information. The
most common formats are the <a href="https://evomics.org/resources/tree-formats/" class="external-link">Newick and Nexus
formats</a>. Both these formats are plain text files storing different
levels of information about the taxa relationship and evolutionary
history. <code>Newick</code> files are the standard for representing
trees in a computer-readable form, as they can be extremely simple and
therefore do not take up much memory. <code>Nexus</code> file, on the
other hand, are composed by different blocks regarding different types
of information and can be used to store DNA alignments, phylogenetic
trees, pre-defined groupings of taxa, or everything at once. Since they
are a step ahead in complexity, we will stick with <code>Newick</code>
files for now.</p>
<p><code>Newick</code> files store the information about the clades in a
tree by representing each clade within a set of parentheses. Sister
clades are separated by <code>,</code>. The notation also requires us to
add the symbol <code>;</code> to represent the end of the information
for that phylogenetic tree.</p>
<p>The basic structure of a tree in a <code>Newick</code> format is
therefore as follows:</p>
<p>((A,B),C);</p>
<p>The notation above indicates that: 1. we have three taxa in our tree,
named <code>A</code>, <code>B</code> and <code>C</code>; 2.
<code>A</code> and <code>B</code> form one clade (A,B); 3. the (A,B)
clade is sister to the <code>C</code> clade (we represent that by adding
another set of parentheses and a <code>,</code> separating (A,B) from
C).</p>
<p>In addition, <code>Newick</code> files can also store information on
the branch leading to it tip and node. We do that by adding
<code>:</code> after each tip/node.</p>
<p>((A:0.5,B:0.5):0.5,C:1);</p>
<p>The notation above indicates that: 1. the branches containing A and B
have each a length = 0.5; 2. the branch that leads to the node
connecting A and B also has length = 0.5; 3. the branch leading to C has
length = 1.</p>
<p>The notation above is what we import into R to start working with and
manipulating our phylogenetic tree. For that goal, we will use the <a href="https://academic.oup.com/bioinformatics/article/20/2/289/204981" class="external-link">ape</a>
package:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span></code></pre>
</div>
<p>To import our tree, we will be using the function
<code>read.tree()</code> from the <code>ape</code> package. In the case
of simple trees as the one above, we could directly create them within R
by giving that notation as a <code>character</code> value to this
function, using the <code>text</code> argument, as shown below:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text <span class="op">=</span> <span class="st">'((A:0.5,B:0.5):0.5,C:1);'</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, we can visually inspect our tree using the <code>plot()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">example_tree</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/phylo-data-rendered-plot-example-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Can you visualize the text notation in that image? We can see the
same information: A is closer related to B than C, and the branches
leading to A and B have half the length of the branch leading to C.</p>
<p>The <code>read.tree()</code> function creates an object of class
<code>phylo</code>. We can further investigate this object by calling it
in our console:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span></span></code></pre>
</div>
<p>The printed information shows us that we have a phylogenetic tree
with 3 tips and 2 internal nodes, where the tip labels are “A, B, C”. We
also are informed that this tree is rooted and has branch lengths.</p>
<p>One way to access the components of this object and better explore it
is to use <code>$</code> after the object name. Here, it will be
important for us to know a little bit more about where the information
about tip labels and branch lengths are stored in that
<code>phylo</code> object. Easy enough, we can access that by calling
<code>tip.labels</code> and <code>edge.length</code> after
<code>$</code>.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span></span>
<span><span class="va">example_tree</span><span class="op">$</span><span class="va">edge.length</span></span></code></pre>
</div>
</div>
</section><section id="cleaning-and-filtering-phylogenetic-data"><h2 class="section-heading">Cleaning and filtering phylogenetic data<a class="anchor" aria-label="anchor" href="#cleaning-and-filtering-phylogenetic-data"></a>
</h2>
<hr class="half-width"><p>Now that we learned how to import and visualize trees in R, let’s
bring the phylogeny for the communities we are working with in this
workshop. Our data so far consists of <a href="abundance-data.html">abundances</a> and <a href="traits-data.html">traits</a> of several taxa of arthropods
collected across three islands in the Hawaiian archipelago. Let’s work
though importing phylogenetic information for these species.</p>
<p>Two common approaches to retrieving a phylogeny for a focal group are
1) relying on a published phylogeny for the group, or 2) surveying
public phylogenetic databases based on your taxa list. A common option
for the latter is the <a href="https://tree.opentreeoflife.org/about/taxonomy-version/ott3.2" class="external-link">Open
Tree of Life Taxonomy</a>, a public database that synthesizes taxonomic
information from different sources. You can even interact with this
database using the R package <code>rotl</code>. A few tutorials to do so
exist online, like <a href="https://mctavishlab.github.io/R_OpenTree_tutorials/" class="external-link">this one</a>.
Using a public database is a good approach when working with taxonomic
groups that are not heavily investigated regarding their phylogenetic
relationships (the well-known <a href="https://doi.org/10.1146/annurev-ecolsys-112414-054400" class="external-link">Darwinian
shortfall</a>). In such cases, databases like OTL will give you a
summary phylogeny already filtered for the taxa you have in hand and
cross-checked for synonyms and misspellings.</p>
<p>For this workshop, since we are using simulated data, we will work
with the first option: a “published” arthropod phylogeny. Let’s load
this phylogeny into R using the function <code>read.tree</code> we
learned earlier.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arthro_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span><span class="st">'LINK_TO_PHYLOGENY'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in file(file, "r"): cannot open file 'LINK_TO_PHYLOGENY': No such file
or directory</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in file(file, "r"): cannot open the connection</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">class</span><span class="op">(</span><span class="va">arthro_tree</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'arthro_tree' not found</code></pre>
</div>
<p>This new <code>phylo</code> object is way larger than the previous
one, being a “real” phylogeny and all. You can inspect it again by
directly calling the object <code>arthro_tree</code>. To plot it, we
will use the <code>type</code> argument to modify how our tree will be
displayed. Here, we used the option <code>'fan'</code>, to display a
circular phylogeny (slightly better to show such a large phylogeny in
the screen). We also set the <code>show.tip.label</code> argument to
<code>False</code>.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">arthro_tree</span>, type <span class="op">=</span> <span class="st">'fan'</span>, show.tip.label <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'arthro_tree' not found</code></pre>
</div>
<p>How do we combine all this information with the community datasets we
have so far for our three islands? First, we will have to perform some
name checking and filtering.</p>
<div class="section level3">
<h3 id="checking-for-missing-tips">Checking for missing tips<a class="anchor" aria-label="anchor" href="#checking-for-missing-tips"></a></h3>
<p><strong>Do we still want to keep this issue? Or simply skip ahead to
pruning the phylogeny?</strong></p>
<p>Since this is a “published” arthropod phylogeny, we will likely not
have any misspelling in the tip names of the object. However, one
important issue that may arise is that, depending on the resolution of
the phylogeny, it may be the case that some taxa in our community
dataset may not be explicitly represented in the phylogeny. For
instance, the tips in our phylogeny might be represented by one species
for each genus, or one genus for each family. This is usually the case
when the goal of the original publication was to elucidate the
phylogenetic relationship for such higher taxonomic levels, with less
focus on the more recent evolutionary history. In these cases, some taxa
in our <a href="abundance-data.html">abundances</a> dataset may not be
found in the phylogeny, which will yield errors when we try to summarize
phylogenetic diversity. Therefore, we will want to assign those “missing
taxa” to one of the tips present in the tree, based on their known genus
and/or family assignments.</p>
<p>Let’s first find out what are those “missing taxa”. For this, we’ll
refer back to our original species list, from the
<code>abundances</code> object:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_names</span> <span class="op">&lt;-</span> <span class="va">abundances</span><span class="op">$</span><span class="va">final_name</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'abundances' not found</code></pre>
</div>
<p>To cross-check this list against the list of names in our phylogeny,
we can use the Boolean operator <code>%in%</code> coupled with
<code>!</code>. This will allow us to check for names in
<code>all_names</code> that are not included in the
<code>arthro_tree$tip.label</code>. In summary, the expression
<code>A %in% B</code> would return the position of the elements in
vector A that are present in vector B. The <code>!</code> (NOT) operator
returns the opposite of that expression, in a way that
<code>!(A %in% B)</code> will return the position of elements in vector
A that are <em>NOT</em> present in vector B (precisely what we need
right now).</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">not_found</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="op">(</span><span class="va">all_names</span> <span class="op">%in%</span> <span class="op">(</span><span class="va">arthro_tree</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'all_names' not found</code></pre>
</div>
<p>The object <code>not_found</code> is now a vector of the positions
(or <em>indexes</em>) of the elements in <code>all_names</code> that are
not found in <code>arthro_tree$tip.label</code>. Let’s then check these
names by using these indexes to subset <code>all_names</code>:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_names</span><span class="op">[</span><span class="va">not_found</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'all_names' not found</code></pre>
</div>
<p><strong>EDIT THIS NEXT PARAGRAPH WHEN PHYLOGENY IS
AVAILABLE</strong></p>
<p>Here, we can decide how we are going to assign these species to tips
in our tree. For instance, we see that we do not have
<strong>XXX</strong> in our tree, but we have <strong>XXX</strong>,
which belongs to the same genus. Similarly…</p>
<p>We can then give new names to these species, based on what tips in
our tree we want them to match to. To do so, we will first duplicate our
<code>all_names</code> vector into the <code>new_names</code> vector.
This way, we can make our changes in <code>new_names</code> while
keeping the original names intact, just in case we need to refer back to
them in the future.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_names</span> <span class="op">&lt;-</span> <span class="va">all_names</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'all_names' not found</code></pre>
</div>
<p>Now we use the indexes saved in <code>not_found</code> to replace the
names in the positions with the ones we want:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_names</span><span class="op">[</span><span class="va">not_found</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'ADD_LIST_HERE'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'new_names' not found</code></pre>
</div>
<p>Now that we have a vector where all names match to a position in our
phylogeny, we need to bring these corrected names back to our abundance
data. Here, we will again first copy our <code>abundances</code> object
into another object, in order to keep the original data intact. Then, we
will change the column <code>final_name</code> based on the vector
<code>new_names</code>.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances_phylo</span> <span class="op">&lt;-</span> <span class="va">abundances</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'abundances' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances_phylo</span><span class="op">$</span><span class="va">final_name</span> <span class="op">&lt;-</span> <span class="va">new_names</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'new_names' not found</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="pruning-our-phylogeny">Pruning our phylogeny<a class="anchor" aria-label="anchor" href="#pruning-our-phylogeny"></a></h3>
<p>Now that we have a modified abundance dataset where all taxa names
are present in our phylogeny, we can work on pruning our phylogeny to
the taxa present in our communities. This is especially important when
working with published phylogenies as these are usually large files
containing several taxa that may not exist in the local community.</p>
<p>We can prune our phylogeny using the function <code>keep.tip</code>
from the <code>ape</code> package. For this function, we provide the
entire phylogeny plus the names of the tips we want to keep. Here, we
will retrieve those names from the new abundances dataset we create,
with names corrected to match the phylogeny.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arthro_tree_pruned</span> <span class="op">&lt;-</span> <span class="fu">keep.tip</span><span class="op">(</span><span class="va">arthro_tree</span>,<span class="va">abundances_phylo</span><span class="op">$</span><span class="va">final_name</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'arthro_tree' not found</code></pre>
</div>
<p>Now, that we have a pruned more manageable phylogeny for our
communities, let’s move on to the next bit of information we need to
summarize phylogenetic diversity: a site-by-species matrix. As we saw in
the <a href="abundance-data.html">abundances</a> episode, such a matrix
can be used to calculate Hill numbers and help us compare patterns
across the different communities. In this episode, we are following the
same approach to calculate Hill numbers, with the addition that, on top
of the abundance, we will also be using phylogenetic information,
i.e. the length of the branches in the phylogenetic tree leading to the
taxa present in each community.</p>
<p>Here, we need to make a site-by-species matrix using the new taxa
names corrected to match the phylogeny. Since we have done this in a
previous episode, we’ll cover this in a challenge.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>Make a site-by-species matrix containing the three islands based on
the new <code>abundance_phylo</code> object that contains the taxa names
corrected to match the phylogeny. Hint: you may want to refer to the <a href="abundance-data.html">abundances</a> episode.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>To create our site-by-species matrix, we will use the
<code>pivot_wider</code> function from the <code>tidyr</code>
package.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_wide</span> <span class="op">&lt;-</span> <span class="fu">pivot_wider</span><span class="op">(</span><span class="va">abundances_phylo</span>,</span>
<span>                          id_cols <span class="op">=</span> <span class="va">site</span>,</span>
<span>                          names_from <span class="op">=</span> <span class="va">final_name</span>,</span>
<span>                          values_from <span class="op">=</span> <span class="va">abundance</span>, </span>
<span>                          values_fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_wider(abundances_phylo, id_cols = site, names_from = final_name, : could not find function "pivot_wider"</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="summarizing-with-hill-numbers"><h2 class="section-heading">Summarizing with hill numbers<a class="anchor" aria-label="anchor" href="#summarizing-with-hill-numbers"></a>
</h2>
<hr class="half-width"><p>In this section, we will extract some summary statistics about the
pattern of phylogenetic diversity (PD) in our communities. As we
discussed above, the relative phylogenetic distance among species and
the distribution of this distance can give insights into processes of
community assembly. Here, we will make use of <a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Hill
numbers</a> to extract summaries of phylogenetic distances. Phylogenetic
hill numbers incorporate information on both the phylogenetic structure
of a system and the abundances of different species. In order to get an
intuition for how these different components influence phylogenetic hill
numbers, we’ll first explore the behavior of this summary statistic
using a few simple examples</p>
<div class="section level3">
<h3 id="example-1--trees-with-different-branch-lenghts">Example 1. Trees with different branch lenghts<a class="anchor" aria-label="anchor" href="#example-1--trees-with-different-branch-lenghts"></a></h3>
<p>For our examples, we’ll assume we have one single community with
eight taxa: A through H. Let’s create two site-by-species matrix for
this community: one denoting even abundance across species, and another
one with uneven abundance. For the even communitu, we create a vector of
the value <code>1</code> repeated 8 times; for the uneven community,
we’ll create a vector where abundance goes up from species A to species
H (for simplicity, we’ll just use values from 1 to 8). We then transform
them to dataframe, and name the columns with the names of the
species.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">even_comm</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># Abundance = 1 for all species</span></span>
<span><span class="va">uneven_comm</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># Abundance equal 1 for species A and goes up to 8 towards species H.</span></span>
<span></span>
<span><span class="co"># We name the columns with the species names</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">even_comm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">uneven_comm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span>,<span class="st">'D'</span>,<span class="st">'E'</span>,<span class="st">'F'</span>,<span class="st">'G'</span>,<span class="st">'H'</span><span class="op">)</span></span></code></pre>
</div>
<p>Now let’s create two different possible trees for these communities:
one with short branch lengths and another with longer branch lengths.
Remember: branch length = amount of evolutionary change</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">short_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(((A:3,B:3):1,(C:3,D:3):1):1,((E:3,F:3):1,(G:3,H:3):1):1);'</span><span class="op">)</span></span>
<span><span class="va">long_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(((A:6,B:6):1,(C:6,D:6):1):1,((E:6,F:6):1,(G:6,H:6):1):1);'</span><span class="op">)</span></span></code></pre>
</div>
<p>If we plot both trees…</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">short_tree</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/phylo-data-rendered-unnamed-chunk-11-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">long_tree</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/phylo-data-rendered-unnamed-chunk-11-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>…we can see that the branches leading to extant taxa are longer for
<code>long_tree</code>, as we intended. This suggests that a greater
amount of evolutionary change is happening in these recent branches of
the longer tree when compared to the shorter tree.</p>
<p>Now, we will calculate phylogenetic hill numbers for both trees using
both even and uneven communities. To store the calculated values, we’ll
a data.frame called <code>even_comm_short_tree</code>. The first column
will be our Hill numbers to be calculated; the second column will be the
Hill number order from 0 to 3, to see how the order affects the values;
the third and fourth column will be the description of our
components</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">even_comm_short_tree</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    hill_nb <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    q <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>    comm <span class="op">=</span> <span class="st">"even"</span>,</span>
<span>    tree <span class="op">=</span> <span class="st">"short"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>Now we will use a <code>for</code> loop to calculate phylogenetic
Hill numbers using the <code>hill_phylo</code> function from the
<code>hillR</code> package. This function takes in a a site-by-species
matrix and phylogeny, and returns phylogenetic Hill numbers for each
site based on which species are present there. We provide to the
function the site-by-species matrix, the phylogenetic tree and the
order.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">even_comm_short_tree</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">even_comm_short_tree</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">even_comm</span>, <span class="va">short_tree</span>, q <span class="op">=</span> <span class="va">even_comm_short_tree</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Let’s repeat this process for the longer tree:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">even_comm_long_tree</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    hill_nb <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    q <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>    comm <span class="op">=</span> <span class="st">"even"</span>,</span>
<span>    tree <span class="op">=</span> <span class="st">"long"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">even_comm_long_tree</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">even_comm_long_tree</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">even_comm</span>, <span class="va">long_tree</span>, q <span class="op">=</span> <span class="va">even_comm_long_tree</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>We can combine both dataframes and plot the values for
comparison:</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">even_comm_nb</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="va">even_comm_short_tree</span>,<span class="va">even_comm_long_tree</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'short'</span><span class="op">]</span>,</span>
<span>     <span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'short'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'red'</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'long'</span><span class="op">]</span>,</span>
<span>     <span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'long'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkred'</span><span class="op">)</span></span></code></pre>
</div>
<p><img src="fig/phylo-data-rendered-unnamed-chunk-15-1.png" style="display: block; margin: auto;" class="figure">
This figure clearly shows that the tree with longer branches (dark red
line) harbors higher evolutionary history, and therefore higher PD, as
calculated by Hill numbers. It also shows that the Hill number value
decreases as the order goes up, since higher orders focus on branch
lengths that are common.</p>
<p>What would happen if the abundance of species in our community was
uneven (a more realistic case)? In this case, both branch lengths and
how abundant each branch is will have an effect on the calculated value.
To visualize, let’s repeat the calculations above for the uneven
community.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Uneven comm with short tree</span></span>
<span><span class="va">uneven_comm_short_tree</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    hill_nb <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    q <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>    comm <span class="op">=</span> <span class="st">"uneven"</span>,</span>
<span>    tree <span class="op">=</span> <span class="st">"short"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">uneven_comm_short_tree</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">uneven_comm_short_tree</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">uneven_comm</span>, <span class="va">short_tree</span>, q <span class="op">=</span> <span class="va">uneven_comm_short_tree</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Uneven comm with long tree</span></span>
<span><span class="va">uneven_comm_long_tree</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    hill_nb <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    q <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>    comm <span class="op">=</span> <span class="st">"uneven"</span>,</span>
<span>    tree <span class="op">=</span> <span class="st">"long"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">uneven_comm_long_tree</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">uneven_comm_long_tree</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">uneven_comm</span>, <span class="va">long_tree</span>, q <span class="op">=</span> <span class="va">uneven_comm_long_tree</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Combining results</span></span>
<span><span class="va">uneven_comm_nb</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="va">uneven_comm_short_tree</span>,<span class="va">uneven_comm_long_tree</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s plot all results together, using red colors for even
communities and blue colors for uneven community. Ligher colors will
represent short trees whereas darker colors will represent long
trees.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'short'</span><span class="op">]</span>,</span>
<span>     <span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'short'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'red'</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">)</span>,<span class="fu">max</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'long'</span><span class="op">]</span>,</span>
<span>     <span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'long'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkred'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'short'</span><span class="op">]</span>,</span>
<span>     <span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'short'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'lightblue'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'long'</span><span class="op">]</span>,</span>
<span>     <span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'long'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkblue'</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/phylo-data-rendered-unnamed-chunk-17-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>From this picture, we can take a few insights: 1. longer branches
still yield higher Hill numbers, regardless of the evenness in the
community abundance; 2. the Hill number for q = 0 remains the same,
regardless of the evenness. 3. As already observed, the value of the
hill numbers drop as the order goes up. In the even community, since
species have the same abundance, this decrease reflects higher orders
focusing on more common values of branch length. In the uneven
community, where species have different abundance, this decrease
reflects higher orders focusing less and less on rare taxa. In other
words, higher orders emphasize what is more common; when community is
even, they emphasize the value of branch length that is common. When
community is uneven, they emphasize the more common taxa (i.e., the
branch that are over represented). 4. For q = 1 to 3, Hill numbers are
always lower for uneven communities. This suggests that although branch
lengths are the same (e.g., both dark red and dark blue lines represent
the long tree), these branches are unevenly represented in the community
due to uneven abundance of species. This unevenness is represented by
the lower values of Hill number on higher order. In other words, some
branches in the tree “dominate” the community more than others. We can
summarize this by saying that the higher the unevenness the lower the
value of the Hill number will be.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>
<code>hillR</code> calculates phylogenetic hill numbers given a
phylogeny and a site by species matrix.</li>
<li>In trees with a similar topology, phylogenetic Hill number of order
0 reflect the sum of branch lengths. Orders of 1 and higher reflect the
sum of branch lengths weighted by the relative abundance of different
species.</li>
</ul></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="example-2--balanced-vs-unbalanced-trees">Example 2. Balanced vs unbalanced trees<a class="anchor" aria-label="anchor" href="#example-2--balanced-vs-unbalanced-trees"></a></h3>
<p>So far, we have learned that Hill numbers are affected by both the
sum of branch lengths and the relative representation of each branch (in
terms of species abundance) in the community. For example 1, we have
used a perfectly balanced tree, i.e., all extant taxa have equal branch
length. In example 2, we will explore the effects of the uneven
distribution of cladogenesis event along the tree, leading to different
phylogeny structures.</p>
<p>Let’s create a totally balanced tree called
<code>even_tree</code>…</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">even_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(((A:1,B:1):1,(C:1,D:1):1):1,((E:1,F:1):1,(G:1,H:1):1):1);'</span><span class="op">)</span></span></code></pre>
</div>
<p>… and a totally unbalanced tree called <code>uneven_tree</code>.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uneven_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(A:7,(B:6,(C:5,(D:4,(E:3,(F:2,(G:1,H:1):1):1):1):1):1):1);'</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s plot both trees for comparison:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">even_tree</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'even_tree' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">uneven_tree</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'uneven_tree' not found</code></pre>
</div>
<p>Notice that here the difference between the trees resides in the fate
of each new lineage at a node. In the unbalanced uneven tree, at each
diversification event one of the lineages always persists till the
present with no change while the other undergoes another round of
diversification. In the even tree, both lineages from each node undergo
a new split. The consequence is that in the even tree, all extant
species result from recent diversification (i.e., they have a short
evolutionary history before coalescing into their ancestor), whereas in
the unbalanced tree we have a mix of old and recent lineages. This means
that the phylogenetic history itself is creating an uneven
representation of branch lengths across the community (even before we
account for species abundance)</p>
<p>To see how such phylogenetic structure influences hill numbers, let’s
repeat the calculations from example 1 with these new trees. First,
let’s focus on the even community (i.e., not introducing the relative
species abundance factor yet):</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">even_comm_even_tree</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    hill_nb <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    q <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>    comm <span class="op">=</span> <span class="st">"even"</span>,</span>
<span>    tree <span class="op">=</span> <span class="st">"even"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">even_comm_even_tree</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">even_comm_even_tree</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">even_comm</span>, <span class="va">even_tree</span>, q <span class="op">=</span> <span class="va">even_comm_even_tree</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'even_tree' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">even_comm_uneven_tree</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    hill_nb <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    q <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>    comm <span class="op">=</span> <span class="st">"even"</span>,</span>
<span>    tree <span class="op">=</span> <span class="st">"uneven"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">even_comm_uneven_tree</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">even_comm_uneven_tree</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">even_comm</span>, <span class="va">uneven_tree</span>, q <span class="op">=</span> <span class="va">even_comm_uneven_tree</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'uneven_tree' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">even_comm_nb</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="va">even_comm_even_tree</span>,<span class="va">even_comm_uneven_tree</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'even'</span><span class="op">]</span>,</span>
<span>     <span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'even'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'red'</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in plot.window(...): NAs not allowed in 'ylim'</code></pre>
</div>
<figure><img src="fig/phylo-data-rendered-unnamed-chunk-21-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">lines</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'uneven'</span><span class="op">]</span>,</span>
<span>     <span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'uneven'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkred'</span><span class="op">)</span></span></code></pre>
</div>
<p>This plot is similar to example 1 in two ways: 1) one of the trees
has higher hill number values, in this case the uneven tree. This
suggests that the unbalanced structure of the tree accounts for a deeper
evolutionary history (i.e., lineages have overall longer branches); 2)
the value of the hill numbers drop as the order goes up. This happens
because higher orders are weighting less and less those branch lengths
that are not so common (like, for instance, the short branch lengths in
the uneven tree).</p>
<p>Something new that we observe here is a more evident difference
between trees in the <strong>rate of change</strong> of the Hill number
value as we increase the order. The uneven tree (darker red) shows a
steeper drop than the even tree (lighter red). This is similar to what
we observed in the <a href="abundance-data.html">abundances</a> episode:
uneven communities show a steeper drop in Hill numbers as order
increases. In this example, instead of species abundance, the unevenness
of the community is represented by the structure of the phylogeny: a
completely unbalanced tree combining long and short branches generates
an unevenness in the distribution of branch lengths. Notice that this
difference in rate of change barely observed in example 1: when plotting
two trees with the same structure (short vs long trees, but both fully
balanced), the lines are very similar In example 1, the difference in
the rate of change only becomes evident when we incorporate unevenness
from the species abundance, but here in example 2 we see this difference
deriving already from the different structures of the trees.</p>
<p>If the tree structure is already introducing some unevenness in our
community, how would the pattern differ when species relative abundance
is included?</p>
<p>In this case, both the structure of the tree and the species
abundance interact to generate a pattern of unevenness. Specifically,
the overall value of PD will be the sum of branch lengths weighted by
their abundance in the community as informed by the species abundance
distribution (instead of by the frequency of the branch length in the
tree). This added information brings new insights. It may be the case,
for instance, that in the uneven tree, even though long branches are
over represented, maybe the species with long branches are actually
super rare in our community, and the short-branch species are actually
super abundant. As you can probably infer, this difference suggests
something about the evolutionary history of our community, in this case
that the most abundant species have a very recent evolutionary history.
Similarly, even though branch lengths are the same in the even tree, if
some of the species is more abundant than others, it suggests that the
evolutionary history of the tree is unevenly represented in our
community.</p>
<p>To visualize these interaction between tree structure and species
relative abundance, let’s redo the calculation for Hill numbers with
even and uneven trees, this time using the uneven community. We’ll plot
all final values together for comparison.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uneven_comm_even_tree</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    hill_nb <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    q <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>    comm <span class="op">=</span> <span class="st">"uneven"</span>,</span>
<span>    tree <span class="op">=</span> <span class="st">"even"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">uneven_comm_even_tree</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">uneven_comm_even_tree</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">uneven_comm</span>, <span class="va">even_tree</span>, q <span class="op">=</span> <span class="va">uneven_comm_even_tree</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'even_tree' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uneven_comm_uneven_tree</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>    hill_nb <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    q <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>    comm <span class="op">=</span> <span class="st">"uneven"</span>,</span>
<span>    tree <span class="op">=</span> <span class="st">"uneven"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">uneven_comm_uneven_tree</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">uneven_comm_uneven_tree</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">uneven_comm</span>, <span class="va">uneven_tree</span>, q <span class="op">=</span> <span class="va">uneven_comm_uneven_tree</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'uneven_tree' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uneven_comm_nb</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="va">uneven_comm_even_tree</span>,<span class="va">uneven_comm_uneven_tree</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'even'</span><span class="op">]</span>,</span>
<span>     <span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'even'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'red'</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">)</span>,<span class="fu">max</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in plot.window(...): NAs not allowed in 'ylim'</code></pre>
</div>
<figure><img src="fig/phylo-data-rendered-unnamed-chunk-22-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">lines</span><span class="op">(</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'uneven'</span><span class="op">]</span>,</span>
<span>     <span class="va">even_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">even_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'uneven'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkred'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'even'</span><span class="op">]</span>,</span>
<span>     <span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'even'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'lightblue'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">q</span><span class="op">[</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'uneven'</span><span class="op">]</span>,</span>
<span>     <span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">hill_nb</span><span class="op">[</span><span class="va">uneven_comm_nb</span><span class="op">$</span><span class="va">tree</span><span class="op">==</span><span class="st">'uneven'</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">'darkblue'</span><span class="op">)</span></span></code></pre>
</div>
<p>Reminders: 1) light colors represent even trees, whereas darker
colors represent uneven trees 2) red lines represent even community
whereas blue lines represent uneven community.</p>
<p>From this second plot, we notice that:</p>
<ol style="list-style-type: decimal"><li><p>Hill number values for the uneven tree (darker colors) are still
higher than those for the even tree (lighter colors). This makes sense
since the relative abundance is the same for both trees, so the overall
pattern (i.e., longer branches in uneven tree) remains;</p></li>
<li><p>the Hill number of order 0 is the same regardless of whether you
have different relative abundances among species. This was also the case
for example 1. Order 0 is simply the sum of all branch lengths (i.e., it
does not account for the relative abundance of branch lengths or
taxa);</p></li>
<li><p>the inclusion of an uneven community accentuates the rate of
decrease in the Hill number value as we increase the order. This is
because higher orders are now affected by the unevenness of the
distribution of branch length as well by the unevenness of species
abundance. This is a good example of how calculating different order of
hill numbers while incorporating species abundance allows to have a
couple summary statistics accounting for evolutionary history and
species abundance distribution that is comparable across different
regions.</p></li>
</ol><div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints2"></a>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>The phylogenetic structure of the community (represented by the
topology of the tree) influences the evenness of the branch length
distribution and contribute to different rates in the decrease of Hill
number values as the order number increases.</li>
<li>The inclusion of community structure (relative abundance of
different species) further accentuates this difference in rate of
change.</li>
<li>Phylogenetic Hill numbers represent the sum of branch lenghts.
Higher orders weight that sum by the distribution of branch lengths and
the species relative abundance.</li>
</ol></div>
</div>
</div>
<p>Now that we acquired an intuition on the information phylogenetic
Hill numbers can give us, let’s move on to our actual data.</p>
</div>
<div class="section level3">
<h3 id="moving-on-to-our-island-communities">Moving on to our island communities<a class="anchor" aria-label="anchor" href="#moving-on-to-our-island-communities"></a></h3>
<p>For this final section, we will be working on the data for the three
island communities to calculate hill numbers from orders of 0 to 3. We
will be using our pruned phylogeny and the site-by-species matrix we
created earlier. Since we covered the workflow of calculating
phylogenetic Hill numbers in the previous section, we will leave this
activity for a challenge. workflow covered in this episode,</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>For this challenge, you should calculate the phylogenetic Hill
numbers of orders 0 to 3 for the three islands we are working with
throughout the workshop The key objects here will be the phylogeny
stored in <code>arthro_tree_pruned</code> and the sites-by-species
matrix stored in <code>phylo_wide</code>. You should also plot the hill
numbers for each island and discuss what the calculated values allow you
to infer regarding the history of each community.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>We can directly calculate Hill numbers using the
<code>hill_phylo</code> function along with the objects
<code>arthro_tree_pruned</code> and <code>phylo_wide</code>. Since we
have three sites, the function will return a vector with tree values,
one Hill number for each site of the order we requested. Let’s create an
empty list to store a vector for each order, and use a <code>for</code>
loop to calculate from orders of 0 to 3.</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_values</span> <span class="op">&lt;-</span> <span class="fu">vector</span><span class="op">(</span><span class="st">'list'</span>,length <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">hill_values</span><span class="op">[[</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">phylo_wide</span>,<span class="va">arthro_tree_pruned</span>, q <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'phylo_wide' not found</code></pre>
</div>
<p>Now, let’s create a data.frame with islands as rows and Hill numbers
of different orders as columns. For that, we will use the function
<code>do.call</code> to collapse our list <code>hill_values</code> using
the function <code>cbind</code>.</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_values</span> <span class="op">&lt;-</span> <span class="fu">do.call</span><span class="op">(</span><span class="va">cbind.data.frame</span>,<span class="va">hill_values</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">hill_values</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'q0'</span>,<span class="st">'q1'</span>,<span class="st">'q2'</span>,<span class="st">'q3'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in names(x) &lt;- value: 'names' attribute [4] must be the same length as the vector [0]</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_phylo_nbs</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>site <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">'HawaiiIsland_01'</span>,<span class="st">'Kauai_01'</span>,<span class="st">'Maui_01'</span><span class="op">)</span>,</span>
<span>                             <span class="va">hill_values</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in data.frame(site = c("HawaiiIsland_01", "Kauai_01", "Maui_01"), : arguments imply differing number of rows: 3, 0</code></pre>
</div>
<p>Finally, we can plot using a similar code to the ones we used in our
examples:</p>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">3</span><span class="op">)</span>,<span class="va">hill_phylo_nbs</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>     type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">"#440154FF"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">'Order'</span>,ylab<span class="op">=</span><span class="st">'Hill values'</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu">range</span><span class="op">(</span><span class="fl">9</span>,<span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'hill_phylo_nbs' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">lines</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">3</span><span class="op">)</span>,<span class="va">hill_phylo_nbs</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>      type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">"#21908CFF"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'hill_phylo_nbs' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">lines</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">3</span><span class="op">)</span>,<span class="va">hill_phylo_nbs</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>      type<span class="op">=</span><span class="st">'b'</span>,col<span class="op">=</span><span class="st">"#FDE725FF"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'hill_phylo_nbs' not found</code></pre>
</div>
</div>
</div>
</div>
</div>

<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="popgen-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="connecting-data-types.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="popgen-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Working with</a>
        <a class="chapter-link float-end" href="connecting-data-types.html" rel="next">Next: Connecting different... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/phylo-data.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/role-model/multidim-biodiv-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.3" class="external-link">sandpaper (0.12.3)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://role-model.github.io/multidim-biodiv-data/phylo-data.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Summarizing phylogenetic data",
  "creativeWorkStatus": "active",
  "url": "https://role-model.github.io/multidim-biodiv-data/phylo-data.html",
  "identifier": "https://role-model.github.io/multidim-biodiv-data/phylo-data.html",
  "dateCreated": "2023-06-14",
  "dateModified": "2023-06-14",
  "datePublished": "2023-06-14"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

