<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Multidimensional biodiversity data: management and analysis: Working with population genetic data</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback." style="text-decoration: unset">
          Â 
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link" style="color: #383838">Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="background: #FFC700; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/popgen-data.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Multidimensional biodiversity data: management and analysis
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Multidimensional biodiversity data: management and analysis
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Multidimensional biodiversity data: management and analysis
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 59%" class="percentage">
    59%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 59%" aria-valuenow="59" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/popgen-data.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="care.html">2. Indigenous Data Sovereignty and the CARE Principles</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="abundance-data.html">3. Abundance Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="traits-data.html">4. Trait data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="phylo-data.html">5. Summarizing phylogenetic data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        6. Working with population genetic data
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#pop-gen-data">Pop gen data</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="finding_data.html">7. Finding multi-dimensional biodiversity data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="care-data-repos.html">8. CARE Principles and Data Repositories</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="phylo-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="finding_data.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="phylo-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Summarizing</a>
        <a class="chapter-link float-end" href="finding_data.html" rel="next">Next: Finding... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Working with population genetic data</h1>
        <p> Last updated on 2023-07-11 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/popgen-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<section id="pop-gen-data"><h2 class="section-heading">Pop gen data<a class="anchor" aria-label="anchor" href="#pop-gen-data"></a>
</h2>
<hr class="half-width"><div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What is genetic diversity?</li>
<li>What does genetic diversity say about community assembly?</li>
<li>What are common summaries of genetic diversity within species and
communities?</li>
<li>What are common sequence file formats?</li>
<li>How do you read in and manipulate sequence data?</li>
<li>How do you visualize and summarize sequence data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal"><li>Understand what genetic diversity is</li>
<li>Visualize genetic diversity distributions</li>
<li>Connect genetic diversity distributions to Hill numbers</li>
<li>Import population genetic data into the R environment</li>
<li>Manipulate population genetic data</li>
<li>Calculate genetic diversity <span class="math inline">\(\pi\)</span>
</li>
<li>Calculate and interpret Hill numbers of genetic diversity</li>
<li>Interpret amplicon sequence variant data and how it relates to data
from gene alignments</li>
</ol></div>
</div>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h1>
<div class="section level2">
<h2 id="what-is-genetic-diversity">What is genetic diversity?<a class="anchor" aria-label="anchor" href="#what-is-genetic-diversity"></a></h2>
<p>GEOBON, in their description of Essential Biodiversity Variables, <a href="https://geobon.org/ebvs/what-are-ebvs/" class="external-link">define intraspecific
genetic diversity</a> as:</p>
<blockquote>
<p>The variation in DNA sequences among individuals of the same
species.</p>
</blockquote>
<p>How is this genetic diversity generated? At a proximal level, genetic
diversity is generated through the processes of recombination and
mutation. These two mechanisms lay the foundation for evolutionary
forces like genetic drift, natural selection, and gene flow to shape the
distribution of genetic diversity within populations and across
communities.</p>
<p>Perhaps one of the most elegant theoretical frameworks for modeling
genetic diversity is coalescent theory. Coalescent theory is a model of
how alleles sampled from a population may have originated from a common
ancestor. Under the assumptions of no recombination, no natural
selection, no gene flow, and no population structure, i.e.Â neutrality,
it estimates the time to the most recent common ancestor (TMRCA) between
two sampled alleles and the variance of this estimate. When a population
is especially large (i.e.Â many breeding individuals), the TMRCA of two
randomly sampled alleles is large. This is because the per-generation
probability of two alleles coalescing is <code>1 / 2Ne</code>, where
<code>Ne</code>, the effective population size, is the number of
breeding individuals in an idealized population that follows the
assumptions stated above. In most, if not all, natural environments
conditions are not ideal and <code>Ne</code> does not perfectly
represent the census size of a population (see <a href="https://www.biorxiv.org/content/10.1101/2021.02.03.429633v2" class="external-link">Lewontinâs
paradox</a>). While the relationship between <code>Ne</code> and census
population size is not perfect, in most cases <code>Ne</code> increases
as population size increases and can be used to understand the relative
sizes of populations or size changes over time. The intuition is best
understood through visualization.</p>
<p>For a population of 10 diploid individuals (20 allele copies), the
probability of two randomly sampled allele coalescing each generation is
1 / (2 * 10), or 1/20. In this single simulation (conducted using <a href="https://github.com/cooplab/popgen-notes/blob/master/Rcode/track_alleles.R" class="external-link">Graham
Coopâs code</a>) , the two sampled alleles, whose genealogy is traced in
blue and red, coalesce 13 generations in the past.</p>
<figure><img src="fig/coop_10ind.png" class="figure mx-auto d-block"></figure><p>However, when the population is smaller (5 individuals, 10 allele
copies), the probability of coalescence each generation is 1 / (2 * 5)
or 1/10. In this simulation, two randomly sampled alleles coalesce after
4 generations.</p>
<figure><img src="fig/coop_5ind.png" class="figure mx-auto d-block"></figure><p>The TMRCA of the two sampled alleles corresponds with the smaller
<code>Ne</code> of this idealized population.</p>
<p>Visualizing this path of ancestry as a tree would look something like
this for 5 sampled allele copies.</p>
<figure><img src="fig/gene%20tree%20no%20mutations.png" class="figure mx-auto d-block"></figure><p>Overlaying mutations on the genealogy results in generally more
mutations occurring on longer branches, although because this is a
stochastic process, the relationship isnât always perfect (note- this is
a tree from a different simulation to the one above).</p>
<figure><img src="fig/genetree_mut_justtree.png" class="figure mx-auto d-block"></figure><p>As empirical biologists, what we have to work with are sequence data.
The accumulation of mutations along a sequence enables us to estimate
the TMRCA of a sample of alleles, leading to an estimate of the
<code>Ne</code> of the population.</p>
<p>One of the simplest and most widespread estimators of <code>Ne</code>
is the average number of pairwise differences between sequences sampled
from a population (ð¹, nucleotide diversity, <a href="https://www.pnas.org/doi/abs/10.1073/pnas.76.10.5269" class="external-link">Nei and Li
1979</a>). The basic idea underlying ð¹ is that more substitutions among
a sample of sequences results from longer branch lengths in the
genealogy of the sample, leading to a longer TMRCA and larger
<code>Ne</code>. The below figure contains the sequence alignment
related to the gene tree beside it. The average number of pairwise
nucleotide differences between s5 and s4 is:</p>
<p>5 (# of nucleotide differences) / 10 (# of basepairs in the sequence)
= 0.5</p>
<p>The average number of pairwise nucleotide differences between s4 and
s1 is:</p>
<p>1 (# of nucleotide differences) / 10 (# of base pairs in the
sequence) = 0.1</p>
<p>The larger number of nucleotide differences between s5 and s4 versus
s4 and s1 are reflected in the gene tree, where there the branches are
longer leading to the common ancestor of s5 and s4 versus s4 and s1. If
we want the genetic diversity of the entire sample of individuals, we
add the average pairwise number of nucleotide differences among all of
the sequences and take the average of those values.</p>
<p>In this case, <span class="math inline">\(\pi\)</span> is 0.183!</p>
<figure><img src="fig/genetree_mut.png" class="figure mx-auto d-block"></figure><p>Given its simplicity and ubiquity, we will cover how to calculate the
average number of pairwise differences between sequences and summarize
this value across assemblages.</p>
<div class="section level4">
<h4 id="resources-for-understanding-the-coalescent">Resources for understanding the coalescent<a class="anchor" aria-label="anchor" href="#resources-for-understanding-the-coalescent"></a></h4>
<ul><li>
<p><a href="http://cooplab.github.io/popgen-notes/" class="external-link">Graham Coopâs
popgen notes</a></p>
<ul><li>awesome resource for population genetics in general</li>
</ul></li>
<li>
<p><a href="https://phytools.shinyapps.io/coalescent-plot/" class="external-link">learnPopGen
app</a></p>
<ul><li>visualize coalescent genealogies with different numbers of
individuals in a population and different numbers of generations</li>
</ul></li>
<li>
<p><a href="https://bedford.io/projects/coaltrace/" class="external-link">coaltrace
app</a></p>
<ul><li>a dynamic and interactive visualization of coalescent genealogies
with the ability to change parameters on the fly</li>
</ul></li>
<li>
<p><a href="https://bedford.io/projects/phylodynamics-lecture/coalescent.html#/" class="external-link">Bedford
lab slides</a></p>
<ul><li>a great set of slides that explains the coalescent in a
complementary way to Graham Coopâs notes</li>
</ul></li>
<li>
<p><a href="https://cdmuir.shinyapps.io/genetree-to-phylogeny/" class="external-link">cdmuir
app</a></p>
<ul><li>uses the coalescent to simulate gene trees and mutations along
sequences</li>
</ul></li>
</ul></div>
</div>
<div class="section level2">
<h2 id="what-does-genetic-diversity-say-about-community-assembly">What does genetic diversity say about community assembly?<a class="anchor" aria-label="anchor" href="#what-does-genetic-diversity-say-about-community-assembly"></a></h2>
<p>Under neutrality and stable demographic histories, variation in
genetic diversity across a community reflects differences in long-term
population sizes across the community. Rather than a snapshot of
present-day abundance, this can be thought of as an average abundance
across generations.</p>
<p>In much the same way as a community with a few, highly abundant
species and many low abundance species may indicate a non-neutral
community assembly process like environmental filtering, a community
that contains a few species with high genetic diversity and many species
with low genetic diversity may indicate similar non-neutral community
assembly processes over a longer timescale.</p>
<p>In addition, genetic diversity is impacted by demographic events and
natural selection. Populations that have undergone a demographic
expansion or contraction in the recent past will have distinct
coalescent genealogies and therefore different genetic diversities. For
instance, a population that has experienced a recent demographic
expansion will have a TMRCA more recent than a population with the same
population size that has remained stable through the same time period,
illustrated below.</p>
<figure><img src="fig/changing_pop_size_1.png" alt="" class="figure mx-auto d-block"><figcaption>Image of coalescent genealogies traced from stable
and growing populations. From the Bedford Lab coalescent slides.</figcaption></figure><p>Distinguishing a small, constant-sized population from a large
population thatâs grown dramatically in the recent past is difficult
with single-locus data typical of large-scale biodiversity studies.
However, when paired with other data like traits and abundances, we can
make more refined interpretations of the shape of genetic diversity
across the community. For instance, the interpretation of a community
with a few highly diverse species paired with many species with low
genetic diversity may reflect a community with a few species that were
not impacted by a historical environmental disturbance alongside many
species that had their populations dramatically reduced, but have
rapidly grown to larger modern population sizes.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>Draw one or more possible scenarios of a pair of coalescent
genealogies drawn from a contracting population. The size of the modern
and historical populations and the number of generations you âsimulateâ
are up to you, but you should draw something similar to the figure
above.</p>
<p>After completing this exercise, do you think a contracting population
would result in more or less genetic diversity?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Contracting populations are funny! If a contraction was recent, the
population may maintain the levels of genetic diversity it had before
the contraction. However, if the contraction happened in the more
distant past and was maintained, genetic diversity is reduced.</p>
<figure><img src="fig/coalescent_contraction_drawing.png" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="hill-numbers-of-genetic-diversity-across-species">Hill numbers of genetic diversity across species<a class="anchor" aria-label="anchor" href="#hill-numbers-of-genetic-diversity-across-species"></a></h3>
<p>In the context of summarizing species abundances, Hill numbers are
the effective number of species in a community, where each exponent
(q-value) increasingly weights common species more heavily. The
interpretation is similar when summarizing genetic diversity. Hill
numbers are the effective number of species in a community, where each
q-value increasingly weights species with higher genetic diversity more
heavily. Rather than measuring an effective diversity of abundances,
this is an effective diversity of genetic diversities.</p>
<p><img src="fig/popgen-data-rendered-example-pi-dist-1.png" width="50%" style="display: block; margin: auto;" class="figure"><img src="fig/popgen-data-rendered-example-pi-dist-2.png" width="50%" style="display: block; margin: auto;" class="figure"></p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="work-with-pop-gen-data">Work with pop gen data<a class="anchor" aria-label="anchor" href="#work-with-pop-gen-data"></a></h1>
<p>The three packages weâre working with are <code>msa</code>,
<code>ape</code>, and <code>hillR</code>. The first two are used to work
with sequencing data, while <code>hillR</code> is what we will use to
calculate Hill numbers. Iâll provide more explanation once we get to
using them.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">msa</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="sequences">Sequences<a class="anchor" aria-label="anchor" href="#sequences"></a></h2>
<div class="section level3">
<h3 id="fasta">FASTA<a class="anchor" aria-label="anchor" href="#fasta"></a></h3>
<p>A FASTA file contains one or more DNA sequences, where the DNA
sequence is preceded by a line with a carat (â&gt;â) followed by an
unique sequence ID:</p>
<pre><code>&gt;seq0010
GATCCCCAATTGGGG</code></pre>
<p>This is perhaps the most common way to store sequence information as
it is simple and has historical inertia. For more information, see the
<a href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat/#:~:text=In%20FASTA%20format%20the%20line,should%20not%20contain%20any%20spaces." class="external-link">the
NCBI page</a>.</p>
</div>
<div class="section level3">
<h3 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a></h3>
<p>Prior to analysis, sequence data must be aligned if they havenât been
already. A sequence alignment is necessary to identify regions of DNA
that are similar or that vary. There are <a href="https://en.wikipedia.org/wiki/List_of_sequence_alignment_software" class="external-link">many
pieces of software to perform alignments</a> and all have their
advantages and disadvantages, but we will choose a single program, <a href="https://www.ebi.ac.uk/Tools/msa/clustalo/" class="external-link">Clustal Omega</a>,
which is fast and robust for short sequence alignments. While Clustal
Omega can be run from the command line, we will use the R package <a href="https://bioconductor.org/packages/release/bioc/html/msa.html" class="external-link">msa</a>
to keep everything in the R environment.</p>
<p>First, a FASTA file containing the sequences needs to be read in
using the <code>readDNAStringSet()</code> function. For most of this
episode, we will be working with sequences from the Kauai island. At the
end, we will summarize the</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_path</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/role-model/multidim-biodiv-data/main/episodes/data/kauai_seqs.fas"</span></span>
<span></span>
<span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">seq_path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqs</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DNAStringSet object of length 130:
      width seq                                             names               
  [1]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  [2]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  [3]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  [4]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  [5]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  ...   ... ...
[126]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...
[127]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...
[128]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...
[129]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...
[130]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...</code></pre>
</div>
<p>To align the sequences using Clustal Omega, we use the
<code>msa</code> function and specify the <code>method</code> as
âClustalOmegaâ. There are more arguments that you can use to fine-tune
the alignment, but these are not necessary for the vast majority of
scenarios.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using Gonnet</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ClustalOmega 1.2.0 

Call:
   msa(seqs, method = "ClustalOmega")

MsaDNAMultipleAlignment with 130 rows and 500 columns
      aln                                                  names
  [1] CGCGCACTCTACCACCCAGACTATC...CCGGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-6
  [2] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGTTAATTCGTG Eudonia_lycopodiae-1
  [3] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-4
  [4] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-3
  [5] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-8
  [6] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-2
  [7] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-5
  [8] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-9
  [9] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-0 
  ... ...
[123] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-1
[124] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-5
[125] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-6
[126] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-7
[127] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-0
[128] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-8
[129] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-4
[130] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-9
  Con CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Consensus </code></pre>
</div>
<p>If after you perform your alignment you find that one or more
individuals do not belong or are negatively impacting the alignment, you
have to manipulate your unaligned sequence data and then re-align the
sequences. For instance, say you didnât want to include the individual
<code>ind_name</code> in the alignment. First, remove the individual
from the <code>seqs</code> object.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove individual</span></span>
<span><span class="va">seqs_filt</span> <span class="op">&lt;-</span> <span class="va">seqs</span><span class="op">[</span><span class="fu">names</span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"Eudonia_lycopodiae-6"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">seqs_filt</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DNAStringSet object of length 129:
      width seq                                             names               
  [1]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  [2]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  [3]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  [4]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  [5]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Proterhinus_punct...
  ...   ... ...
[125]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...
[126]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...
[127]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...
[128]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...
[129]   500 CGCGCACTCTACCACCCAGACT...TGACTCCTATCGTTAATTCGTG Campsicnemus_nigr...</code></pre>
</div>
<p>As a reminder, <code>!=</code> means âdoes not equalâ, so
<code>names(seqs) != "Eudonia_lycopodiae-6"</code> returns
<code>TRUE</code> when the sequence name does not match
âEudonia_lycopodiae-6â.</p>
<p>Following the filtering, you need to realign the new set of
sequences.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs_filt</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using Gonnet</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ClustalOmega 1.2.0 

Call:
   msa(seqs_filt, method = "ClustalOmega")

MsaDNAMultipleAlignment with 129 rows and 500 columns
      aln                                                  names
  [1] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGTTAATTCGTG Eudonia_lycopodiae-1
  [2] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-4
  [3] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-3
  [4] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-8
  [5] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-2
  [6] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-5
  [7] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-9
  [8] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-0
  [9] CGCGCACTCTACCACCCAGACTATC...CCCGACTCCTATCGATAATTCGTG Eudonia_lycopodiae-7 
  ... ...
[122] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-1
[123] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-5
[124] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-6
[125] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-7
[126] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-0
[127] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-8
[128] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-4
[129] CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Nesodynerus_mimus-9
  Con CGCGCACTCTACCACCCAGACTATC...CCTGACTCCTATCGTTAATTCGTG Consensus </code></pre>
</div>
<p>There are quite a few packages in R to work with sequence data, for
example <a href="https://cran.r-project.org/web/packages/seqinr/index.html" class="external-link">seqinr</a>,
<a href="https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.html" class="external-link">phangorn</a>,
and <a href="https://cran.r-project.org/web/packages/ape/index.html" class="external-link">ape</a>.
<code>ape</code> is the most versatile of these. But before we can use
<code>ape</code>, the alignment must be converted to the
<code>ape::DNAbin</code> format.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ape_align</span> <span class="op">&lt;-</span> <span class="fu">msaConvert</span><span class="op">(</span><span class="va">alignment</span>, type <span class="op">=</span> <span class="st">"ape::DNAbin"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ape_align</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>129 DNA sequences in binary format stored in a matrix.

All sequences of same length: 500 

Labels:
Eudonia_lycopodiae-1
Eudonia_lycopodiae-4
Eudonia_lycopodiae-3
Eudonia_lycopodiae-8
Eudonia_lycopodiae-2
Eudonia_lycopodiae-5
...

Base composition:
    a     c     g     t 
0.260 0.245 0.234 0.261 
(Total: 64.5 kb)</code></pre>
</div>
<p>Although not necessary for this workshop (the sequence data is
already aligned), itâs generally advisable to write your DNA alignment
to a new file to use later or with other programs.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ape</span><span class="fu">::</span><span class="fu">write.FASTA</span><span class="op">(</span><span class="va">ape_align</span>, <span class="st">"seqs_aligned.fas"</span><span class="op">)</span></span></code></pre>
</div>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>This FASTA file contains multiple species. However, you may only want
to align and analyze a single species for your analysis. Using what you
learned about the <code>%in%</code> operator in the Phylogenetics
episode, filter your <code>seqs</code> object to only contain sequences
for <em>Hyposmocoma sagittata</em>.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># you can also make a vector with the c() function and list each individual separately, but this requires a little less typing</span></span>
<span><span class="va">hypo_inds</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="st">"Hyposmocoma_sagittata-"</span>, <span class="fl">0</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqs_hypo</span> <span class="op">&lt;-</span> <span class="va">seqs</span><span class="op">[</span><span class="fu">names</span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">hypo_inds</span><span class="op">]</span></span>
<span></span>
<span><span class="va">seqs_hypo</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DNAStringSet object of length 10:
     width seq                                              names               
 [1]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
 [2]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
 [3]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
 [4]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
 [5]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
 [6]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
 [7]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
 [8]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
 [9]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...
[10]   500 CGCGCACTCTACCACCCAGACTA...TGACTCCTATCGTTAATTCGTG Hyposmocoma_sagit...</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="ape">ape<a class="anchor" aria-label="anchor" href="#ape"></a></h3>
<div class="section level4">
<h4 id="check-alignment">Check alignment<a class="anchor" aria-label="anchor" href="#check-alignment"></a></h4>
<p><code>ape</code> has a useful utility that allows you perform a
series of diagnostics on your alignment to make sure nothing is fishy.
The function can output multiple plots, but letâs just visualize the
alignment plot. If the colors seem jumbled in a region, this indicates
that an alignment error may have occurred and to inspect your sequences
for any problems.</p>
<p>In addition, the function outputs a series of helpful statistics to
your console. Ideally, your alignment should have few gaps and a number
of segregating sites that is reasonable for the number of sites in your
alignment and the evolutionary divergence of the sequences included in
the alignment, e.g., for an alignment with 970 sites of a single bird
species, around 40 segregating sites is reasonable, but 200 segregating
sites may indicate that something went wrong with your alignment or a
species was misidentified. In addition, for diploid species a vast
majority, if not all of the sites should contain one or two observed
bases.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">checkAlignment</span><span class="op">(</span><span class="va">ape_align</span>, what <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Number of sequences: 129 
Number of sites: 500 

No gap in alignment.

Number of segregating sites (including gaps): 126
Number of sites with at least one substitution: 126
Number of sites with 1, 2, 3 or 4 observed bases:
  1   2   3   4 
374  66  56   4 </code></pre>
</div>
<figure><img src="fig/popgen-data-rendered-ape-filt-inds-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level4">
<h4 id="calculate-genetic-diversity">Calculate genetic diversity<a class="anchor" aria-label="anchor" href="#calculate-genetic-diversity"></a></h4>
<p>To calculate genetic diversity for each species, itâs easiest to
split the alignment by species before calculating the average number of
pairwise nucleotide differences. Since weâre not concerned about
alignment quality, we can just split the alignment, rather than
splitting the original <code>msa</code> sequence object and
re-aligning.</p>
<p>First, you need to get the names of the species. Thankfully, the
genus and species is included in the sequences names, where the format
is âGenus_species-numberâ. To get the species names, you need to use the
<code>gsub()</code> function, which replaces all of the occurrences of a
particular string pattern in a string. In our case, we need to remove
the â-numberâ individual identifier at the end of the sequence name, so
we only have âGenus_speciesâ. We will use a short regular expression, or
<a href="https://r4ds.had.co.nz/strings.html" class="external-link">regex</a>, which is a
sequence of characters that specifies a pattern that a string search
algorithm like <code>gsub()</code> will search for.</p>
<p>The pattern <code>"-[0-9]"</code> means âmatch any pattern in the
string that is a dash followed by any occurrence of a digitâ. The hard
brackets <code>[ ]</code> specify a character set, where any occurrence
of the characters inside the brackets will be matched. In this case, we
specify zero through nine, so any digits that follow a dash will be
matched. The second argument contains the string to replace the pattern
with. We want to remove the pattern, so we supply an empty string to be
replaced.</p>
<p>Following the string removal, we only want unique instances of the
species names, so we use the <code>unique()</code> function to retain
unique species names.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp_names</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"-[0-9]"</span>, <span class="st">""</span>, <span class="fu">names</span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sp_names</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="va">sp_names</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sp_names</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Proterhinus_punctipennis" "Xyletobius_collingei"    
 [3] "Metrothorax_deverilli"    "Nesodynerus_mimus"       
 [5] "Scaptomyza_vagabunda"     "Lucilia_graphita"        
 [7] "Eudonia_lycopodiae"       "Atelothrus_depressus"    
 [9] "Mecyclothorax_longulus"   "Laupala_pruna"           
[11] "Hylaeus_sphecodoides"     "Hyposmocoma_sagittata"   
[13] "Campsicnemus_nigricollis"</code></pre>
</div>
<p>Next, we are going to build an R function line-by-line to facilitate
creating a list of species alignments. User-defined functions are very
useful for when you want to complete a set of tasks repeatedly.
Copy-pasting chunks of code and changing things slightly for each
situation can lead to errors, headaches, and general pain. The worst bit
of pain is when you need to change something in that chunk of code and
you have to go back and change it for every instance you copy-pasted it
before! When you have a function, you only need to change the code once.
Iâll go through each line of code, then weâll wrap up the code in a
function.</p>
<p>First, we need to turn the <code>ape</code> alignment into a list
because the original <code>DNAbin</code> object doesnât allow
subsetting.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">align_list</span> <span class="op">&lt;-</span> <span class="fu">as.list</span><span class="op">(</span><span class="va">ape_align</span><span class="op">)</span></span></code></pre>
</div>
<p>The <code>grepl()</code> function searches for a string pattern in
the text, then returns <code>TRUE</code> if it finds it and
<code>FALSE</code> if it doesnât. Weâll use the resulting boolean
(<code>TRUE/FALSE</code>) vector to extract the species from the
alignment. Here is an example with <em>Proterhinus punctipennis</em>.
Youâll see the 100-110 slots are <code>TRUE</code>, indicating that
those are the locations in the alignment that contain <em>P.
punctipennis</em> sequences.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp_boolean</span> <span class="op">&lt;-</span> <span class="fu">grepl</span><span class="op">(</span><span class="st">"Proterhinus_punctipennis"</span>, <span class="fu">names</span><span class="op">(</span><span class="va">align_list</span><span class="op">)</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sp_boolean</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [25] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [37] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [49] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [61] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [73] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [85] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [97] FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
[109]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[121] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</code></pre>
</div>
<p>Following the conversion, we use bracket indexing to subset the
alignment by the <code>sp_boolean</code> boolean vector. Now we have an
alignment of only <em>Proterhinus punctipennis</em> sequences!</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">align_sp</span> <span class="op">&lt;-</span> <span class="va">align_list</span><span class="op">[</span><span class="va">sp_boolean</span><span class="op">]</span></span>
<span></span>
<span><span class="va">align_sp</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Proterhinus_punctipennis-1
Proterhinus_punctipennis-4
Proterhinus_punctipennis-2
Proterhinus_punctipennis-5
Proterhinus_punctipennis-6
Proterhinus_punctipennis-9
...

Base composition:
    a     c     g     t 
0.262 0.242 0.237 0.259 
(Total: 5 kb)</code></pre>
</div>
<p>Now, we need to wrap these bits of code into a function so we can
create separate alignments for all species. The structure of a function
looks like this.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">do_something</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="op">{</span></span>
<span>    </span>
<span><span class="op">}</span></span></code></pre>
</div>
<p><code>function()</code> defines the <code>do_something</code> object
as a function, while <code>x</code> and <code>y</code> are the arguments
you are supplying to the function. The curly brackets <code>{ }</code>
contain the code you want to run in the function.</p>
<p>For our function <code>split_align</code>, we need two arguments- the
species name we want to extract from the alignment (<code>spname</code>)
and the alignment itself (<code>alignment</code>). We then add our code
inside the curly brackets <code>{ }</code>, replacing
<code>"Proterhinus_punctipennis"</code> with <code>spname</code> and
<code>ape_align</code> with <code>alignment</code>. At the end, we
specify that we want to return the <code>align_sp</code> object with the
<code>return()</code> function. Although not always necessary, using the
<code>return()</code> function is generally good practice for code
readability.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">split_align</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">spname</span>, <span class="va">alignment</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># convert ape object to list</span></span>
<span>    <span class="va">align_list</span> <span class="op">&lt;-</span> <span class="fu">as.list</span><span class="op">(</span><span class="va">alignment</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># get boolean vector for species</span></span>
<span>    <span class="va">sp_boolean</span> <span class="op">&lt;-</span> <span class="fu">grepl</span><span class="op">(</span><span class="va">spname</span>, <span class="fu">names</span><span class="op">(</span><span class="va">align_list</span><span class="op">)</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># subset alignment for species</span></span>
<span>    <span class="va">align_sp</span> <span class="op">&lt;-</span> <span class="va">align_list</span><span class="op">[</span><span class="va">sp_boolean</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># return the alignment</span></span>
<span>    <span class="kw">return</span><span class="op">(</span><span class="va">align_sp</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Now we can check that the function works by running it on
<em>Proterhinus punctipennis</em>. It does! (or it should).</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">split_align</span><span class="op">(</span>spname <span class="op">=</span> <span class="st">"Proterhinus_punctipennis"</span>, alignment <span class="op">=</span> <span class="va">ape_align</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Proterhinus_punctipennis-1
Proterhinus_punctipennis-4
Proterhinus_punctipennis-2
Proterhinus_punctipennis-5
Proterhinus_punctipennis-6
Proterhinus_punctipennis-9
...

Base composition:
    a     c     g     t 
0.262 0.242 0.237 0.259 
(Total: 5 kb)</code></pre>
</div>
<p>If your code throws an error or the output isnât what you expect, you
can place the <code>browser()</code> function at a particular line in
your code to enable you to run it line by line. This helps for
diagnosing where exactly something went wrong.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">split_align</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">spname</span>, <span class="va">alignment</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">browser</span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="co"># convert ape object to list</span></span>
<span>    <span class="va">align_list</span> <span class="op">&lt;-</span> <span class="fu">as.list</span><span class="op">(</span><span class="va">alignment</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># get boolean vector for species</span></span>
<span>    <span class="va">sp_boolean</span> <span class="op">&lt;-</span> <span class="fu">grepl</span><span class="op">(</span><span class="va">spname</span>, <span class="fu">names</span><span class="op">(</span><span class="va">align_list</span><span class="op">)</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># subset alignment for species</span></span>
<span>    <span class="va">align_sp</span> <span class="op">&lt;-</span> <span class="va">align_list</span><span class="op">[</span><span class="va">sp_boolean</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># return the alignment</span></span>
<span>    <span class="kw">return</span><span class="op">(</span><span class="va">align_sp</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Now to apply our function to the vector of species names! Weâre using
the <code>lapply()</code> function to apply the <code>split_seqs</code>
function to every element in <code>sp_names</code> and return a list.
<code>lapply()</code> automatically populates the first argument in the
function with the vector or list you supply it, but you can supply
further arguments after entering your function. So, after adding
<code>split_seqs</code>, we specify
<code>alignment = ape_align</code>.</p>
<p>For good measure, weâll also specify the names of the list elements
as the species names, so theyâre easy to access later.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">align_split</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">sp_names</span>, <span class="va">split_align</span>, alignment <span class="op">=</span> <span class="va">ape_align</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">align_split</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sp_names</span></span>
<span></span>
<span><span class="va">align_split</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$Proterhinus_punctipennis
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Proterhinus_punctipennis-1
Proterhinus_punctipennis-4
Proterhinus_punctipennis-2
Proterhinus_punctipennis-5
Proterhinus_punctipennis-6
Proterhinus_punctipennis-9
...

Base composition:
    a     c     g     t 
0.262 0.242 0.237 0.259 
(Total: 5 kb)

$Xyletobius_collingei
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Xyletobius_collingei-2
Xyletobius_collingei-5
Xyletobius_collingei-0
Xyletobius_collingei-1
Xyletobius_collingei-3
Xyletobius_collingei-4
...

Base composition:
    a     c     g     t 
0.262 0.244 0.236 0.259 
(Total: 5 kb)

$Metrothorax_deverilli
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Metrothorax_deverilli-2
Metrothorax_deverilli-8
Metrothorax_deverilli-0
Metrothorax_deverilli-1
Metrothorax_deverilli-3
Metrothorax_deverilli-4
...

Base composition:
    a     c     g     t 
0.260 0.246 0.232 0.262 
(Total: 5 kb)

$Nesodynerus_mimus
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Nesodynerus_mimus-2
Nesodynerus_mimus-3
Nesodynerus_mimus-1
Nesodynerus_mimus-5
Nesodynerus_mimus-6
Nesodynerus_mimus-7
...

Base composition:
    a     c     g     t 
0.254 0.247 0.237 0.262 
(Total: 5 kb)

$Scaptomyza_vagabunda
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Scaptomyza_vagabunda-4
Scaptomyza_vagabunda-6
Scaptomyza_vagabunda-7
Scaptomyza_vagabunda-8
Scaptomyza_vagabunda-9
Scaptomyza_vagabunda-2
...

Base composition:
    a     c     g     t 
0.260 0.247 0.233 0.260 
(Total: 5 kb)

$Lucilia_graphita
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Lucilia_graphita-5
Lucilia_graphita-6
Lucilia_graphita-8
Lucilia_graphita-1
Lucilia_graphita-7
Lucilia_graphita-4
...

Base composition:
    a     c     g     t 
0.262 0.240 0.237 0.261 
(Total: 5 kb)

$Eudonia_lycopodiae
9 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Eudonia_lycopodiae-1
Eudonia_lycopodiae-4
Eudonia_lycopodiae-3
Eudonia_lycopodiae-8
Eudonia_lycopodiae-2
Eudonia_lycopodiae-5
...

Base composition:
    a     c     g     t 
0.258 0.248 0.231 0.262 
(Total: 4.5 kb)

$Atelothrus_depressus
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Atelothrus_depressus-0
Atelothrus_depressus-1
Atelothrus_depressus-2
Atelothrus_depressus-3
Atelothrus_depressus-4
Atelothrus_depressus-5
...

Base composition:
    a     c     g     t 
0.260 0.246 0.232 0.262 
(Total: 5 kb)

$Mecyclothorax_longulus
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Mecyclothorax_longulus-2
Mecyclothorax_longulus-9
Mecyclothorax_longulus-7
Mecyclothorax_longulus-8
Mecyclothorax_longulus-3
Mecyclothorax_longulus-6
...

Base composition:
    a     c     g     t 
0.260 0.244 0.235 0.261 
(Total: 5 kb)

$Laupala_pruna
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Laupala_pruna-1
Laupala_pruna-2
Laupala_pruna-4
Laupala_pruna-8
Laupala_pruna-9
Laupala_pruna-0
...

Base composition:
    a     c     g     t 
0.260 0.247 0.233 0.260 
(Total: 5 kb)

$Hylaeus_sphecodoides
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Hylaeus_sphecodoides-7
Hylaeus_sphecodoides-1
Hylaeus_sphecodoides-2
Hylaeus_sphecodoides-3
Hylaeus_sphecodoides-4
Hylaeus_sphecodoides-5
...

Base composition:
    a     c     g     t 
0.258 0.248 0.234 0.260 
(Total: 5 kb)

$Hyposmocoma_sagittata
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Hyposmocoma_sagittata-0
Hyposmocoma_sagittata-1
Hyposmocoma_sagittata-2
Hyposmocoma_sagittata-3
Hyposmocoma_sagittata-4
Hyposmocoma_sagittata-5
...

Base composition:
    a     c     g     t 
0.260 0.246 0.232 0.262 
(Total: 5 kb)

$Campsicnemus_nigricollis
10 DNA sequences in binary format stored in a list.

All sequences of same length: 500 

Labels:
Campsicnemus_nigricollis-0
Campsicnemus_nigricollis-1
Campsicnemus_nigricollis-2
Campsicnemus_nigricollis-3
Campsicnemus_nigricollis-4
Campsicnemus_nigricollis-5
...

Base composition:
    a     c     g     t 
0.260 0.246 0.232 0.262 
(Total: 5 kb)</code></pre>
</div>
<p>Now weâre going to calculate the average number of pairwise genetic
differences among these sequences. <code>ape</code> has the
<code>dist.dna()</code> function to calculate various genetic distance
metrics for an alignment.</p>
<p>You use the argument <code>model = "raw"</code> to specify that you
want to use the average number of pairwise differences as your metric.
When calculating the number of pairwise genetic differences, you have to
choose whether you want missing data to be deleted across the entire
alignment (<code>pairwise.deletion = FALSE</code>) or only between the
two sequences being compared (<code>pairwise.deletion = TRUE</code>).
Since we want to take advantage of as much data as possible when
computing ð, weâll set <code>pairwise.deletion = TRUE</code>.</p>
<p>Weâll do this for a single species, then youâll get a chance to write
your own function to do it across species!</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendist</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">align_split</span><span class="op">$</span><span class="va">Proterhinus_punctipennis</span>, model <span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fas_gendist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                           Proterhinus_punctipennis-1
Proterhinus_punctipennis-4                      0.010
Proterhinus_punctipennis-2                      0.006
Proterhinus_punctipennis-5                      0.006
Proterhinus_punctipennis-6                      0.006
Proterhinus_punctipennis-9                      0.006
Proterhinus_punctipennis-7                      0.008
Proterhinus_punctipennis-0                      0.026
Proterhinus_punctipennis-3                      0.028
Proterhinus_punctipennis-8                      0.024
                           Proterhinus_punctipennis-4
Proterhinus_punctipennis-4                           
Proterhinus_punctipennis-2                      0.004
Proterhinus_punctipennis-5                      0.004
Proterhinus_punctipennis-6                      0.004
Proterhinus_punctipennis-9                      0.004
Proterhinus_punctipennis-7                      0.006
Proterhinus_punctipennis-0                      0.024
Proterhinus_punctipennis-3                      0.026
Proterhinus_punctipennis-8                      0.022
                           Proterhinus_punctipennis-2
Proterhinus_punctipennis-4                           
Proterhinus_punctipennis-2                           
Proterhinus_punctipennis-5                      0.000
Proterhinus_punctipennis-6                      0.000
Proterhinus_punctipennis-9                      0.000
Proterhinus_punctipennis-7                      0.002
Proterhinus_punctipennis-0                      0.020
Proterhinus_punctipennis-3                      0.022
Proterhinus_punctipennis-8                      0.018
                           Proterhinus_punctipennis-5
Proterhinus_punctipennis-4                           
Proterhinus_punctipennis-2                           
Proterhinus_punctipennis-5                           
Proterhinus_punctipennis-6                      0.000
Proterhinus_punctipennis-9                      0.000
Proterhinus_punctipennis-7                      0.002
Proterhinus_punctipennis-0                      0.020
Proterhinus_punctipennis-3                      0.022
Proterhinus_punctipennis-8                      0.018
                           Proterhinus_punctipennis-6
Proterhinus_punctipennis-4                           
Proterhinus_punctipennis-2                           
Proterhinus_punctipennis-5                           
Proterhinus_punctipennis-6                           
Proterhinus_punctipennis-9                      0.000
Proterhinus_punctipennis-7                      0.002
Proterhinus_punctipennis-0                      0.020
Proterhinus_punctipennis-3                      0.022
Proterhinus_punctipennis-8                      0.018
                           Proterhinus_punctipennis-9
Proterhinus_punctipennis-4                           
Proterhinus_punctipennis-2                           
Proterhinus_punctipennis-5                           
Proterhinus_punctipennis-6                           
Proterhinus_punctipennis-9                           
Proterhinus_punctipennis-7                      0.002
Proterhinus_punctipennis-0                      0.020
Proterhinus_punctipennis-3                      0.022
Proterhinus_punctipennis-8                      0.018
                           Proterhinus_punctipennis-7
Proterhinus_punctipennis-4                           
Proterhinus_punctipennis-2                           
Proterhinus_punctipennis-5                           
Proterhinus_punctipennis-6                           
Proterhinus_punctipennis-9                           
Proterhinus_punctipennis-7                           
Proterhinus_punctipennis-0                      0.022
Proterhinus_punctipennis-3                      0.024
Proterhinus_punctipennis-8                      0.020
                           Proterhinus_punctipennis-0
Proterhinus_punctipennis-4                           
Proterhinus_punctipennis-2                           
Proterhinus_punctipennis-5                           
Proterhinus_punctipennis-6                           
Proterhinus_punctipennis-9                           
Proterhinus_punctipennis-7                           
Proterhinus_punctipennis-0                           
Proterhinus_punctipennis-3                      0.014
Proterhinus_punctipennis-8                      0.010
                           Proterhinus_punctipennis-3
Proterhinus_punctipennis-4                           
Proterhinus_punctipennis-2                           
Proterhinus_punctipennis-5                           
Proterhinus_punctipennis-6                           
Proterhinus_punctipennis-9                           
Proterhinus_punctipennis-7                           
Proterhinus_punctipennis-0                           
Proterhinus_punctipennis-3                           
Proterhinus_punctipennis-8                      0.008</code></pre>
</div>
<p>The function returns a distance matrix containing the pairwise
differences among all sequences in the data set. To take a quick peak at
their distribution, we can use the <code>hist()</code> function on the
distance matrix.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">fas_gendist</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/popgen-data-rendered-fas-hist-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Since we want the average of these differences, we need to take their
average. However, you may have noticed the distribution of genetic
distances is non-normal, with mostly low genetic diversities. This is
the case with most genetic data sets. So, instead of taking the mean,
the <code>median()</code> will provide a better estimate of the center
of the distribution. The average number of pairwise genetic differences,
aka ð, aka genetic diversity of <em>Proterhinus punctipennis</em> is
0.01!</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendiv</span> <span class="op">&lt;-</span> <span class="fu">median</span><span class="op">(</span><span class="va">fas_gendist</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fas_gendiv</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.01</code></pre>
</div>
<div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge3"></a>
</h3>
<div class="callout-content">
<p>Write your own function to calculate ð for each of the alignments!
After writing the function, try applying the function to every alignment
in <code>align_split</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">calc_pi</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">alignment</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># get genetic distances</span></span>
<span>    <span class="va">dists</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">alignment</span>, model <span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="co"># get average</span></span>
<span>    <span class="va">pi</span> <span class="op">&lt;-</span> <span class="fu">median</span><span class="op">(</span><span class="va">dists</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="kw">return</span><span class="op">(</span><span class="va">pi</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>I havenât introduced it yet, but <code>sapply()</code> is similar to
<code>lapply()</code>, except it returns a vector instead of a list. It
saves an extra step when you know your function is returning single
values of the same type.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># solution 1. This returns a vector</span></span>
<span><span class="va">pi_across_sp</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">align_split</span>, <span class="va">calc_pi</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># solution 2</span></span>
<span><span class="va">pi_across_sp</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">align_split</span>, <span class="va">calc_pi</span><span class="op">)</span></span>
<span><span class="co">## convert to a vector</span></span>
<span><span class="va">pi_across_sp</span> <span class="op">&lt;-</span> <span class="fu">unlist</span><span class="op">(</span><span class="va">pi_across_sp</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Letâs plot the ranked genetic diversities of Kauai! Weâll use
plotting techniques similar to what youâve used in the Abundance and
Traits episodes.</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">pi_across_sp</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, xlab <span class="op">=</span> <span class="st">"Rank"</span>, ylab <span class="op">=</span> <span class="st">"Genetic diversity"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/popgen-data-rendered-plot-ranked-gendiv-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
<div class="section level3">
<h3 id="hill-numbers">Hill numbers<a class="anchor" aria-label="anchor" href="#hill-numbers"></a></h3>
<p>Now, letâs calculate the first five Hill numbers (q = 1-5) of this
distribution. For this, we will use the <code>hillR</code> package.
<code>hillR</code> does not have a function specifically for calculating
Hill numbers from genetic diversities, but the function you have used
for calculating Hill numbers from abundances, <code>hill_taxa</code>
will work the same.</p>
<p>To turn your vector of genetic diversities into a site by species
matrix, you just need to transpose <code>pi_across_sp</code> with the
<code>t()</code> function. This turns <code>pi_across_sp</code> into a
matrix with a column for each species and a row for each site, which in
this case is 1.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pi_t</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="va">pi_across_sp</span><span class="op">)</span></span>
<span><span class="va">pi_t</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     Proterhinus_punctipennis Xyletobius_collingei Metrothorax_deverilli
[1,]                     0.01                0.006                 0.002
     Nesodynerus_mimus Scaptomyza_vagabunda Lucilia_graphita Eudonia_lycopodiae
[1,]             0.004                0.004            0.016              0.012
     Atelothrus_depressus Mecyclothorax_longulus Laupala_pruna
[1,]                    0                  0.002         0.002
     Hylaeus_sphecodoides Hyposmocoma_sagittata Campsicnemus_nigricollis
[1,]                    0                     0                        0</code></pre>
</div>
<p>Finally, you can feed this matrix into <code>hill_taxa()</code> for
each q value weâre interested in. To make this efficient, letâs use
<code>sapply()</code>!</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_pi</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="va">hill_taxa</span>, comm <span class="op">=</span> <span class="va">pi_t</span><span class="op">)</span></span></code></pre>
</div>
<p>A quick barplot shows what we expect- Hill decreases with increasing
q because it is increasingly weighted by low genetic diversities.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">barplot</span><span class="op">(</span><span class="va">hill_pi</span>, names.arg <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"q=1"</span>, <span class="st">"q=2"</span>, <span class="st">"q=3"</span>, <span class="st">"q=4"</span>, <span class="st">"q=5"</span><span class="op">)</span>, ylab <span class="op">=</span> <span class="st">"Hill diversity"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/popgen-data-rendered-hill-barplot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level3">
<h3 id="comparison-across-islands">Comparison across islands<a class="anchor" aria-label="anchor" href="#comparison-across-islands"></a></h3>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge4"></a>
</h3>
<div class="callout-content">
<p>Lets go through the whole workflow for the Big Island and Maui! Go
from the raw sequences to per-species Hill numbers of genetic diversity
for both islands. After the challenge weâll plot genetic diversities and
Hill numbers across all three islands and compare!</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># big island fasta</span></span>
<span><span class="va">seqs_bi</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/role-model/multidim-biodiv-data/main/episodes/data/big_island_seqs.fas"</span><span class="op">)</span></span>
<span><span class="co"># maui fasta</span></span>
<span><span class="va">seqs_maui</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/role-model/multidim-biodiv-data/main/episodes/data/maui_seqs.fas"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>If you had more islands or localities, you could make a mega-function
to get the genetic diversities and/or Hill numbers of each of the
localities, rather than copy-pasting this code repeatedly. Functions are
very useful!</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># big island</span></span>
<span><span class="co">## align sequences</span></span>
<span><span class="va">align_bi</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs_bi</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using Gonnet</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## convert to DNAbin</span></span>
<span><span class="va">ape_align_bi</span> <span class="op">&lt;-</span> <span class="fu">msaConvert</span><span class="op">(</span><span class="va">align_bi</span>, type <span class="op">=</span> <span class="st">"ape::DNAbin"</span><span class="op">)</span></span>
<span><span class="co">## unique names to create alignment list</span></span>
<span><span class="va">sp_bi</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">gsub</span><span class="op">(</span><span class="st">"-[0-9]"</span>, <span class="st">""</span>, <span class="fu">names</span><span class="op">(</span><span class="va">seqs_bi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## split alignment into list of each species</span></span>
<span><span class="va">align_list_bi</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">sp_bi</span>, <span class="va">split_align</span>, alignment <span class="op">=</span> <span class="va">ape_align_bi</span><span class="op">)</span></span>
<span><span class="co">## add names to alignment list</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">align_list_bi</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sp_bi</span></span>
<span><span class="co">## get genetic diversities</span></span>
<span><span class="va">pi_bi</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">align_list_bi</span>, <span class="va">calc_pi</span><span class="op">)</span></span>
<span><span class="co">## get Hill numbers 1:5</span></span>
<span><span class="va">hill_bi</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="va">hill_taxa</span>, comm <span class="op">=</span> <span class="fu">t</span><span class="op">(</span><span class="va">pi_bi</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># maui</span></span>
<span><span class="co">## align sequences</span></span>
<span><span class="va">align_maui</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs_maui</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>using Gonnet</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## convert to DNAbin</span></span>
<span><span class="va">ape_align_maui</span> <span class="op">&lt;-</span> <span class="fu">msaConvert</span><span class="op">(</span><span class="va">align_maui</span>, type <span class="op">=</span> <span class="st">"ape::DNAbin"</span><span class="op">)</span></span>
<span><span class="co">## unique names to create alignment list</span></span>
<span><span class="va">sp_maui</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">gsub</span><span class="op">(</span><span class="st">"-[0-9]"</span>, <span class="st">""</span>, <span class="fu">names</span><span class="op">(</span><span class="va">seqs_maui</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## split alignment into list of each species</span></span>
<span><span class="va">align_list_maui</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">sp_maui</span>, <span class="va">split_align</span>, alignment <span class="op">=</span> <span class="va">ape_align_maui</span><span class="op">)</span></span>
<span><span class="co">## add names to alignment list</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">align_list_maui</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sp_maui</span></span>
<span><span class="co">## get genetic diversities</span></span>
<span><span class="va">pi_maui</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">align_list_maui</span>, <span class="va">calc_pi</span><span class="op">)</span></span>
<span><span class="co">## get Hill numbers 1:5</span></span>
<span><span class="va">hill_maui</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="va">hill_taxa</span>, comm <span class="op">=</span> <span class="fu">t</span><span class="op">(</span><span class="va">pi_maui</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Now, letâs plot the ranked genetic diversities of each island on the
same plot to compare their shapes. A line plot shows the general shape
of the trend a little clearer, so weâll set <code>type = "l"</code>. We
will use Maui as the base plot, since it has the most species. Then,
weâll add the other two islands to the plot using the
<code>lines()</code> function.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">pi_maui</span>, <span class="cn">TRUE</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"l"</span>, xlab <span class="op">=</span> <span class="st">"Rank"</span>, ylab <span class="op">=</span> <span class="st">"Genetic diversity"</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">pi_bi</span>, <span class="cn">TRUE</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="va">pi_across_sp</span>, <span class="cn">TRUE</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="st">"topright"</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"Maui"</span>, <span class="st">"Big Island"</span>, <span class="st">"Kauai"</span><span class="op">)</span>, fill <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/popgen-data-rendered-island-rank-plots-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Now, letâs compare Hill q=1 across the three islands. An interesting
pattern that we see is that even though Maui has more species, the Hill
diversities of the Big Island and Kauai are larger! What is a potential
interpretation of these data, especially in light of the abundance and
traits data?</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">barplot</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">hill_maui</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">hill_bi</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">hill_pi</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>, names.arg <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Maui"</span>, <span class="st">"Big Island"</span>, <span class="st">"Kauai"</span><span class="op">)</span>, ylab <span class="op">=</span> <span class="st">"Hill q=1"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/popgen-data-rendered-hill-across-islands-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
<div class="section level2">
<h2 id="asvs-and-metabarcoding">ASVs and Metabarcoding<a class="anchor" aria-label="anchor" href="#asvs-and-metabarcoding"></a></h2>
<p>Metabarcoding is rapidly growing DNA sequencing method used for
biodiversity assessment. It involves sequencing a short, variable region
of DNA sequence across a sample that may contain many taxa. This can be
used to rapidly assess the diversity of a particular sampling site with
low sampling effort. Some common applications include <a href="https://en.wikipedia.org/wiki/Metabarcoding" class="external-link">microbiome diversity
assessments, environmental DNA, and community DNA</a>. In many
applications where a data base of species identities for sequence
variants is not available or desired, the diversity of sequences is
inferred, rather than the diversity of species. Additionally, while <a href="https://onlinelibrary-wiley-com.ezproxy.gc.cuny.edu/doi/full/10.1111/1755-0998.12902" class="external-link">there
are caveats</a>, the relative abundance of sequences can be considered
analogous to the relative abundance of species across a community.</p>
<p>The basic unit of diversity used in most studies is the <a href="https://en.wikipedia.org/wiki/Amplicon_sequence_variant" class="external-link">amplicon
sequence variant (ASV)</a>. An ASV is any one of the sequences inferred
from a high-throughput sequencing experiment. These sequences are free
from errors and may diverge by as little as a single base pair from each
other. They are considered a substitute for operational taxonomic units
(OTUs) for evaluating ecological diversity given their higher
resolution, reproducibility, and no need for a reference database.</p>
<div class="section level3">
<h3 id="sequence-table">Sequence table<a class="anchor" aria-label="anchor" href="#sequence-table"></a></h3>
<p>The basic unit of analysis used for metabarcoding studies is the
sequence table. It is a matrix with named rows corresponding to the
samples and named columns corresponding to the ASVs. Each cell contains
the count of each variant in each sample. Hereâs an example with 5
samples and 4 sequence variants:</p>
<table class="table"><thead><tr class="header"><th></th>
<th>asv_1</th>
<th>asv_2</th>
<th>asv_3</th>
<th>asv_4</th>
</tr></thead><tbody><tr class="odd"><td>sample_1</td>
<td>0</td>
<td>276</td>
<td>200</td>
<td>10</td>
</tr><tr class="even"><td>sample_2</td>
<td>10</td>
<td>1009</td>
<td>0</td>
<td>0</td>
</tr><tr class="odd"><td>sample_3</td>
<td>0</td>
<td>0</td>
<td>234</td>
<td>15</td>
</tr><tr class="even"><td>sample_4</td>
<td>129</td>
<td>1992</td>
<td>0</td>
<td>485</td>
</tr><tr class="odd"><td>sample_5</td>
<td>1001</td>
<td>10</td>
<td>1847</td>
<td>178</td>
</tr></tbody></table><p>sample_1 may correspond to a soil sample collected in a forest, while
the variants represent different ASVs, which could look like this:</p>
<p>asv_1: GTTCA</p>
<p>asv_2: GTACA</p>
<p>asv_3: GATCA</p>
<p>asv_4: CTTCA</p>
<p>This looks similiar to a site by species matrix, where the rows are a
sampling locality, the columns are the species, and the cell values are
the abundance of a species at a particular site.</p>
<p>Weâll briefly go over how the data makes its way from raw sequencing
data to the sequence table.</p>
</div>
<div class="section level3">
<h3 id="dada2">DADA2<a class="anchor" aria-label="anchor" href="#dada2"></a></h3>
<p><a href="https://benjjneb.github.io/dada2/index.html" class="external-link">DADA2</a> is
one of the most powerful and commonly used software packages for
processing raw ASV sequencing data to turn it into a table of ASVs for
each sample. This workshop assumes you are using the output from this
workflow for further analysis, but for clarity the steps to the workflow
include:</p>
<ol style="list-style-type: decimal"><li>read quality assessment</li>
<li>filtering and trimming FASTQ files</li>
<li>learning sequencing error rates, so the algorithm can correct for
them</li>
<li>infer the sequence variants</li>
<li>merging paired reads (if sequencing was done with forward and
reverse read pairs)</li>
<li>constructing a sequence table</li>
</ol><p>The raw reads are typically in the FASTQ format, which is similar to
FASTA, but with some added metadata about the sequences. For more
in-depth discussion about FASTQ format, you can scroll to the
<strong>Extra</strong> section!</p>
<p>There are more available to further assess the quality of the data
and map the data to taxonomic databases, but the core output of DADA2
and similar software pipelines is the sequence table.</p>
</div>
<div class="section level3">
<h3 id="comparison-with-sequence-alignment-workflow">Comparison with sequence alignment workflow<a class="anchor" aria-label="anchor" href="#comparison-with-sequence-alignment-workflow"></a></h3>
<p>You can analyze ASV data in two ways. You can use the ASV sequencing
table directly and use abundances of each ASV as input into the
<code>hill_taxa()</code> function, considering each ASV as its own
âspeciesâ. This has the advantage of using the ASVâs quality as a
minimal reproducible taxonomic unit.</p>
<p>Alternatively, if you have a taxonomic reference database, you can
map each ASV to its species identity, convert the table into a FASTA
alignment, then calculate metrics like the average number of pairwise
nucleotide differences per species. This has the advantage of being able
to directly compare genetic diversity metrics with other dimensions of
biodiversity like traits and abundances among species. However, the
sequence lengths used for ASVs are typically very short and may not
capture sufficient variation to characterize species diversity.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li><p>Genetic diversity is a useful and fundamental unit of
biodiversity</p></li>
<li><p>Population genetic data may come in a variety of formats</p></li>
<li><p>The Hill number framework is also an informative approach to
summarizing genetic diversity across a community</p></li>
</ul></div>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="extra">Extra<a class="anchor" aria-label="anchor" href="#extra"></a></h1>
<div class="section level2">
<h2 id="fastq">FASTQ<a class="anchor" aria-label="anchor" href="#fastq"></a></h2>
<p>The FASTQ file format is similar to the FASTA format introduced
earlier, but includes additional data about the contained sequences.
This enables more flexible data manipulation and filtering capabilities,
and is popular with high-throughput sequencing technologies. When
working with ASVs, this is the most common file format you will
encounter.</p>
<p>A FASTA formatted sequence may look like this,</p>
<pre><code>&gt;seq1
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT</code></pre>
<p>while a FASTQ version of the same sequence may look like this:</p>
<pre><code>@seq1
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT
+
!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65</code></pre>
<p>If your eyes glaze over at the seemingly random numbers below the
sequence, donât fret! They mean something, but usually your machine does
the interpreting for you. A FASTQ file is formatted as follows:</p>
<ol style="list-style-type: decimal"><li><p>An â@â symbol followed by a sequence identifier and an optional
description</p></li>
<li><p>The raw sequence letters</p></li>
<li><p>A â+â character that is optionally followed by the sequence
identifier and more description</p></li>
<li><p>Quality encodings for the raw sequence letters in field 2. There
must be the same number of encoding symbols as there are sequence
letters.</p></li>
</ol><p>The symbols are bytes representing quality, ranging from 0x21 (lowest
quality; â!â in ASCII) to 0x7e (highest quality; â~â in ASCII). Here,
theyâre arranged in order of quality from lowest on the left to highest
on the right:</p>
<pre><code>!"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~</code></pre>
<p>Fortunately, whatever bioinformatic program you use will generally do
the interpreting for you.</p>
<p>Different sequencing technologies (e.g.Â Illumina) have standardized
description strings which contain more information about the sequences
and can be useful when filtering sequences or performing QC. For
examples, see <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">the
Wikipedia page</a>.</p>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="phylo-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="finding_data.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="phylo-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Summarizing</a>
        <a class="chapter-link float-end" href="finding_data.html" rel="next">Next: Finding... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/popgen-data.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/role-model/multidim-biodiv-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:biodataworkshop2023@gmail.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper" class="external-link">sandpaper (0.12.5)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://role-model.github.io/multidim-biodiv-data/popgen-data.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Working with population genetic data",
  "creativeWorkStatus": "active",
  "url": "https://role-model.github.io/multidim-biodiv-data/popgen-data.html",
  "identifier": "https://role-model.github.io/multidim-biodiv-data/popgen-data.html",
  "dateCreated": "2023-06-18",
  "dateModified": "2023-07-11",
  "datePublished": "2023-08-08"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

