<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Multidimensional biodiversity data: management and analysis: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="manifest" href="../site.webmanifest">
<link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link" style="color: #383838">Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="background: #FFC700; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Multidimensional biodiversity data: management and analysis
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Multidimensional biodiversity data: management and analysis
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul>
</li>
      </ul>
</div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
<input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset>
</form>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Multidimensional biodiversity data: management and analysis
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="care.html">2. The CARE principles and open data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="abundance-data.html">3. Abundance Data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="traits-data.html">4. Trait data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="popgen-data.html">5. Working with population genetic data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="phylo-data.html">6. Summarizing phylogenetic data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="connecting-data-types.html">7. Connecting different data types</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="finding_data.html">8. Finding multi-dimensional biodiversity data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Introduction</a></p>
<hr>
<p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 30 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ol style="list-style-type: decimal">
<li>What do we mean by “multidimensional biodiversity data?</li>
<li>What are the four types of data we’ll be covering in this
workshop?</li>
<li>What types of questions can we explore with these data?</li>
<li>How will this workshop be structured? Where do I find course
materials, etc?</li>
</ol>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants should be able to…</p>
<ol style="list-style-type: decimal">
<li>Articulate a shared understanding of what “multidimensional
biodiversity data” is</li>
<li>List the types of data to be covered in this workshop, and potential
applications of MBDB</li>
<li>Locate course resources.</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<section id="welcome"><h2 class="section-heading">Welcome<a class="anchor" aria-label="anchor" href="#welcome"></a>
</h2>
<hr class="half-width">
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor1" aria-labelledby="headingInstructor1">
<div class="accordion-body">
<p>Given that this curriculum deals explicitly with Indigenous data
sovereignty and governance, it is advisable to start with a land
acknowledgement.</p>
<p>For Albuquerque, NM (the location of the first offering of this
curriculum) here is info UNM’s land acknowledgement and the process
leading up to it: - <a href="https://diverse.unm.edu/about-dei-at-unm/land-labor-acknowledgement.html" class="external-link uri">https://diverse.unm.edu/about-dei-at-unm/land-labor-acknowledgement.html</a>
- <a href="https://diverse.unm.edu/assets/documents/unm-indigenous-peoples-land-and-territory-acknowledgment-brown-paper.pdf" class="external-link uri">https://diverse.unm.edu/assets/documents/unm-indigenous-peoples-land-and-territory-acknowledgment-brown-paper.pdf</a></p>
<p>Any land acknowledgement should be done in the context of growing
awareness that the intentions of many land acknowledgements, while
aspiring to be good intentions, where ill-informed and the
acknowledgements have led to harm in some cases. Here are resources to
read about that: - <a href="https://nativegov.org/news/beyond-land-acknowledgment-guide/" class="external-link uri">https://nativegov.org/news/beyond-land-acknowledgment-guide/</a>
- <a href="https://hyperallergic.com/769024/your-land-acknowledgment-is-not-enough/" class="external-link uri">https://hyperallergic.com/769024/your-land-acknowledgment-is-not-enough/</a>
- <a href="https://www.pbs.org/newshour/nation/analysis-how-well-meaning-land-acknowledgements-can-erase-indigenous-people-and-sanitize-history" class="external-link uri">https://www.pbs.org/newshour/nation/analysis-how-well-meaning-land-acknowledgements-can-erase-indigenous-people-and-sanitize-history</a></p>
<p>It is becoming agreed upon protocol that land acknowledgements should
be given only in the context of how the event or institution presenting
the land acknowledgement engages with activities beyond an otherwise
hollow “acknowledgement” of unceded lands.</p>
</div>
</div>
</div>
</div>
<p>Welcome everyone! This workshop will engage with ethics around
biodiversity data through the lens of Indigenous data sovereignty and
governance. With that in mind, many of the locations where this workshop
might run will likely be unceded Indigenous lands. As part of our
introductions we’ll think about where we are, where we come from, and
our positionalities. <a href="https://native-land.ca/" class="external-link">Native Land
Digital</a> is a great resource for understanding where we are
geographically, and here is a great video by Dr. Kat Gardner-Vandy and
Daniella Scalice demonstrating classical academic introductions and
relational introductions that center positionality:</p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/sY3LYVTTiH4?start=414&amp;end=811" data-external="1" allowfullscreen>
</iframe>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor2" aria-labelledby="headingInstructor2">
<div class="accordion-body">
<p>A note on Native Land Digital map: the map loads with a disclaimer
that is important to read: the geospatial polygons are not necessarily
approved by the groups they purport to represent. Native Land Digital is
Native-led and has a mechanism to verify polygons, but there might be
more accurate map representations you can use for your areas.</p>
<p>A note on relational intros: if giving this curriculum live you could
choose to demonstrate yourselves (as instructors) the different types of
introductions modeled by Dr. Kat Gardner-Vandy and Daniella Scalice, or
you could show the video. Note that it’s a legit consideration about the
length of the relational introduction to allow time and space for
everybody to get through their introductions; more people = more
succinct intros.</p>
</div>
</div>
</div>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>Introductions! Please use this list to help guide your introduction
of yourself:</p>
<ul>
<li>Your name</li>
<li>Your pronouns</li>
<li>Your positionality</li>
<li>Your institution</li>
<li>Something you’re excited or curious about for the workshop</li>
</ul>
</div>
</div>
</div>
</section><section id="four-dimensions-of-biodiversity-data"><h2 class="section-heading">Four dimensions of biodiversity data<a class="anchor" aria-label="anchor" href="#four-dimensions-of-biodiversity-data"></a>
</h2>
<hr class="half-width">
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<ul>
<li>What does “multidimensional biodiversity” data mean to you?</li>
<li>What types of data do you encounter in your work?</li>
<li>Have you ever looked at multiple types simultaneously?</li>
</ul>
</div>
</div>
</div>
<p>In this workshop, we’ll be working with four types of biodiversity
data: species <strong>abundances</strong>, <strong>traits</strong>,
<strong>population genetics</strong>, and
<strong>phylogenetics</strong>. In this workshop, we’ll cover skills for
working with each data type separately, and conclude with what we can
accomplish when we integrate multiple data types into the same
analysis.</p>
<p>To make this a little more interesting, we’ll ground our learning
with a “real-world” (or, close to real-world) example!</p>
<div id="accordionInstructor3" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor3" aria-expanded="false" aria-controls="collapseInstructor3">
  <h3 class="accordion-header" id="headingInstructor3">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor3" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor3" aria-labelledby="headingInstructor3">
<div class="accordion-body">
<p>Transition to data narrative.</p>
</div>
</div>
</div>
</div>
<p>For this workshop, we’ll be working with some <em>simulated</em> data
based on real insect species found in the Hawaiian Islands The data
contain real taxonomic names (so we can use workflows designed for
dealing with real taxonomic data) but the abundances, traits, genetics,
and phylogeny are simulated using a <em>process model</em> (more on that
in Part II of this workshop).</p>
</section><section id="hawaiian-sovereignty"><h2 class="section-heading">Hawaiian sovereignty<a class="anchor" aria-label="anchor" href="#hawaiian-sovereignty"></a>
</h2>
<hr class="half-width">
<div id="accordionInstructor4" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor4" aria-expanded="false" aria-controls="collapseInstructor4">
  <h3 class="accordion-header" id="headingInstructor4">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor4" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor4" aria-labelledby="headingInstructor4">
<div class="accordion-body">
<p>Instead of “…the Indigenous land where we gather…” you could
reference the specific communities.</p>
</div>
</div>
</div>
</div>
<p>Building from our acknowledgement of the Indigenous land where we
gather for this workshop, we need to acknowledge that over a century of
settler research has been conducted across the Hawaiian Islands without
the consent of the Kanaka Maoli, the Native Hawaiians. The aina, or
lands, of the Kanaka Maoli were stolen, their sovereignty illegally
stripped, their culture, communities, and bodies killed. Sovereignty and
the return of control of <em>data</em> to Indigenous people and
communities is one of the learning objectives of this workshop. To start
us on that journey, here are the words (<a href="https://youtu.be/RwWNigoZ5ro?t=89" class="external-link">watch on youtube</a>) of
Dr. Haunani-Kay Trask, a leader in the Hawaiian sovereignty
movement:</p>
<blockquote>
<p>We are not Americans! We will die as Hawaiians! We will never be
Americans! I am here to explain what sovereignty is. Sovereignty, as
many people say, is a feeling. The other day in the paper I read
sovereignty is aloha, it’s love. Later on someone said it’s pride.
No. Sovereignty is government. Sovereignty is government! It is an
attribute of nationhood. We already have aloha. We already have pride.
We already know who we are. Are we sovereign? No! Because we don’t have
a government. Because we don’t control a land base. Sovereignty is not a
feeling, it is the power of government. It is political power!</p>
</blockquote>
<figure><img src="https://d1l18ops95qbzp.cloudfront.net/wp-content/2021/07/03134707/manoa-trask-h-uprichard.jpeg" alt="" class="figure mx-auto d-block"><figcaption>A portrait of Dr. Haunani-Kay Trask, a leader in the
Hawaiian soveirgnity movement. Photo by Brett Uprichard</figcaption></figure><p>One of the contributions we as researchers can make to rematriation
and repatriation of resources is to cede control of legacy biological
data back to Indigenous communities, and preferably to engage in
co-produced research from the beginning, where consent, governance, and
mutual benefit are agreed upon before research begins. In Episode 3:
CARE and FAIR principles, we will be learning about protocols developed
by <a href="https://www.enrich-hub.org/" class="external-link">ENRICH</a> and <a href="https://localcontexts.org/" class="external-link">Local Contexts</a> that center
Indigenous data sovereignty and governance.</p>
</section><section id="more-about-our-simulated-data"><h2 class="section-heading">More about our simulated data<a class="anchor" aria-label="anchor" href="#more-about-our-simulated-data"></a>
</h2>
<hr class="half-width">
<p>The Hawaiian Islands formed over a volcanic hotspot, as such Kauai is
about 5 million years old, while the Big Island is still forming.</p>
<figure><img src="../intro_img/intro_01.png" alt="" class="figure mx-auto d-block"><figcaption>A painting by John D. Dawson shows how the ecosystems
of the Hawaiian Islands change across the the chronosequence, while an
overlaid false-color map shows the substrate ages of the islands, with
the substrates of the Big Island being youngest, progressing in
increasing age toward Kauai</figcaption></figure><p>This <em>chronosequence</em> allows us to observe in the Modern (as
opposed to the fossil record) what we might hypothesize to be different
eco-evolutionary stages, or “snapshots,” of community assembly. We might
further hypothesize that different assembly processes are more or less
prevalent at different snapshots, such as increased importance of
immigration early on, and greater importance of <em>in situ</em>
diversification later on.</p>
<figure><img src="../intro_img/intro_02.png" alt="" class="figure mx-auto d-block"><figcaption>A figure showing how hypothesized mechanisms might
trade-off across the chronosequence: in communities on older substrates
we might expect more time for evolution and consequently assembly by
speciation and competitive coexistence being important. In contrast on
younger substrates we might expect less time for evolution and thus
assembly by immigration and neutral ecological drift being important</figcaption></figure><div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion3"></a>
</h3>
<div class="callout-content">
<p>What data could we bring to bear on exploring these hypothesized
changes in assembly processes?</p>
</div>
</div>
</div>
<figure><img src="../intro_img/intro_03.png" alt="" class="figure mx-auto d-block"><figcaption>A figure showing a cartoon of organisms (shown as
dots) with different colors representing different species, and
different sizes of the dots representing body sizes—a common and useful
trait.</figcaption></figure><div id="accordionInstructor5" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor5" aria-expanded="false" aria-controls="collapseInstructor5">
  <h3 class="accordion-header" id="headingInstructor5">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor5" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor5" aria-labelledby="headingInstructor5">
<div class="accordion-body">
<p>Breakout groups or collaborate on the white board.</p>
</div>
</div>
</div>
</div>
<div id="discussion4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion4"></a>
</h3>
<div class="callout-content">
<p>How do we express the patterns in these data - What patterns do you
see in the data? - How could you summarize these data?</p>
<p>How would you expect genetic diversity, species diversity, trait, or
phylogenetic diversity to vary with island age?</p>
</div>
</div>
</div>
</section><section id="workshop-logistics-and-preview"><h2 class="section-heading">Workshop logistics and preview<a class="anchor" aria-label="anchor" href="#workshop-logistics-and-preview"></a>
</h2>
<hr class="half-width">
<p>For the rest of the workshop, we’ll take a tour of the data types and
then bring them together.</p>
<div class="section level3">
<h3 id="helpful-links">Helpful links<a class="anchor" aria-label="anchor" href="#helpful-links"></a>
</h3>
<p>Course website: <a href="https://role-model.github.io/multidim-biodiv-data/" class="external-link uri">https://role-model.github.io/multidim-biodiv-data/</a></p>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></section><section id="aio-care"><p>Content from <a href="care.html">The CARE principles and open data</a></p>
<hr>
<p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/care.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is Indigenous data sovereignty?</li>
<li>What are the CARE principles?</li>
<li>What is Local Contexts Hub, and how can we use it to implement the
CARE principles in our work?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal">
<li>Articulate the mission of Local Contexts in the context of the CARE
and FAIR principles</li>
<li>Apply Bio-Cultural and Traditional Knowledge Notices appropriately
to datasets</li>
<li>Create an account on Local Contexts Hub</li>
<li>Create a project on Local Contexts Hub</li>
<li>Query the Local Contexts Hub API to link a dataset with project IDs
on Local Contexts Hub</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<section id="indigenous-data-soverginty"><h2 class="section-heading">Indigenous data soverginty<a class="anchor" aria-label="anchor" href="#indigenous-data-soverginty"></a>
</h2>
<hr class="half-width">
<p>To aid in our learning about and discussion of Indigenous data
soverginty we will watch this video from <a href="https://localcontexts.org/" class="external-link">Local Contexts</a>:</p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/g7-RuXmLewk" data-external="1" allowfullscreen>
</iframe>
</section><section id="the-care-principles"><h2 class="section-heading">The CARE Principles<a class="anchor" aria-label="anchor" href="#the-care-principles"></a>
</h2>
<hr class="half-width">
<p>The <a href="">CARE principles</a> developed by <a href="">xyz <em>et
al.</em></a> codify Indigenous data sovereignty into actionable
standards for data governance. The CARE principles are designed to live
alongside and make more just the FAIR principles. The FAIR principles
are seen as also being a necessary of Indigenous data sovereignty and
governance</p>
<figure><img src="https://images.squarespace-cdn.com/content/v1/5d3799de845604000199cd24/1567592451970-2R8XFL670QNMAGDRV7ED/Be+FAIR+and+CARE.png?format=2500w" class="figure mx-auto d-block"></figure><div class="section level3">
<h3 id="collective-benefit">Collective benefit<a class="anchor" aria-label="anchor" href="#collective-benefit"></a>
</h3>
<p>Collective benefit means that the communities from which data are
gathered should share in any benefit derived from those data.</p>
</div>
<div class="section level3">
<h3 id="authority-to-control">Authority to control<a class="anchor" aria-label="anchor" href="#authority-to-control"></a>
</h3>
<p>Authority to control means that communities have the right and
authority to govern how data pertaining to themselves (their cultures,
knowledges, bodies, environments) is gathered, used, shared, and
re-used.</p>
</div>
<div class="section level3">
<h3 id="responsibility">Responsibility<a class="anchor" aria-label="anchor" href="#responsibility"></a>
</h3>
<p>Responsibility refers to the requirement that researchers engage with
respect in the process of working with local communities, including the
ethics of those communities</p>
</div>
<div class="section level3">
<h3 id="ethics">Ethics<a class="anchor" aria-label="anchor" href="#ethics"></a>
</h3>
<p>Ethics speaks to the fact that the ethics of local communities must
be respected and privileged when making governing decisions about
data</p>
</div>
</section><section id="implementing-the-care-principles-through-the-local-contexts-hub"><h2 class="section-heading">Implementing the CARE Principles through the Local Contexts Hub<a class="anchor" aria-label="anchor" href="#implementing-the-care-principles-through-the-local-contexts-hub"></a>
</h2>
<hr class="half-width">
<p>Local Contexts Hub serves Traditional Knowledge <a href="https://localcontexts.org/labels/traditional-knowledge-labels/" class="external-link">Labels</a>
and <a href="https://localcontexts.org/notice/tk-notice/" class="external-link">Notices</a>,
and Biocultural <a href="https://localcontexts.org/labels/biocultural-labels/" class="external-link">Labels</a>
and <a href="https://localcontexts.org/notice/bc-notice/" class="external-link">Notices</a>.
We learned about those in the video from Local Contexts, now we’ll learn
how Labels and Notices can operationalize the CARE principles.</p>
<p>Follow along with this <a href="">slide deck</a>.</p>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>Give the slide show</p>
</div>
</div>
</div>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>Set up your own account on <a href="https://localcontextshub.org/register/" class="external-link">Local Contexts
Hub</a>!</p>
</div>
</div>
</div>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" aria-labelledby="headingInstructor2" data-bs-parent="#accordionInstructor2">
<div class="accordion-body">
<p>Give a tour of your own Local Contexts Hub account</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>“Be FAIR and CARE”: The CARE principles (<strong>C</strong>ollective
benefit, <strong>A</strong>uthority to control,
<strong>R</strong>esponsibility, and <strong>E</strong>thics) are part
of a framework emphasizing Indigenous rights and interests in the
context of data sharing ecosystems.</li>
<li>Local Contexts Hub provides Labels and Notices for learning more
about and engaging with Indigenous data in accordance with the CARE
principles.</li>
<li>Work remains to be done in order to spread the use of TK and BC
Labels/Notices by researchers and data repositories</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-abundance-data"><p>Content from <a href="abundance-data.html">Abundance Data</a></p>
<hr>
<p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/abundance-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is abundance data, and how can we use it to gain insights about
a system?</li>
<li>How do I clean and manipulate abundance data to prepare for
analyses?</li>
<li>How do I calculate summary statistics and relate these to ecological
pattern?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants should be able to…</p>
<ol style="list-style-type: decimal">
<li>Describe ecological abundance data and what it can tell us about a
system</li>
<li>Import and examine organismal abundance data from .csv files</li>
<li>Clean taxonomic names</li>
<li>Manipulate abundance data into different formats</li>
<li>Generate species abundance distribution plots from abundance
data</li>
<li>Summarize species abundance data using Hill numbers</li>
<li>Interpret how different Hill numbers correspond to different
signatures of species diversity</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="introduction-to-abundance-data">Introduction to abundance data<a class="anchor" aria-label="anchor" href="#introduction-to-abundance-data"></a>
</h1>
<p>Abundance data is one of the most widely-collected biodiversity data
types. It generally keeps track of how many individuals of each species
(or taxonomic unit) are present at a site. It may also be recorded as
relative abundances or percent cover, depending on the group.</p>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>How have you used species abundance data in your work?</p>
</div>
</div>
</div>
<p>There’s a huge diversity of applications of abundance data. Here,
we’ll focus on one of the most generally-applicable approaches, which is
to look at the <em>diversity</em> of a system. By diversity, we mean how
abundance is distributed among the different taxonomic groups present in
that system. This incorporates both species richness, and the
<em>evenness</em> of how abundances is portioned among different
species.</p>
<p>Let’s make this more concrete. If we plot the abundances of all the
species in a system, sorted from most to least abundant, we end up with
something like this:</p>
<p>This is the <em>species abundance distribution</em>, or SAD.</p>
<p>The SAD can show how even or uneven our system is.</p>
<p>Ok, great. What if we want to make comparisons between different
systems? For that we need a quantifiable metric.</p>
<p>There are dozens - literally - of summary statistics for ecological
diversity.</p>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>What diversity metrics have you encountered?</p>
</div>
</div>
</div>
<p>For this workshop, we’re going to focus on <strong>Hill
numbers</strong>.</p>
<p>Hill numbers are a family of diversity indices that can be described
verbally as the “effective number of species”. That is, they describe
how many species of <em>equal abundances</em> would be present in a
system to generate the corresponding diversity index.</p>
<div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion3"></a>
</h3>
<div class="callout-content">
<p>Let’s get into Hill numbers more on the whiteboard.</p>
</div>
</div>
</div>
<p>Hill numbers are similar to other diversity indices you might have
encountered. In fact, they mathematically converge with Shannon,
Simpson, and species evenness.</p>
</div>
<div class="section level1">
<h1 id="working-with-abundance-data">Working with abundance data<a class="anchor" aria-label="anchor" href="#working-with-abundance-data"></a>
</h1>
<p>Ok, so how about we do some actual coding?</p>
<p>For this episode, we’ll be working with a couple of specialized
packages for biological data.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">vegan</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">tidyr</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<p>We’ll be working with ecological species abundance data stored in
<code>.csv</code> files. For help working with other storage formats, <a href="https://datacarpentry.org/R-ecology-lesson/05-r-and-databases.html" class="external-link">the
Carpentries’ Ecological Data lesson materials on databases</a> are a
great place to start!</p>
<p>Let’s load the data:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/role-model/multidim-biodiv-data/main/episodes/data/abundances_raw.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>And look at what we’ve got:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span></span></code></pre>
</div>
<p><code>abundances</code> is a data frame with columns for
<code>island</code>, <code>site</code>, <code>GenSp</code>, and
<code>abundance</code>. Each row tells us how many individuals of each
species (GenSp) were recorded at a given site on each island. This is a
common format - imagine you are censusing a plant quadrat, counting up
how many individuals of each species you see.</p>
</div>
<div class="section level2">
<h2 id="cleaning-taxonomic-names">Cleaning taxonomic names<a class="anchor" aria-label="anchor" href="#cleaning-taxonomic-names"></a>
</h2>
<p>The first thing we’ll want to do is check for human error wherever we
can, in this case in the form of typos in data entry.</p>
<p>The <code>taxize</code> R package can help identify and resolve
simple typos in taxonomic names.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_list</span> <span class="op">&lt;-</span> <span class="va">abundances</span><span class="op">$</span><span class="va">GenSp</span></span>
<span></span>
<span><span class="va">name_resolve</span> <span class="op">&lt;-</span> <span class="fu">gnr_resolve</span><span class="op">(</span><span class="va">species_list</span>, best_match_only <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                            canonical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># returns only name, not authority</span></span></code></pre>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor1" aria-labelledby="headingInstructor1">
<div class="accordion-body">
<p>Sometimes <code>gnr_resolve</code> doesnt work. From the
<code>taxize</code> documentation:</p>
<blockquote>
<p>503 Service Unavailable: This is typically a temporary problem; often
given when a server is handling too many requests, and is briefly
down.</p>
</blockquote>
</div>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">name_resolve</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mismatches</span> <span class="op">&lt;-</span> <span class="va">name_resolve</span><span class="op">[</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span> <span class="op">!=</span></span>
<span>                                <span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">mismatches</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"user_supplied_name"</span>, <span class="st">"matched_name2"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<p>Four of these are just typos. But <code>Agrotis chersotoides</code>
in our data is resolved only to <code>Agrotis</code> and
<code>Metrothorax deverilli</code> is resolved to
<code>Metrothorax</code>. What’s up there?</p>
<div id="discussion4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion4"></a>
</h3>
<div class="callout-content">
<p><em>Agrotis</em> and <em>Metrothorax</em> taxonomic resolution</p>
</div>
</div>
</div>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor2" aria-labelledby="headingInstructor2">
<div class="accordion-body">
<p>Go to google and you’ll see that Agrotis chersotoides is a synonym of
Peridroma chersotoides, while Metrothorax deverilli is a valid species
with little information about it, thus it dosen’t show up in GNR.</p>
</div>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span><span class="op">[</span></span>
<span>    <span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span> <span class="op">==</span> <span class="st">"Agrotis chersotoides"</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>    <span class="st">"Peridroma chersotoides"</span></span>
<span></span>
<span><span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span><span class="op">[</span></span>
<span>    <span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span> <span class="op">==</span> <span class="st">"Metrothorax deverilli"</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>    <span class="st">"Metrothorax deverilli"</span></span></code></pre>
</div>
<p>Now we need to add the newly-resolved names to our
<code>abundances</code> data. For this, we’ll use a function called
<code>left_join</code>.</p>
<div id="discussion5" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion5"></a>
</h3>
<div class="callout-content">
<p>Visual of how <code>left_join</code> works</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">abundances</span>, <span class="va">name_resolve</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"GenSp"</span> <span class="op">=</span> <span class="st">"user_supplied_name"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">abundances</span><span class="op">$</span><span class="va">final_name</span> <span class="op">&lt;-</span> <span class="va">abundances</span><span class="op">$</span><span class="va">matched_name2</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-species-abundance-distributions">Visualizing species abundance distributions<a class="anchor" aria-label="anchor" href="#visualizing-species-abundance-distributions"></a>
</h2>
<p>Now that we have cleaned data, we can generate plots of how abundance
is distributed.</p>
<p>Because we’ll want to look at each island separately, we’ll use the
<code>split</code> command to break the <code>abundances</code> data
frame apart by island. <code>split</code> will split a dataframe into
groups defined by the <code>f</code> argument (for “factor”) - in this
case, the different values of the <code>island</code> column of the
<code>abundances</code> data frame:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">island_abundances</span> <span class="op">&lt;-</span> <span class="fu">split</span><span class="op">(</span><span class="va">abundances</span>, f <span class="op">=</span> <span class="va">abundances</span><span class="op">$</span><span class="va">island</span><span class="op">)</span></span></code></pre>
</div>
<p>Usual practice is to plot distributions of abundance as the species
abundance on the y-axis and the <em>rank</em> of that species (from
most-to-least-abundant) on the x-axis. This allows us to make
comparisons between sites that don’t have any species in common.</p>
<p>Now, we’ll construct a plot with lines for the abundances of species
on each island.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># figure out max number of species at a site for axis limit setting below</span></span>
<span><span class="va">max_sp</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">island_abundances</span>, <span class="va">nrow</span><span class="op">)</span></span>
<span><span class="va">max_sp</span> <span class="op">&lt;-</span> <span class="fu">max</span><span class="op">(</span><span class="va">max_sp</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">island_abundances</span><span class="op">$</span><span class="va">Kauai</span><span class="op">$</span><span class="va">abundance</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Species abundances at each site"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Rank"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Abundance"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#440154FF"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="va">max_sp</span><span class="op">)</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fu">max</span><span class="op">(</span><span class="va">abundances</span><span class="op">$</span><span class="va">abundance</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    log <span class="op">=</span> <span class="st">'y'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">island_abundances</span><span class="op">$</span><span class="va">Maui</span><span class="op">$</span><span class="va">abundance</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#21908CFF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">island_abundances</span><span class="op">$</span><span class="va">BigIsland</span><span class="op">$</span><span class="va">abundance</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#FDE725FF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">legend</span><span class="op">(</span></span>
<span>    <span class="st">"topright"</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Kauai"</span>, <span class="st">"Maui"</span>, <span class="st">"Hawaii"</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    pt.lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#FDE725FF"</span><span class="op">)</span>,</span>
<span>    bty <span class="op">=</span> <span class="st">"n"</span>, </span>
<span>    cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/abundance-data-rendered-base-R-SAD-plots-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="discussion6" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion6"></a>
</h3>
<div class="callout-content">
<p>What do you notice about the SADs on the different islands?</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="quantifying-diversity-using-hill-numbers">Quantifying diversity using Hill numbers<a class="anchor" aria-label="anchor" href="#quantifying-diversity-using-hill-numbers"></a>
</h2>
<p>Let’s calculate Hill numbers to put some numbers to these shapes.</p>
<p>For this, we’ll need what’s known as a <strong>site by species
matrix</strong>. This is a very common data format for ecological
diversity data.</p>
<div class="section level3">
<h3 id="site-by-species-matrix">Site-by-species matrix<a class="anchor" aria-label="anchor" href="#site-by-species-matrix"></a>
</h3>
<p>A site by species sites as rows and species as columns. We can get
there using the <code>pivot_wider</code> function from the package
<code>tidyr</code>:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="fu">pivot_wider</span><span class="op">(</span><span class="va">abundances</span>, id_cols <span class="op">=</span> <span class="va">site</span>,</span>
<span>                              names_from <span class="op">=</span> <span class="va">final_name</span>,</span>
<span>                              values_from <span class="op">=</span> <span class="va">abundance</span>, </span>
<span>                              values_fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<p>We’ll want this data to have <em>row names</em> based on the sites,
so we’ll need some more steps:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">row.names</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span><span class="op">$</span><span class="va">site</span></span>
<span></span>
<span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="calculating-hill-numbers-with-hillr">Calculating Hill numbers with <code>hillR</code>
<a class="anchor" aria-label="anchor" href="#calculating-hill-numbers-with-hillr"></a>
</h3>
<p>The <code>hillR</code> package allows us to calculate Hill
numbers.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_0</span> <span class="op">&lt;-</span> <span class="fu">hill_taxa</span><span class="op">(</span><span class="va">abundance_wide</span>, q <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hill_0</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_1</span> <span class="op">&lt;-</span> <span class="fu">hill_taxa</span><span class="op">(</span><span class="va">abundance_wide</span>, q <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hill_1</span></span></code></pre>
</div>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>Calculate the hill numbers for q = 2.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_2</span> <span class="op">&lt;-</span> <span class="fu">hill_taxa</span><span class="op">(</span><span class="va">abundance_wide</span>, q <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hill_2</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="relating-hill-numbers-to-patterns-in-diversity">Relating Hill numbers to patterns in diversity<a class="anchor" aria-label="anchor" href="#relating-hill-numbers-to-patterns-in-diversity"></a>
</h2>
<p>Let’s revisit the SAD plots we generated before, and think about
these in terms of Hill numbers.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">island_abundances</span><span class="op">$</span><span class="va">Kauai</span><span class="op">$</span><span class="va">abundance</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Species abundances at each site"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Rank"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Abundance"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#440154FF"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="va">max_sp</span><span class="op">)</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fu">max</span><span class="op">(</span><span class="va">abundances</span><span class="op">$</span><span class="va">abundance</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    log <span class="op">=</span> <span class="st">'y'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">island_abundances</span><span class="op">$</span><span class="va">Maui</span><span class="op">$</span><span class="va">abundance</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#21908CFF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">island_abundances</span><span class="op">$</span><span class="va">BigIsland</span><span class="op">$</span><span class="va">abundance</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#FDE725FF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">legend</span><span class="op">(</span></span>
<span>    <span class="st">"topright"</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Kauai"</span>, <span class="st">"Maui"</span>, <span class="st">"Hawaii"</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    pt.lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#FDE725FF"</span><span class="op">)</span>,</span>
<span>    bty <span class="op">=</span> <span class="st">"n"</span>, </span>
<span>    cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/abundance-data-rendered-render-SAD-plots-again-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hill_numbers</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">hill_0</span>, <span class="va">hill_1</span>, <span class="va">hill_2</span><span class="op">)</span></span>
<span><span class="va">hill_numbers</span></span></code></pre>
</div>
</div>
</div>
<div class="section level1">
<h1 id="recap">Recap<a class="anchor" aria-label="anchor" href="#recap"></a>
</h1>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Organismal abundance data are a fundamental data type for population
and community ecology.</li>
<li>The <code>taxise</code> package can help with data cleaning, but
quality checks are often ultimately dataset-specific.</li>
<li>The species abundance distribution (SAD) summarizes site-specific
abundance information to facilitate cross-site or over-time
comparisons.</li>
<li>We can quantify the shape of the SAD using <em>summary
statistics</em>. Specifically, <em>Hill numbers</em> provide a unified
framework for describing the diversity of a system.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section><section id="aio-traits-data"><p>Content from <a href="traits-data.html">Trait data</a></p>
<hr>
<p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/traits-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What information is captured via trait data?</li>
<li>How do you read, clean, and visualize trait data?</li>
<li>What do Hill numbers convey in the context of trait data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants should be able to…</p>
<ol style="list-style-type: decimal">
<li>Import trait data in a CSV format into the R environment</li>
<li>Clean taxonomic names using the <code>taxize</code> package</li>
<li>Aggregate traits</li>
<li>Visualize trait distributions</li>
<li>Calculate Hill numbers</li>
<li>Interpret Hill numbers using ranked trait plots</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<section id="introduction-to-traits"><h2 class="section-heading">Introduction to traits<a class="anchor" aria-label="anchor" href="#introduction-to-traits"></a>
</h2>
<hr class="half-width">
<p>Trait data is both old and new in ecology.</p>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>What have been your experiences with trait data? What are examples of
traits you’ve looked at, or you imagine might be important?</p>
</div>
</div>
</div>
<p>Similarly to abundance, investigating trait data can give insights
into the ecological processes shaping different communities.
Specifically, it would be interesting to answer questions like: 1) how
much does a specific trait vary in a community?; 2) Are all species
converging into a similar trait value or dispersing into a wide range of
values for the trait?; 3) do those patterns of variation change across
different communities? Answering these questions can help us understand
how species are interacting with the environment and/or with each
other.</p>
<p>Defining what exactly constitutes a trait, and measuring traits, can
get complicated in ecology. (For example, there are lots of things we
can measure about even just a leaf - but it’s another story to
understand which of those attributes are telling us something
<em>ecologically</em> meaningful). Ideally, we’d want to focus on a
trait that is easily measureable and also linked to ecological features
of interest. One such trait is <em>body size</em>. Body size is
correlated with a whole host of ecologically meaningful traits, and is
one of the easiest things to measure about an organism.</p>
<p>Similarly to how we look at community-wide distributions of
abundance, we can generate and interpret community-wide <em>trait</em>
distributions.</p>
<p>And, similar to how we can look at species abundance diversity using
Hill numbers, we can look at trait diversity using a <em>trait</em> Hill
number.</p>
</section><section id="working-with-trait-data"><h2 class="section-heading">Working with trait data<a class="anchor" aria-label="anchor" href="#working-with-trait-data"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="importing-and-cleaning-trait-data">Importing and cleaning trait data<a class="anchor" aria-label="anchor" href="#importing-and-cleaning-trait-data"></a>
</h3>
<p>Here, we will work with a dataset containing values of <strong>body
mass</strong> for the species in the Hawaiian islands. Unlike to the
abundance data, each row represents one specimen: in this dataset, each
row contains the body mass measured for that specimen. Our overall goal
here is to clean this data and attach it to the species data, so we can
investigate trait patterns per community.</p>
<p>We’re starting with a challenge!</p>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>In the abundance-data episode, you learned how to read in a
<code>CSV</code> file with the <code>read.csv()</code> function. In
addition, you learned how to clean and standardize the taxonomic
information in your data set using the <code>gnr_resolve()</code>
function from the <code>taxize</code> package.</p>
<p>You’re going to use those skills to import and clean the trait data
set. The trait data set contains two columns: <code>GenSp</code>, which
contains the binomial species names, and <code>mass_g</code>, which
contains the individual’s mass in grams, the metric of body size chosen
for the study. The data also contain typos and a problem with the
taxonomy that you must correct and investigate. Make sure to glance at
the data before cleaning!</p>
<p>Your trait data is located at: <a href="https://raw.githubusercontent.com/role-model/multidim-biodiv-data/main/episodes/data/body_size_data.csv" class="external-link uri">https://raw.githubusercontent.com/role-model/multidim-biodiv-data/main/episodes/data/body_size_data.csv</a></p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Read in the traits data using the <code>read.csv</code> function and
supplying it with the path to the data.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">'https://raw.githubusercontent.com/role-model/multidim-biodiv-data/main/episodes/data/body_size_data.csv'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span></code></pre>
</div>
<p>Check the names of the traits using the <code>gnr_resolve()</code>
function. To streamline your efforts, supply only the unique species
names in your data to <code>gnr_resolve()</code>.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="co"># only need to check the unique names</span></span>
<span><span class="va">species_list</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="va">traits</span><span class="op">$</span><span class="va">GenSp</span><span class="op">)</span></span>
<span></span>
<span><span class="va">name_resolve</span> <span class="op">&lt;-</span> <span class="fu">gnr_resolve</span><span class="op">(</span><span class="va">species_list</span>, best_match_only <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                            canonical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">name_resolve</span><span class="op">)</span></span></code></pre>
</div>
<p>To quickly see which taxa are in conflict with <code>taxize</code>’s,
use bracket subsetting and boolean matching.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mismatches_traits</span> <span class="op">&lt;-</span> </span>
<span>    <span class="va">name_resolve</span><span class="op">[</span><span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span> <span class="op">!=</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span>, </span>
<span>                 <span class="fu">c</span><span class="op">(</span><span class="st">"user_supplied_name"</span>, <span class="st">"matched_name2"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">mismatches_traits</span></span></code></pre>
</div>
<p>Fixing the names comes in two steps. First, we join the
<code>traits</code> dataframe with the <code>name_resolve</code>
dataframe using the <code>left_join()</code> function. Note, we indicate
that the <code>GenSp</code> and <code>user_supplied_name</code> columns
have the same information by supplying a named vector to the
<code>by =</code> argument.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">traits</span>, </span>
<span>                    <span class="va">name_resolve</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"user_supplied_name"</span>, <span class="st">"matched_name2"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                    by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"GenSp"</span> <span class="op">=</span> <span class="st">"user_supplied_name"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span></code></pre>
</div>
<p>Then, to fix <em>Agrotis chersotoides</em> and <em>Metrothorax
deverilli</em>, as in abundance-data episode, use bracketed indexing,
boolean matching, and assignment.</p>
<p>In addition, although not necessary, changing the column name
<code>matched_name2</code> to <code>final_name</code> to give it a more
sensible name for later use is good practice.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span><span class="op">$</span><span class="va">matched_name2</span><span class="op">[</span><span class="va">traits</span><span class="op">$</span><span class="va">matched_name2</span> <span class="op">==</span> <span class="st">"Agrotis"</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>    <span class="st">"Peridroma chersotoides"</span></span>
<span></span>
<span><span class="va">traits</span><span class="op">$</span><span class="va">matched_name2</span><span class="op">[</span><span class="va">traits</span><span class="op">$</span><span class="va">matched_name2</span> <span class="op">==</span> <span class="st">"Metrothorax"</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>    <span class="st">"Metrothorax deverilli"</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span><span class="op">[</span><span class="fu">colnames</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span> <span class="op">==</span> <span class="st">"matched_name2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"final_name"</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="summarizing-and-cleaning-trait-data">Summarizing and cleaning trait data<a class="anchor" aria-label="anchor" href="#summarizing-and-cleaning-trait-data"></a>
</h3>
<p>When analyzing trait data, it often needs to be summarized at a
higher level than the individual. For instance, many community assembly
analyses require species-level summaries, rather than individual
measurements. So, we often want to calculate summary statistics of
traits for each species. For numeric measurements, body size, statistics
like the mean, median, and standard deviation give information about the
center and spread of the distribution of traits for the species. For
this section, you will aggregate the trait data to the species level,
calculating the mean, median, and mode of body size for each
species.</p>
<p>While there are methods to aggregate data in base R, the <a href="https://dplyr.tidyverse.org/" class="external-link">dplyr</a> makes this task and other
data wrangling tasks much more intuitive. The function
<code>group_by()</code> groups your data frame by a variable. The first
argument to <code>group_by()</code> is your data frame, and all
following unnamed arguments are the variables you want to group the data
by. In our case, we want to group by species name, so we’ll supply
<code>final_name</code>.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># group the data frame by species</span></span>
<span><span class="va">traits_sumstats</span> <span class="op">&lt;-</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">traits</span>, <span class="va">final_name</span><span class="op">)</span></span></code></pre>
</div>
<p><code>group_by()</code> just adds an index that tells any following
<code>dplyr</code> functions to perform their calculations on the group
the data frame was indexed by. So, to perform the actual calculations,
we will use the <code>summarize()</code> function
(<code>summarise()</code> works too for non-Americans). The first
argument you supply is the data frame, following by the calculations you
want to perform on the grouped data and what you want to name the
resulting variables. The structure is
<code>new_var_name = function(original_var_name)</code> for each
calculation. Here, we’re using the <code>mean()</code>,
<code>median()</code>, and <code>sd()</code> functions. Then, we’ll take
a look at the new data set with the <code>head()</code> function.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># summarize the grouped data frame, so you're calculating the summary statistics for each species</span></span>
<span><span class="va">traits_sumstats</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">summarize</span><span class="op">(</span></span>
<span>        <span class="va">traits_sumstats</span>,</span>
<span>        mean_mass_g <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">mass_g</span><span class="op">)</span>,</span>
<span>        median_mass_g <span class="op">=</span> <span class="fu">median</span><span class="op">(</span><span class="va">mass_g</span><span class="op">)</span>,</span>
<span>        sd_mass_g <span class="op">=</span> <span class="fu">sd</span><span class="op">(</span><span class="va">mass_g</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_sumstats</span><span class="op">)</span></span></code></pre>
</div>
<p>Finally, you need to add the aggregated species-level information
back to the abundance data, so you have the summary statistics for each
species at each site.</p>
<p>First, re-read in the abundance data you used in the abundance
episode.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/role-model/multidim-biodiv-data/main/episodes/data/abundances_resolved.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span></span></code></pre>
</div>
<p>To join the data, you’ll use <code>left_join()</code>, which is
hopefully getting familiar to you now!</p>
<p><code>abundances</code> should be on the left side, since we want the
species-information for each species in each community.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_sumstats</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">abundances</span>, </span>
<span>                             <span class="va">traits_sumstats</span>, </span>
<span>                             by <span class="op">=</span> <span class="st">"final_name"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_sumstats</span><span class="op">)</span></span></code></pre>
</div>
</div>
</section><section id="visualizing-trait-distributions"><h2 class="section-heading">Visualizing trait distributions<a class="anchor" aria-label="anchor" href="#visualizing-trait-distributions"></a>
</h2>
<hr class="half-width">
<p>Histograms and density plots can help give you a quick look at the
distribution of your data. For quality control purposes, they are useful
to see if there are any suspicious values due to human error. Let’s
overlay the histogram and density plots to get an idea of what
aggregated (histogram) and smoothed (density) representations of the
data look like.</p>
<p>The first argument to the <code>hist()</code> function is your data.
<code>breaks</code> sets the number of histogram bars, which can be
tuned heuristically to find a number of bars that best represents the
distribution of your data. I found that <code>breaks = 40</code> is a
reasonable value. <code>xlab</code>, <code>ylab</code>, and
<code>main</code> are all used to specify labels for the x-axis, y-axis,
and title, respectively. When scaling the plot, sometimes base R
plotting functions aren’t as responsive to the data as we like. To fix
this, you can use the <code>xlim</code> or <code>ylim</code> functions
to set the value range of the x-axis and y-axis, respectively. Here, we
set <code>ylim</code> to a vector <code>c(0, 0.11)</code>, which
specifies the y-axis to range from 0 to 0.11.</p>
<p>If you’re used to seeing histograms, the y-axis may look unfamiliar.
To get the two plots to use the same scale, you set
<code>freq = FALSE</code>, which converts the histogram counts to a
density value. The density scales the distribution from 0-1, so the bar
height (and line height on the density plot) is a fraction of the total
distribution.</p>
<p>Finally, to add the density line to the plot, you calculate the
density of the trait distribution with the <code>density()</code>
function and wrap it with the <code>lines()</code> function.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span></span>
<span>    <span class="va">traits_sumstats</span><span class="op">$</span><span class="va">mean_mass_g</span>,</span>
<span>    breaks <span class="op">=</span> <span class="fl">40</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Average mass (g)"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"All species"</span>,</span>
<span>    freq <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="fu">density</span><span class="op">(</span><span class="va">traits_sumstats</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>In addition to quality control, knowing the distribution of trait
data lends itself towards questions of what processes are shaping the
communities under study. Are the traits overdispersed? Underdispersed?
Does the average vary across communities? To get at these questions, you
need to plot each site and compare among sites. To facilitate
comparison, you’re going to layer the distributions on top of each other
in a single plot. Histograms get cluttered quickly, so let’s use density
plots for this task.</p>
<p>To plot each site, you need a separate dataframe for each site. The
easiest way to do this is with the <code>split()</code> function, which
takes a dataframe followed by a variable to split the dataframe by. It
then returns a list of data frames for each element in the variable.</p>
<p>Running <code>head()</code> on the list returns a lot of output, so
you can use the <code>names()</code> function to make sure the names of
the dataframes in your list are the sites you intended to split by.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># split by site</span></span>
<span><span class="va">traits_sumstats_split</span> <span class="op">&lt;-</span> <span class="fu">split</span><span class="op">(</span><span class="va">traits_sumstats</span>, <span class="va">traits_sumstats</span><span class="op">$</span><span class="va">site</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">)</span></span></code></pre>
</div>
<p>To plot the data, you first need to initialize the plot with the
<code>plot()</code> function of a single site’s density. Here is also
where you add the plot aesthetics (labels, colors, line characteristics,
etc.). You’re labeling the title and axis labels with <code>main</code>,
<code>xlab</code>, and <code>ylab</code>. The <code>lwd</code>
(“linewidth”) function specifies the size of the line, and I found a
value of two is reasonable. You then supply the color of the line
(<code>col</code>) with a color name or hex code. I chose a hex
representing a purple color from the <a href="https://sjmgarnier.github.io/viridisLite/index.html" class="external-link">viridis</a>
color palette. Finally, you need to specify the x-axis limits and y-axis
limits with <code>xlim</code> and <code>ylim</code>. While a negative
mass doesn’t make biological sense, it is an artifact of the smoothing
done by the <code>density</code> function.</p>
<p>To add density lines to the plot, you use the <code>lines()</code>
function with the data for the desired site as the first argument. You
also need to specify the line-specific aesthetics for each line, which
are the <code>lwd</code> and <code>col</code>. I chose the green and
yellow colors from the <em>viridis</em> palette, but you can pick
whatever your heart desires!</p>
<p>Finally, you need a legend for the plot. Within the
<code>legend()</code> function, you specify the position of the legend
(“topright”), the legend labels, the line width, and colors to assign to
the labels.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span></span>
<span>    <span class="fu">density</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">KA_01</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Average mass per study site"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Average mass (g)"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#440154FF"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span></span>
<span>    <span class="fu">density</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">MA_01</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#21908CFF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lines</span><span class="op">(</span></span>
<span>    <span class="fu">density</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">BI_01</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#FDE725FF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">legend</span><span class="op">(</span></span>
<span>    <span class="st">"topright"</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Kauai"</span>, <span class="st">"Maui"</span>, <span class="st">"Big Island"</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#FDE725FF"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>It looks like Kauai insects have a smoother distribution of body
masses, while Maui and Big Island show more clumped distributions,
possibly due to higher immigration on those islands.</p>
</section><div class="section level1">
<h1 id="hill-numbers">Hill numbers<a class="anchor" aria-label="anchor" href="#hill-numbers"></a>
</h1>
<p>Hill numbers are a useful and informative summary of trait diversity,
in addition to other biodiversity variables (<a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Gaggiotti
et al. 2018</a>). They contain signatures of the processes underlying
community assembly. To calculate Hill numbers for traits, you’re going
to use the <code>hill_func()</code> function from the <code>hillR</code>
package. It requires two inputs- the site by species abundance dataframe
and a traits dataframe that has species as rows and traits as
columns.</p>
<div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion3"></a>
</h3>
<div class="callout-content">
<p>Your first challenge, if you don’t have the site by species abundance
dataframe, is to recreate it here. Remember, you’ll need to use the
<code>pivot_wider()</code> function and the <code>abundances</code>
data!</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">tidyr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="fu">pivot_wider</span><span class="op">(</span></span>
<span>    <span class="va">abundances</span>,</span>
<span>    id_cols <span class="op">=</span> <span class="va">site</span>,</span>
<span>    names_from <span class="op">=</span> <span class="va">final_name</span>,</span>
<span>    values_from <span class="op">=</span> <span class="va">abundance</span>,</span>
<span>    values_fill <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># tibbles don't like row names</span></span>
<span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">row.names</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span><span class="op">$</span><span class="va">site</span></span>
<span></span>
<span><span class="co"># remove the site column</span></span>
<span><span class="va">abundance_wide</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Now, to create the traits dataframe for Hill number calculations
purposed, we just need the taxonomic names and trait mean.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_simple</span> <span class="op">&lt;-</span> <span class="va">traits_sumstats</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"final_name"</span>, <span class="st">"mean_mass_g"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span></span></code></pre>
</div>
<p>Next, you need to filter for unique species in the dataframe.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_simple</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span></span></code></pre>
</div>
<p>Finally, you need to set the species names to be row names and remove
the <code>final_name</code> column. Note, you have to use the
<code>drop = FALSE</code> argument because R has the funny behavior that
it likes to convert single-column dataframes into a vector.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">row.names</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">traits_simple</span><span class="op">$</span><span class="va">final_name</span></span>
<span></span>
<span><span class="va">traits_simple</span> <span class="op">&lt;-</span> <span class="va">traits_simple</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">)</span></span></code></pre>
</div>
<p>Next, you’ll use the <code>hill_func()</code> function from the
<code>hillR</code> package to calculate Hill numbers 0-2 of body size
across sites.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span></span>
<span><span class="va">traits_hill_0</span> <span class="op">&lt;-</span> <span class="fu">hill_func</span><span class="op">(</span>comm <span class="op">=</span> <span class="va">abundance_wide</span>, traits <span class="op">=</span> <span class="va">traits_simple</span>, q <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="va">traits_hill_1</span> <span class="op">&lt;-</span> <span class="fu">hill_func</span><span class="op">(</span>comm <span class="op">=</span> <span class="va">abundance_wide</span>, traits <span class="op">=</span> <span class="va">traits_simple</span>, q <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">traits_hill_2</span> <span class="op">&lt;-</span> <span class="fu">hill_func</span><span class="op">(</span>comm <span class="op">=</span> <span class="va">abundance_wide</span>, traits <span class="op">=</span> <span class="va">traits_simple</span>, q <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<p>The output of <code>hill_func()</code> returns quite a few Hill
number options, which are defined in <a href="https://www.annualreviews.org/doi/abs/10.1146/annurev-ecolsys-120213-091540" class="external-link">Chao
et al. 2014</a>. For simplicity’s sake, we will look at
<code>D_q</code>, which from the documentation is the:</p>
<pre><code>&gt; functional Hill number, the effective number of equally abundant and functionally equally distinct species
</code></pre>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_hill_1</span></span></code></pre>
</div>
<p>To gain an intuition for what this means, let’s plot our data. First,
you will wrangle the output into a single dataframe to work with. Since
Hill q = 0 is species richness, let’s focus on Hill q = 1 and Hill q =
2.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits_hill</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>q0 <span class="op">=</span> <span class="va">traits_hill_0</span><span class="op">[</span><span class="fl">3</span>, <span class="op">]</span>, </span>
<span>                          q1 <span class="op">=</span> <span class="va">traits_hill_1</span><span class="op">[</span><span class="fl">3</span>, <span class="op">]</span>, </span>
<span>                          q2 <span class="op">=</span> <span class="va">traits_hill_2</span><span class="op">[</span><span class="fl">3</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># I don't like rownames for plotting, so making the rownames a column</span></span>
<span><span class="va">traits_hill</span><span class="op">$</span><span class="va">site</span> <span class="op">&lt;-</span> <span class="fu">row.names</span><span class="op">(</span><span class="va">traits_hill</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">row.names</span><span class="op">(</span><span class="va">traits_hill</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre>
</div>
<p>Let’s look at how Hill q = 1 compare across sites.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="fu">factor</span><span class="op">(</span><span class="va">traits_hill</span><span class="op">$</span><span class="va">site</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"KA_01"</span>, <span class="st">"MA_01"</span>, <span class="st">"BI_01"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>     <span class="va">traits_hill</span><span class="op">$</span><span class="va">q1</span>, ylab <span class="op">=</span> <span class="st">"Hill q = 1"</span><span class="op">)</span></span></code></pre>
</div>
<p>Hill q = 1 is smallest on the Big Island, largest on Maui, and in the
middle on Kauai. But what does this mean? To gain an intuition for what
a higher or lower trait Hill number means, you will make rank plots of
the trait data, to make a plot analogous to the species abundance
distribution.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># figure out max number of species at a site for axis limit setting below</span></span>
<span><span class="va">max_sp</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">traits_sumstats_split</span>, <span class="va">nrow</span><span class="op">)</span></span>
<span><span class="va">max_sp</span> <span class="op">&lt;-</span> <span class="fu">max</span><span class="op">(</span><span class="va">max_sp</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">KA_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Average mass per study site"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Rank"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Average mass (mg)"</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#440154FF"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="va">max_sp</span><span class="op">)</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fu">max</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">MA_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#21908CFF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">BI_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#FDE725FF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">legend</span><span class="op">(</span></span>
<span>    <span class="st">"topright"</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Kauai"</span>, <span class="st">"Maui"</span>, <span class="st">"Big Island"</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#FDE725FF"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>You can see clearly that the trait distributions differ across
islands, where Maui has more species than the other islands and a
seemingly more even distribution than Kauai. Kauai seems to have a
greater spread of trait values, however, species abundance strongly
influences the calculation of a Hill number. To visualize the role of
abundance, we can replicate each species mean by the abundance of that
species</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># replicate abundances</span></span>
<span><span class="va">biIndividualsTrt</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">BI_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, </span>
<span>                        <span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">BI_01</span><span class="op">$</span><span class="va">abundance</span><span class="op">)</span></span>
<span><span class="va">maIndividualsTrt</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">MA_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, </span>
<span>                        <span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">MA_01</span><span class="op">$</span><span class="va">abundance</span><span class="op">)</span></span>
<span><span class="va">kaIndividualsTrt</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">KA_01</span><span class="op">$</span><span class="va">mean_mass_g</span>, </span>
<span>                        <span class="va">traits_sumstats_split</span><span class="op">$</span><span class="va">KA_01</span><span class="op">$</span><span class="va">abundance</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">kaIndividualsTrt</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Average mass across individuals"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Rank"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Average mass (mg)"</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#440154FF"</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fu">max</span><span class="op">(</span><span class="va">traits_simple</span><span class="op">$</span><span class="va">mean_mass_g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">maIndividualsTrt</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#21908CFF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">points</span><span class="op">(</span></span>
<span>    <span class="fu">sort</span><span class="op">(</span><span class="va">biIndividualsTrt</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#FDE725FF"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">legend</span><span class="op">(</span></span>
<span>    <span class="st">"topright"</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Kauai"</span>, <span class="st">"Maui"</span>, <span class="st">"Big Island"</span><span class="op">)</span>,</span>
<span>    pch <span class="op">=</span> <span class="fl">19</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#FDE725FF"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>What we can see from the individuals-level plot is that much of the
range of trait values on Kauai are broght by very rare species, these
rare species do not contribute substantially to the diversity as
expressed by Hill numbers.</p>
<!-- To factor out the influence -->
<!-- of the number of species from the Hill number and get a stronger -->
<!-- indicator of "evenness", you can divide the Hill number by the number of -->
<!-- species, which you will do for Hill q = 1. -->
<!-- ```{r hill-divide} -->
<!-- sp_counts <- -->
<!--     c( -->
<!--         nrow(traits_sumstats_split$HawaiiIsland_01), -->
<!--         nrow(traits_sumstats_split$Kauai_01), -->
<!--         nrow(traits_sumstats_split$Maui_01) -->
<!--     ) -->
<!-- traits_hill$q1 / sp_counts -->
<!-- ``` -->
<!-- Interestingly, we see that Kauai actually has a more even distribution -->
<!-- of trait values than Maui when accounting for species richness! -->
<div id="discussion4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion4"></a>
</h3>
<div class="callout-content">
<p>What are some other interpretations from the plots, summary
statistics, and what you know from your work?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Traits data contain more information about ecologially significant
traits than just species IDs.</li>
<li>Traits data can be analyed with Hill numbers and visualized with
rank plots similarly to abundance data.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section><section id="aio-popgen-data"><p>Content from <a href="popgen-data.html">Working with population genetic data</a></p>
<hr>
<p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/popgen-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<section id="pop-gen-data"><h2 class="section-heading">Pop gen data<a class="anchor" aria-label="anchor" href="#pop-gen-data"></a>
</h2>
<hr class="half-width">
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is genetic diversity?</li>
<li>What does genetic diversity say about community assembly?</li>
<li>What are common summaries of genetic diversity within species and
communities?</li>
<li>What are common sequence file formats?</li>
<li>How do you read in and manipulate sequence data?</li>
<li>How do you visualize and summarize sequence data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal">
<li>Understand what genetic diversity is</li>
<li>Visualize genetic diversity distributions</li>
<li>Connect genetic diversity distributions to Hill numbers</li>
<li>Identify key features of different sequence data formats</li>
<li>Import population genetic data in multiple formats into the R
environment</li>
<li>Manipulate pop gen data by adding and removing sequences</li>
<li>Calculate genetic diversity <span class="math inline">\(\pi\)</span>
</li>
<li>Calculate Hill numbers</li>
<li>Interpret Hill numbers</li>
</ol>
</div>
</div>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<div class="section level2">
<h2 id="what-is-genetic-diversity">What is genetic diversity?<a class="anchor" aria-label="anchor" href="#what-is-genetic-diversity"></a>
</h2>
<p>GEOBON, in their description of Essential Biodiversity Variables, <a href="https://geobon.org/ebvs/what-are-ebvs/" class="external-link">define intraspecific
genetic diversity</a> as:</p>
<blockquote>
<p>The variation in DNA sequences among individuals of the same
species.</p>
</blockquote>
<p>How is this genetic diversity generated? At a proximal level, genetic
diversity is generated through the processes of recombination and
mutation. These two mechanisms lay the foundation for evolutionary
forces like genetic drift, natural selection, and gene flow to shape the
distribution of genetic diversity within populations and across
communities.</p>
<p>Perhaps one of the most elegant theoretical frameworks for modeling
genetic diversity is coalescent theory. Coalescent theory is a model of
how alleles sampled from a population may have originated from a common
ancestor. Under the assumptions of no recombination, no natural
selection, no gene flow, and no population structure, i.e. neutrality,
it estimates the time to the most recent common ancestor (TMRCA) between
two sampled alleles and the variance of this estimate. When a population
is especially large (i.e. many breeding individuals), the TMRCA of two
randomly sampled alleles is large. This is because the per-generation
probability of two alleles coalescing is <code>1 / 2Ne</code>, where
<code>Ne</code>, the effective population size, is the number of
breeding individuals in an idealized population that follows the
assumptions stated above. In most, if not all, natural environments
conditions are not ideal and <code>Ne</code> does not perfectly
represent the census size of a population (see <a href="https://www.biorxiv.org/content/10.1101/2021.02.03.429633v2" class="external-link">Lewontin’s
paradox</a>). While the relationship between <code>Ne</code> and census
population size is not perfect, in most cases <code>Ne</code> increases
as population size increases and can be used to understand the relative
sizes of populations or size changes over time. The intuition is best
understood through visualization.</p>
<p>For a population of 10 diploid individuals (20 allele copies), the
probability of two randomly sampled allele coalescing each generation is
1 / (2 * 10), or 1/20. In this single simulation (conducted using <a href="https://github.com/cooplab/popgen-notes/blob/master/Rcode/track_alleles.R" class="external-link">Graham
Coop’s code</a>) , the two sampled alleles, whose genealogy is traced in
blue and red, coalesce 16 generations in the past.</p>
<figure><img src="http://127.0.0.1:26201/graphics/10d9d9fc-fe25-474e-8d09-ff197b99e224.png" class="figure mx-auto d-block"></figure><p>However, when the population is smaller (5 individuals, 10 allele
copies), the probability of coalescence each generation is 1 / (2 * 5)
or 1/10. In this simulation, two randomly sampled alleles coalesce after
6 generations.</p>
<figure><img src="http://127.0.0.1:26201/graphics/21826093-5257-4379-a9df-aba895384e41.png" alt="" class="figure mx-auto d-block"><figure><img src="http://127.0.0.1:26201/graphics/21826093-5257-4379-a9df-aba895384e41.png" class="figure mx-auto d-block"></figure></figure><p>The TMRCA of the two sampled alleles corresponds with the smaller
<code>Ne</code> of this idealized population.</p>
<p>Visualizing this path of ancestry as a tree would look something like
this for 5 sampled allele copies.</p>
<p><a href="https://cdmuir.shinyapps.io/genetree-to-phylogeny/" class="external-link"><img src="../images/gene%20tree%20no%20mutations.png" class="figure"></a></p>
<p>Overlaying mutations on the genealogy results in generally more
mutations occurring on longer branches, although because this is a
stochastic process, the relationship isn’t always perfect (note- this is
a tree from a different simulation to the one above).</p>
<p><a href="https://cdmuir.shinyapps.io/genetree-to-phylogeny/" class="external-link"><img src="../images/genetree%20with%20mutations-01.png" alt="This represents a genealogy of 5 sampled allele copies from a population, with mutations represented as red boxes. Multiple numbers inside the boxes indicate more mutations." class="figure"></a></p>
<p>As empirical biologists, what we have to work with are sequence data.
The accumulation of mutations along a sequence enables us to estimate
the TMRCA of a sample of alleles, leading to an estimate of the
<code>Ne</code> of the population.</p>
<p>One of the simplest and most widespread estimators of <code>Ne</code>
is the average number of pairwise differences between sequences sampled
from a population (𝞹, nucleotide diversity, <a href="https://www.pnas.org/doi/abs/10.1073/pnas.76.10.5269" class="external-link">Nei and Li
1979</a>). The basic idea underlying 𝞹 is that more sequence differences
among a sample of sequences results from longer branch lengths in the
genealogy of the sample, leading to a longer TMRCA and larger
<code>Ne</code>. Given its simplicity and ubiquity, we will cover how to
calculate the average number of pairwise differences between sequences
and summarize this value across assemblages.</p>
<div class="section level4">
<h4 id="resources-for-understanding-the-coalescent">Resources for understanding the coalescent<a class="anchor" aria-label="anchor" href="#resources-for-understanding-the-coalescent"></a>
</h4>
<ul>
<li>
<a href="http://cooplab.github.io/popgen-notes/" class="external-link">Graham Coop’s
popgen notes</a>
<ul>
<li>awesome resource for population genetics in general</li>
</ul>
</li>
<li>
<a href="https://phytools.shinyapps.io/coalescent-plot/" class="external-link">learnPopGen
app</a>
<ul>
<li>visualize coalescent genealogies with different numbers of
individuals in a population and different numbers of generations</li>
</ul>
</li>
<li>
<a href="https://bedford.io/projects/coaltrace/" class="external-link">coaltrace app</a>
<ul>
<li>a dynamic and interactive visualization of coalescent genealogies
with the ability to change parameters on the fly</li>
</ul>
</li>
<li>
<a href="https://bedford.io/projects/phylodynamics-lecture/coalescent.html#/" class="external-link">Bedford
lab slides</a>
<ul>
<li>a great set of slides that explains the coalescent in a
complementary way to Graham Coop’s notes</li>
</ul>
</li>
<li>
<a href="https://cdmuir.shinyapps.io/genetree-to-phylogeny/" class="external-link">cdmuir
app</a>
<ul>
<li>uses the coalescent to simulate gene trees and mutations along
sequences</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="what-does-genetic-diversity-say-about-community-assembly">What does genetic diversity say about community assembly?<a class="anchor" aria-label="anchor" href="#what-does-genetic-diversity-say-about-community-assembly"></a>
</h2>
<p>Under neutrality and stable demographic histories, variation in
genetic diversity across a community reflects differences in long-term
population sizes across the community. Rather than a snapshot of
present-day abundance, this can be thought of as an average abundance
across generations.</p>
<p>In much the same way as a community with a few, highly abundant
species and many low abundance species may indicate a non-neutral
community assembly process like environmental filtering, a community
that contains a few species with high genetic diversity and many species
with low genetic diversity may indicate similar non-neutral community
assembly processes over a longer timescale.</p>
<p>In addition, genetic diversity is impacted by demographic events and
natural selection. Populations that have undergone a demographic
expansion or contraction in the recent past will have distinct
coalescent genealogies and therefore different genetic diversities. For
instance, a population that has experienced a recent demographic
expansion will have a TMRCA more recent than a population with the same
population size that has remained stable through the same time period,
illustrated below.</p>
<p><a href="https://bedford.io/projects/phylodynamics-lecture/coalescent.html#/20" class="external-link"><img src="../images/changing_pop_size_1.png" alt="Image of coalescent genealogies traced from stable and growing populations. From the Bedford Lab coalescent slides." class="figure"></a></p>
<p>Distinguishing a small, constant-sized population from a large
population that’s grown dramatically in the recent past is difficult
with single-locus data typical of large-scale biodiversity studies.
However, when paired with other data like traits and abundances, we can
make more refined interpretations of the shape of genetic diversity
across the community. For instance, the interpretation of a community
with a few highly diverse species paired with many species with low
genetic diversity may reflect a community with a few species that were
not impacted by a historical environmental disturbance alongside many
species that had their populations dramatically reduced, but have
rapidly grown to larger modern population sizes.</p>
</div>
</div>
<div class="section level1">
<h1 id="summarizing-genetic-diversity-across-species">Summarizing genetic diversity across species<a class="anchor" aria-label="anchor" href="#summarizing-genetic-diversity-across-species"></a>
</h1>
<div class="section level3">
<h3 id="genetic-diversity-distributions">Genetic diversity distributions<a class="anchor" aria-label="anchor" href="#genetic-diversity-distributions"></a>
</h3>
<p>{{pi across species neutral viz}}</p>
<p>{{pi across species non-neutral viz}}</p>
</div>
<div class="section level3">
<h3 id="hill-numbers-of-genetic-diversity-across-species">Hill numbers of genetic diversity across species<a class="anchor" aria-label="anchor" href="#hill-numbers-of-genetic-diversity-across-species"></a>
</h3>
<p>In the context of summarizing species abundances, Hill numbers are
the effective number of species in a community, where each exponent
(q-value) increasingly weights rare species more heavily. The
interpretation is similar when summarizing genetic diversity. Hill
numbers are the effective number of species in a community, where each
q-value increasingly weights species with lower genetic diversity more
heavily. Rather than measuring an effective diversity of abundances,
this is an effective diversity of genetic diversities.</p>
</div>
</div>
<div class="section level1">
<h1 id="work-with-pop-gen-data">Work with pop gen data<a class="anchor" aria-label="anchor" href="#work-with-pop-gen-data"></a>
</h1>
<div class="section level2">
<h2 id="sequences">Sequences<a class="anchor" aria-label="anchor" href="#sequences"></a>
</h2>
<div class="section level3">
<h3 id="fasta">FASTA<a class="anchor" aria-label="anchor" href="#fasta"></a>
</h3>
<p>A FASTA file contains one or more DNA sequences, where the DNA
sequence is preceded by a line with a carat (“&gt;”) followed by an
unique sequence ID:</p>
<pre><code>&gt;seq0010
GATCCCCAATTGGGG</code></pre>
<p>This is perhaps the most common way to store sequence information as
it is simple and has historical inertia. For more information, see the
<a href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat/#:~:text=In%20FASTA%20format%20the%20line,should%20not%20contain%20any%20spaces." class="external-link">the
NCBI page</a>.</p>
</div>
<div class="section level3">
<h3 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h3>
<p>Prior to analysis, sequence data must be aligned if they haven’t been
already. A sequence alignment is necessary to identify regions of DNA
that are similar or that vary. There are <a href="https://en.wikipedia.org/wiki/List_of_sequence_alignment_software" class="external-link">many
pieces of software to perform alignments</a> and all have their
advantages and disadvantages, but we will choose a single program, <a href="https://www.ebi.ac.uk/Tools/msa/clustalo/" class="external-link">Clustal Omega</a>,
which is fast and robust for short sequence alignments. While Clustal
Omega can be run from the command line, we will use the R package <a href="https://bioconductor.org/packages/release/bioc/html/msa.html" class="external-link">msa</a>
to keep everything in the R environment.</p>
<p>First, a FASTA file containing the sequences needs to be read in
using the <code>readDNAStringSet()</code> function.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">msa</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seq_path</span> <span class="op">&lt;-</span> <span class="st">"/Users/connorfrench/Dropbox/Old_Mac/School_Stuff/CUNY/Courses/Spring-2019/Machine-Learning/project/fasta/hypsiboas_seqs_aligned.fas"</span></span>
<span></span>
<span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">seq_path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqs</span></span></code></pre>
</div>
<p>To align the sequences using Clustal Omega, we use the
<code>msa</code> function and specify the <code>method</code> as
“ClustalOmega”. There are more arguments that you can use to fine-tune
the alignment, but these are not necessary for the vast majority of
scenarios.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">alignment</span></span></code></pre>
</div>
<p>If after you perform your alignment you find that an individual does
not belong or is negatively impacting the alignment, you have to
manipulate your unaligned sequence data and then re-align the sequences.
For instance, say you didn’t want to include the individual
<code>ind_name</code> in the alignment. First, remove the individual
from the <code>seqs</code> object.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove individual</span></span>
<span><span class="va">seqs_filt</span> <span class="op">&lt;-</span> <span class="va">seqs</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span> <span class="op">%in%</span> <span class="st">"Itar128"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">seqs_filt</span></span></code></pre>
</div>
<p>This nifty filtering approach takes advantage of the powerful
<code>%in%</code> function. In English, this function takes the vector
on the left side of the <code>%in%</code> and checks if each element
matches any of the elements on the right side. The left side contains
the elements in the data you want to check, while the right side
contains what you want to compare them against. So in this case,</p>
<p><code>names(seqs) %in% "Itar128"</code></p>
<p>checks to see if “Itar128” is present in any of the sequence names,
returns <code>TRUE</code> if so and <code>FALSE</code> if not. Since we
want to remove the “Itar128” sequence, we want the opposite of this
result. To do this, we wrap this expression with the bang
(<code>!</code>) operator, which negates the result of a boolean
expression. So, the following expression returns <code>TRUE</code> for
all sequence names that are not “Itar128” and <code>FALSE</code>
otherwise.</p>
<p><code>!(names(seqs) %in% "Itar128"</code></p>
<p>Using bracket indexing, you then return all sequences that do not
have the name “Itar128”.</p>
<p><code>seqs[!(names(seqs) %in% "Itar128"</code></p>
<p>Following the filtering, you need to realign the new set of
sequences.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs_filt</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">alignment</span></span></code></pre>
</div>
<p>There are quite a few packages in R to work with sequence data, for
example <a href="https://cran.r-project.org/web/packages/seqinr/index.html" class="external-link">seqinr</a>,
<a href="https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.html" class="external-link">phangorn</a>,
and <a href="https://cran.r-project.org/web/packages/ape/index.html" class="external-link">ape</a>.
<code>ape</code> is the most versatile of these. But before we can use
<code>ape</code>, the alignment must be converted to the
<code>ape::DNAbin</code> format.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ape_align</span> <span class="op">&lt;-</span> <span class="fu">msaConvert</span><span class="op">(</span><span class="va">alignment</span>, type <span class="op">=</span> <span class="st">"ape::DNAbin"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ape_align</span></span></code></pre>
</div>
<p>Although not necessary for this workshop, it’s generally advisable to
write your DNA alignment to a new file to use later or with other
programs.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ape</span><span class="fu">::</span><span class="fu">write.FASTA</span><span class="op">(</span><span class="va">ape_align</span>, <span class="st">"path/to/alignment.fas"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="ape">ape<a class="anchor" aria-label="anchor" href="#ape"></a>
</h3>
<div class="section level4">
<h4 id="check-alignment">Check alignment<a class="anchor" aria-label="anchor" href="#check-alignment"></a>
</h4>
<p><code>ape</code> has a useful utility that allows you perform a
series of diagnostics on your alignment to make sure nothing is fishy.
The function can output multiple plots, but let’s just visualize the
alignment plot. If the colors seem jumbled in a region, this indicates
that an alignment error may have occurred and to inspect your sequences
for any problems.</p>
<p>In addition, the function outputs a series of helpful statistics to
your console. Ideally, your alignment should have few gaps and a number
of segregating sites that is reasonable for the number of sites in your
alignment and the evolutionary divergence of the sequences included in
the alignment, e.g., for an alignment with 970 sites of a single bird
species, around 40 segregating sites is reasonable, but 200 segregating
sites may indicate that something went wrong with your alignment or a
species was misidentified. In addition, for diploid species a vast
majority, if not all of the sites should contain one or two observed
bases.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">checkAlignment</span><span class="op">(</span><span class="va">ape_align</span>, what <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="calculate-genetic-diversity">Calculate genetic diversity<a class="anchor" aria-label="anchor" href="#calculate-genetic-diversity"></a>
</h4>
<p>Now we’re going to calculate the average number of pairwise genetic
differences among these sequences. <code>ape</code> has the
<code>dist.dna()</code> function to calculate various genetic distance
metrics for an alignment.</p>
<p>You use the argument <code>model = "raw"</code> to specify that you
want to use the average number of pairwise differences as your metric.
When calculating the number of pairwise genetic differences, you have to
choose whether you want missing data to be deleted across the entire
alignment (<code>pairwise.deletion = FALSE</code>) or only between the
two sequences being compared (<code>pairwise.deletion = TRUE</code>).
Since we want to take advantage of as much data as possible when
computing 𝜋, we’ll set <code>pairwise.deletion = TRUE</code>.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendist</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">ape_align</span>, model <span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fas_gendist</span></span></code></pre>
</div>
<p>The function returns a distance matrix containing the pairwise
differences among all sequences in the data set. To take a quick peak at
their distribution, we can use the <code>hist()</code> function on the
distance matrix.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">fas_gendist</span><span class="op">)</span></span></code></pre>
</div>
<p>Since we want the average of these differences, we need to take their
average. However, you may have noticed the distribution of genetic
distances is non-normal, with mostly low genetic diversities. This is
the case with most genetic data sets. So, instead of taking the mean,
the <code>median()</code> will provide a better estimate of the center
of the distribution.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendiv</span> <span class="op">&lt;-</span> <span class="fu">median</span><span class="op">(</span><span class="va">fas_gendist</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fas_gendiv</span></span></code></pre>
</div>
<p>To calculate this for multiple populations…</p>
<p>read in fasta alignments…</p>
<p><code>lapply</code> explanation…</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_paths</span> <span class="op">&lt;-</span> <span class="fu">list.files</span><span class="op">(</span><span class="st">"..."</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fas_alignments</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">fas_paths</span>, <span class="va">read.dna</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can use <code>lapply</code> to calculate the pairwise genetic
distances for each species...</p>
<p><code>sapply</code> is similar to <code>lapply</code>, except it
returns a vector instead of a list. This is useful when you know the
function is returning the same data type, like in our case where a
number is being returned.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendist_pop</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">fas_alignments</span>, <span class="va">dist.dna</span>, model <span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">fas_gendiv_pop</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">fas_gendist_pop</span>, <span class="va">median</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">fas_gendiv_pop</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
<div class="section level3">
<h3 id="hill-numbers">Hill numbers<a class="anchor" aria-label="anchor" href="#hill-numbers"></a>
</h3>
<p><strong>How are we calculating genetic Hill numbers?</strong></p>
</div>
</div>
<div class="section level2">
<h2 id="asvs-and-metabarcoding">ASVs and Metabarcoding<a class="anchor" aria-label="anchor" href="#asvs-and-metabarcoding"></a>
</h2>
<p>Metabarcoding is rapidly growing DNA sequencing method used for
biodiversity assessment. It involves sequencing a short, variable region
of DNA sequence across a sample that may contain many taxa. This can be
used to rapidly assess the diversity of a particular sampling site with
low sampling effort. Some common applications include <a href="https://en.wikipedia.org/wiki/Metabarcoding" class="external-link">microbiome diversity
assessments, environmental DNA, and community DNA</a>. In many
applications where a data base of species identities for sequence
variants is not available or desired, the diversity of sequences is
inferred, rather than the diversity of species. Additionally, while <a href="https://onlinelibrary-wiley-com.ezproxy.gc.cuny.edu/doi/full/10.1111/1755-0998.12902" class="external-link">there
are caveats</a>, the relative abundance of sequences can be considered
analogous to the relative abundance of species across a community.</p>
<p>The basic unit of diversity used in most studies is the <a href="https://en.wikipedia.org/wiki/Amplicon_sequence_variant" class="external-link">amplicon
sequence variant (ASV)</a>. An ASV is any one of the sequences inferred
from a high-throughput sequencing experiment. These sequences are free
from errors and may diverge by as little as a single base pair from each
other. They are considered a substitute for operational taxonomic units
(OTUs) for evaluating ecological diversity given their higher
resolution, reproducibility, and no need for a reference database.</p>
<div class="section level3">
<h3 id="sequence-table">Sequence table<a class="anchor" aria-label="anchor" href="#sequence-table"></a>
</h3>
<p>The basic unit of analysis used for metabarcoding studies is the
sequence table. It is a matrix with named rows corresponding to the
samples and named columns corresponding to the ASVs. Each cell contains
the count of each variant in each sample. Here’s an example with 5
samples and 4 sequence variants:</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th>asv_1</th>
<th>asv_2</th>
<th>asv_3</th>
<th>asv_4</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>sample_1</td>
<td>0</td>
<td>276</td>
<td>200</td>
<td>10</td>
</tr>
<tr class="even">
<td>sample_2</td>
<td>10</td>
<td>1009</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td>sample_3</td>
<td>0</td>
<td>0</td>
<td>234</td>
<td>15</td>
</tr>
<tr class="even">
<td>sample_4</td>
<td>129</td>
<td>1992</td>
<td>0</td>
<td>485</td>
</tr>
<tr class="odd">
<td>sample_5</td>
<td>1001</td>
<td>10</td>
<td>1847</td>
<td>178</td>
</tr>
</tbody>
</table>
<p>sample_1 may correspond to a soil sample collected in a forest, while
the variants represent different ASVs, which could look like this:</p>
<p>asv_1: GTTCA</p>
<p>asv_2: GTACA</p>
<p>asv_3: GATCA</p>
<p>asv_4: CTTCA</p>
<p>This looks similiar to a site-species matrix, where the rows are a
sampling locality, the columns are the species, and the cell values are
the abundance of a species at a particular site.</p>
<p>While for this workshop we will focus on the analysis of a sequence
table, we’ll briefly go over how the data makes its way from raw
sequencing data to the sequence table.</p>
</div>
<div class="section level3">
<h3 id="fastq">FASTQ<a class="anchor" aria-label="anchor" href="#fastq"></a>
</h3>
<p>The FASTQ file format is similar to the FASTA format introduced
earlier, but includes additional data about the contained sequences.
This enables more flexible data manipulation and filtering capabilities,
and is popular with high-throughput sequencing technologies. When
working with ASVs, this is the most common file format you will
encounter.</p>
<p>A FASTA formatted sequence may look like this,</p>
<pre><code>&gt;seq1
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT</code></pre>
<p>while a FASTQ version of the same sequence may look like this:</p>
<pre><code>@seq1
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT
+
!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65</code></pre>
<p>If your eyes glaze over at the seemingly random numbers below the
sequence, don’t fret! They mean something, but usually your machine does
the interpreting for you. A FASTQ file is formatted as follows:</p>
<ol style="list-style-type: decimal">
<li><p>An ‘@’ symbol followed by a sequence identifier and an optional
description</p></li>
<li><p>The raw sequence letters</p></li>
<li><p>A ‘+’ character that is optionally followed by the sequence
identifier and more description</p></li>
<li><p>Quality encodings for the raw sequence letters in field 2. There
must be the same number of encoding symbols as there are sequence
letters.</p></li>
</ol>
<p>The symbols are bytes representing quality, ranging from 0x21 (lowest
quality; ‘!’ in ASCII) to 0x7e (highest quality; ‘~’ in ASCII). Here,
they’re arranged in order of quality from lowest on the left to highest
on the right:</p>
<pre><code>!"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~</code></pre>
<p>Fortunately, whatever bioinformatic program you use will generally do
the interpreting for you.</p>
<p>Different sequencing technologies (e.g. Illumina) have standardized
description strings which contain more information about the sequences
and can be useful when filtering sequences or performing QC. For
examples, see <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">the
Wikipedia page</a>.</p>
</div>
<div class="section level3">
<h3 id="dada2">DADA2<a class="anchor" aria-label="anchor" href="#dada2"></a>
</h3>
<p><a href="https://benjjneb.github.io/dada2/index.html" class="external-link">DADA2</a> is
one of the most powerful and commonly used software packages for
processing raw ASV sequencing data to turn it into a table of ASVs for
each sample. This workshop assumes you are using the output from this
workflow for further analysis, but for clarity the steps to the workflow
include:</p>
<ol style="list-style-type: decimal">
<li>read quality assessment</li>
<li>filtering and trimming FASTQ files</li>
<li>learning sequencing error rates, so the algorithm can correct for
them</li>
<li>infer the sequence variants</li>
<li>merging paired reads (if sequencing was done with forward and
reverse read pairs)</li>
<li>constructing a sequence table</li>
</ol>
<p>There are more available to further assess the quality of the data
and map the data to taxonomic databases, but the core output of DADA2
and similar software pipelines is the sequence table.</p>
</div>
<div class="section level3">
<h3 id="hill-numbers-1">Hill numbers<a class="anchor" aria-label="anchor" href="#hill-numbers-1"></a>
</h3>
<p>I assume we’re calculating Hill numbers using hillR::hill_taxa, but
is that correct?</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="extra">Extra<a class="anchor" aria-label="anchor" href="#extra"></a>
</h1>
<div class="section level2">
<h2 id="snps">SNPs<a class="anchor" aria-label="anchor" href="#snps"></a>
</h2>
<div class="section level3">
<h3 id="vcf">VCF<a class="anchor" aria-label="anchor" href="#vcf"></a>
</h3>
<p>Here are a few things about a VCF file… tab delimited… fields…
size…</p>
<div class="section level4">
<h4 id="vcfr">vcfR<a class="anchor" aria-label="anchor" href="#vcfr"></a>
</h4>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">vcfR</span><span class="op">)</span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">read.vcfR</span><span class="op">(</span><span class="va">vcf_path</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="manipulate-data">Manipulate data<a class="anchor" aria-label="anchor" href="#manipulate-data"></a>
</h4>
<p>Filter individuals Filter based on quality, missingness</p>
</div>
<div class="section level4">
<h4 id="export-to-genind">Export to genind<a class="anchor" aria-label="anchor" href="#export-to-genind"></a>
</h4>
<p>A bit about genind and <code>adegenet</code>… If you have many SNPs,
genlight may be a more appropriate format…</p>
<p><code>vcfR2genind</code> also takes arguments for the function
<code>df2genind</code> from the <code>adegenet</code> package. There are
a couple things we can do to speed up the conversion, like specifying
the ploidy of the file and setting the <code>check.ploidy</code> flag to
<code>FALSE</code>.</p>
<p>To add populations…</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">adegenet</span><span class="op">)</span></span>
<span><span class="va">snps_genind</span> <span class="op">&lt;-</span> <span class="fu">vcfR2genind</span><span class="op">(</span></span>
<span>    <span class="va">snps</span>, </span>
<span>    n.cores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    ploidy <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    check.ploidy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    pop <span class="op">=</span> <span class="va">pops</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">snps_genind</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="calculate-genetic-diversity-1">Calculate genetic diversity<a class="anchor" aria-label="anchor" href="#calculate-genetic-diversity-1"></a>
</h4>
<p>If you didn’t specify populations when converting or reading in the
genind file, do so now…</p>
<p>Individual-based genetic distances… Maybe mention PCA?</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_gendist_ind</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="va">snps_genind</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span></code></pre>
</div>
<p>Nei’s genetic distance…</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_genpop</span> <span class="op">&lt;-</span> <span class="fu">genind2genpop</span><span class="op">(</span><span class="va">snps_genind</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snps_gendist_pop</span> <span class="op">&lt;-</span> <span class="fu">dist.genpop</span><span class="op">(</span><span class="va">snps_genpop</span>, method <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<p>FST…</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hierfstat</span><span class="op">)</span></span>
<span><span class="va">snps_hierfstat</span> <span class="op">&lt;-</span> <span class="fu">genind2hierfstat</span><span class="op">(</span><span class="va">snps_genind</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snps_fst</span> <span class="op">&lt;-</span> <span class="fu">pairwise.fst</span><span class="op">(</span><span class="va">snps_hierfstat</span>, diploid <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">snps_fst</span><span class="op">[</span><span class="fu">upper.tri</span><span class="op">(</span><span class="va">snps_fst</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="summarize-genetic-diversity">Summarize genetic diversity<a class="anchor" aria-label="anchor" href="#summarize-genetic-diversity"></a>
</h1>
<div class="section level2">
<h2 id="calculate-hill-numbers-within-speciespopulations">Calculate Hill numbers within species/populations<a class="anchor" aria-label="anchor" href="#calculate-hill-numbers-within-speciespopulations"></a>
</h2>
<p><em>* Going to need a SNP matrix</em></p>
<p>Hill numbers of intraspecific allele frequencies are the
<em>effective number of SNPs</em> in the population/species.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># columns are SNPs, rows are individuals</span></span>
<span><span class="co"># treat it like abundances </span></span>
<span><span class="va">intra_hill_1</span> <span class="op">&lt;-</span> <span class="fu">hillR</span><span class="fu">::</span><span class="fu">hill_taxa</span><span class="op">(</span><span class="va">snp_matrix</span>, q <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="interpret-refer-back-to-visualizations">Interpret (refer back to visualizations)<a class="anchor" aria-label="anchor" href="#interpret-refer-back-to-visualizations"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="calculate-hill-numbers-across-species">Calculate Hill numbers across species<a class="anchor" aria-label="anchor" href="#calculate-hill-numbers-across-species"></a>
</h2>
<p><em>Need species alignments from ape</em></p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate pairwise distance matrices</span></span>
<span><span class="va">per_species_dist</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">species_alignments</span>, <span class="fu">ape</span><span class="fu">::</span><span class="va">dist.dna</span>, model <span class="op">=</span> <span class="st">"raw"</span>, as.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mat_avg</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="fu">upper.tri</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span>    <span class="kw">return</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">per_species_pi</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">per_species_dist</span>, <span class="va">mat_avg</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">per_species_pi</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="interpret-refer-back-to-visualizations-1">Interpret (refer back to visualizations)<a class="anchor" aria-label="anchor" href="#interpret-refer-back-to-visualizations-1"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="within-species">Within-species<a class="anchor" aria-label="anchor" href="#within-species"></a>
</h2>
<div class="section level3">
<h3 id="sfs-visualization">SFS visualization<a class="anchor" aria-label="anchor" href="#sfs-visualization"></a>
</h3>
<p>The site frequency spectrum (SFS) is a summary of genetic variants
across a population, similar to how abundances are summarized across a
community. Specifically, it is the distribution of the allele
frequencies of a given set of loci (generally single nucleotide
polymorphisms, SNPs), in a population or sample. The best way to gain an
intuition for what the SFS is, is to visualize it.</p>
<p>The first bar of this plot indicates derived alleles (i.e. not
ancestral) that appear once across the population. Approximately 35% of
the sampled SNPs in the genome alignment are these “singletons”. The
second bar indicates derived alleles that appear twice across the
population, where ~18% of the sampled SNPs are these “doubletons” and so
forth.</p>
<p>The shape of the SFS can tell us important information about the
demographic history of a population and their history of selection. The
above plot corresponds with a population that has experienced
demographic stability. A couple other examples of the impact of
demographic history on the SFS include:</p>
<p>There are a variety of summary statistics that describe the shape of
the SFS. The average number of pairwise nucleotide differences, a.k.a.
Tajima’s/Nei’s <span class="math inline">\(\pi\)</span>, where a high
value indicates a higher level of genetic diversity in the population,
describes the average of the distribution. However, as you’ve seen
above, the shape of the distribution contains important information
about the population’s history!</p>
</div>
<div class="section level3">
<h3 id="hill-numbers-of-the-sfs">Hill numbers of the SFS<a class="anchor" aria-label="anchor" href="#hill-numbers-of-the-sfs"></a>
</h3>
<p>In the context of summarizing species abundances, Hill numbers are
the effective number of species in a community, where each exponent
(q-value) increasingly weights rare species more heavily. The
interpretation is similar when summarizing within-population genetic
diversity. Hill numbers of within-population allele frequencies convey
the <em>effective number of alleles</em> in a population, where each
q-value increasingly weights rarer alleles more heavily.</p>
<p><em>Something about when we’d use</em> <em>pi vs Hill</em></p>
</div>
</div>
<div class="section level2">
<h2 id="section"></h2>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</div></section><section id="aio-phylo-data"><p>Content from <a href="phylo-data.html">Summarizing phylogenetic data</a></p>
<hr>
<p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/phylo-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What information is contained in phylogenies?</li>
<li>How to summarize and interpret phylogenies using Hill numbers?</li>
<li>What are the common phylogeny data formats?</li>
<li>How do you import and manipulate phylogeny data in R?</li>
<li>How do you visualize phylogenies in R?</li>
<li>How do you calculate Hill numbers with phylogenetic data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants will be able to:</p>
<ol style="list-style-type: decimal">
<li>Identify key features of the Newick format</li>
<li>Import and manipulate phylogenetic data in the R environment</li>
<li>Visualize phylogenies</li>
<li>Calculate and interpret phylogenetic Hill numbers</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<section id="introduction-to-phylogenetic-data"><h2 class="section-heading">Introduction to phylogenetic data<a class="anchor" aria-label="anchor" href="#introduction-to-phylogenetic-data"></a>
</h2>
<hr class="half-width">
<p>In this episode, we will explore how to extract information from
phylogenetic trees in order to complement our hypotheses and inferences
about processes shaping biodiversity patterns. Phylogenetic trees have
information about how the species in a taxonomic group are related to
each other and how much relative evolutionary change has accumulated
among them. Since local communities differ in their phylogenetic
composition, this information can give insights on why communities are
how they are.</p>
<p>[HERE INCLUDE VISUALS OF PHYLOGENIES]</p>
<p>The relative phylogenetic distance among species in a community as
well as the distribution of the amount of evolutionary history
(represented by the length of the branches in a phylogeny) are a result
of different factors such as the age since the initial formation of the
community and the rate of macroevolutionary processes such as speciation
and extinction. For instance, young communities that are dominated by
closely related species and show very short branch lengths may suggest a
short history with few colonization events and high rates of local
speciation; alternatively, if the same young communities harbor
distantly related species with longer branch lengths, it may suggest
that most of the local diversity was generated by speciation elsewhere
followed by colonization events involving distantly related species.
Coupled with information on ecological traits and rates of
macroevolutionary processes, these patterns also allow to test for
hypotheses regarding, for instance, ecological filtering or niche
conservatism.</p>
<p>Summarizing this phylogenetic information (i.e., phylogenetic
distance and distribution of branch lengths) is therefore important for
inference. As we have seen in previous episodes, the use of <a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Hill
numbers</a> is an informative approach to summarize biodiversity.</p>
<p>[MORE HERE ON MEANING OF PHYLOGENETIC HILL, RELATING TO PHYLO
CHARACTERISTICS]</p>
</section><section id="working-with-phylogenetic-data-in-r"><h2 class="section-heading">Working with phylogenetic data in R<a class="anchor" aria-label="anchor" href="#working-with-phylogenetic-data-in-r"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="importing-phylogenetic-data">Importing phylogenetic data<a class="anchor" aria-label="anchor" href="#importing-phylogenetic-data"></a>
</h3>
<p>Several file formats exist to store phylogenetic information. The
most common formats are the <a href="https://evomics.org/resources/tree-formats/" class="external-link">Newick and Nexus
formats</a>. Both these formats are plain text files storing different
levels of information about the taxa relationship and evolutionary
history. <code>Newick</code> files are the standard for representing
trees in a computer-readable form, as they can be extremely simple and
therefore do not take up much memory. <code>Nexus</code> file, on the
other hand, are composed by different blocks regarding different types
of information and can be used to store DNA alignments, phylogenetic
trees, pre-defined groupings of taxa, or everything at once. Since they
are a step ahead in complexity, we will stick with <code>Newick</code>
files for now.</p>
<p><code>Newick</code> files store the information about the clades in a
tree by representing each clade within a set of parentheses. Sister
clades are separated by <code>,</code>. The notation also requires us to
add the symbol <code>;</code> to represent the end of the information
for that phylogenetic tree.</p>
<p>The basic structure of a tree in a <code>Newick</code> format is
therefore as follows:</p>
<p><code>((A,B),C);</code></p>
<p>The notation above indicates that: 1. we have three taxa in our tree,
named <code>A</code>, <code>B</code> and <code>C</code>; 2.
<code>A</code> and <code>B</code> form one clade (A,B); 3. the (A,B)
clade is sister to the <code>C</code> clade (we represent that by adding
another set of parentheses and a <code>,</code> separating (A,B) from
C).</p>
<p>In addition, <code>Newick</code> files can also store information on
the branch leading to it tip and node. We do that by adding
<code>:</code> after each tip/node.</p>
<p><code>((A:0.5,B:0.5):0.5,C:1);</code></p>
<p>The notation above indicates that: 1. the branches containing A and B
have each a length = 0.5; 2. the branch that leads to the node
connecting A and B also has length = 0.5; 3. the branch leading to C has
length = 1.</p>
<p>The notation above is what we import into R to start working with and
manipulating our phylogenetic tree. For that goal, we will use the <a href="https://academic.oup.com/bioinformatics/article/20/2/289/204981" class="external-link">ape</a>
package:</p>
<p>To import our tree, we will be using the function
<code>read.tree()</code> from the <code>ape</code> package. In the case
of simple trees as the one above, we could directly create them within R
by giving that notation as a <code>character</code> value to this
function, using the <code>text</code> argument, as shown below:</p>
<p>Now, we can visually inspect our tree using the <code>plot()</code>
function:</p>
<p>Can you visualize the text notation in that image? We can see the
same information: A is closer related to B than C, and the branches
leading to A and B have half the length of the branch leading to C.</p>
<p>The <code>read.tree()</code> function creates an object of class
<code>phylo</code>. We can further investigate this object by calling it
in our console:</p>
<p>The printed information shows us that we have a phylogenetic tree
with 3 tips and 2 internal nodes, where the tip labels are “A, B, C”. We
also are informed that this tree is rooted and has branch lengths.</p>
<p>One way to access the components of this object and better explore it
is to use <code>$</code> after the object name. Here, it will be
important for us to know a little bit more about where the information
about tip labels and branch lengths are stored in that
<code>phylo</code> object. Easy enough, we can access that by calling
<code>tip.labels</code> and <code>edge.length</code> after
<code>$</code>.</p>
</div>
</section><section id="cleaning-and-filtering-phylogenetic-data"><h2 class="section-heading">Cleaning and filtering phylogenetic data<a class="anchor" aria-label="anchor" href="#cleaning-and-filtering-phylogenetic-data"></a>
</h2>
<hr class="half-width">
<p>Now that we learned how to import and visualize trees in R, let’s
bring the phylogeny for the communities we are working with in this
workshop. Our data so far consists of <a href="abundance-data.html">abundances</a> and <a href="traits-data.html">traits</a> of several taxa of arthropods
collected across three islands in the Hawaiian archipelago. Let’s work
though importing phylogenetic information for these species.</p>
<p>Two common approaches to retrieving a phylogeny for a focal group are
1) relying on a published phylogeny for the group, or 2) surveying
public phylogenetic databases based on your taxa list. A common option
for the latter is the <a href="https://tree.opentreeoflife.org/about/taxonomy-version/ott3.2" class="external-link">Open
Tree of Life Taxonomy</a>, a public database that synthesizes taxonomic
information from different sources. You can even interact with this
database using the R package <code>rotl</code>. A few tutorials to do so
exist online, like <a href="https://mctavishlab.github.io/R_OpenTree_tutorials/" class="external-link">this one</a>.
Using a public database is a good approach when working with taxonomic
groups that are not heavily investigated regarding their phylogenetic
relationships (the well-known <a href="https://doi.org/10.1146/annurev-ecolsys-112414-054400" class="external-link">Darwinian
shortfall</a>). In such cases, databases like OTL will give you a
summary phylogeny already filtered for the taxa you have in hand and
cross-checked for synonyms and misspellings.</p>
<p>For this workshop, since we are using simulated data, we will work
with the first option: a “published” arthropod phylogeny. Let’s load
this phylogeny into R using the function <code>read.tree</code> we
learned earlier.</p>
<p>This new <code>phylo</code> object is way larger than the previous
one, being a “real” phylogeny and all. You can inspect it again by
directly calling the object <code>arthro_tree</code>. To plot it, we
will use the <code>type</code> argument to modify how our tree will be
displayed. Here, we used the option <code>'fan'</code>, to display a
circular phylogeny (slightly better to show such a large phylogeny in
the screen). We also set the <code>show.tip.label</code> argument to
<code>False</code>.</p>
<p>How do we combine all this information with the community datasets we
have so far for our three islands? First, we will have to perform some
name checking and filtering.</p>
<div class="section level3">
<h3 id="cleaning-and-checking-phylogeny-taxa">Cleaning and checking phylogeny taxa<a class="anchor" aria-label="anchor" href="#cleaning-and-checking-phylogeny-taxa"></a>
</h3>
<p>The issue of underscore:</p>
<p><strong>Do we still want to keep this issue? Or simply skip ahead to
pruning the phylogeny?</strong></p>
<p>Since this is a “published” arthropod phylogeny, we will likely not
have any misspelling in the tip names of the object. However, one
important issue that may arise is that, depending on the resolution of
the phylogeny, it may be the case that some taxa in our community
dataset may not be explicitly represented in the phylogeny. For
instance, the tips in our phylogeny might be represented by one species
for each genus, or one genus for each family. This is usually the case
when the goal of the original publication was to elucidate the
phylogenetic relationship for such higher taxonomic levels, with less
focus on the more recent evolutionary history. In these cases, some taxa
in our <a href="abundance-data.html">abundances</a> dataset may not be
found in the phylogeny, which will yield errors when we try to summarize
phylogenetic diversity. Therefore, we will want to assign those “missing
taxa” to one of the tips present in the tree, based on their known genus
and/or family assignments.</p>
<p>Let’s first find out what are those “missing taxa”. For this, we’ll
refer back to our original species list, from the
<code>abundances</code> object:</p>
<p>To cross-check this list against the list of names in our phylogeny,
we can use the Boolean operator <code>%in%</code> coupled with
<code>!</code>. This will allow us to check for names in
<code>all_names</code> that are not included in the
<code>arthro_tree$tip.label</code>. In summary, the expression
<code>A %in% B</code> would return the position of the elements in
vector A that are present in vector B. The <code>!</code> (NOT) operator
returns the opposite of that expression, in a way that
<code>!(A %in% B)</code> will return the position of elements in vector
A that are <em>NOT</em> present in vector B (precisely what we need
right now).</p>
<p>The object <code>not_found</code> is now a vector of the positions
(or <em>indexes</em>) of the elements in <code>all_names</code> that are
not found in <code>arthro_tree$tip.label</code>. Let’s then check these
names by using these indexes to subset <code>all_names</code>:</p>
<p><strong>EDIT THIS NEXT PARAGRAPH WHEN PHYLOGENY IS
AVAILABLE</strong></p>
<p>Here, we can decide how we are going to assign these species to tips
in our tree. For instance, we see that we do not have
<strong>XXX</strong> in our tree, but we have <strong>XXX</strong>,
which belongs to the same genus. Similarly…</p>
<p>We can then give new names to these species, based on what tips in
our tree we want them to match to. To do so, we will first duplicate our
<code>all_names</code> vector into the <code>new_names</code> vector.
This way, we can make our changes in <code>new_names</code> while
keeping the original names intact, just in case we need to refer back to
them in the future.</p>
<p>Now we use the indexes saved in <code>not_found</code> to replace the
names in the positions with the ones we want:</p>
<ul>
<li>Here I am changing the abundance to match the phylogeny. IT SHOULD
BE THE OTHER WAY AROUND.*</li>
<li>Make change tip labels a challenge. *</li>
</ul>
<p>Now that we have a vector where all names match to a position in our
phylogeny, we need to bring these corrected names back to our abundance
data. Here, we will again first copy our <code>abundances</code> object
into another object, in order to keep the original data intact. Then, we
will change the column <code>final_name</code> based on the vector
<code>new_names</code>.</p>
</div>
<div class="section level3">
<h3 id="pruning-our-phylogeny">Pruning our phylogeny<a class="anchor" aria-label="anchor" href="#pruning-our-phylogeny"></a>
</h3>
<p>Now that we have a modified abundance dataset where all taxa names
are present in our phylogeny, we can work on pruning our phylogeny to
the taxa present in our communities. This is especially important when
working with published phylogenies as these are usually large files
containing several taxa that may not exist in the local community.</p>
<p>We can prune our phylogeny using the function <code>keep.tip</code>
from the <code>ape</code> package. For this function, we provide the
entire phylogeny plus the names of the tips we want to keep. Here, we
will retrieve those names from the new abundances dataset we create,
with names corrected to match the phylogeny.</p>
<p>Now, that we have a pruned more manageable phylogeny for our
communities, let’s move on to the next bit of information we need to
summarize phylogenetic diversity: a site-by-species matrix. As we saw in
the <a href="abundance-data.html">abundances</a> episode, such a matrix
can be used to calculate Hill numbers and help us compare patterns
across the different communities. In this episode, we are following the
same approach to calculate Hill numbers, with the addition that, on top
of the abundance, we will also be using phylogenetic information,
i.e. the length of the branches in the phylogenetic tree leading to the
taxa present in each community.</p>
<p>Here, we need to make a site-by-species matrix using the new taxa
names corrected to match the phylogeny. Since we have done this in a
previous episode, we’ll cover this in a challenge.</p>
<ul>
<li>REMOVE THIS CHALLENGE *</li>
</ul>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>Make a site-by-species matrix containing the three islands based on
the new <code>abundance_phylo</code> object that contains the taxa names
corrected to match the phylogeny. Hint: you may want to refer to the <a href="abundance-data.html">abundances</a> episode.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>To create our site-by-species matrix, we will use the
<code>pivot_wider</code> function from the <code>tidyr</code>
package.</p>
</div>
</div>
</div>
</div>
</div>
</section><section id="summarizing-with-hill-numbers"><h2 class="section-heading">Summarizing with hill numbers<a class="anchor" aria-label="anchor" href="#summarizing-with-hill-numbers"></a>
</h2>
<hr class="half-width">
<p>In this section, we will extract some summary statistics about the
pattern of phylogenetic diversity (PD) in our communities. As we
discussed above, the relative phylogenetic distance among species and
the distribution of this distance can give insights into processes of
community assembly. Here, we will make use of <a href="https://onlinelibrary.wiley.com/doi/10.1111/eva.12593" class="external-link">Hill
numbers</a> to extract summaries of phylogenetic distances. Phylogenetic
hill numbers incorporate information on both the phylogenetic structure
of a system and the abundances of different species. In order to get an
intuition for how these different components influence phylogenetic hill
numbers, we’ll first explore the behavior of this summary statistic
using a few simple examples</p>
<div class="section level3">
<h3 id="example-1--trees-with-different-branch-lenghts">Example 1. Trees with different branch lenghts<a class="anchor" aria-label="anchor" href="#example-1--trees-with-different-branch-lenghts"></a>
</h3>
<p>For our examples, we’ll assume we have one single community with
eight taxa: A through H. Let’s create two site-by-species matrix for
this community: one denoting even abundance across species, and another
one with uneven abundance. For the even communitu, we create a vector of
the value <code>1</code> repeated 8 times; for the uneven community,
we’ll create a vector where abundance goes up from species A to species
H (for simplicity, we’ll just use values from 1 to 8). We then transform
them to dataframe, and name the columns with the names of the
species.</p>
<p>Now let’s create two different possible trees for these communities:
one with short branch lengths and another with longer branch lengths.
Remember: branch length = amount of evolutionary change</p>
<p>If we plot both trees…</p>
<p>…we can see that the branches leading to extant taxa are longer for
<code>long_tree</code>, as we intended. This suggests that a greater
amount of evolutionary change is happening in these recent branches of
the longer tree when compared to the shorter tree.</p>
<p>Now, we will calculate phylogenetic hill numbers for both trees using
both even and uneven communities. To store the calculated values, we’ll
a data.frame called <code>even_comm_short_tree</code>. The first column
will be our Hill numbers to be calculated; the second column will be the
Hill number order from 0 to 3, to see how the order affects the values;
the third and fourth column will be the description of our
components</p>
<p>Now we will use a <code>for</code> loop to calculate phylogenetic
Hill numbers using the <code>hill_phylo</code> function from the
<code>hillR</code> package. This function takes in a a site-by-species
matrix and phylogeny, and returns phylogenetic Hill numbers for each
site based on which species are present there. We provide to the
function the site-by-species matrix, the phylogenetic tree and the
order.</p>
<p>Let’s repeat this process for the longer tree:</p>
<p>We can combine both dataframes and plot the values for
comparison:</p>
<p>This figure clearly shows that the tree with longer branches (dark
red line) harbors higher evolutionary history, and therefore higher PD,
as calculated by Hill numbers. It also shows that the Hill number value
decreases as the order goes up, since higher orders focus on branch
lengths that are more common.</p>
<p>What would happen if the abundance of species in our community was
uneven (a more realistic scenario)? In this case, both branch lengths
and how abundant each branch is will have an effect on the calculated
value. To visualize, let’s repeat the calculations above for the uneven
community.</p>
<p>Let’s plot all results together, using red colors for even
communities and blue colors for uneven community. Ligher colors will
represent short trees whereas darker colors will represent long
trees.</p>
<p>From this picture, we can take a few insights: 1. longer branches
still yield higher Hill numbers, regardless of the evenness in the
community abundance; 2. the Hill number for q = 0 remains the same,
regardless of the evenness. 3. As already observed, the value of the
hill numbers drop as the order goes up. In the even community, since
species have the same abundance, this decrease reflects higher orders
focusing on more common values of branch lengths. In the uneven
community, where species have different abundances, this decrease
reflects higher orders focusing less and less on branches leading to
rare taxa. 4. For q = 1 to 3, Hill numbers are always lower for uneven
communities. This suggests that although branch lengths are the same
(e.g., both dark red and dark blue lines represent the long tree), these
branches are unevenly represented in the community due to uneven
abundance of species. This unevenness is represented by the lower values
of Hill number on higher orders. In other words, some branches in the
tree “dominate” the community more than others. We can summarize this by
saying that the higher the unevenness the lower the value of the Hill
number will be.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>
<code>hillR</code> calculates phylogenetic hill numbers given a
phylogeny and a site by species matrix.</li>
<li>In trees with a similar topology, phylogenetic Hill number of order
0 reflect the sum of branch lengths. Orders of 1 and higher reflect the
sum of branch lengths weighted by the relative abundance of different
species.</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="example-2--balanced-vs-unbalanced-trees">Example 2. Balanced vs unbalanced trees<a class="anchor" aria-label="anchor" href="#example-2--balanced-vs-unbalanced-trees"></a>
</h3>
<p>So far, we have learned that Hill numbers are affected by both the
sum of branch lengths and the relative representation of each branch (in
terms of species abundance) in the community. For example 1, we have
used a perfectly balanced tree, i.e., all extant taxa have equal branch
length. In example 2, we will explore the effects of the uneven
distribution of cladogenesis event along the tree, leading to different
phylogeny structures.</p>
<p>Let’s create a totally balanced tree…</p>
<p>… and a totally unbalanced tree.</p>
<p>Let’s plot both trees for comparison:</p>
<p>Notice that here the difference between the trees resides in the fate
of each new lineage at a node. In the unbalanced uneven tree, at each
diversification event one of the lineages always persists till the
present with no change while the other undergoes another round of
diversification. In the even tree, both lineages from each node undergo
a new split. The consequence is that in the even tree, all extant
species result from recent diversification (i.e., they have a short
evolutionary history before coalescing into their ancestor), whereas in
the unbalanced tree we have a mix of old and recent lineages. This means
that the phylogenetic history itself is creating an uneven
representation of branch lengths across the community (even before we
account for species abundance)</p>
<p>To see how such phylogenetic structure influences hill numbers, let’s
repeat the calculations from example 1 with these new trees. First,
let’s focus on the even community (i.e., not introducing the relative
species abundance factor yet):</p>
<p>This plot is similar to example 1 in two ways: 1) one of the trees
has higher hill number values, in this case the uneven tree. This
suggests that the unbalanced structure of the tree accounts for a deeper
evolutionary history (i.e., lineages have overall longer branches); 2)
the value of the hill numbers drop as the order goes up. This happens
because higher orders are weighting less and less those branch lengths
that are not so common (like, for instance, the short branch lengths in
the uneven tree).</p>
<p>Something new that we observe here is a more evident difference
between trees in the <strong>rate of change</strong> of the Hill number
value as we increase the order. The uneven tree (darker red) shows a
steeper drop than the even tree (lighter red). This is similar to what
we observed in the <a href="abundance-data.html">abundances</a> episode:
uneven communities show a steeper drop in Hill numbers as order
increases. In this example, instead of species abundance, the unevenness
of the community is represented by the structure of the phylogeny: a
completely unbalanced tree combining long and short branches generates
an unevenness in the distribution of branch lengths. Notice that this
difference in rate of change barely observed in example 1: when plotting
two trees with the same structure (short vs long trees, but both fully
balanced), the lines are very similar In example 1, the difference in
the rate of change only becomes evident when we incorporate unevenness
from the species abundance, but here in example 2 we see this difference
deriving already from the different structures of the trees.</p>
<p>If the tree structure is already introducing some unevenness in our
community, how would the pattern differ when species relative abundance
is included?</p>
<p>In this case, both the structure of the tree and the species
abundance interact to generate a pattern of unevenness. Specifically,
the overall value of PD will be the sum of branch lengths weighted by
their abundance in the community as informed by the species abundance
distribution (instead of by the frequency of the branch length in the
tree). This added information brings new insights. It may be the case,
for instance, that in the uneven tree, even though long branches are
over represented, maybe the species with long branches are actually
super rare in our community, and the short-branch species are actually
super abundant. As you can probably infer, this difference suggests
something about the evolutionary history of our community, in this case
that the most abundant species have a very recent evolutionary history.
Similarly, even though branch lengths are the same in the even tree, if
some of the species is more abundant than others, it suggests that the
evolutionary history of the tree is unevenly represented in our
community.</p>
<p>To visualize these interaction between tree structure and species
relative abundance, let’s redo the calculation for Hill numbers with
even and uneven trees, this time using the uneven community. We’ll plot
all final values together for comparison.</p>
<p>Reminders: 1) light colors represent even trees, whereas darker
colors represent uneven trees 2) red lines represent even community
whereas blue lines represent uneven community.</p>
<p>From this second plot, we notice that:</p>
<ol style="list-style-type: decimal">
<li><p>Hill number values for the uneven tree (darker colors) are still
higher than those for the even tree (lighter colors). This makes sense
since the relative abundance is the same for both trees, so the overall
pattern (i.e., longer branches in uneven tree) remains;</p></li>
<li><p>the Hill number of order 0 is the same regardless of whether you
have different relative abundances among species. This was also the case
for example 1. Order 0 is simply the sum of all branch lengths (i.e., it
does not account for the relative abundance of branch lengths or
taxa);</p></li>
<li><p>the inclusion of an uneven community accentuates the rate of
decrease in the Hill number value as we increase the order. This is
because higher orders are now affected by the unevenness of the
distribution of branch length as well by the unevenness of species
abundance. This is a good example of how calculating different order of
hill numbers while incorporating species abundance allows to have a
couple summary statistics accounting for evolutionary history and
species abundance distribution that is comparable across different
regions.</p></li>
</ol>
<div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints2"></a>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>The phylogenetic structure of the community (represented by the
topology of the tree) influences the evenness of the branch length
distribution and contribute to different rates in the decrease of Hill
number values as the order number increases.</li>
<li>The inclusion of community structure (relative abundance of
different species) further accentuates this difference in rate of
change.</li>
<li>Phylogenetic Hill numbers represent the sum of branch lenghts.
Higher orders weight that sum by the distribution of branch lengths and
the species relative abundance.</li>
</ol>
</div>
</div>
</div>
<p>Now that we acquired an intuition on the information phylogenetic
Hill numbers can give us, let’s move on to our actual data.</p>
</div>
<div class="section level3">
<h3 id="moving-on-to-our-island-communities">Moving on to our island communities<a class="anchor" aria-label="anchor" href="#moving-on-to-our-island-communities"></a>
</h3>
<p>For this final section, we will be working on the data for the three
island communities to calculate hill numbers from orders of 0 to 3. We
will be using our pruned phylogeny and the site-by-species matrix we
created earlier. Since we covered the workflow of calculating
phylogenetic Hill numbers in the previous section, we will leave this
activity for a challenge. workflow covered in this episode,</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>For this challenge, you should calculate the phylogenetic Hill
numbers of orders 0 to 3 for the three islands we are working with
throughout the workshop The key objects here will be the phylogeny
stored in <code>arthro_tree_pruned</code> and the sites-by-species
matrix stored in <code>phylo_wide</code>. You should also plot the hill
numbers for each island and discuss what the calculated values allow you
to infer regarding the history of each community.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>We can directly calculate Hill numbers using the
<code>hill_phylo</code> function along with the objects
<code>arthro_tree_pruned</code> and <code>phylo_wide</code>. Since we
have three sites, the function will return a vector with tree values,
one Hill number for each site of the order we requested. Let’s create an
empty list to store a vector for each order, and use a <code>for</code>
loop to calculate from orders of 0 to 3.</p>
<p>Now, let’s create a data.frame with islands as rows and Hill numbers
of different orders as columns. For that, we will use the function
<code>do.call</code> to collapse our list <code>hill_values</code> using
the function <code>cbind</code>.</p>
<p>Finally, we can plot using a similar code to the ones we used in our
examples:</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor1" aria-labelledby="headingInstructor1">
<div class="accordion-body">
<p>A few points of discussion…</p>
</div>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></section><section id="aio-connecting-data-types"><p>Content from <a href="connecting-data-types.html">Connecting different data types</a></p>
<hr>
<p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/connecting-data-types.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What questions and analyses can we explore by linking multiple data
types together?</li>
<li>How do we determine if our data can be linked?</li>
<li>How do we actually link different data types together?</li>
<li>How do we analyze, visualize, and interpret multiple data types
concurrently?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal">
<li>Generate ideas for scientific questions that can be answered using 2
or more of the data types we have covered in previous episodes</li>
<li>Identify considerations regarding the cross-compatibility of data
from different sources</li>
<li>Use spatial and taxonomic tools to link data of different types</li>
<li>Calculate and visualize joint diversity indices for multiple data
types simultaneously</li>
<li>Interpret these results in scientific context</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="motivation-why-combine-data-types">Motivation: Why combine data types?<a class="anchor" aria-label="anchor" href="#motivation-why-combine-data-types"></a>
</h1>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>Think-pair-share</p>
<p>Segueing into the simulated Hawaii data</p>
</div>
</div>
</div>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>What kinds of questions can we imagine answering using multipe data
types at once?</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="show-demo-visualization-of-hawaii-sim-data">Show demo visualization of Hawaii sim data<a class="anchor" aria-label="anchor" href="#show-demo-visualization-of-hawaii-sim-data"></a>
</h2>
</div>
</div>
<div class="section level1">
<h1 id="linking-different-data-types">Linking different data types<a class="anchor" aria-label="anchor" href="#linking-different-data-types"></a>
</h1>
<div class="section level2">
<h2 id="checking-the-compatibility-of-data-types">Checking the compatibility of data types<a class="anchor" aria-label="anchor" href="#checking-the-compatibility-of-data-types"></a>
</h2>
<div class="section level3">
<h3 id="taxonomic-overlap">Taxonomic overlap<a class="anchor" aria-label="anchor" href="#taxonomic-overlap"></a>
</h3>
</div>
<div class="section level3">
<h3 id="taxonomic-consistency">Taxonomic consistency<a class="anchor" aria-label="anchor" href="#taxonomic-consistency"></a>
</h3>
</div>
<div class="section level3">
<h3 id="spatial-extent">Spatial extent<a class="anchor" aria-label="anchor" href="#spatial-extent"></a>
</h3>
</div>
<div class="section level3">
<h3 id="others">Others??<a class="anchor" aria-label="anchor" href="#others"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="linking-data-types">Linking data types<a class="anchor" aria-label="anchor" href="#linking-data-types"></a>
</h2>
<div class="section level3">
<h3 id="resources-for-resolving-inconsistencies">Resources for resolving inconsistencies<a class="anchor" aria-label="anchor" href="#resources-for-resolving-inconsistencies"></a>
</h3>
</div>
<div class="section level3">
<h3 id="left_join">left_join<a class="anchor" aria-label="anchor" href="#left_join"></a>
</h3>
<p>(being slightly flip here, but this is the heart of it, right?)</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="analyzing-and-visualizing-joint-data-types">Analyzing and visualizing joint data types<a class="anchor" aria-label="anchor" href="#analyzing-and-visualizing-joint-data-types"></a>
</h1>
<div class="section level2">
<h2 id="hill-numbers-for-multidim-data">Hill numbers for multidim data<a class="anchor" aria-label="anchor" href="#hill-numbers-for-multidim-data"></a>
</h2>
<div class="section level3">
<h3 id="discussionlecture">Discussion/lecture<a class="anchor" aria-label="anchor" href="#discussionlecture"></a>
</h3>
</div>
<div class="section level3">
<h3 id="calculating-hills">Calculating Hills<a class="anchor" aria-label="anchor" href="#calculating-hills"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-multiple-data-types">Visualizing multiple data types<a class="anchor" aria-label="anchor" href="#visualizing-multiple-data-types"></a>
</h2>
<div class="section level3">
<h3 id="joing-vis-plots">Joing vis plots<a class="anchor" aria-label="anchor" href="#joing-vis-plots"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h2>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>Discuss Hawaii interp</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="challenge-using-other-data-sets">Challenge: using other data sets<a class="anchor" aria-label="anchor" href="#challenge-using-other-data-sets"></a>
</h1>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" aria-labelledby="headingInstructor2" data-bs-parent="#accordionInstructor2">
<div class="accordion-body">
<p>Here I was thinking to generate a few more toy data sets using RoLE
as foreshadowing of days 3-4.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="recap">Recap<a class="anchor" aria-label="anchor" href="#recap"></a>
</h1>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Considering multiple data types at once can answer a unique and
rapidly evolving set of questions</li>
<li>Data types do need to be (minimally) compatible; there are
resources/approaches for reconciling not-fully-compatible data</li>
<li>Hill numbers provide a unified framework for the analysis and
visualization of different data types as part of one question</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section><section id="aio-finding_data"><p>Content from <a href="finding_data.html">Finding multi-dimensional biodiversity data</a></p>
<hr>
<p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/finding_data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Where do multi-dimensional biodiversity data live online?</li>
<li>How can we use R APIs to quickly and reproducibly retrieve
data?</li>
<li>What about data that’s not in a database?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal">
<li>Effectively search for and find trait, sequence, and abundance data
online given a taxon or research question</li>
<li>Retrieve data from NCBI &amp; GBIF using their R APIs and apply API
principles to other databases</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="sources-of-multidimensional-biodiversity-data-large-open-access-databases">Sources of multidimensional biodiversity data: large open-access
databases<a class="anchor" aria-label="anchor" href="#sources-of-multidimensional-biodiversity-data-large-open-access-databases"></a>
</h1>
<p>Large open-access biological databases have been growing at an
incredible pace in recent years. Often they accumulate data from many
projects and organizations. These platforms are then used to share and
standardize data for wide use.</p>
<p>Many such databases exist, but here are some listed examples:</p>
<ul>
<li><p><a href="https://www.gbif.org/" class="external-link">GBIF</a> contains occurrence
records of individuals assembled from dozens of smaller
databases.</p></li>
<li><p><a href="https://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a> (National Center
for Biotechnology Information) database, which includes <a href="https://www.ncbi.nlm.nih.gov/genbank/" class="external-link">GenBank</a> is the largest
repository of genetic data in the world, but includes a lot of
biomedical data.</p></li>
<li><p><a href="https://tree.opentreeoflife.org/opentree/argus/opentree14.7@ott93302" class="external-link">OTOL</a>
(Open Tree of Life) is a database that combines numerous published trees
and taxonomies into one supertree</p></li>
<li><p><a href="https://geome-db.org/" class="external-link">GEOME</a> (Genomic Observatories
MetaDatabase) contains genetic data associated with specimens or samples
from the field.</p></li>
<li><p><a href="https://boldsystems.org/" class="external-link">BOLD</a> (Barcode of Life) is
a reference library and database for DNA barcoding</p></li>
<li><p><a href="https://eol.org/" class="external-link">EOL</a> (Encyclopedia of Life) is a
species-level platform containing data like taxonomic descriptions,
distribution maps, and images</p></li>
</ul>
<div class="section level2">
<h2 id="whats-an-api">What’s an API?<a class="anchor" aria-label="anchor" href="#whats-an-api"></a>
</h2>
<p>An API is a set of tools that allows users to interact with a
database, such as reading, writing, updating, and deleting data. We
can’t go over all of these databases in detail, but let’s cover some
examples that illustrate some principles of how these APIs work for
different types of biodiversity data.</p>
<p>We can do this using the example of Hawaiian arthropods. Hawaii is
known for its diversity of <em>Tetragnatha</em>, a genus of orb weaver
spiders. With this target in mind let’s download some
<em>occurrence</em> data from GBIF, some <em>sequence</em> data from
NCBI, and a <em>phylogeny</em> for <em>Tetragnatha</em> spiders from
OTOL.</p>
<div class="section level3">
<h3 id="using-apis-setup">Using APIs: Setup<a class="anchor" aria-label="anchor" href="#using-apis-setup"></a>
</h3>
<p>For this episode, we’ll be working with some R packages that wrap the
APIs of online repositories. Let’s set those up now.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">spocc</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rentrez</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rotl</span><span class="op">)</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'rotl'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:spocc':

    inspect</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'taxize'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:rotl':

    synonyms, tax_name, tax_rank</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="getting-consistent-species-names-using-taxize">Getting consistent species names using <code>taxize</code>
<a class="anchor" aria-label="anchor" href="#getting-consistent-species-names-using-taxize"></a>
</h3>
<p>We will want to source records of the genus <em>Tetragnatha</em>, but
some APIs will require species names rather than the names of broader
taxa like genus or family. We can use the <code>taxize</code> package
for this. <code>taxize</code> links to a genus identification number,
drawn from NCBI, for each genus. We use the <code>get_uid()</code>
function to get this number:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="va">uid</span> <span class="op">&lt;-</span> <span class="fu">get_uid</span><span class="op">(</span><span class="st">"Tetragnatha"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>No ENTREZ API key provided
 Get one via taxize::use_entrez()
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>══  1 queries  ═══════════════</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Retrieving data for taxon 'Tetragnatha'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>✔  Found:  Tetragnatha
══  Results  ═════════════════

• Total: 1 
• Found: 1 
• Not Found: 0</code></pre>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor1" aria-labelledby="headingInstructor1">
<div class="accordion-body">
<p>You will be prompted to get an ENTREZ API key. This is a good idea to
do in order to make larger queries, but it won’t be neccesary for this
workshop.</p>
</div>
</div>
</div>
</div>
<p>Then, we use the <code>downstream()</code> function to get a list of
all species in <em>Tetragnatha</em>.</p>
<div id="callout1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout1"></a>
</h3>
<div class="callout-content">
<p>This also works on larger taxa like families and orders.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_uids</span> <span class="op">&lt;-</span> <span class="fu">downstream</span><span class="op">(</span><span class="va">uid</span>,downto<span class="op">=</span><span class="st">"species"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>No ENTREZ API key provided
 Get one via taxize::use_entrez()
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/
No ENTREZ API key provided
 Get one via taxize::use_entrez()
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/
No ENTREZ API key provided
 Get one via taxize::use_entrez()
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="va">species_uids</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">childtaxa_name</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Tetragnatha paludicola" "Tetragnatha boydi"      "Tetragnatha cavaleriei"
[4] "Tetragnatha virescens"  "Tetragnatha josephi"    "Tetragnatha gui"       </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="occurrence-data-from-gbif">Occurrence data from GBIF<a class="anchor" aria-label="anchor" href="#occurrence-data-from-gbif"></a>
</h3>
<p>The <code>spocc</code> package is all about occurrence data and lets
you download occurrences from all major occurrence databases. For GBIF
access using the spocc package, we use the <code>occ()</code>
function.</p>
<p>We’ll put the species names we retrieved as the <code>query</code>
and specify <code>gbif</code> as our target database in the
<code>from</code> field.</p>
<p>Since the occurrences from spocc are organized as a list of
dataframes, one per species, we use the bind_rows() function from dplyr
to combine this into a single dataframe.</p>
<p>We can also specify a geographic bounding box of where we’d like to
look for occrences. Here we’re using a bounding box that covers the
Hawaiian Islands.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">spocc</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">occurrences</span> <span class="op">&lt;-</span> <span class="fu">occ</span><span class="op">(</span>query <span class="op">=</span> <span class="va">species_names</span>, from <span class="op">=</span> <span class="st">'gbif'</span>, has_coords<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                   gbifopts<span class="op">=</span><span class="fu">list</span><span class="op">(</span><span class="st">"decimalLatitude"</span><span class="op">=</span><span class="st">'18.910361,28.402123'</span>,</span>
<span>                                 <span class="st">"decimalLongitude"</span><span class="op">=</span><span class="st">'-178.334698,-154.806773'</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># extract the data from gbif</span></span>
<span><span class="va">occurrences_gbif</span> <span class="op">&lt;-</span> <span class="va">occurrences</span><span class="op">$</span><span class="va">gbif</span><span class="op">$</span><span class="va">data</span> </span>
<span></span>
<span><span class="co"># the results in `$data` are a list with one element per species, </span></span>
<span><span class="co"># so we combine all those elements</span></span>
<span><span class="va">occurrences_df</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">occurrences_gbif</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">occurrences_df</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 96
  name           longitude latitude issues prov  key   scientificName datasetKey
  &lt;chr&gt;              &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;          &lt;chr&gt;     
1 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
2 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
3 Tetragnatha p…     -156.     20.8 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
4 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
5 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
6 Tetragnatha p…     -157.     20.9 cdc,o… gbif  1060… Tetragnatha p… 5d283bb6-…
# ℹ 88 more variables: publishingOrgKey &lt;chr&gt;, installationKey &lt;chr&gt;,
#   hostingOrganizationKey &lt;chr&gt;, publishingCountry &lt;chr&gt;, protocol &lt;chr&gt;,
#   lastCrawled &lt;chr&gt;, lastParsed &lt;chr&gt;, crawlId &lt;int&gt;, basisOfRecord &lt;chr&gt;,
#   individualCount &lt;int&gt;, occurrenceStatus &lt;chr&gt;, taxonKey &lt;int&gt;,
#   kingdomKey &lt;int&gt;, phylumKey &lt;int&gt;, classKey &lt;int&gt;, orderKey &lt;int&gt;,
#   familyKey &lt;int&gt;, genusKey &lt;int&gt;, speciesKey &lt;int&gt;, acceptedTaxonKey &lt;int&gt;,
#   acceptedScientificName &lt;chr&gt;, kingdom &lt;chr&gt;, phylum &lt;chr&gt;, order &lt;chr&gt;, …</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="phylogenetic-trees-from-the-open-tree-of-life">Phylogenetic trees from the Open Tree of Life<a class="anchor" aria-label="anchor" href="#phylogenetic-trees-from-the-open-tree-of-life"></a>
</h3>
<p>First we need to get a clean list of taxonomic names from the GBIF
results:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hiTetragnatha</span> <span class="op">&lt;-</span> <span class="fu">unique</span><span class="op">(</span><span class="va">occurrences_df</span><span class="op">$</span><span class="va">name</span><span class="op">)</span></span>
<span><span class="va">hiTetClean</span> <span class="op">&lt;-</span> <span class="fu">gnr_resolve</span><span class="op">(</span><span class="va">hiTetragnatha</span>, best_match_only <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                          canonical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">hiTetragnatha</span> <span class="op">&lt;-</span> <span class="va">hiTetClean</span><span class="op">$</span><span class="va">matched_name2</span></span>
<span><span class="va">hiTetragnatha</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Tetragnatha paludicola"   "Tetragnatha stelarobusta"
 [3] "Tetragnatha perkinsi"     "Tetragnatha filiciphilia"
 [5] "Tetragnatha eurychasma"   "Tetragnatha macracantha" 
 [7] "Tetragnatha anuenue"      "Tetragnatha acuta"       
 [9] "Tetragnatha pilosa"       "Tetragnatha quasimodo"   
[11] "Tetragnatha tantalus"     "Tetragnatha polychromata"
[13] "Tetragnatha restricta"    "Tetragnatha brevignatha" 
[15] "Tetragnatha kamakou"      "Tetragnatha versicolor"  
[17] "Tetragnatha kauaiensis"   "Tetragnatha waikamoi"    
[19] "Tetragnatha hawaiensis"  </code></pre>
</div>
<p>Now we can use the <code>rotl</code> package. The <code>rotl</code>
package provides an interface to the Open Tree of Life. To get a tree
from this database, we have to match our species names to OTOL’s
ids.</p>
<p>We first use <code>trns_match_names()</code> to match our species
names to OTOL’s taxonomy.</p>
<p>Then we use <code>ott_id()</code> to get OTOL ids for these matched
names.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resolved_names</span> <span class="op">&lt;-</span> <span class="fu">tnrs_match_names</span><span class="op">(</span><span class="va">hiTetragnatha</span><span class="op">)</span></span>
<span><span class="va">otol_ids</span> <span class="op">&lt;-</span> <span class="fu">ott_id</span><span class="op">(</span><span class="va">resolved_names</span><span class="op">)</span></span></code></pre>
</div>
<p>Finally, we get the tree containing these IDs as tips.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu">tol_induced_subtree</span><span class="op">(</span>ott_ids <span class="op">=</span> <span class="fu">ott_id</span><span class="op">(</span><span class="va">resolved_names</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tr</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/finding_data-rendered-unnamed-chunk-6-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Unfortunately we can see that the phylogenetic relationships between
these species aren’t resolved in OTOL, so we may have to look elsewhere
for a phylogeny. This is an example of how not all data are in large
repositories.</p>
</div>
<div class="section level3">
<h3 id="sequence-data-from-ncbi">Sequence data from NCBI<a class="anchor" aria-label="anchor" href="#sequence-data-from-ncbi"></a>
</h3>
<p>Let’s continue with the <code>rentrez</code> package, which provides
an interface to NCBI.</p>
<p>NCBI’s API works with genera and larger ranks, but has a harder time
with lists of species.</p>
<p>For this reason, let’s just search for <em>Tetragnatha</em> using the
‘term’ argument, which takes a string in a specific format describing
the query.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># note we're using `hiTetragnatha` that we made previously to make a long</span></span>
<span><span class="co"># string of names that will be sent to NCBI via the ENTREZ API</span></span>
<span><span class="va">termstring</span> <span class="op">&lt;-</span> <span class="fu">paste</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">'%s[ORGN]'</span>, <span class="va">hiTetragnatha</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">' OR '</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">termstrig</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'termstrig' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">search_results</span> <span class="op">&lt;-</span> <span class="fu">entrez_search</span><span class="op">(</span>db<span class="op">=</span><span class="st">"nucleotide"</span>, term <span class="op">=</span> <span class="va">termstring</span><span class="op">)</span></span></code></pre>
</div>
<p><em>Unlike</em> <code>spocc</code> but <em>similarly</em> to
<code>rotl</code>, this first search returns a set of IDs.</p>
<p>We then have to use <code>entrez_fetch()</code> on the IDs in
search_results to obtain the actual record information.</p>
<p>We specify ‘db’ as ‘nucleotide’ and ‘rettype’ as ‘fasta’ to get FASTA
sequences</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu">entrez_fetch</span><span class="op">(</span>db <span class="op">=</span> <span class="st">"nucleotide"</span>, id <span class="op">=</span> <span class="va">search_results</span><span class="op">$</span><span class="va">ids</span>, </span>
<span>                          rettype <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># just look at the first part of this *long* returned string</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="fu">substr</span><span class="op">(</span><span class="va">sequences</span>, <span class="fl">1</span>, <span class="fl">1148</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&gt;MG509184.1 Tetragnatha versicolor voucher BIOUG19074-C11 cytochrome oxidase subunit 1 (COI) gene, partial cds; mitochondrial
AGTTTATATTTTTTATTTGGAGTGTGATCAGNTATAGTAGGGACAGCTATAAGAGTTTTAATTCGAATTG
AATTAGGTCAATCTGGCAAATTTCTTGGGGATGACCAATTATATAATGTTATTGTGACTGCTCATGCTTT
TGTAATAATTTTTTTTATGGTTATACCAATTTTAATTGGGGGGTTTGGGAATTGATTAGTCCCATTAATA
TTAGGGGCACCAGATATGGCTTTCCCCCGGATAAATAACTTAAGTTTTTGGTTGTTGCCCCCTTCTCTTT
TTATATTGTTTATTTCGTCTATAGTGGATGTGGGAGTTGGAGCAGGTTGANCCNNTTATCCCCCTTTATC
TTCTTTAGAAGGGCATTCAGGCAGATCTGTAGATTTTGCTATTTTTTCACTTCATTTAGCTGGGGCTTCT
TCAATTATAGGGGCTATTAATTTTATTTCTACTATCCTTAATATGCGGATAAGAGGAATTTCTATAGAAA
AGGTACCTCTTTTTGTGTGATCTGTTTTGATTACAGCG

&gt;MG511976.1 Tetragnatha versicolor voucher BIOUG23383-C01 cytochrome oxidase subunit 1 (COI) gene, partial cds; mitochondrial
AGTTTATATTTTTTATTTGGAGTGTGATCAGCTATAGTGGGGACAGCTATAAGAGTTTTAATTCGAATTG
AGTTAGGTCAATCTGGGAAATTTCTTGGGGATGACCAATTATACAATGTAATTGTAACTGCTCATGCTTT
TGTAATAATTTTTTTTATGGTGATACCAATTTTAATTGGGGGCTTTGGAAATTGATTAGTGCCCTTAATA
TTAGGAGCGCCAGATATAGCTTTTCCTCGGATAAATAACTTAAGTTTTTGGTTGCTACCCCCTTCTCTTT
TTATGTTGTTTATTTCATCTATAGTAGATGTGGGAGTTGGGGCAGGTTGAACTGTCTATCCCCCTTTATC
T

&gt;</code></pre>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>What are some similarities and differences in how data is obtained
using the R APIs of GBIF, NCBI, and OTOL?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>NCBI and OTOL both require an initial search to obtain record IDs,
then a second fetch to obtain the actual record data for each ID</p>
<p>NCBI uses a string-based query where the search is a
specially-formated string, while GBIF uses an argument-based query where
the search is based on the various arguments you give to the R API
function</p>
<p>NCBI is capable of taking a genus and returning the records of all
its species, but OTOL and GBIF require obtaining the species names first
using a package like taxize</p>
<p>NCBI can take a state in the query, GBIF can take a country but not a
state.</p>
<p>GBIF data is formatted as a list of occurrences for each species
contained in the object we get from the query.</p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="sources-of-mdbd-databases-without-apis">Sources of MDBD: Databases without APIs<a class="anchor" aria-label="anchor" href="#sources-of-mdbd-databases-without-apis"></a>
</h1>
<p>Many excellent databases exist without specialized APIs or R packages
to facilitate their use.</p>
<p>For example, the BioTIME database is a compilation of timeseries of
species ocurrence, abundance, and biomass data. It is <a href="https://biotime.st-andrews.ac.uk/" class="external-link">openly available</a> via the
University of St. Andrews. The North American Breeding Bird Survey, run
by the USGS, contains data on bird species abundances across the United
States and Canada going back more than 50 years. These data are hosted
on <a href="https://www.sciencebase.gov/catalog/item/52b1dfa8e4b0d9b325230cd9" class="external-link">ScienceBase</a>.</p>
<p>Some of these databases can be accessed via the <a href="https://retriever.readthedocs.io/en/latest/" class="external-link">Data Retriever and
accompanying R package the <code>rdataretriever</code>.</a></p>
<p>Some are findable via <a href="https://search.dataone.org/data" class="external-link">DataONE</a> and the <a href="https://edirepository.org/" class="external-link">Ecological Data Initiative</a>.</p>
<p>Others can be downloaded directly from the source.</p>
</div>
<div class="section level1">
<h1 id="sources-of-mdbd-small-data-attached-to-papers">Sources of MDBD: “small data” attached to papers<a class="anchor" aria-label="anchor" href="#sources-of-mdbd-small-data-attached-to-papers"></a>
</h1>
<p>Finally, A lot of useful data is not held in databases, and is
instead attached to papers. Searching manuscript databases or Google
Scholar can be an effective way to find such data. You can filter Web of
Science to find entries with “associated data”.</p>
<p>However, this can still be a somewhat heterogeneous and
time-consuming strategy!</p>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor2" aria-labelledby="headingInstructor2">
<div class="accordion-body">
<p>Time-permitting, this could be an opportunity for a 15-minute
breakout exercise where individuals or groups pick a database or website
and search for data, then report back on what they found and what the
process was like.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="recap">Recap<a class="anchor" aria-label="anchor" href="#recap"></a>
</h1>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Many sources exist online where multidimensional biodiversity data
can be obtained</li>
<li>APIs allow for fast and reproducible querying and downloading</li>
<li>While there are patterns in how these APIs work, there will always
be differences between databases that make using each API a bit
different.</li>
<li>Standalone databases and even the supplemental data in manuscripts
can also be re-used.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/role-model/multidim-biodiv-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:biodataworkshop2023@gmail.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.3" class="external-link">sandpaper (0.12.3)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://role-model.github.io/multidim-biodiv-data/instructor/aio.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://role-model.github.io/multidim-biodiv-data/instructor/aio.html",
  "identifier": "https://role-model.github.io/multidim-biodiv-data/instructor/aio.html",
  "dateCreated": "2023-06-15",
  "dateModified": "2023-06-15",
  "datePublished": "2023-06-15"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

