<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Multidimensional biodiversity data: management and analysis: Summarizing phylogenetic data</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../phylo-data.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Multidimensional biodiversity data: management and analysis
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Multidimensional biodiversity data: management and analysis
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Multidimensional biodiversity data: management and analysis
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 50%" class="percentage">
    50%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 50%" aria-valuenow="50" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../phylo-data.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="abundance-data.html">2. Abundance Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="traits-data.html">3. Trait data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="popgen-data.html">4. Working with population genetic data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        Summarizing phylogenetic data
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#visualization-of-phylo-hill-numbers">Visualization of phylo Hill numbers</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="connecting-data-types.html">6. Connecting different data types</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="finding_data.html">7. Finding multi-dimensional biodiversity data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="care.html">8. The CARE principles and open data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/popgen-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/connecting-data-types.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/popgen-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Working with</a>
        <a class="chapter-link float-end" href="../instructor/connecting-data-types.html" rel="next">Next: Connecting different... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Summarizing phylogenetic data</h1>
        <p> Last updated on 2023-05-17 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/phylo-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What are the common phylogeny data formats?</li>
<li>How do you import and manipulate phylogeny data in R?</li>
<li>How do you visualize phylogenies in R?</li>
<li>How to summarize and interpret phylogenies using Hill numbers?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, participants will be able to:</p>
<ol style="list-style-type: decimal"><li>Identify key features of different phylogenetic data formats</li>
<li>Import, clean and manipulate phylogenetic data in the R
environment</li>
<li>Visualize phylogenies</li>
<li>Calculate and interpret phylogenetic Hill numbers</li>
</ol></div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h3>
<p>In this episode, we will explore how to extract information from
phylogenetic trees in order to complement our hypotheses and inferences
about processes shaping biodiversity patterns. Phylogenetic trees have
information about how the species in a taxonomic group are related to
each other and how much relative evolutionary change has accumulated
among them. Since local communities differ in their phylogenetic
composition, this information can give insights on why communities are
how they are.</p>
<p>The relative phylogenetic distance among species in a community as
well as the distribution of the amount of evolutionary history
(represented by the length of the branches in a phylogeny) are a result
of different factors such as the age since the initial formation of the
community and the rate of macroevolutionary processes such as speciation
and extinction. For instance, young communities that are dominated by
closely related species and show very short branch lengths may suggest a
short history with few colonization events and high rates of local
speciation; alternatively, if the same young communities harbor
distantly related species with longer branch lengths, it may suggest
that most of the local diversity was generated by speciation elsewhere,
along with many colonization by species from different points in the
phylogeny. Coupled with information on ecological traits and rates of
macroevolutionary processes, these patterns also allow to test for
hypotheses regarding, for instance, ecological filtering or niche
conservatism.</p>
<!-- Maybe bring in the paragraph above explicit connections to our young vs older Hawaian islands examples. -->
<p>Summarizing this phylogenetic information (i.e., phylogenetic
distance and distribution of branch lenghts) is therefore crucial for
inference. As we have seen in previous episodes, the use of <a href="">Hill numbers</a> is an informative approach to summarize
biodiversity. In this episode, we will then go over 1) how to import and
manipulate phylogenetic data in R; 2) how to calculate and interpret
Hill numbers for different communities.</p>
</div>
<div class="section level3">
<h3 id="importing-phylogenetic-data">Importing phylogenetic data<a class="anchor" aria-label="anchor" href="#importing-phylogenetic-data"></a></h3>
<p>Several file formats exist to store phylogenetic information. The
most common formats are the <a href="">Newick</a> and <a href="">Nexus</a> formats. Both these formats are plain text files
storing different levels of information about the taxa relationship and
evolutionary history. <code>Newick</code> files are the standard for
representing trees in a computer-readable form, as they can be extremely
simple and therefore do not take up much memory. <code>Nexus</code>
file, on the other hand, are composed by different blocks regarding
different types of information and can be used to store DNA alignments,
phylogenetic trees, pre-defined groupings of taxa, or everything at
once. Since they are a step ahead in complexity, we will stick with
<code>Newick</code> files for now.</p>
<p><code>Newick</code> files store the information about the clades in a
tree by representing each clade within two parentheses, with tips
separated by <code>,</code>. We also add the symbol <code>;</code> at
the end to represent the end of the information for that phylogenetic
tree.</p>
<p>The basic structure of a tree in a <code>Newick</code> format is
therefore as follows:</p>
<p>((A,B),C);</p>
<p>The notation above indicates that: 1) we have three taxa in our tree,
named <code>A</code>, <code>B</code> and <code>C</code>; 2)
<code>A</code> and <code>B</code> form one clade (A,B); 3) the (A,B)
clade is sister to the <code>C</code> clade (we represent that by adding
another set of parentheses and a <code>,</code> separating (A,B) from
C).</p>
<p>In addition, <code>Newick</code> files can also store information on
the branch leading to it tip and node. We do that by adding
<code>:</code> after each tip/node.</p>
<p>((A:0.5,B:0.5):0.5,C:1);</p>
<p>The notation above indicates that: 1) the branches containing A and B
have each a length = 0.5; 2) the branch that leads to the node
connecting A and B also has length = 0.5; 3) the branch leading to C has
length = 1.</p>
<p>The notation above is what we import into R to start working with and
manipulating our phylogenetic tree. For that goal, we will use the <a href="package">ape</a>:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span></code></pre>
</div>
<p>To import our tree, we will be using the function
<code>read.tree()</code> from the <code>ape</code> package. In the case
of simple trees as the one above, we could directly create them within R
by giving that notation as a <code>character</code> value to this
function, using the <code>text</code> argument, as shown below:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text <span class="op">=</span> <span class="st">'((A:0.5,B:0.5):0.5,C:1);'</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, we can visually inspect our tree using the <code>plot()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">example_tree</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/phylo-data-rendered-plot-example-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Can you visualize the text notation in that image? We can see the
same information: A is closer related to B than C, and the branches
leading to A and B have half the length of the branch leading to C.</p>
<p>The <code>read.tree()</code> function creates an object of class
<code>phylo</code>. We can further investigate this object by calling it
in our console:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Phylogenetic tree with 3 tips and 2 internal nodes.

Tip labels:
  A, B, C

Rooted; includes branch lengths.</code></pre>
</div>
<p>The printed information shows us that we have a phylogenetic tree
with 3 tips and 2 internal nodes, where the tip labels are “A, B, C”. We
also are informed that this tree is <a href="">rooted</a> and has branch
lengths.</p>
<p>One way to access the components of this object and better explore it
is to use <code>$</code> after the object name. Here, it will be
important for us to know a little bit more about where the information
about tip labels and branch lengths are stored in that
<code>phylo</code> objecy. Easy enough, we can access that by calling
<code>tip.labels</code> and <code>edge.length</code> after
<code>$</code>.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span><span class="op">$</span><span class="va">tip.label</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "B" "C"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_tree</span><span class="op">$</span><span class="va">edge.length</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.5 0.5 0.5 1.0</code></pre>
</div>
<p>Now that we learned how to import and visualize trees in R, let’s
bring the phylogeny for the communities we are working with in this
workshop. Our data so far consists of <a href="abundance-data.html">abundances</a> and <a href="traits-data.html">traits</a> of several taxa of arthropods
collected across three islands in the Hawaian archipelago. Here, we will
import a global phylogeny of arthropods, located at: &lt;&gt;</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylogeny</span> <span class="op">&lt;-</span> <span class="fu">read_tree</span><span class="op">(</span><span class="st">''</span><span class="op">)</span></span></code></pre>
</div>
<p>This new object is way larger than the previous once, being a “real”
phylogeny and all. You can inspect it again by directly calling the
object <code>phylogeny</code>. To plot it, we will use the
<code>type</code> argument to modify how our tree will be displayed.
Here, we used the option <code>'fan'</code>, to display a circular
phylogeny (slightly better to show such a large phylogeny in the
screen).</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">phylogeny</span>, type <span class="op">=</span> <span class="st">'fan'</span><span class="op">)</span></span></code></pre>
</div>
<p>How do we combine all this information with the community datasets we
have so far for our three islands? First, we will have perform some
cleaning.</p>
</div>
<div class="section level3">
<h3 id="cleaning-phylogenetic-data">Cleaning phylogenetic data<a class="anchor" aria-label="anchor" href="#cleaning-phylogenetic-data"></a></h3>
<p>Often, when we want to include the phylogenetic relatedness of
organisms in our biodiversity analyses, we will run into published
phylogenetic trees for our focal taxon (that is the case of the tree we
are using here). A couple issues arise in these cases. First, and maybe
most importantly, we need to make sure that the taxonomic names used
across the different datasets in our analyses (phylogenies included) are
the same. A second issue is that, depending on the resolution of the
published phylogeny, it may be the case that some taxa in our <a href="">community dataset</a> may not be explicitly represented in the
phylogeny. In these cases, we will want to assign those “missing taxa”
to one of the tips present in the tree, based on their known genus
and/or family assignments.</p>
<p>In this section, we will deal with these issues in order to prepare
our data for community inferences. Our first issue will be dealt with in
a challenge!</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>In previous episodes, we have used the package <code>taxize</code> to
clean and standardize the taxonomic names in out dataset. Here, we will
apply the same approach to the tip names of our phylogeny. We need to
make sure the names we will accept from <code>taxize</code> will match
the names we have accepted in previous episodes. Finally, after we
decide on the names, we need to add them to the <code>phylogeny</code>
objects, where the tip names are stored.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>First, we check the names using the <code>gnr_resolve()</code>
function. In <code>phylogeny</code>, tip names are stored in
<code>phylogeny$tip.label</code>, so we provide that vector to our
function.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">taxize</span><span class="op">)</span></span>
<span><span class="va">name_resolve</span> <span class="op">&lt;-</span> <span class="fu">gnr_resolve</span><span class="op">(</span><span class="va">phylogeny</span><span class="op">$</span><span class="va">tip.label</span>, best_match_only <span class="op">=</span> <span class="cn">TRUE</span>, canonical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">name_resolve</span><span class="op">)</span></span></code></pre>
</div>
<p>We can see what taxa are in conflict with the names retrieved by
<code>taxize</code> by subsetting <code>name_resolve</code> based on
boolean matching.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mismatches_phylo</span> <span class="op">&lt;-</span> <span class="va">name_resolve</span><span class="op">[</span><span class="va">name_resolve</span><span class="op">$</span><span class="va">user_supplied_name</span> <span class="op">!=</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"user_supplied_name"</span>, <span class="st">"matched_name2"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">mismatches_phylo</span></span></code></pre>
</div>
<p>From here, we can see that apparently there are no misspellings in
our tree (as expected from a published phylogeny). Most of the
mismatches seem to be related to changes in the taxonomic treatment of
species. This is a common situation, since taxonomic treatments are
often revised after new trees are published; it might be the case that
<code>nameA</code> in our dataset was treated as <code>nameB</code> at
the time the phylogeny was created.</p>
<p>In the previous episodes, we used the names provided by
<code>taxize</code> directly. Therefore, here we can just add those
names directly to the phylogeny.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylogeny</span><span class="op">$</span><span class="va">tip.label</span> <span class="op">==</span> <span class="va">name_resolve</span><span class="op">$</span><span class="va">matched_name2</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The next issue we need to address is the fact that we may have
species in our dataset that are not present in the phylogenetic tree we
are using. We therefore need to 1) find out which species are not in the
tree; 2) assign them to tips that are present in the tree.</p>
<p>To check which species are not in the tree, we can use any of the
objects we created so far that has the list of all the species in our
communities. These could be: 1) the abundances we calculated per site
and island in the <a href="">abundances episode</a>, stored in the
<code>abundance_tallies</code> object; 2) the <em>site by species
matrix</em> we have created in the abundance episode (stored in the
<code>abundance_wide</code> object); 3) the list of species with their
respective trait values, created in the <a href="">traits episode</a>
and stored in the <code>traits_simple</code> object. Since we directly
use the site-by-species matrix to calculate Hill numbers (next section),
let’s focus on this object and retrieve the species name for it.</p>
<p>We do that by accessing the <code>colnames()</code> of that object.
We also remove the first element, since it is just the header for the
column with the <code>site</code> name.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">abundance_wide</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<p>Since the taxonomic names in both the abundance data and in the
phylogenetic tree were corrected using the same database from
<code>taxize</code>, we expect that any names from our abundance data
that are nof found in the tree simply represent taxa that are not
included in this phylogeny. We can use the boolean operator
<code>%in%</code> coupled with <code>!</code> to check
<code>species_names</code> that are not included in the
<code>phylogeny$tip.label</code>. In summary, the expression
<code>A %in% B</code> would return what elements of A are present in B.
The <code>!</code> (NOT) operator returnts the opposite of that
expression, in a way that <code>!(A %in% B)</code> will return the
elements of A that are <em>NOT</em> present in B (which is what we
want)</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">not_present</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="op">(</span><span class="va">species_names</span> <span class="op">%in%</span> <span class="va">phylogeny</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span></span></code></pre>
</div>
<p>Now let’s check those species that were not in the tree:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_name</span><span class="op">[</span><span class="va">not_present</span><span class="op">]</span></span></code></pre>
</div>
<p>Here, we can decide how we are going to assign these species to tips
in our tree. For instance, we see that we do not have <em>XXX</em> in
our tree, but we have <em>XXX</em>, which belongs to the same genus.
Similarly…</p>
<p>We can then rename the species with the name of those that are
present in the species.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">species_name</span><span class="op">[</span><span class="va">not_present</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>And now we can rename the species in out site-by-species matrix using
these new names. For this, we will first transfer our matrix to another
object (in order to keep the abundance matrix intact), and then we will
rename the columns of this new object based on the modified names.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_wide</span> <span class="op">&lt;-</span> <span class="va">abundance_wide</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">phylo_wide</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">species_name</span></span></code></pre>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">SH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode sh" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Another</span> issue that can arise with phylogenies is that when we assign species to a tip, we may end up with duplicates in the site-by-species matrix. We will not show this here, but it might be something important to bring up.</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>After solving these issues, we now have 1) a phylogenetic tree with
tip labels taxonomically cleaned; 2) a site-by-species matrix where all
species are represented in out tree. These two objects are the two bits
of information that we need to summarize the phylogenetic diversity and
the distribution of branch lenghts across our focal communities.</p>
</div>
<div class="section level3">
<h3 id="summarizing-with-hill-numbers">Summarizing with hill numbers<a class="anchor" aria-label="anchor" href="#summarizing-with-hill-numbers"></a></h3>
<blockquote>
<p>Recap the idea of hill numbers. How they apply to phylogenies</p>
</blockquote>
<p>As we discussed above,</p>
<p>“The relative phylogenetic distance among species in a community as
well as the distribution of the amount of evolutionary history
(represented by the length of the branches in a phylogeny) are a result
of different factors such as the age since the initial formation of the
community and the rate of macroevolutionary processes such as speciation
and extinction. For instance, young communities that are dominated by
closely related species and show very short branch lengths may suggest a
short history with few colonization events and high rates of local
speciation; alternatively, if the same young communities harbor
distantly related species with longer branch lengths, it may suggest
that most of the local diversity was generated by speciation elsewhere,
along with many colonization by species from different points in the
phylogeny. Coupled with information on ecological traits and rates of
macroevolutionary processes, these patterns also allow to test for
hypotheses regarding, for instance, ecological filtering or niche
conservatism.”</p>
<p>here we are calculating branch lengths and the distribution of those
lengths across the community.</p>
<p>Let’s first visualize what this information mean by comparing two
trees with different pattern of relationships.</p>
<blockquote>
<p>Play aroud with toy trees: ladder phylogenies and perfectly balanced
trees Totally Balanced - short and long Totally unbalanced 6 clades</p>
</blockquote>
<p>Show here how to manually retrieve branch lengths and check the
evenness of the distribution, as a prelude to hill numbers.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tre1</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'(a:3,(b:2,(c:1,d:1):1):1);'</span><span class="op">)</span></span>
<span><span class="va">tre2</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text<span class="op">=</span><span class="st">'((a:1,b:1):1,(c:1,d:1):1);'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## ADD HERE CODE TO RETRIEVE BRANCH LENGHTS MANUALLY</span></span></code></pre>
</div>
<blockquote>
<p>Maybe show here a SAD-like distribution of branch lengths to give an
intuition of the evenness?</p>
</blockquote>
<p>Values of hill numbers can give an idea of the relative length of the
branches across the community and help answer the questions above.</p>
<blockquote>
<p><code>hill_phylo</code> function - show how the function
automatically calculates phylo hill numbers for each community based on
a global tree. Show example in toy trees first.</p>
</blockquote>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">site_m</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">site_m</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'a'</span>,<span class="st">'b'</span>,<span class="st">'c'</span>,<span class="st">'d'</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span><span class="fu">hill_phylo</span><span class="op">(</span><span class="va">site_m</span>,<span class="va">tre1</span><span class="op">)</span></span>
<span><span class="fu">hill_phylo</span><span class="op">(</span><span class="va">site_m</span>,<span class="va">tre2</span><span class="op">)</span></span></code></pre>
</div>
<p>Doing it in your own tree leaving to a challenge. Discussion on what
is found?</p>
<p>Do the calculating for all islands in our simulated dataset using the
species-by-site matrix created in previous episodes.</p>
</div>
<section id="visualization-of-phylo-hill-numbers"><h2 class="section-heading">Visualization of phylo Hill numbers<a class="anchor" aria-label="anchor" href="#visualization-of-phylo-hill-numbers"></a>
</h2>
<hr class="half-width"><p>==== Personal code for me to better understand phylogenetic hill
numbers:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Creating a site matrix with one site and four taxa all present in the one site</span></span>
<span><span class="va">site_m</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">site_m</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="st">'t'</span>,<span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Six random trees</span></span>
<span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu">rmtopology</span><span class="op">(</span><span class="fl">6</span>,<span class="fl">4</span>,rooted <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">trees</span>,FUN <span class="op">=</span> <span class="va">force.ultrametric</span><span class="op">)</span></span>
<span><span class="va">hill_data</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span><span class="op">(</span><span class="va">trees</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">hill0</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">site_m</span>, <span class="va">trees</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, q <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="va">hill1</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">site_m</span>, <span class="va">trees</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, q <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="va">hill2</span> <span class="op">&lt;-</span> <span class="fu">hill_phylo</span><span class="op">(</span><span class="va">site_m</span>, <span class="va">trees</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, q <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="va">hill_data</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">hill_data</span>,<span class="fu">c</span><span class="op">(</span><span class="va">hill0</span>,<span class="va">hill1</span>,<span class="va">hill2</span>,<span class="va">trees</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">edge.length</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">hill_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'hill_0'</span>,<span class="st">'hill_1'</span>,<span class="st">'hill_2'</span>,<span class="fu">paste0</span><span class="op">(</span><span class="st">'edge_'</span>,<span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting SAD of branch lenghts for each tree</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,<span class="fu">sort</span><span class="op">(</span><span class="va">hill_data</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,<span class="fu">sort</span><span class="op">(</span><span class="va">hill_data</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,<span class="fu">sort</span><span class="op">(</span><span class="va">hill_data</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,<span class="fu">sort</span><span class="op">(</span><span class="va">hill_data</span><span class="op">[</span><span class="fl">4</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,<span class="fu">sort</span><span class="op">(</span><span class="va">hill_data</span><span class="op">[</span><span class="fl">5</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,<span class="fu">sort</span><span class="op">(</span><span class="va">hill_data</span><span class="op">[</span><span class="fl">6</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Personal notes: * Order 0 is just the sum of branch lengths… From
order 1 and above, you weight by the abundance of that branch length in
the tree… Like, how many branches have high length vs how many branchs
have low length? In order one, if you have one branch that is super long
and all others are super short, your hill number drops: your PD is
driven by only one branch… If branch length is evenly distributed
through the phylogeny (i.e., several branches with similar lenghts),
then your PD increases… That is a Shannon entropy index basically… * As
you increase the order of the hill number, you are giving more weight to
highly abundant branch lenghts… Order 0 is just the sum of lenghts, 1 is
weight by evenness, 2 is weighted by evenness giving more weight to
values of branch lengths that are more abundant, 3 is the same but
focusing even more on the most abundance values (i.e., as you increase
the order, you decrease your focus on rare values)… So, like, if you
have a phylogeny dminated by few branches that are very long, your hill
number drops really fast as you move up in the orders… Because your hill
number becomes dictated by only a few branches very fast… Whereas if you
have a more even distribution of branch lenghts (several branch with
similar values of length), the hill number doesn’t change as much as you
move higher in the order, because even high orders still focus on most
of the branches (as they are the most abundant, since they are equally
abundant) * If all branches had the exact same branch length, then hill
numbers wouldn’t change as you go up in the orders (i.e, evenness is
perfect)</p>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/popgen-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/connecting-data-types.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/popgen-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Working with</a>
        <a class="chapter-link float-end" href="../instructor/connecting-data-types.html" rel="next">Next: Connecting different... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/phylo-data.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/role-model/multidim-biodiv-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.11.17" class="external-link">sandpaper (0.11.17)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.16" class="external-link">varnish (0.2.16)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://role-model.github.io/multidim-biodiv-data/instructor/phylo-data.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Summarizing phylogenetic data",
  "creativeWorkStatus": "active",
  "url": "https://role-model.github.io/multidim-biodiv-data/instructor/phylo-data.html",
  "identifier": "https://role-model.github.io/multidim-biodiv-data/instructor/phylo-data.html",
  "dateCreated": "2023-05-17",
  "dateModified": "2023-05-17",
  "datePublished": "2023-05-17"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

