<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Multidimensional biodiversity data: management and analysis: Working with population genetic data</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback." style="text-decoration: unset">
          ¬†
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link" style="color: #383838">Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="background: #FFC700; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../popgen-data.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Multidimensional biodiversity data: management and analysis
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Multidimensional biodiversity data: management and analysis
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Multidimensional biodiversity data: management and analysis
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 58%" class="percentage">
    58%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 58%" aria-valuenow="58" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../popgen-data.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="care.html">2. The CARE principles and open data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="abundance-data.html">3. Abundance Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="traits-data.html">4. Trait data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        5. Working with population genetic data
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#pop-gen-data">Pop gen data</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="phylo-data.html">6. Summarizing phylogenetic data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="connecting-data-types.html">7. Connecting different data types</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="finding_data.html">8. Finding multi-dimensional biodiversity data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/traits-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/phylo-data.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/traits-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Trait data</a>
        <a class="chapter-link float-end" href="../instructor/phylo-data.html" rel="next">Next: Summarizing... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Working with population genetic data</h1>
        <p> Last updated on 2023-06-15 | 
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/popgen-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<section id="pop-gen-data"><h2 class="section-heading">Pop gen data<a class="anchor" aria-label="anchor" href="#pop-gen-data"></a>
</h2>
<hr class="half-width"><div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What is genetic diversity?</li>
<li>What does genetic diversity say about community assembly?</li>
<li>What are common summaries of genetic diversity within species and
communities?</li>
<li>What are common sequence file formats?</li>
<li>How do you read in and manipulate sequence data?</li>
<li>How do you visualize and summarize sequence data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>After following this episode, we intend that participants will be
able to:</p>
<ol style="list-style-type: decimal"><li>Understand what genetic diversity is</li>
<li>Visualize genetic diversity distributions</li>
<li>Connect genetic diversity distributions to Hill numbers</li>
<li>Identify key features of different sequence data formats</li>
<li>Import population genetic data in multiple formats into the R
environment</li>
<li>Manipulate pop gen data by adding and removing sequences</li>
<li>Calculate genetic diversity <span class="math inline">\(\pi\)</span>
</li>
<li>Calculate Hill numbers</li>
<li>Interpret Hill numbers</li>
</ol></div>
</div>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h1>
<div class="section level2">
<h2 id="what-is-genetic-diversity">What is genetic diversity?<a class="anchor" aria-label="anchor" href="#what-is-genetic-diversity"></a></h2>
<p>GEOBON, in their description of Essential Biodiversity Variables, <a href="https://geobon.org/ebvs/what-are-ebvs/" class="external-link">define intraspecific
genetic diversity</a> as:</p>
<blockquote>
<p>The variation in DNA sequences among individuals of the same
species.</p>
</blockquote>
<p>How is this genetic diversity generated? At a proximal level, genetic
diversity is generated through the processes of recombination and
mutation. These two mechanisms lay the foundation for evolutionary
forces like genetic drift, natural selection, and gene flow to shape the
distribution of genetic diversity within populations and across
communities.</p>
<p>Perhaps one of the most elegant theoretical frameworks for modeling
genetic diversity is coalescent theory. Coalescent theory is a model of
how alleles sampled from a population may have originated from a common
ancestor. Under the assumptions of no recombination, no natural
selection, no gene flow, and no population structure, i.e.¬†neutrality,
it estimates the time to the most recent common ancestor (TMRCA) between
two sampled alleles and the variance of this estimate. When a population
is especially large (i.e.¬†many breeding individuals), the TMRCA of two
randomly sampled alleles is large. This is because the per-generation
probability of two alleles coalescing is <code>1 / 2Ne</code>, where
<code>Ne</code>, the effective population size, is the number of
breeding individuals in an idealized population that follows the
assumptions stated above. In most, if not all, natural environments
conditions are not ideal and <code>Ne</code> does not perfectly
represent the census size of a population (see <a href="https://www.biorxiv.org/content/10.1101/2021.02.03.429633v2" class="external-link">Lewontin‚Äôs
paradox</a>). While the relationship between <code>Ne</code> and census
population size is not perfect, in most cases <code>Ne</code> increases
as population size increases and can be used to understand the relative
sizes of populations or size changes over time. The intuition is best
understood through visualization.</p>
<p>For a population of 10 diploid individuals (20 allele copies), the
probability of two randomly sampled allele coalescing each generation is
1 / (2 * 10), or 1/20. In this single simulation (conducted using <a href="https://github.com/cooplab/popgen-notes/blob/master/Rcode/track_alleles.R" class="external-link">Graham
Coop‚Äôs code</a>) , the two sampled alleles, whose genealogy is traced in
blue and red, coalesce 16 generations in the past.</p>
<figure><img src="http://127.0.0.1:26201/graphics/10d9d9fc-fe25-474e-8d09-ff197b99e224.png" class="figure mx-auto d-block"></figure><p>However, when the population is smaller (5 individuals, 10 allele
copies), the probability of coalescence each generation is 1 / (2 * 5)
or 1/10. In this simulation, two randomly sampled alleles coalesce after
6 generations.</p>
<figure><img src="http://127.0.0.1:26201/graphics/21826093-5257-4379-a9df-aba895384e41.png" alt="" class="figure mx-auto d-block"><figure><img src="http://127.0.0.1:26201/graphics/21826093-5257-4379-a9df-aba895384e41.png" class="figure mx-auto d-block"></figure></figure><p>The TMRCA of the two sampled alleles corresponds with the smaller
<code>Ne</code> of this idealized population.</p>
<p>Visualizing this path of ancestry as a tree would look something like
this for 5 sampled allele copies.</p>
<p><a href="https://cdmuir.shinyapps.io/genetree-to-phylogeny/" class="external-link"><img src="../images/gene%20tree%20no%20mutations.png" class="figure"></a></p>
<p>Overlaying mutations on the genealogy results in generally more
mutations occurring on longer branches, although because this is a
stochastic process, the relationship isn‚Äôt always perfect (note- this is
a tree from a different simulation to the one above).</p>
<p><a href="https://cdmuir.shinyapps.io/genetree-to-phylogeny/" class="external-link"><img src="../images/genetree%20with%20mutations-01.png" alt="This represents a genealogy of 5 sampled allele copies from a population, with mutations represented as red boxes. Multiple numbers inside the boxes indicate more mutations." class="figure"></a></p>
<p>As empirical biologists, what we have to work with are sequence data.
The accumulation of mutations along a sequence enables us to estimate
the TMRCA of a sample of alleles, leading to an estimate of the
<code>Ne</code> of the population.</p>
<p>One of the simplest and most widespread estimators of <code>Ne</code>
is the average number of pairwise differences between sequences sampled
from a population (ùûπ, nucleotide diversity, <a href="https://www.pnas.org/doi/abs/10.1073/pnas.76.10.5269" class="external-link">Nei and Li
1979</a>). The basic idea underlying ùûπ is that more sequence differences
among a sample of sequences results from longer branch lengths in the
genealogy of the sample, leading to a longer TMRCA and larger
<code>Ne</code>. Given its simplicity and ubiquity, we will cover how to
calculate the average number of pairwise differences between sequences
and summarize this value across assemblages.</p>
<div class="section level4">
<h4 id="resources-for-understanding-the-coalescent">Resources for understanding the coalescent<a class="anchor" aria-label="anchor" href="#resources-for-understanding-the-coalescent"></a></h4>
<ul><li>
<a href="http://cooplab.github.io/popgen-notes/" class="external-link">Graham Coop‚Äôs
popgen notes</a>
<ul><li>awesome resource for population genetics in general</li>
</ul></li>
<li>
<a href="https://phytools.shinyapps.io/coalescent-plot/" class="external-link">learnPopGen
app</a>
<ul><li>visualize coalescent genealogies with different numbers of
individuals in a population and different numbers of generations</li>
</ul></li>
<li>
<a href="https://bedford.io/projects/coaltrace/" class="external-link">coaltrace app</a>
<ul><li>a dynamic and interactive visualization of coalescent genealogies
with the ability to change parameters on the fly</li>
</ul></li>
<li>
<a href="https://bedford.io/projects/phylodynamics-lecture/coalescent.html#/" class="external-link">Bedford
lab slides</a>
<ul><li>a great set of slides that explains the coalescent in a
complementary way to Graham Coop‚Äôs notes</li>
</ul></li>
<li>
<a href="https://cdmuir.shinyapps.io/genetree-to-phylogeny/" class="external-link">cdmuir
app</a>
<ul><li>uses the coalescent to simulate gene trees and mutations along
sequences</li>
</ul></li>
</ul></div>
</div>
<div class="section level2">
<h2 id="what-does-genetic-diversity-say-about-community-assembly">What does genetic diversity say about community assembly?<a class="anchor" aria-label="anchor" href="#what-does-genetic-diversity-say-about-community-assembly"></a></h2>
<p>Under neutrality and stable demographic histories, variation in
genetic diversity across a community reflects differences in long-term
population sizes across the community. Rather than a snapshot of
present-day abundance, this can be thought of as an average abundance
across generations.</p>
<p>In much the same way as a community with a few, highly abundant
species and many low abundance species may indicate a non-neutral
community assembly process like environmental filtering, a community
that contains a few species with high genetic diversity and many species
with low genetic diversity may indicate similar non-neutral community
assembly processes over a longer timescale.</p>
<p>In addition, genetic diversity is impacted by demographic events and
natural selection. Populations that have undergone a demographic
expansion or contraction in the recent past will have distinct
coalescent genealogies and therefore different genetic diversities. For
instance, a population that has experienced a recent demographic
expansion will have a TMRCA more recent than a population with the same
population size that has remained stable through the same time period,
illustrated below.</p>
<p><a href="https://bedford.io/projects/phylodynamics-lecture/coalescent.html#/20" class="external-link"><img src="../images/changing_pop_size_1.png" alt="Image of coalescent genealogies traced from stable and growing populations. From the Bedford Lab coalescent slides." class="figure"></a></p>
<p>Distinguishing a small, constant-sized population from a large
population that‚Äôs grown dramatically in the recent past is difficult
with single-locus data typical of large-scale biodiversity studies.
However, when paired with other data like traits and abundances, we can
make more refined interpretations of the shape of genetic diversity
across the community. For instance, the interpretation of a community
with a few highly diverse species paired with many species with low
genetic diversity may reflect a community with a few species that were
not impacted by a historical environmental disturbance alongside many
species that had their populations dramatically reduced, but have
rapidly grown to larger modern population sizes.</p>
</div>
</div>
<div class="section level1">
<h1 id="summarizing-genetic-diversity-across-species">Summarizing genetic diversity across species<a class="anchor" aria-label="anchor" href="#summarizing-genetic-diversity-across-species"></a></h1>
<div class="section level3">
<h3 id="genetic-diversity-distributions">Genetic diversity distributions<a class="anchor" aria-label="anchor" href="#genetic-diversity-distributions"></a></h3>
<p>{{pi across species neutral viz}}</p>
<p>{{pi across species non-neutral viz}}</p>
</div>
<div class="section level3">
<h3 id="hill-numbers-of-genetic-diversity-across-species">Hill numbers of genetic diversity across species<a class="anchor" aria-label="anchor" href="#hill-numbers-of-genetic-diversity-across-species"></a></h3>
<p>In the context of summarizing species abundances, Hill numbers are
the effective number of species in a community, where each exponent
(q-value) increasingly weights rare species more heavily. The
interpretation is similar when summarizing genetic diversity. Hill
numbers are the effective number of species in a community, where each
q-value increasingly weights species with lower genetic diversity more
heavily. Rather than measuring an effective diversity of abundances,
this is an effective diversity of genetic diversities.</p>
</div>
</div>
<div class="section level1">
<h1 id="work-with-pop-gen-data">Work with pop gen data<a class="anchor" aria-label="anchor" href="#work-with-pop-gen-data"></a></h1>
<div class="section level2">
<h2 id="sequences">Sequences<a class="anchor" aria-label="anchor" href="#sequences"></a></h2>
<div class="section level3">
<h3 id="fasta">FASTA<a class="anchor" aria-label="anchor" href="#fasta"></a></h3>
<p>A FASTA file contains one or more DNA sequences, where the DNA
sequence is preceded by a line with a carat (‚Äú&gt;‚Äù) followed by an
unique sequence ID:</p>
<pre><code>&gt;seq0010
GATCCCCAATTGGGG</code></pre>
<p>This is perhaps the most common way to store sequence information as
it is simple and has historical inertia. For more information, see the
<a href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat/#:~:text=In%20FASTA%20format%20the%20line,should%20not%20contain%20any%20spaces." class="external-link">the
NCBI page</a>.</p>
</div>
<div class="section level3">
<h3 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a></h3>
<p>Prior to analysis, sequence data must be aligned if they haven‚Äôt been
already. A sequence alignment is necessary to identify regions of DNA
that are similar or that vary. There are <a href="https://en.wikipedia.org/wiki/List_of_sequence_alignment_software" class="external-link">many
pieces of software to perform alignments</a> and all have their
advantages and disadvantages, but we will choose a single program, <a href="https://www.ebi.ac.uk/Tools/msa/clustalo/" class="external-link">Clustal Omega</a>,
which is fast and robust for short sequence alignments. While Clustal
Omega can be run from the command line, we will use the R package <a href="https://bioconductor.org/packages/release/bioc/html/msa.html" class="external-link">msa</a>
to keep everything in the R environment.</p>
<p>First, a FASTA file containing the sequences needs to be read in
using the <code>readDNAStringSet()</code> function.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">msa</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seq_path</span> <span class="op">&lt;-</span> <span class="st">"/Users/connorfrench/Dropbox/Old_Mac/School_Stuff/CUNY/Courses/Spring-2019/Machine-Learning/project/fasta/hypsiboas_seqs_aligned.fas"</span></span>
<span></span>
<span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">seq_path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqs</span></span></code></pre>
</div>
<p>To align the sequences using Clustal Omega, we use the
<code>msa</code> function and specify the <code>method</code> as
‚ÄúClustalOmega‚Äù. There are more arguments that you can use to fine-tune
the alignment, but these are not necessary for the vast majority of
scenarios.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">alignment</span></span></code></pre>
</div>
<p>If after you perform your alignment you find that an individual does
not belong or is negatively impacting the alignment, you have to
manipulate your unaligned sequence data and then re-align the sequences.
For instance, say you didn‚Äôt want to include the individual
<code>ind_name</code> in the alignment. First, remove the individual
from the <code>seqs</code> object.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove individual</span></span>
<span><span class="va">seqs_filt</span> <span class="op">&lt;-</span> <span class="va">seqs</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span> <span class="op">%in%</span> <span class="st">"Itar128"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">seqs_filt</span></span></code></pre>
</div>
<p>This nifty filtering approach takes advantage of the powerful
<code>%in%</code> function. In English, this function takes the vector
on the left side of the <code>%in%</code> and checks if each element
matches any of the elements on the right side. The left side contains
the elements in the data you want to check, while the right side
contains what you want to compare them against. So in this case,</p>
<p><code>names(seqs) %in% "Itar128"</code></p>
<p>checks to see if ‚ÄúItar128‚Äù is present in any of the sequence names,
returns <code>TRUE</code> if so and <code>FALSE</code> if not. Since we
want to remove the ‚ÄúItar128‚Äù sequence, we want the opposite of this
result. To do this, we wrap this expression with the bang
(<code>!</code>) operator, which negates the result of a boolean
expression. So, the following expression returns <code>TRUE</code> for
all sequence names that are not ‚ÄúItar128‚Äù and <code>FALSE</code>
otherwise.</p>
<p><code>!(names(seqs) %in% "Itar128"</code></p>
<p>Using bracket indexing, you then return all sequences that do not
have the name ‚ÄúItar128‚Äù.</p>
<p><code>seqs[!(names(seqs) %in% "Itar128"</code></p>
<p>Following the filtering, you need to realign the new set of
sequences.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alignment</span> <span class="op">&lt;-</span> <span class="fu">msa</span><span class="op">(</span><span class="va">seqs_filt</span>, method <span class="op">=</span> <span class="st">"ClustalOmega"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">alignment</span></span></code></pre>
</div>
<p>There are quite a few packages in R to work with sequence data, for
example <a href="https://cran.r-project.org/web/packages/seqinr/index.html" class="external-link">seqinr</a>,
<a href="https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.html" class="external-link">phangorn</a>,
and <a href="https://cran.r-project.org/web/packages/ape/index.html" class="external-link">ape</a>.
<code>ape</code> is the most versatile of these. But before we can use
<code>ape</code>, the alignment must be converted to the
<code>ape::DNAbin</code> format.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ape_align</span> <span class="op">&lt;-</span> <span class="fu">msaConvert</span><span class="op">(</span><span class="va">alignment</span>, type <span class="op">=</span> <span class="st">"ape::DNAbin"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ape_align</span></span></code></pre>
</div>
<p>Although not necessary for this workshop, it‚Äôs generally advisable to
write your DNA alignment to a new file to use later or with other
programs.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ape</span><span class="fu">::</span><span class="fu">write.FASTA</span><span class="op">(</span><span class="va">ape_align</span>, <span class="st">"path/to/alignment.fas"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="ape">ape<a class="anchor" aria-label="anchor" href="#ape"></a></h3>
<div class="section level4">
<h4 id="check-alignment">Check alignment<a class="anchor" aria-label="anchor" href="#check-alignment"></a></h4>
<p><code>ape</code> has a useful utility that allows you perform a
series of diagnostics on your alignment to make sure nothing is fishy.
The function can output multiple plots, but let‚Äôs just visualize the
alignment plot. If the colors seem jumbled in a region, this indicates
that an alignment error may have occurred and to inspect your sequences
for any problems.</p>
<p>In addition, the function outputs a series of helpful statistics to
your console. Ideally, your alignment should have few gaps and a number
of segregating sites that is reasonable for the number of sites in your
alignment and the evolutionary divergence of the sequences included in
the alignment, e.g., for an alignment with 970 sites of a single bird
species, around 40 segregating sites is reasonable, but 200 segregating
sites may indicate that something went wrong with your alignment or a
species was misidentified. In addition, for diploid species a vast
majority, if not all of the sites should contain one or two observed
bases.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">checkAlignment</span><span class="op">(</span><span class="va">ape_align</span>, what <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="calculate-genetic-diversity">Calculate genetic diversity<a class="anchor" aria-label="anchor" href="#calculate-genetic-diversity"></a></h4>
<p>Now we‚Äôre going to calculate the average number of pairwise genetic
differences among these sequences. <code>ape</code> has the
<code>dist.dna()</code> function to calculate various genetic distance
metrics for an alignment.</p>
<p>You use the argument <code>model = "raw"</code> to specify that you
want to use the average number of pairwise differences as your metric.
When calculating the number of pairwise genetic differences, you have to
choose whether you want missing data to be deleted across the entire
alignment (<code>pairwise.deletion = FALSE</code>) or only between the
two sequences being compared (<code>pairwise.deletion = TRUE</code>).
Since we want to take advantage of as much data as possible when
computing ùúã, we‚Äôll set <code>pairwise.deletion = TRUE</code>.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendist</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">ape_align</span>, model <span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fas_gendist</span></span></code></pre>
</div>
<p>The function returns a distance matrix containing the pairwise
differences among all sequences in the data set. To take a quick peak at
their distribution, we can use the <code>hist()</code> function on the
distance matrix.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">fas_gendist</span><span class="op">)</span></span></code></pre>
</div>
<p>Since we want the average of these differences, we need to take their
average. However, you may have noticed the distribution of genetic
distances is non-normal, with mostly low genetic diversities. This is
the case with most genetic data sets. So, instead of taking the mean,
the <code>median()</code> will provide a better estimate of the center
of the distribution.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendiv</span> <span class="op">&lt;-</span> <span class="fu">median</span><span class="op">(</span><span class="va">fas_gendist</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fas_gendiv</span></span></code></pre>
</div>
<p>To calculate this for multiple populations‚Ä¶</p>
<p>read in fasta alignments‚Ä¶</p>
<p><code>lapply</code> explanation‚Ä¶</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_paths</span> <span class="op">&lt;-</span> <span class="fu">list.files</span><span class="op">(</span><span class="st">"..."</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fas_alignments</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">fas_paths</span>, <span class="va">read.dna</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can use <code>lapply</code> to calculate the pairwise genetic
distances for each species...</p>
<p><code>sapply</code> is similar to <code>lapply</code>, except it
returns a vector instead of a list. This is useful when you know the
function is returning the same data type, like in our case where a
number is being returned.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fas_gendist_pop</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">fas_alignments</span>, <span class="va">dist.dna</span>, model <span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">fas_gendiv_pop</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">fas_gendist_pop</span>, <span class="va">median</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">fas_gendiv_pop</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
<div class="section level3">
<h3 id="hill-numbers">Hill numbers<a class="anchor" aria-label="anchor" href="#hill-numbers"></a></h3>
<p><strong>How are we calculating genetic Hill numbers?</strong></p>
</div>
</div>
<div class="section level2">
<h2 id="asvs-and-metabarcoding">ASVs and Metabarcoding<a class="anchor" aria-label="anchor" href="#asvs-and-metabarcoding"></a></h2>
<p>Metabarcoding is rapidly growing DNA sequencing method used for
biodiversity assessment. It involves sequencing a short, variable region
of DNA sequence across a sample that may contain many taxa. This can be
used to rapidly assess the diversity of a particular sampling site with
low sampling effort. Some common applications include <a href="https://en.wikipedia.org/wiki/Metabarcoding" class="external-link">microbiome diversity
assessments, environmental DNA, and community DNA</a>. In many
applications where a data base of species identities for sequence
variants is not available or desired, the diversity of sequences is
inferred, rather than the diversity of species. Additionally, while <a href="https://onlinelibrary-wiley-com.ezproxy.gc.cuny.edu/doi/full/10.1111/1755-0998.12902" class="external-link">there
are caveats</a>, the relative abundance of sequences can be considered
analogous to the relative abundance of species across a community.</p>
<p>The basic unit of diversity used in most studies is the <a href="https://en.wikipedia.org/wiki/Amplicon_sequence_variant" class="external-link">amplicon
sequence variant (ASV)</a>. An ASV is any one of the sequences inferred
from a high-throughput sequencing experiment. These sequences are free
from errors and may diverge by as little as a single base pair from each
other. They are considered a substitute for operational taxonomic units
(OTUs) for evaluating ecological diversity given their higher
resolution, reproducibility, and no need for a reference database.</p>
<div class="section level3">
<h3 id="sequence-table">Sequence table<a class="anchor" aria-label="anchor" href="#sequence-table"></a></h3>
<p>The basic unit of analysis used for metabarcoding studies is the
sequence table. It is a matrix with named rows corresponding to the
samples and named columns corresponding to the ASVs. Each cell contains
the count of each variant in each sample. Here‚Äôs an example with 5
samples and 4 sequence variants:</p>
<table class="table"><thead><tr class="header"><th></th>
<th>asv_1</th>
<th>asv_2</th>
<th>asv_3</th>
<th>asv_4</th>
</tr></thead><tbody><tr class="odd"><td>sample_1</td>
<td>0</td>
<td>276</td>
<td>200</td>
<td>10</td>
</tr><tr class="even"><td>sample_2</td>
<td>10</td>
<td>1009</td>
<td>0</td>
<td>0</td>
</tr><tr class="odd"><td>sample_3</td>
<td>0</td>
<td>0</td>
<td>234</td>
<td>15</td>
</tr><tr class="even"><td>sample_4</td>
<td>129</td>
<td>1992</td>
<td>0</td>
<td>485</td>
</tr><tr class="odd"><td>sample_5</td>
<td>1001</td>
<td>10</td>
<td>1847</td>
<td>178</td>
</tr></tbody></table><p>sample_1 may correspond to a soil sample collected in a forest, while
the variants represent different ASVs, which could look like this:</p>
<p>asv_1: GTTCA</p>
<p>asv_2: GTACA</p>
<p>asv_3: GATCA</p>
<p>asv_4: CTTCA</p>
<p>This looks similiar to a site-species matrix, where the rows are a
sampling locality, the columns are the species, and the cell values are
the abundance of a species at a particular site.</p>
<p>While for this workshop we will focus on the analysis of a sequence
table, we‚Äôll briefly go over how the data makes its way from raw
sequencing data to the sequence table.</p>
</div>
<div class="section level3">
<h3 id="fastq">FASTQ<a class="anchor" aria-label="anchor" href="#fastq"></a></h3>
<p>The FASTQ file format is similar to the FASTA format introduced
earlier, but includes additional data about the contained sequences.
This enables more flexible data manipulation and filtering capabilities,
and is popular with high-throughput sequencing technologies. When
working with ASVs, this is the most common file format you will
encounter.</p>
<p>A FASTA formatted sequence may look like this,</p>
<pre><code>&gt;seq1
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT</code></pre>
<p>while a FASTQ version of the same sequence may look like this:</p>
<pre><code>@seq1
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT
+
!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65</code></pre>
<p>If your eyes glaze over at the seemingly random numbers below the
sequence, don‚Äôt fret! They mean something, but usually your machine does
the interpreting for you. A FASTQ file is formatted as follows:</p>
<ol style="list-style-type: decimal"><li><p>An ‚Äò@‚Äô symbol followed by a sequence identifier and an optional
description</p></li>
<li><p>The raw sequence letters</p></li>
<li><p>A ‚Äò+‚Äô character that is optionally followed by the sequence
identifier and more description</p></li>
<li><p>Quality encodings for the raw sequence letters in field 2. There
must be the same number of encoding symbols as there are sequence
letters.</p></li>
</ol><p>The symbols are bytes representing quality, ranging from 0x21 (lowest
quality; ‚Äò!‚Äô in ASCII) to 0x7e (highest quality; ‚Äò~‚Äô in ASCII). Here,
they‚Äôre arranged in order of quality from lowest on the left to highest
on the right:</p>
<pre><code>!"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~</code></pre>
<p>Fortunately, whatever bioinformatic program you use will generally do
the interpreting for you.</p>
<p>Different sequencing technologies (e.g.¬†Illumina) have standardized
description strings which contain more information about the sequences
and can be useful when filtering sequences or performing QC. For
examples, see <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">the
Wikipedia page</a>.</p>
</div>
<div class="section level3">
<h3 id="dada2">DADA2<a class="anchor" aria-label="anchor" href="#dada2"></a></h3>
<p><a href="https://benjjneb.github.io/dada2/index.html" class="external-link">DADA2</a> is
one of the most powerful and commonly used software packages for
processing raw ASV sequencing data to turn it into a table of ASVs for
each sample. This workshop assumes you are using the output from this
workflow for further analysis, but for clarity the steps to the workflow
include:</p>
<ol style="list-style-type: decimal"><li>read quality assessment</li>
<li>filtering and trimming FASTQ files</li>
<li>learning sequencing error rates, so the algorithm can correct for
them</li>
<li>infer the sequence variants</li>
<li>merging paired reads (if sequencing was done with forward and
reverse read pairs)</li>
<li>constructing a sequence table</li>
</ol><p>There are more available to further assess the quality of the data
and map the data to taxonomic databases, but the core output of DADA2
and similar software pipelines is the sequence table.</p>
</div>
<div class="section level3">
<h3 id="hill-numbers-1">Hill numbers<a class="anchor" aria-label="anchor" href="#hill-numbers-1"></a></h3>
<p>I assume we‚Äôre calculating Hill numbers using hillR::hill_taxa, but
is that correct?</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="extra">Extra<a class="anchor" aria-label="anchor" href="#extra"></a></h1>
<div class="section level2">
<h2 id="snps">SNPs<a class="anchor" aria-label="anchor" href="#snps"></a></h2>
<div class="section level3">
<h3 id="vcf">VCF<a class="anchor" aria-label="anchor" href="#vcf"></a></h3>
<p>Here are a few things about a VCF file‚Ä¶ tab delimited‚Ä¶ fields‚Ä¶
size‚Ä¶</p>
<div class="section level4">
<h4 id="vcfr">vcfR<a class="anchor" aria-label="anchor" href="#vcfr"></a></h4>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">vcfR</span><span class="op">)</span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">read.vcfR</span><span class="op">(</span><span class="va">vcf_path</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="manipulate-data">Manipulate data<a class="anchor" aria-label="anchor" href="#manipulate-data"></a></h4>
<p>Filter individuals Filter based on quality, missingness</p>
</div>
<div class="section level4">
<h4 id="export-to-genind">Export to genind<a class="anchor" aria-label="anchor" href="#export-to-genind"></a></h4>
<p>A bit about genind and <code>adegenet</code>‚Ä¶ If you have many SNPs,
genlight may be a more appropriate format‚Ä¶</p>
<p><code>vcfR2genind</code> also takes arguments for the function
<code>df2genind</code> from the <code>adegenet</code> package. There are
a couple things we can do to speed up the conversion, like specifying
the ploidy of the file and setting the <code>check.ploidy</code> flag to
<code>FALSE</code>.</p>
<p>To add populations‚Ä¶</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">adegenet</span><span class="op">)</span></span>
<span><span class="va">snps_genind</span> <span class="op">&lt;-</span> <span class="fu">vcfR2genind</span><span class="op">(</span></span>
<span>    <span class="va">snps</span>, </span>
<span>    n.cores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    ploidy <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    check.ploidy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    pop <span class="op">=</span> <span class="va">pops</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">snps_genind</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="calculate-genetic-diversity-1">Calculate genetic diversity<a class="anchor" aria-label="anchor" href="#calculate-genetic-diversity-1"></a></h4>
<p>If you didn‚Äôt specify populations when converting or reading in the
genind file, do so now‚Ä¶</p>
<p>Individual-based genetic distances‚Ä¶ Maybe mention PCA?</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_gendist_ind</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="va">snps_genind</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span></code></pre>
</div>
<p>Nei‚Äôs genetic distance‚Ä¶</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_genpop</span> <span class="op">&lt;-</span> <span class="fu">genind2genpop</span><span class="op">(</span><span class="va">snps_genind</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snps_gendist_pop</span> <span class="op">&lt;-</span> <span class="fu">dist.genpop</span><span class="op">(</span><span class="va">snps_genpop</span>, method <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<p>FST‚Ä¶</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hierfstat</span><span class="op">)</span></span>
<span><span class="va">snps_hierfstat</span> <span class="op">&lt;-</span> <span class="fu">genind2hierfstat</span><span class="op">(</span><span class="va">snps_genind</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snps_fst</span> <span class="op">&lt;-</span> <span class="fu">pairwise.fst</span><span class="op">(</span><span class="va">snps_hierfstat</span>, diploid <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">snps_fst</span><span class="op">[</span><span class="fu">upper.tri</span><span class="op">(</span><span class="va">snps_fst</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="summarize-genetic-diversity">Summarize genetic diversity<a class="anchor" aria-label="anchor" href="#summarize-genetic-diversity"></a></h1>
<div class="section level2">
<h2 id="calculate-hill-numbers-within-speciespopulations">Calculate Hill numbers within species/populations<a class="anchor" aria-label="anchor" href="#calculate-hill-numbers-within-speciespopulations"></a></h2>
<p><em>* Going to need a SNP matrix</em></p>
<p>Hill numbers of intraspecific allele frequencies are the
<em>effective number of SNPs</em> in the population/species.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">hillR</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># columns are SNPs, rows are individuals</span></span>
<span><span class="co"># treat it like abundances </span></span>
<span><span class="va">intra_hill_1</span> <span class="op">&lt;-</span> <span class="fu">hillR</span><span class="fu">::</span><span class="fu">hill_taxa</span><span class="op">(</span><span class="va">snp_matrix</span>, q <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="interpret-refer-back-to-visualizations">Interpret (refer back to visualizations)<a class="anchor" aria-label="anchor" href="#interpret-refer-back-to-visualizations"></a></h3>
</div>
</div>
<div class="section level2">
<h2 id="calculate-hill-numbers-across-species">Calculate Hill numbers across species<a class="anchor" aria-label="anchor" href="#calculate-hill-numbers-across-species"></a></h2>
<p><em>Need species alignments from ape</em></p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate pairwise distance matrices</span></span>
<span><span class="va">per_species_dist</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">species_alignments</span>, <span class="fu">ape</span><span class="fu">::</span><span class="va">dist.dna</span>, model <span class="op">=</span> <span class="st">"raw"</span>, as.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mat_avg</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="fu">upper.tri</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span>    <span class="kw">return</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">per_species_pi</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">per_species_dist</span>, <span class="va">mat_avg</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">per_species_pi</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="interpret-refer-back-to-visualizations-1">Interpret (refer back to visualizations)<a class="anchor" aria-label="anchor" href="#interpret-refer-back-to-visualizations-1"></a></h3>
</div>
</div>
<div class="section level2">
<h2 id="within-species">Within-species<a class="anchor" aria-label="anchor" href="#within-species"></a></h2>
<div class="section level3">
<h3 id="sfs-visualization">SFS visualization<a class="anchor" aria-label="anchor" href="#sfs-visualization"></a></h3>
<p>The site frequency spectrum (SFS) is a summary of genetic variants
across a population, similar to how abundances are summarized across a
community. Specifically, it is the distribution of the allele
frequencies of a given set of loci (generally single nucleotide
polymorphisms, SNPs), in a population or sample. The best way to gain an
intuition for what the SFS is, is to visualize it.</p>
<p>The first bar of this plot indicates derived alleles (i.e.¬†not
ancestral) that appear once across the population. Approximately 35% of
the sampled SNPs in the genome alignment are these ‚Äúsingletons‚Äù. The
second bar indicates derived alleles that appear twice across the
population, where ~18% of the sampled SNPs are these ‚Äúdoubletons‚Äù and so
forth.</p>
<p>The shape of the SFS can tell us important information about the
demographic history of a population and their history of selection. The
above plot corresponds with a population that has experienced
demographic stability. A couple other examples of the impact of
demographic history on the SFS include:</p>
<p>There are a variety of summary statistics that describe the shape of
the SFS. The average number of pairwise nucleotide differences, a.k.a.
Tajima‚Äôs/Nei‚Äôs <span class="math inline">\(\pi\)</span>, where a high
value indicates a higher level of genetic diversity in the population,
describes the average of the distribution. However, as you‚Äôve seen
above, the shape of the distribution contains important information
about the population‚Äôs history!</p>
</div>
<div class="section level3">
<h3 id="hill-numbers-of-the-sfs">Hill numbers of the SFS<a class="anchor" aria-label="anchor" href="#hill-numbers-of-the-sfs"></a></h3>
<p>In the context of summarizing species abundances, Hill numbers are
the effective number of species in a community, where each exponent
(q-value) increasingly weights rare species more heavily. The
interpretation is similar when summarizing within-population genetic
diversity. Hill numbers of within-population allele frequencies convey
the <em>effective number of alleles</em> in a population, where each
q-value increasingly weights rarer alleles more heavily.</p>
<p><em>Something about when we‚Äôd use</em> <em>pi vs Hill</em></p>
</div>
</div>
<div class="section level2">
<h2 id="section"></h2>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/traits-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/phylo-data.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/traits-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Trait data</a>
        <a class="chapter-link float-end" href="../instructor/phylo-data.html" rel="next">Next: Summarizing... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/role-model/multidim-biodiv-data/edit/main/episodes/popgen-data.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/role-model/multidim-biodiv-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/role-model/multidim-biodiv-data/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:biodataworkshop2023@gmail.com">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.3" class="external-link">sandpaper (0.12.3)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://role-model.github.io/multidim-biodiv-data/instructor/popgen-data.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Working with population genetic data",
  "creativeWorkStatus": "active",
  "url": "https://role-model.github.io/multidim-biodiv-data/instructor/popgen-data.html",
  "identifier": "https://role-model.github.io/multidim-biodiv-data/instructor/popgen-data.html",
  "dateCreated": "2023-06-15",
  "dateModified": "2023-06-15",
  "datePublished": "2023-06-15"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

